{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "declared-lewis",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "formed-valuation",
   "metadata": {},
   "outputs": [],
   "source": [
    "project='cpTree_v4'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "varied-concert",
   "metadata": {},
   "outputs": [],
   "source": [
    "def tree_taxo(df):\n",
    "    print(df.shape[0], 'o:',df.order.nunique(),'f:',df.family.nunique(),'g:',df.genus.nunique(),'s:',df.sci_name.nunique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "diagnostic-slovenia",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "174\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\ProgramData\\Anaconda3\\lib\\site-packages\\IPython\\core\\interactiveshell.py:3147: DtypeWarning: Columns (16,18,19,26,32,33,35,41,43,55,66,69) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    }
   ],
   "source": [
    "blacklist = pd.read_csv('../PAFTOL_DB/2021-10-01_paftol_export.csv')\n",
    "blacklist = blacklist[blacklist.Blacklisted==1]\n",
    "print(blacklist.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "hungarian-puppy",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PAFTOL: 8916 < 8561 > 8426\n",
      "SRA: 1718 < 1661 > 1661\n",
      "GAP: 1455 < 774 > 774\n",
      "1KP: 744 < 744 > 744\n",
      "Li et al: 1327 < 1327 > 1327\n",
      "RefSeq: 5176 < 5176 > 5176\n",
      "INSDC: 2916 < 2916 > 2916\n",
      "Root (Gymnosperms): 10 < 11 > 10\n"
     ]
    }
   ],
   "source": [
    "### List all samples for the tree\n",
    "# Load paftol_pt\n",
    "paftol_pt = pd.read_csv('../Organelles/PAFTOL/PAFTOL_Organelle_Recovery.csv').rename(columns={'Order':'order',\n",
    "                                            'Family':'family','Genus':'genus','Sum_len_pt':'length'})\n",
    "paftol_pt['species'] = paftol_pt.sci_name.str.split(' ').str[1:].str.join(' ')\n",
    "print('PAFTOL:',paftol_pt.shape[0],end=' < ')\n",
    "paftol_ls = [file.replace('_pt.fasta','') for file in os.listdir('../Organelles/PAFTOL/fasta_pt/') if file.endswith('.fasta')]\n",
    "paftol_pt = paftol_pt[paftol_pt.Sample_Name.isin(paftol_ls)]\n",
    "paftol_pt = paftol_pt[paftol_pt.idSequencing.isin(blacklist.idSequencing)==False] #Remove blacklisted samples\n",
    "paftol_pt['DataSource']='PAFTOL'\n",
    "print(len(paftol_ls), '>',paftol_pt.shape[0])\n",
    "\n",
    "# Load sra_pt\n",
    "sra_pt = pd.read_csv('../Organelles/SRA/SRA_Organelle_Recovery.csv').rename(columns={'Sum_len_pt':'length'})\n",
    "sra_pt['species'] = sra_pt.sci_name.str.split(' ').str[1:].str.join(' ')\n",
    "print('SRA:',sra_pt.shape[0],end=' < ')\n",
    "SRA_ls = [file.replace('_pt.fasta','') for file in os.listdir('../Organelles/SRA/fasta_pt/') if file.endswith('.fasta')]\n",
    "sra_pt = sra_pt[sra_pt.Sample_Name.isin(SRA_ls)]\n",
    "sra_pt = sra_pt[sra_pt.idSequencing.isin(blacklist.idSequencing)==False] #Remove blacklisted samples\n",
    "sra_pt['DataSource']='SRA'\n",
    "print(len(SRA_ls), '>',sra_pt.shape[0])\n",
    "\n",
    "# Load gap_pt\n",
    "gap_pt = pd.read_csv('../Organelles/GAP/GAP_Organelle_Recovery.csv').rename(columns={'Sum_len_pt':'length'})\n",
    "gap_pt['species'] = gap_pt.sci_name.str.split(' ').str[1:].str.join(' ')\n",
    "print('GAP:',gap_pt.shape[0],end=' < ')\n",
    "GAP_ls = [file.replace('_pt.fasta','') for file in os.listdir('../Organelles/GAP/fasta_pt/') if file.endswith('.fasta')]\n",
    "gap_pt = gap_pt[gap_pt.Sample_Name.isin(GAP_ls)]\n",
    "gap_pt = gap_pt[gap_pt.idSequencing.isin(blacklist.idSequencing)==False] #Remove blacklisted samples\n",
    "gap_pt['DataSource']='GAP'\n",
    "print(len(GAP_ls), '>',gap_pt.shape[0])\n",
    "\n",
    "# Load 1kp_pt\n",
    "okp_pt = pd.read_csv('1KP/okp_pt.csv').rename(columns={'Sample':'Sample_Name',\"('len', 'sum_len')\":'length'})\n",
    "okp_pt['species'] = okp_pt.sci_name.str.split(' ').str[1:].str.join(' ')\n",
    "print('1KP:',okp_pt.shape[0],end=' < ')\n",
    "okp_ls = [file.replace('_CPgenes.fasta','') for file in os.listdir('1KP/1kp_fasta_pt/') if file.endswith('.fasta')]\n",
    "okp_pt = okp_pt[okp_pt.Sample_Name.isin(okp_ls)]\n",
    "okp_pt['DataSource']='OneKP'\n",
    "print(len(okp_ls), '>',okp_pt.shape[0])\n",
    "\n",
    "# Load Li et al\n",
    "Lietal_pt = pd.read_csv('Li_et_al/Li_et_al_samples.csv').rename(columns={'Sample':'Sample_Name'})\n",
    "print('Li et al:',Lietal_pt.shape[0],end=' < ')\n",
    "lietal_ls = [file.replace('_CPgenes.fasta','') for file in os.listdir('Li_et_al/Li_et_al-Data/') if file.endswith('.fasta')]\n",
    "Lietal_pt = Lietal_pt[Lietal_pt.Sample_Name.isin(lietal_ls)]\n",
    "Lietal_pt['DataSource']='Lietal';\n",
    "print(len(lietal_ls), '>',Lietal_pt.shape[0])\n",
    "\n",
    "# Load NCBI refseq\n",
    "refseq_pt = pd.read_csv('NCBI/RefSeq/refseq_sp.csv').rename(columns={'SeqID':'Sample_Name'})\n",
    "print('RefSeq:',refseq_pt.shape[0],end=' < ')\n",
    "refseq_ls = [file.replace('_pt.fasta','') for file in os.listdir('NCBI/RefSeq/Refseq_fasta_pt/') if file.endswith('.fasta')]\n",
    "refseq_pt = refseq_pt[refseq_pt.Sample_Name.isin(refseq_ls)]\n",
    "refseq_pt['DataSource']='REFSEQ'\n",
    "print(len(refseq_ls), '>',refseq_pt.shape[0])\n",
    "\n",
    "# Load NCBI INSDC (other plastomes, partial)\n",
    "insdc_pt = pd.read_csv('NCBI/INSDC/insdc_sp.csv').rename(columns={'SeqID':'Sample_Name'})\n",
    "print('INSDC:',insdc_pt.shape[0],end=' < ')\n",
    "insdc_ls = [file.replace('_pt.fasta','') for file in os.listdir('NCBI/INSDC/INSDC_fasta_pt/') if file.endswith('.fasta')]\n",
    "insdc_pt = insdc_pt[insdc_pt.Sample_Name.isin(insdc_ls)]\n",
    "insdc_pt['DataSource']='INSDC'\n",
    "print(len(insdc_ls), '>',insdc_pt.shape[0])\n",
    "\n",
    "# List gymno\n",
    "gymno_pt = pd.read_csv('NCBI/Gymnosperms/Gymnosperms_raw_SeqID.csv').rename(columns={'SeqID':'Sample_Name'})\n",
    "gymno_pt['genus'] = gymno_pt.sci_name.str.split(' ').str[0]\n",
    "gymno_pt['species'] = gymno_pt.sci_name.str.split(' ').str[1:].str.join(' ')\n",
    "print('Root (Gymnosperms):',gymno_pt.shape[0],end=' < ')\n",
    "gymno_ls = [file.replace('_pt.fasta','') for file in os.listdir('NCBI/Gymnosperms/') if file.endswith('.fasta')]\n",
    "gymno_pt = gymno_pt[gymno_pt.Sample_Name.isin(gymno_ls)]\n",
    "gymno_pt['DataSource']='Gymno'\n",
    "print(len(gymno_ls), '>',gymno_pt.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "filled-touch",
   "metadata": {},
   "outputs": [],
   "source": [
    "col_keep=['Sample_Name','family','genus','species','sci_name','DataSource','length']\n",
    "pd.concat([ paftol_pt[col_keep], sra_pt[col_keep], gap_pt[col_keep], okp_pt[col_keep], Lietal_pt[col_keep],\n",
    "                    refseq_pt[col_keep], insdc_pt[col_keep], gymno_pt[col_keep]], ignore_index=True)\\\n",
    "    .to_csv(project + '/AllSamples_' + project + '.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "tender-librarian",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{1: 1313, 2: 288, 3: 125, 4: 80, 5: 51, 6: 28, 7: 22, 8: 13, 9: 15, 10: 13, 11: 3, 12: 9, 13: 3, 14: 8, 15: 5, 16: 3, 17: 6, 18: 4, 19: 2, 20: 3, 21: 1, 22: 3, 23: 3, 24: 1, 25: 1, 27: 2, 28: 1, 29: 1, 31: 3, 32: 1, 34: 2, 35: 2, 36: 2, 41: 1, 46: 1, 47: 1, 54: 2, 66: 1}\n",
      "{1: 1672, 2: 376, 3: 169, 4: 103, 5: 51, 6: 61, 7: 40, 8: 20, 9: 21, 10: 21, 11: 11, 12: 8, 13: 8, 14: 5, 15: 9, 16: 6, 17: 11, 18: 4, 19: 10, 20: 2, 21: 2, 22: 6, 23: 4, 24: 3, 25: 1, 26: 3, 27: 2, 28: 4, 29: 1, 30: 1, 31: 2, 32: 1, 33: 1, 34: 3, 36: 1, 38: 2, 40: 2, 41: 1, 44: 1, 45: 1, 48: 1, 49: 1, 50: 1, 53: 2, 54: 1, 56: 1, 66: 1, 70: 1, 75: 1, 81: 1, 88: 1, 101: 1, 117: 1}\n"
     ]
    }
   ],
   "source": [
    "# Reduce public accessions to one by species, with priority to Refseq\n",
    "NCBI_pt = refseq_pt.sort_values('length',ascending=False).groupby('sci_name').head(1)\n",
    "NCBI_genus = NCBI_pt.groupby('genus').genus.count().to_frame().rename(columns={'genus':'N'}).reset_index()\n",
    "print(NCBI_genus.groupby('N').size().to_dict())\n",
    "NCBI_pt = pd.concat([NCBI_pt, insdc_pt[(insdc_pt.Sample_Name.isin(NCBI_pt.Sample_Name)==False)] ])\n",
    "NCBI_pt = NCBI_pt.groupby('sci_name').head(1)\n",
    "NCBI_genus = NCBI_pt.groupby('genus').genus.count().to_frame().rename(columns={'genus':'N'}).reset_index()\n",
    "print(NCBI_genus.groupby('N').size().to_dict())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "aging-madonna",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "666 665\n"
     ]
    }
   ],
   "source": [
    "# Li et al only for species not in REFSEQ/INSDC and one by species\n",
    "Lietal_pt = Lietal_pt[(Lietal_pt.sci_name.isin(NCBI_pt.sci_name)==False)]\\\n",
    "            .sort_values('length',ascending=False).groupby('sci_name').head(1)\n",
    "print(Lietal_pt.shape[0],Lietal_pt.sci_name.nunique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "seasonal-elizabeth",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "738 738\n"
     ]
    }
   ],
   "source": [
    "# 1KP one by species\n",
    "okp_pt = okp_pt.sort_values('length',ascending=False).groupby('sci_name').head(1)\n",
    "print(okp_pt.shape[0],okp_pt.sci_name.nunique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "comprehensive-emperor",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20367 s: 17226 g: 8507\n"
     ]
    }
   ],
   "source": [
    "# Merge lists of samples\n",
    "col_keep=['Sample_Name','genus','species','sci_name','DataSource','length']\n",
    "all_pt = pd.concat([ paftol_pt[col_keep], sra_pt[col_keep], gap_pt[col_keep], okp_pt[col_keep], Lietal_pt[col_keep],\n",
    "                    NCBI_pt[col_keep], gymno_pt[col_keep]], ignore_index=True)\n",
    "print(all_pt.shape[0],'s:',all_pt.sci_name.nunique(),'g:',all_pt.genus.nunique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "smooth-qatar",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20 o: 5 f: 9 g: 17 s: 20\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>Sample_Name</th>\n",
       "      <th>TaxID</th>\n",
       "      <th>mol_type</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>sci_name</th>\n",
       "      <th>DataSource</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Cycadales</td>\n",
       "      <td>Cycadaceae</td>\n",
       "      <td>XZUY</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cycas</td>\n",
       "      <td>micholitzii</td>\n",
       "      <td>Cycas micholitzii</td>\n",
       "      <td>OneKP</td>\n",
       "      <td>36702.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Ginkgoales</td>\n",
       "      <td>Ginkgoaceae</td>\n",
       "      <td>SGTW</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ginkgo</td>\n",
       "      <td>biloba</td>\n",
       "      <td>Ginkgo biloba</td>\n",
       "      <td>OneKP</td>\n",
       "      <td>36327.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        order       family Sample_Name  TaxID mol_type   genus      species  \\\n",
       "0   Cycadales   Cycadaceae        XZUY    NaN      NaN   Cycas  micholitzii   \n",
       "1  Ginkgoales  Ginkgoaceae        SGTW    NaN      NaN  Ginkgo       biloba   \n",
       "\n",
       "            sci_name DataSource   length  \n",
       "0  Cycas micholitzii      OneKP  36702.0  \n",
       "1      Ginkgo biloba      OneKP  36327.0  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Define root samples\n",
    "root_1kp = pd.read_table('../Phylogenetic_Tree/Release_1.0/OneKP_root.txt',sep=' ',header=None).rename(columns={\n",
    "    0:'order',1:'family',2:'genus',3:'species',4:'Sample_Name'})\n",
    "root = pd.concat([root_1kp,gymno_pt]).drop(columns=['genus','species','sci_name','DataSource','length'])\n",
    "root = pd.merge(root,all_pt,on='Sample_Name',how='inner')\n",
    "print(root.shape[0], 'o:',root.order.nunique(),'f:',root.family.nunique(),'g:',root.genus.nunique(),'s:',root.sci_name.nunique())\n",
    "root[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "classified-vietnamese",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'all_pt' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-12-11f05daa5bd0>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      2\u001b[0m \u001b[0mwcvp\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'../PAFTOL_DB/wcvp_genus.csv'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m;\u001b[0m \u001b[0mwcvp\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcolumns\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mmap\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mstr\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mlower\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mwcvp\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      3\u001b[0m \u001b[1;31m# wcvp = wcvp.sort_values('nentries',ascending=False).groupby('genus').head(1).sort_values('genus')\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 4\u001b[1;33m \u001b[0mall_pt\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mmerge\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mall_pt\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mwcvp\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mhow\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m'inner'\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mon\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m'genus'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      5\u001b[0m \u001b[0mprint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mall_pt\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mgroupby\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'class'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0msize\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mto_dict\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      6\u001b[0m \u001b[0mall_pt\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mconcat\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m[\u001b[0m \u001b[0mall_pt\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mroot\u001b[0m \u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mNameError\u001b[0m: name 'all_pt' is not defined"
     ]
    }
   ],
   "source": [
    "# Add class, order and genus info from WCVP\n",
    "wcvp = pd.read_csv('../PAFTOL_DB/wcvp_genus.csv'); wcvp.columns = map(str.lower, wcvp.columns)\n",
    "# wcvp = wcvp.sort_values('nentries',ascending=False).groupby('genus').head(1).sort_values('genus')\n",
    "all_pt = pd.merge(all_pt,wcvp,how='inner',on='genus')\n",
    "print(all_pt.groupby('class').size().to_dict())\n",
    "all_pt = pd.concat([ all_pt, root ])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "conceptual-police",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "o: 69 f: 426 g: 8405 s: 17106\n",
      "{'GAP': 772, 'Gymno': 10, 'INSDC': 2905, 'Lietal': 665, 'OneKP': 671, 'PAFTOL': 8343, 'REFSEQ': 5169, 'SRA': 1656}\n",
      "{'Sample_Name': 0, 'genus': 0, 'species': 103, 'sci_name': 2, 'DataSource': 0, 'length': 0, 'order': 0, 'class': 20, 'family': 0, 'nentries': 20, 'TaxID': 20181, 'mol_type': 20181}\n",
      "20191 > 20170 > 20065\n",
      "{'Sample_Name': 0, 'genus': 0, 'species': 0, 'sci_name': 0, 'DataSource': 0, 'length': 0, 'order': 0, 'class': 20, 'family': 0, 'nentries': 20, 'TaxID': 20055, 'mol_type': 20055}\n",
      "{'GAP': 772, 'Gymno': 10, 'INSDC': 2851, 'Lietal': 624, 'OneKP': 671, 'PAFTOL': 8339, 'REFSEQ': 5162, 'SRA': 1636}\n",
      "o: 69 f: 426 g: 8396 s: 17016\n"
     ]
    }
   ],
   "source": [
    "print('o:',all_pt.order.nunique(),'f:',all_pt.family.nunique(),'g:',all_pt.genus.nunique(),'s:',all_pt.sci_name.nunique())\n",
    "print(all_pt.groupby('DataSource').size().to_dict())\n",
    "print(all_pt.isna().sum().to_dict())\n",
    "print(all_pt.shape[0],end=' > ')\n",
    "all_pt = all_pt.groupby('Sample_Name').head(1) # Make sure IDs are unique\n",
    "print(all_pt.shape[0],end=' > ')\n",
    "all_pt = all_pt[(all_pt.order.notnull()) & (all_pt.family.notnull()) & (all_pt.genus!='NILL')\n",
    "               & (all_pt.species.notnull()) & (all_pt.species!='NILL')]\n",
    "print(all_pt.shape[0])\n",
    "print(all_pt.isna().sum().to_dict())\n",
    "print(all_pt.groupby('DataSource').size().to_dict())\n",
    "all_pt = all_pt.sort_values(['DataSource','order','family','genus']).reset_index(drop=True)\n",
    "print('o:',all_pt.order.nunique(),'f:',all_pt.family.nunique(),'g:',all_pt.genus.nunique(),'s:',all_pt.sci_name.nunique())\n",
    "# Make list\n",
    "all_pt.to_csv(project + '/' + project + '_preValidation.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "waiting-france",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "64 / 64 orders\n",
      "417 / 417 families\n",
      "missing families: []\n",
      "8379 / 13549 genera\n"
     ]
    }
   ],
   "source": [
    "# wcvp = wcvp[wcvp['class']=='Magnoliopsida']\n",
    "print(wcvp[wcvp.order.isin(all_pt.order)].order.nunique(),'/',wcvp.order.nunique(),'orders')\n",
    "print(wcvp[wcvp.family.isin(all_pt.family)].family.nunique(),'/',wcvp.family.nunique(),'families')\n",
    "print('missing families:',wcvp[wcvp.family.isin(all_pt.family)==False].family.unique())\n",
    "print(wcvp[wcvp.genus.isin(all_pt.genus)].genus.nunique(),'/',wcvp.genus.nunique(),'genera')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "portuguese-cricket",
   "metadata": {},
   "source": [
    "## Pre-Validation. Remove Samples that failed Barcode Validation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "standing-athletics",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20055\n",
      "Barcode Validation: {'Confirmed': 17918, 'Inconclusive': 1917, 'Rejected': 220}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>sci_name</th>\n",
       "      <th>DataSource</th>\n",
       "      <th>length</th>\n",
       "      <th>order</th>\n",
       "      <th>class</th>\n",
       "      <th>family</th>\n",
       "      <th>nentries</th>\n",
       "      <th>...</th>\n",
       "      <th>VATfam</th>\n",
       "      <th>VATgen</th>\n",
       "      <th>Vfam_pc</th>\n",
       "      <th>Vgen_pc</th>\n",
       "      <th>best matching family</th>\n",
       "      <th>N best family</th>\n",
       "      <th>best matching genus</th>\n",
       "      <th>N best genus</th>\n",
       "      <th>Validation</th>\n",
       "      <th>Validation_genus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GAP_026547</td>\n",
       "      <td>Albidella</td>\n",
       "      <td>oligococca</td>\n",
       "      <td>Albidella oligococca</td>\n",
       "      <td>GAP</td>\n",
       "      <td>59917</td>\n",
       "      <td>Alismatales</td>\n",
       "      <td>monocots</td>\n",
       "      <td>Alismataceae</td>\n",
       "      <td>4.0</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>83.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>Butomaceae</td>\n",
       "      <td>1</td>\n",
       "      <td>Ranalisma</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Confirmed</td>\n",
       "      <td>Inconclusive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GAP_028289</td>\n",
       "      <td>Astonia</td>\n",
       "      <td>australiensis</td>\n",
       "      <td>Astonia australiensis</td>\n",
       "      <td>GAP</td>\n",
       "      <td>8956</td>\n",
       "      <td>Alismatales</td>\n",
       "      <td>monocots</td>\n",
       "      <td>Alismataceae</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Confirmed</td>\n",
       "      <td>Inconclusive</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Sample      genus        species               sci_name DataSource  \\\n",
       "0  GAP_026547  Albidella     oligococca   Albidella oligococca        GAP   \n",
       "1  GAP_028289    Astonia  australiensis  Astonia australiensis        GAP   \n",
       "\n",
       "   length        order     class        family  nentries  ...  VATfam  VATgen  \\\n",
       "0   59917  Alismatales  monocots  Alismataceae       4.0  ...       5       1   \n",
       "1    8956  Alismatales  monocots  Alismataceae       1.0  ...       1       0   \n",
       "\n",
       "   Vfam_pc  Vgen_pc  best matching family  N best family  best matching genus  \\\n",
       "0     83.0     50.0            Butomaceae              1            Ranalisma   \n",
       "1    100.0      NaN                   NaN              0                  NaN   \n",
       "\n",
       "   N best genus  Validation Validation_genus  \n",
       "0           1.0   Confirmed     Inconclusive  \n",
       "1           NaN   Confirmed     Inconclusive  \n",
       "\n",
       "[2 rows x 25 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "BV = pd.read_excel(project + '/' + project + '_Barcode_Validation.xlsx',sheet_name='Validation_Results')\n",
    "print(BV.shape[0])\n",
    "print('Barcode Validation:',BV.groupby('Validation').size().to_dict())\n",
    "BV[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "civil-messenger",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19835\n"
     ]
    }
   ],
   "source": [
    "valid_pt = all_pt[all_pt.Sample_Name.isin(BV[BV.Validation.isin(['Confirmed','Inconclusive'])].Sample)]\n",
    "print(valid_pt.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "reserved-narrative",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\kle10kg\\AppData\\Roaming\\Python\\Python37\\site-packages\\ipykernel_launcher.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  This is separate from the ipykernel package so we can avoid doing imports until\n",
      "C:\\ProgramData\\Anaconda3\\lib\\site-packages\\pandas\\core\\generic.py:5491: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self[name] = value\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Sample_Name': 0, 'genus': 0, 'species': 0, 'sci_name': 0, 'DataSource': 0, 'length': 0, 'order': 0, 'class': 10, 'family': 0, 'nentries': 10, 'TaxID': 19835, 'mol_type': 19835, 'Label': 0}\n"
     ]
    }
   ],
   "source": [
    "# Make list and labels\n",
    "valid_pt['Label'] = valid_pt.order + '_' + valid_pt.family + '_' + valid_pt.genus + '_' + valid_pt.species.str.replace(' ','-') + '_' \\\n",
    "     + valid_pt.Sample_Name.str.replace('_','')\n",
    "valid_pt.Label=valid_pt.Label.str.replace(' ','')#.str.replace('.','_')\n",
    "print(valid_pt.isna().sum().to_dict())\n",
    "valid_pt.to_csv(project + '/' + project + '_pretree.csv',index=False)\n",
    "valid_pt[['Sample_Name','Label']].to_csv(project + '/' + project + '_pretree_labels.txt',index=False,header=None,sep=' ')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "transsexual-destruction",
   "metadata": {},
   "source": [
    "# Tree samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "jewish-accused",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19835 o: 69 f: 424 g: 8319 s: 16865\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_Name</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>sci_name</th>\n",
       "      <th>DataSource</th>\n",
       "      <th>length</th>\n",
       "      <th>order</th>\n",
       "      <th>class</th>\n",
       "      <th>family</th>\n",
       "      <th>nentries</th>\n",
       "      <th>TaxID</th>\n",
       "      <th>mol_type</th>\n",
       "      <th>Label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GAP_026547</td>\n",
       "      <td>Albidella</td>\n",
       "      <td>oligococca</td>\n",
       "      <td>Albidella oligococca</td>\n",
       "      <td>GAP</td>\n",
       "      <td>59917.0</td>\n",
       "      <td>Alismatales</td>\n",
       "      <td>monocots</td>\n",
       "      <td>Alismataceae</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Alismatales_Alismataceae_Albidella_oligococca_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GAP_028289</td>\n",
       "      <td>Astonia</td>\n",
       "      <td>australiensis</td>\n",
       "      <td>Astonia australiensis</td>\n",
       "      <td>GAP</td>\n",
       "      <td>8956.0</td>\n",
       "      <td>Alismatales</td>\n",
       "      <td>monocots</td>\n",
       "      <td>Alismataceae</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Alismatales_Alismataceae_Astonia_australiensis...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Sample_Name      genus        species               sci_name DataSource  \\\n",
       "0  GAP_026547  Albidella     oligococca   Albidella oligococca        GAP   \n",
       "1  GAP_028289    Astonia  australiensis  Astonia australiensis        GAP   \n",
       "\n",
       "    length        order     class        family  nentries  TaxID  mol_type  \\\n",
       "0  59917.0  Alismatales  monocots  Alismataceae       4.0    NaN       NaN   \n",
       "1   8956.0  Alismatales  monocots  Alismataceae       1.0    NaN       NaN   \n",
       "\n",
       "                                               Label  \n",
       "0  Alismatales_Alismataceae_Albidella_oligococca_...  \n",
       "1  Alismatales_Alismataceae_Astonia_australiensis...  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "valid_pt = pd.read_csv(project + '/' + project + '_pretree.csv')\n",
    "print(valid_pt.shape[0], 'o:',valid_pt.order.nunique(),'f:',valid_pt.family.nunique(),\n",
    "      'g:',valid_pt.genus.nunique(),'s:',valid_pt.sci_name.nunique())\n",
    "valid_pt[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "looking-sentence",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "869083 14063 72\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_Name</th>\n",
       "      <th>gene</th>\n",
       "      <th>len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>45619</td>\n",
       "      <td>72</td>\n",
       "      <td>52602</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>45621</td>\n",
       "      <td>72</td>\n",
       "      <td>52564</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Sample_Name  gene    len\n",
       "0       45619    72  52602\n",
       "1       45621    72  52564"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tree_samples = pd.read_csv(project + '/AllSamples_Allgenes_SeqTable.csv')\n",
    "print(tree_samples.shape[0],tree_samples.Sample_Name.nunique(),tree_samples.gene.nunique())\n",
    "tree_samples = tree_samples.groupby('Sample_Name').agg({'gene':'count','len':'sum'}).reset_index()\n",
    "tree_samples[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "excessive-marshall",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "14063 o: 69 f: 419 g: 5500 s: 12571\n",
      "{'GAP': 345, 'INSDC': 2809, 'Lietal': 617, 'OneKP': 663, 'PAFTOL': 3490, 'REFSEQ': 5134, 'SRA': 1005}\n"
     ]
    }
   ],
   "source": [
    "tree_pt = pd.merge(valid_pt,tree_samples,on='Sample_Name',how='inner')\n",
    "print(tree_pt.shape[0], 'o:',tree_pt.order.nunique(),'f:',tree_pt.family.nunique(),\n",
    "      'g:',tree_pt.genus.nunique(),'s:',tree_pt.sci_name.nunique())\n",
    "tree_pt.to_csv(project + '/' + project + '_tree_samples.csv',index=False)\n",
    "print(tree_pt.groupby('DataSource').size().to_dict())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "framed-hearts",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "64 / 64 orders\n",
      "412 / 417 families\n",
      "missing families: ['Siparunaceae' 'Rafflesiaceae' 'Apodanthaceae' 'Tetracarpaeaceae'\n",
      " 'Tiganophytaceae']\n",
      "5490 / 13549 genera\n"
     ]
    }
   ],
   "source": [
    "print(wcvp[wcvp.order.isin(tree_pt.order)].order.nunique(),'/',wcvp.order.nunique(),'orders')\n",
    "print(wcvp[wcvp.family.isin(tree_pt.family)].family.nunique(),'/',wcvp.family.nunique(),'families')\n",
    "print('missing families:',wcvp[wcvp.family.isin(tree_pt.family)==False].family.unique())\n",
    "print(wcvp[wcvp.genus.isin(tree_pt.genus)].genus.nunique(),'/',wcvp.genus.nunique(),'genera')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "selected-ukraine",
   "metadata": {},
   "source": [
    "## Monophyly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "welcome-performer",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from ete3 import Tree, NodeStyle, TreeStyle, TextFace, AttrFace, faces\n",
    "import warnings; warnings.filterwarnings('ignore')\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "lightweight-china",
   "metadata": {},
   "outputs": [],
   "source": [
    "def Treetaxo(df):\n",
    "    print('N:',df.shape[0],'o:',df.order.nunique(),'f:',df.family.nunique(),'g:',df.genus.nunique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "optimum-ferry",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "14488\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>sci_name</th>\n",
       "      <th>DataSource</th>\n",
       "      <th>length</th>\n",
       "      <th>order</th>\n",
       "      <th>class</th>\n",
       "      <th>family</th>\n",
       "      <th>Label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GAP_026547</td>\n",
       "      <td>Albidella</td>\n",
       "      <td>oligococca</td>\n",
       "      <td>Albidella oligococca</td>\n",
       "      <td>GAP</td>\n",
       "      <td>59917</td>\n",
       "      <td>Alismatales</td>\n",
       "      <td>monocots</td>\n",
       "      <td>Alismataceae</td>\n",
       "      <td>Alismatales_Alismataceae_Albidella_oligococca_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GAP_028289</td>\n",
       "      <td>Astonia</td>\n",
       "      <td>australiensis</td>\n",
       "      <td>Astonia australiensis</td>\n",
       "      <td>GAP</td>\n",
       "      <td>8956</td>\n",
       "      <td>Alismatales</td>\n",
       "      <td>monocots</td>\n",
       "      <td>Alismataceae</td>\n",
       "      <td>Alismatales_Alismataceae_Astonia_australiensis...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Sample      genus        species               sci_name DataSource  \\\n",
       "0  GAP_026547  Albidella     oligococca   Albidella oligococca        GAP   \n",
       "1  GAP_028289    Astonia  australiensis  Astonia australiensis        GAP   \n",
       "\n",
       "   length        order     class        family  \\\n",
       "0   59917  Alismatales  monocots  Alismataceae   \n",
       "1    8956  Alismatales  monocots  Alismataceae   \n",
       "\n",
       "                                               Label  \n",
       "0  Alismatales_Alismataceae_Albidella_oligococca_...  \n",
       "1  Alismatales_Alismataceae_Astonia_australiensis...  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Release_df = pd.read_csv(project + '/' + project + '.csv').rename(columns={'Sample_Name':'Sample'})\n",
    "print(Release_df.shape[0])\n",
    "Release_df[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "global-player",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8330\n",
      "Phylogenetic Validation: {'Confirmed': 8162, 'Inconclusive': 56, 'Rejected': 112}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Tidx</th>\n",
       "      <th>Label</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>Sample</th>\n",
       "      <th>Phylogenetic_Validation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>Apiales_Araliaceae_Motherwellia_haplosciadea_G...</td>\n",
       "      <td>Apiales</td>\n",
       "      <td>Araliaceae</td>\n",
       "      <td>Motherwellia</td>\n",
       "      <td>haplosciadea</td>\n",
       "      <td>GAP022353</td>\n",
       "      <td>Inconclusive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>Apiales_Pittosporaceae_Auranticarpa_rhombifoli...</td>\n",
       "      <td>Apiales</td>\n",
       "      <td>Pittosporaceae</td>\n",
       "      <td>Auranticarpa</td>\n",
       "      <td>rhombifolia</td>\n",
       "      <td>GAP026913</td>\n",
       "      <td>Confirmed</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Tidx                                              Label    order  \\\n",
       "0     0  Apiales_Araliaceae_Motherwellia_haplosciadea_G...  Apiales   \n",
       "1     1  Apiales_Pittosporaceae_Auranticarpa_rhombifoli...  Apiales   \n",
       "\n",
       "           family         genus       species     Sample  \\\n",
       "0      Araliaceae  Motherwellia  haplosciadea  GAP022353   \n",
       "1  Pittosporaceae  Auranticarpa   rhombifolia  GAP026913   \n",
       "\n",
       "  Phylogenetic_Validation  \n",
       "0            Inconclusive  \n",
       "1               Confirmed  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PV = pd.read_csv(project + '/' + project + '_0.2_iqtree_labelled_table.csv').reset_index().rename(columns={\n",
    "    'index':'Tidx','label':'Label'})\n",
    "# PV['Sample'] = PV['Sample'].str.replace('GAP','GAP_').str.replace('PAFTOL','PAFTOL_').str.replace('NC','NC_').str.replace('AC','AC_')\n",
    "print(PV.shape[0])\n",
    "print('Phylogenetic Validation:',PV.groupby('Phylogenetic_Validation').size().to_dict())\n",
    "PV[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "subsequent-appeal",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8329\n"
     ]
    }
   ],
   "source": [
    "Release_df = pd.merge(left = Release_df, right = PV[['Label','Phylogenetic_Validation','Tidx']].rename(columns={\n",
    "    'Phylogenetic_Validation':'PhylogeneticValidation'}),\n",
    "                     how = 'inner', on = 'Label')\n",
    "Release_df = pd.merge(left = Release_df, right = BV[['Sample','Validation']].rename(columns={'Validation':'BarcodeValidation'}),\n",
    "                     how = 'left', on = 'Sample').sort_values('Tidx').reset_index(drop=True)\n",
    "print(Release_df.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "typical-durham",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Sample': 0, 'genus': 0, 'species': 0, 'sci_name': 0, 'DataSource': 0, 'length': 0, 'order': 0, 'class': 20, 'family': 0, 'Label': 0, 'PhylogeneticValidation': 0, 'Tidx': 0, 'BarcodeValidation': 10}\n",
      "      Tidx                                         Label       order  \\\n",
      "3071  3071  Sapindales_Rutaceae_Citrus_Ã—-aurantium_UHJR  Sapindales   \n",
      "\n",
      "        family   genus       species Sample Phylogenetic_Validation  \n",
      "3071  Rutaceae  Citrus  Ã—-aurantium   UHJR               Confirmed  \n"
     ]
    }
   ],
   "source": [
    "print(Release_df.isna().sum().to_dict())\n",
    "print(PV[PV.Label.isin(Release_df.Label)==False])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "injured-video",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>BarcodeValidation</th>\n",
       "      <th>Confirmed</th>\n",
       "      <th>Inconclusive</th>\n",
       "      <th>Rejected</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PhylogeneticValidation</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Confirmed</th>\n",
       "      <td>8026</td>\n",
       "      <td>89</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Inconclusive</th>\n",
       "      <td>50</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rejected</th>\n",
       "      <td>16</td>\n",
       "      <td>1</td>\n",
       "      <td>95</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "BarcodeValidation       Confirmed  Inconclusive  Rejected\n",
       "PhylogeneticValidation                                   \n",
       "Confirmed                    8026            89        37\n",
       "Inconclusive                   50             2         3\n",
       "Rejected                       16             1        95"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Release_df['Decision']=np.nan\n",
    "Release_df.loc[(Release_df.PhylogeneticValidation=='Confirmed') & (Release_df.BarcodeValidation=='Confirmed'), \n",
    "           'Decision']='Include'\n",
    "Release_df.loc[(Release_df.PhylogeneticValidation=='Confirmed') & (Release_df.BarcodeValidation=='Inconclusive'), \n",
    "           'Decision']='Include'\n",
    "Release_df.loc[(Release_df.PhylogeneticValidation=='Confirmed') & (Release_df.BarcodeValidation=='Rejected'), \n",
    "           'Decision']='Review'\n",
    "\n",
    "Release_df.loc[(Release_df.PhylogeneticValidation=='Inconclusive') & (Release_df.BarcodeValidation=='Confirmed'), \n",
    "           'Decision']='Review'\n",
    "Release_df.loc[(Release_df.PhylogeneticValidation=='Inconclusive') & (Release_df.BarcodeValidation=='Inconclusive'), \n",
    "           'Decision']='Review'\n",
    "Release_df.loc[(Release_df.PhylogeneticValidation=='Inconclusive') & (Release_df.BarcodeValidation=='Rejected'), \n",
    "           'Decision']='Exclude'\n",
    "\n",
    "Release_df.loc[(Release_df.PhylogeneticValidation=='Rejected') & (Release_df.BarcodeValidation=='Confirmed'), \n",
    "           'Decision']='Review'\n",
    "Release_df.loc[(Release_df.PhylogeneticValidation=='Rejected') & (Release_df.BarcodeValidation=='Inconclusive'), \n",
    "           'Decision']='Exclude'\n",
    "Release_df.loc[(Release_df.PhylogeneticValidation=='Rejected') & (Release_df.BarcodeValidation=='Rejected'), \n",
    "           'Decision']='Exclude'\n",
    "# barcode validation against phylogenetic validation\n",
    "Release_df.groupby(['PhylogeneticValidation','BarcodeValidation']).size().to_frame().reset_index().rename(columns={0:'Count'})\\\n",
    "    .pivot(index='PhylogeneticValidation', columns='BarcodeValidation', values='Count').to_clipboard()\n",
    "Release_df.groupby(['PhylogeneticValidation','BarcodeValidation']).size().to_frame().reset_index().rename(columns={0:'Count'})\\\n",
    "    .pivot(index='PhylogeneticValidation', columns='BarcodeValidation', values='Count')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "federal-avenue",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "N: 8329 o: 72 f: 396 g: 4406\n",
      "{'GAP': 163, 'Gymno': 10, 'INSDC': 326, 'Lietal': 902, 'OneKP': 666, 'PAFTOL': 2507, 'REFSEQ': 3118, 'SRA': 637}\n",
      "{'Exclude': 99, 'Include': 8115, 'Review': 105}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>sci_name</th>\n",
       "      <th>DataSource</th>\n",
       "      <th>length</th>\n",
       "      <th>order</th>\n",
       "      <th>class</th>\n",
       "      <th>family</th>\n",
       "      <th>Label</th>\n",
       "      <th>PhylogeneticValidation</th>\n",
       "      <th>Tidx</th>\n",
       "      <th>BarcodeValidation</th>\n",
       "      <th>Decision</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GAP_022353</td>\n",
       "      <td>Motherwellia</td>\n",
       "      <td>haplosciadea</td>\n",
       "      <td>Motherwellia haplosciadea</td>\n",
       "      <td>GAP</td>\n",
       "      <td>120594</td>\n",
       "      <td>Apiales</td>\n",
       "      <td>campanulids</td>\n",
       "      <td>Araliaceae</td>\n",
       "      <td>Apiales_Araliaceae_Motherwellia_haplosciadea_G...</td>\n",
       "      <td>Inconclusive</td>\n",
       "      <td>0</td>\n",
       "      <td>Confirmed</td>\n",
       "      <td>Review</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GAP_026913</td>\n",
       "      <td>Auranticarpa</td>\n",
       "      <td>rhombifolia</td>\n",
       "      <td>Auranticarpa rhombifolia</td>\n",
       "      <td>GAP</td>\n",
       "      <td>79637</td>\n",
       "      <td>Apiales</td>\n",
       "      <td>campanulids</td>\n",
       "      <td>Pittosporaceae</td>\n",
       "      <td>Apiales_Pittosporaceae_Auranticarpa_rhombifoli...</td>\n",
       "      <td>Confirmed</td>\n",
       "      <td>1</td>\n",
       "      <td>Confirmed</td>\n",
       "      <td>Include</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Sample         genus       species                   sci_name  \\\n",
       "0  GAP_022353  Motherwellia  haplosciadea  Motherwellia haplosciadea   \n",
       "1  GAP_026913  Auranticarpa   rhombifolia   Auranticarpa rhombifolia   \n",
       "\n",
       "  DataSource  length    order        class          family  \\\n",
       "0        GAP  120594  Apiales  campanulids      Araliaceae   \n",
       "1        GAP   79637  Apiales  campanulids  Pittosporaceae   \n",
       "\n",
       "                                               Label PhylogeneticValidation  \\\n",
       "0  Apiales_Araliaceae_Motherwellia_haplosciadea_G...           Inconclusive   \n",
       "1  Apiales_Pittosporaceae_Auranticarpa_rhombifoli...              Confirmed   \n",
       "\n",
       "   Tidx BarcodeValidation Decision  \n",
       "0     0         Confirmed   Review  \n",
       "1     1         Confirmed  Include  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Treetaxo(Release_df)\n",
    "print(Release_df.groupby('DataSource').size().to_dict())\n",
    "print(Release_df.groupby('Decision').size().to_dict())\n",
    "Release_df[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "prescription-bench",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "N: 8309 o: 64 f: 387 g: 4389\n"
     ]
    }
   ],
   "source": [
    "root_orders = ['Pinales','Gnetales','Cupressales','Ginkgoales','Ephedrales','Cupressales','Araucariales','Cycadales','Taxales']\n",
    "Treetaxo(Release_df[Release_df.order.isin(root_orders)==False])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "fatal-longitude",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "% of Exclusion or Review by source\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "DataSource\n",
       "GAP         3.1\n",
       "Gymno     100.0\n",
       "INSDC       0.6\n",
       "Lietal      1.7\n",
       "OneKP       2.1\n",
       "PAFTOL      4.5\n",
       "REFSEQ      1.3\n",
       "SRA         2.2\n",
       "dtype: float64"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print('% of Exclusion or Review by source')\n",
    "round(Release_df[Release_df.Decision!='Include'].groupby('DataSource').size()/Release_df.groupby('DataSource').size()*100,1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "olympic-article",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Output file for review process\n",
    "review_columns=['Sample', 'Label', 'DataSource', 'order', 'family', 'genus', 'species','length',\n",
    "                 'PhylogeneticValidation', 'BarcodeValidation','Decision']\n",
    "review_df = pd.concat( [Release_df[review_columns],\n",
    "                    pd.DataFrame(columns=['DecisionReview','DecisionReason','ValidationComments']) ] )\n",
    "review_df.to_excel(project + '/' + project + '_Review.xlsx',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "approved-people",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Tree for review\n",
    "# https://www.biostars.org/p/172568/\n",
    "from ete3 import Tree, NodeStyle, TreeStyle, TextFace, AttrFace, faces\n",
    "\n",
    "tree_file = project + '/' + project + '_0.2_iqtree_labelled_rooted.nwk'\n",
    "T=Tree(tree_file)\n",
    "#Convert to ultrametric\n",
    "most_distant_leaf, tree_length = T.get_farthest_leaf()\n",
    "current_dist = 0\n",
    "for postorder, node in T.iter_prepostorder():\n",
    "    if postorder:\n",
    "        current_dist -= node.dist\n",
    "    else:\n",
    "        if node.is_leaf():\n",
    "            node.dist += tree_length - (current_dist + node.dist)\n",
    "        elif node.up: # node is internal\n",
    "            current_dist += node.dist\n",
    "            \n",
    "ts = TreeStyle()\n",
    "# # ts.show_branch_length = False\n",
    "# # ts.scale = 2\n",
    "# # ts.mode = \"r\"\n",
    "# ts.show_branch_support = True\n",
    "# ts.layout_fn = layout\n",
    "ts.show_leaf_name = True\n",
    "            \n",
    "for n in T.iter_leaves():\n",
    "#     print(n.name, end=',')\n",
    "    nstyle = NodeStyle()\n",
    "    print(n.name, end=',')\n",
    "    if n.name == '':\n",
    "        nstyle[\"size\"] = 0\n",
    "        n.set_style(nstyle)\n",
    "    else:\n",
    "        tmp=Release_df[Release_df.Label==n.name]\n",
    "#         print(tmp)\n",
    "\n",
    "        ## FIX point style should be based on Barcode Validation\n",
    "        if tmp.shape[0]>0:\n",
    "            if (tmp.Decision.isin(['Include']).bool()==True):\n",
    "                nstyle[\"fgcolor\"] = \"green\"\n",
    "                nstyle[\"size\"] = 10\n",
    "            elif (tmp.Decision.isin(['Exclude']).bool()==True):\n",
    "                nstyle[\"fgcolor\"] = \"red\"\n",
    "                nstyle[\"size\"] = 10\n",
    "            elif (tmp.Decision.isin(['Review']).bool()==True):\n",
    "                nstyle[\"fgcolor\"] = \"black\"\n",
    "                nstyle[\"size\"] = 10\n",
    "            else:\n",
    "                nstyle[\"fgcolor\"] = \"blue\"\n",
    "                nstyle[\"size\"] = 8\n",
    "\n",
    "            if (tmp.BarcodeValidation.isin(['Confirmed']).bool()==True):\n",
    "                nstyle[\"shape\"] = \"circle\"\n",
    "            elif (tmp.BarcodeValidation.isin(['Rejected']).bool()==True):\n",
    "                nstyle[\"shape\"] = \"circle\"\n",
    "            elif (tmp.BarcodeValidation.isin(['Inconclusive']).bool()==True):\n",
    "                nstyle[\"shape\"] = \"square\"\n",
    "            \n",
    "            n.set_style(nstyle)\n",
    "        else:\n",
    "            nstyle[\"size\"] = 0\n",
    "            n.set_style(nstyle)\n",
    "T.render(project + '/' + project + \"_pretree.svg\", w=150, units=\"mm\",dpi=150,tree_style=ts);"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "czech-rider",
   "metadata": {},
   "source": [
    "## Final tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "sublime-lafayette",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8329\n",
      "N: 8329 o: 72 f: 396 g: 4406\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Label</th>\n",
       "      <th>DataSource</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>length</th>\n",
       "      <th>PhylogeneticValidation</th>\n",
       "      <th>BarcodeValidation</th>\n",
       "      <th>Decision</th>\n",
       "      <th>DecisionReview</th>\n",
       "      <th>DecisionReason</th>\n",
       "      <th>ValidationComments</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PAFTOL_008669</td>\n",
       "      <td>Commelinales_Commelinaceae_Floscopa_glomerata_...</td>\n",
       "      <td>PAFTOL</td>\n",
       "      <td>Commelinales</td>\n",
       "      <td>Commelinaceae</td>\n",
       "      <td>Floscopa</td>\n",
       "      <td>glomerata</td>\n",
       "      <td>40762</td>\n",
       "      <td>Confirmed</td>\n",
       "      <td>Rejected</td>\n",
       "      <td>Review</td>\n",
       "      <td>Include</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GAP_026913</td>\n",
       "      <td>Apiales_Pittosporaceae_Auranticarpa_rhombifoli...</td>\n",
       "      <td>GAP</td>\n",
       "      <td>Apiales</td>\n",
       "      <td>Pittosporaceae</td>\n",
       "      <td>Auranticarpa</td>\n",
       "      <td>rhombifolia</td>\n",
       "      <td>79637</td>\n",
       "      <td>Confirmed</td>\n",
       "      <td>Confirmed</td>\n",
       "      <td>Include</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Sample                                              Label  \\\n",
       "0  PAFTOL_008669  Commelinales_Commelinaceae_Floscopa_glomerata_...   \n",
       "1     GAP_026913  Apiales_Pittosporaceae_Auranticarpa_rhombifoli...   \n",
       "\n",
       "  DataSource         order          family         genus      species  length  \\\n",
       "0     PAFTOL  Commelinales   Commelinaceae      Floscopa    glomerata   40762   \n",
       "1        GAP       Apiales  Pittosporaceae  Auranticarpa  rhombifolia   79637   \n",
       "\n",
       "  PhylogeneticValidation BarcodeValidation Decision DecisionReview  \\\n",
       "0              Confirmed          Rejected   Review        Include   \n",
       "1              Confirmed         Confirmed  Include            NaN   \n",
       "\n",
       "   DecisionReason ValidationComments  \n",
       "0             NaN                NaN  \n",
       "1             NaN                NaN  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read review file\n",
    "Review = pd.read_excel(project + '/' + project + '_pretree_Review.xlsx')\n",
    "print(Review.shape[0])\n",
    "Treetaxo(Review)\n",
    "Review[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "desirable-bunch",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "N: 8222 o: 72 f: 396 g: 4346\n",
      "N: 8202 o: 64 f: 387 g: 4329\n",
      "{'Sample': 0, 'Label': 0, 'DataSource': 0, 'order': 0, 'family': 0, 'genus': 0, 'species': 0, 'length': 0, 'PhylogeneticValidation': 0, 'BarcodeValidation': 10, 'Decision': 10, 'DecisionReview': 8115, 'DecisionReason': 8222, 'ValidationComments': 8162}\n",
      "DataSource\n",
      "GAP        2.0\n",
      "Gymno      0.1\n",
      "INSDC      3.9\n",
      "Lietal    10.9\n",
      "OneKP      8.1\n",
      "PAFTOL    29.5\n",
      "REFSEQ    37.8\n",
      "SRA        7.7\n",
      "dtype: float64\n"
     ]
    }
   ],
   "source": [
    "Review = Review[(Review.Decision=='Include') | (Review.DecisionReview=='Include')]\n",
    "Treetaxo(Review)\n",
    "root_orders = ['Pinales','Gnetales','Cupressales','Ginkgoales','Ephedrales','Cupressales','Araucariales','Cycadales','Taxales']\n",
    "Treetaxo(Review[Review.order.isin(root_orders)==False])\n",
    "print(Review.isna().sum().to_dict())\n",
    "print(round(Review.groupby('DataSource').size()/Review.shape[0]*100,1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "under-mercy",
   "metadata": {},
   "outputs": [],
   "source": [
    "Review.to_csv(project + '/' + project + '_final.csv',index=False)\n",
    "Review[['Sample','Label']].to_csv(project + '/' + project + '_final_labels.txt',index=False,header=None,sep=' ')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "entertaining-target",
   "metadata": {},
   "source": [
    "## Table of samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "retired-duncan",
   "metadata": {},
   "outputs": [],
   "source": [
    "root_orders = ['Pinales','Gnetales','Cupressales','Ginkgoales','Ephedrales','Cupressales','Araucariales','Cycadales','Taxales']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "mighty-crime",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20045 o: 64 f: 417 g: 8379 s: 16996\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>sci_name</th>\n",
       "      <th>DataSource</th>\n",
       "      <th>length</th>\n",
       "      <th>order</th>\n",
       "      <th>class</th>\n",
       "      <th>family</th>\n",
       "      <th>nentries</th>\n",
       "      <th>TaxID</th>\n",
       "      <th>mol_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GAP_026547</td>\n",
       "      <td>Albidella</td>\n",
       "      <td>oligococca</td>\n",
       "      <td>Albidella oligococca</td>\n",
       "      <td>GAP</td>\n",
       "      <td>59917.0</td>\n",
       "      <td>Alismatales</td>\n",
       "      <td>monocots</td>\n",
       "      <td>Alismataceae</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GAP_028289</td>\n",
       "      <td>Astonia</td>\n",
       "      <td>australiensis</td>\n",
       "      <td>Astonia australiensis</td>\n",
       "      <td>GAP</td>\n",
       "      <td>8956.0</td>\n",
       "      <td>Alismatales</td>\n",
       "      <td>monocots</td>\n",
       "      <td>Alismataceae</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Sample      genus        species               sci_name DataSource  \\\n",
       "0  GAP_026547  Albidella     oligococca   Albidella oligococca        GAP   \n",
       "1  GAP_028289    Astonia  australiensis  Astonia australiensis        GAP   \n",
       "\n",
       "    length        order     class        family  nentries  TaxID mol_type  \n",
       "0  59917.0  Alismatales  monocots  Alismataceae       4.0    NaN      NaN  \n",
       "1   8956.0  Alismatales  monocots  Alismataceae       1.0    NaN      NaN  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pre_val = pd.read_csv(project + '/' + project + '_preValidation.csv').rename(columns={'Sample_Name':'Sample'})\n",
    "pre_val=pre_val[(pre_val.DataSource!='Gymno') & (pre_val.order.isin(root_orders)==False)]\n",
    "tree_taxo(pre_val)\n",
    "pre_val[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "korean-arbitration",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19825 o: 64 f: 417 g: 8309 s: 16855\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>sci_name</th>\n",
       "      <th>DataSource</th>\n",
       "      <th>length</th>\n",
       "      <th>order</th>\n",
       "      <th>class</th>\n",
       "      <th>family</th>\n",
       "      <th>nentries</th>\n",
       "      <th>TaxID</th>\n",
       "      <th>mol_type</th>\n",
       "      <th>Label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GAP_026547</td>\n",
       "      <td>Albidella</td>\n",
       "      <td>oligococca</td>\n",
       "      <td>Albidella oligococca</td>\n",
       "      <td>GAP</td>\n",
       "      <td>59917.0</td>\n",
       "      <td>Alismatales</td>\n",
       "      <td>monocots</td>\n",
       "      <td>Alismataceae</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Alismatales_Alismataceae_Albidella_oligococca_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GAP_028289</td>\n",
       "      <td>Astonia</td>\n",
       "      <td>australiensis</td>\n",
       "      <td>Astonia australiensis</td>\n",
       "      <td>GAP</td>\n",
       "      <td>8956.0</td>\n",
       "      <td>Alismatales</td>\n",
       "      <td>monocots</td>\n",
       "      <td>Alismataceae</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Alismatales_Alismataceae_Astonia_australiensis...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Sample      genus        species               sci_name DataSource  \\\n",
       "0  GAP_026547  Albidella     oligococca   Albidella oligococca        GAP   \n",
       "1  GAP_028289    Astonia  australiensis  Astonia australiensis        GAP   \n",
       "\n",
       "    length        order     class        family  nentries  TaxID  mol_type  \\\n",
       "0  59917.0  Alismatales  monocots  Alismataceae       4.0    NaN       NaN   \n",
       "1   8956.0  Alismatales  monocots  Alismataceae       1.0    NaN       NaN   \n",
       "\n",
       "                                               Label  \n",
       "0  Alismatales_Alismataceae_Albidella_oligococca_...  \n",
       "1  Alismatales_Alismataceae_Astonia_australiensis...  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "post_val = pd.read_csv(project + '/' + project + '_pretree.csv').rename(columns={'Sample_Name':'Sample'})\n",
    "post_val=post_val[(post_val.DataSource!='Gymno') & (post_val.order.isin(root_orders)==False)]\n",
    "tree_taxo(post_val)\n",
    "post_val[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "single-circulation",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "14053 o: 64 f: 412 g: 5490 s: 12561\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>sci_name</th>\n",
       "      <th>DataSource</th>\n",
       "      <th>length</th>\n",
       "      <th>order</th>\n",
       "      <th>class</th>\n",
       "      <th>family</th>\n",
       "      <th>nentries</th>\n",
       "      <th>TaxID</th>\n",
       "      <th>mol_type</th>\n",
       "      <th>Label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GAP_026547</td>\n",
       "      <td>Albidella</td>\n",
       "      <td>oligococca</td>\n",
       "      <td>Albidella oligococca</td>\n",
       "      <td>GAP</td>\n",
       "      <td>59917.0</td>\n",
       "      <td>Alismatales</td>\n",
       "      <td>monocots</td>\n",
       "      <td>Alismataceae</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Alismatales_Alismataceae_Albidella_oligococca_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>GAP_026933</td>\n",
       "      <td>Stuckenia</td>\n",
       "      <td>pectinata</td>\n",
       "      <td>Stuckenia pectinata</td>\n",
       "      <td>GAP</td>\n",
       "      <td>31766.0</td>\n",
       "      <td>Alismatales</td>\n",
       "      <td>monocots</td>\n",
       "      <td>Potamogetonaceae</td>\n",
       "      <td>10.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Alismatales_Potamogetonaceae_Stuckenia_pectina...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Sample      genus     species              sci_name DataSource  \\\n",
       "0  GAP_026547  Albidella  oligococca  Albidella oligococca        GAP   \n",
       "6  GAP_026933  Stuckenia   pectinata   Stuckenia pectinata        GAP   \n",
       "\n",
       "    length        order     class            family  nentries  TaxID  \\\n",
       "0  59917.0  Alismatales  monocots      Alismataceae       4.0    NaN   \n",
       "6  31766.0  Alismatales  monocots  Potamogetonaceae      10.0    NaN   \n",
       "\n",
       "   mol_type                                              Label  \n",
       "0       NaN  Alismatales_Alismataceae_Albidella_oligococca_...  \n",
       "6       NaN  Alismatales_Potamogetonaceae_Stuckenia_pectina...  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tree_samples = pd.read_csv(project + '/AllSamples_Allgenes_SeqTable.csv')\n",
    "final_df = post_val[post_val.Sample.isin(tree_samples.Sample_Name)]\n",
    "tree_taxo(final_df)\n",
    "final_df[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "alert-applicant",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DataSource</th>\n",
       "      <th>Sample</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>sci_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Total</td>\n",
       "      <td>20045</td>\n",
       "      <td>417</td>\n",
       "      <td>8379</td>\n",
       "      <td>16996</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  DataSource Sample family genus sci_name\n",
       "0      Total  20045    417  8379    16996"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def stats_total(df):\n",
    "    tmp = {'DataSource':'Total','Sample':df.shape[0],'family':df.family.nunique(),\n",
    "           'genus':df.genus.nunique(),'sci_name':df.sci_name.nunique()}\n",
    "    return pd.DataFrame.from_dict(tmp,orient='index').transpose()\n",
    "stats_total(all_pt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "secret-moldova",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DataSource</th>\n",
       "      <th>Sample</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>sci_name</th>\n",
       "      <th>stage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GAP</td>\n",
       "      <td>772</td>\n",
       "      <td>102</td>\n",
       "      <td>754</td>\n",
       "      <td>770</td>\n",
       "      <td>Pre-Validation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>INSDC</td>\n",
       "      <td>2851</td>\n",
       "      <td>179</td>\n",
       "      <td>1129</td>\n",
       "      <td>2851</td>\n",
       "      <td>Pre-Validation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Lietal</td>\n",
       "      <td>624</td>\n",
       "      <td>290</td>\n",
       "      <td>516</td>\n",
       "      <td>624</td>\n",
       "      <td>Pre-Validation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OneKP</td>\n",
       "      <td>661</td>\n",
       "      <td>249</td>\n",
       "      <td>524</td>\n",
       "      <td>661</td>\n",
       "      <td>Pre-Validation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PAFTOL</td>\n",
       "      <td>8339</td>\n",
       "      <td>361</td>\n",
       "      <td>6053</td>\n",
       "      <td>7307</td>\n",
       "      <td>Pre-Validation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>REFSEQ</td>\n",
       "      <td>5162</td>\n",
       "      <td>277</td>\n",
       "      <td>2011</td>\n",
       "      <td>5162</td>\n",
       "      <td>Pre-Validation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SRA</td>\n",
       "      <td>1636</td>\n",
       "      <td>199</td>\n",
       "      <td>1309</td>\n",
       "      <td>1520</td>\n",
       "      <td>Pre-Validation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Total</td>\n",
       "      <td>20045</td>\n",
       "      <td>417</td>\n",
       "      <td>8379</td>\n",
       "      <td>16996</td>\n",
       "      <td>Pre-Validation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GAP</td>\n",
       "      <td>760</td>\n",
       "      <td>101</td>\n",
       "      <td>742</td>\n",
       "      <td>758</td>\n",
       "      <td>Post-validation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>INSDC</td>\n",
       "      <td>2840</td>\n",
       "      <td>176</td>\n",
       "      <td>1122</td>\n",
       "      <td>2840</td>\n",
       "      <td>Post-validation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Lietal</td>\n",
       "      <td>623</td>\n",
       "      <td>290</td>\n",
       "      <td>515</td>\n",
       "      <td>623</td>\n",
       "      <td>Post-validation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OneKP</td>\n",
       "      <td>657</td>\n",
       "      <td>247</td>\n",
       "      <td>520</td>\n",
       "      <td>657</td>\n",
       "      <td>Post-validation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PAFTOL</td>\n",
       "      <td>8194</td>\n",
       "      <td>358</td>\n",
       "      <td>5987</td>\n",
       "      <td>7211</td>\n",
       "      <td>Post-validation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>REFSEQ</td>\n",
       "      <td>5143</td>\n",
       "      <td>275</td>\n",
       "      <td>2001</td>\n",
       "      <td>5143</td>\n",
       "      <td>Post-validation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SRA</td>\n",
       "      <td>1608</td>\n",
       "      <td>198</td>\n",
       "      <td>1291</td>\n",
       "      <td>1495</td>\n",
       "      <td>Post-validation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Total</td>\n",
       "      <td>19825</td>\n",
       "      <td>417</td>\n",
       "      <td>8309</td>\n",
       "      <td>16855</td>\n",
       "      <td>Post-validation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GAP</td>\n",
       "      <td>345</td>\n",
       "      <td>67</td>\n",
       "      <td>340</td>\n",
       "      <td>345</td>\n",
       "      <td>Tree</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>INSDC</td>\n",
       "      <td>2809</td>\n",
       "      <td>174</td>\n",
       "      <td>1118</td>\n",
       "      <td>2809</td>\n",
       "      <td>Tree</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Lietal</td>\n",
       "      <td>617</td>\n",
       "      <td>287</td>\n",
       "      <td>510</td>\n",
       "      <td>617</td>\n",
       "      <td>Tree</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OneKP</td>\n",
       "      <td>653</td>\n",
       "      <td>246</td>\n",
       "      <td>516</td>\n",
       "      <td>653</td>\n",
       "      <td>Tree</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PAFTOL</td>\n",
       "      <td>3490</td>\n",
       "      <td>259</td>\n",
       "      <td>2728</td>\n",
       "      <td>3157</td>\n",
       "      <td>Tree</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>REFSEQ</td>\n",
       "      <td>5134</td>\n",
       "      <td>273</td>\n",
       "      <td>1995</td>\n",
       "      <td>5134</td>\n",
       "      <td>Tree</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SRA</td>\n",
       "      <td>1005</td>\n",
       "      <td>133</td>\n",
       "      <td>832</td>\n",
       "      <td>957</td>\n",
       "      <td>Tree</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Total</td>\n",
       "      <td>14053</td>\n",
       "      <td>412</td>\n",
       "      <td>5490</td>\n",
       "      <td>12561</td>\n",
       "      <td>Tree</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  DataSource Sample family genus sci_name            stage\n",
       "0        GAP    772    102   754      770   Pre-Validation\n",
       "1      INSDC   2851    179  1129     2851   Pre-Validation\n",
       "2     Lietal    624    290   516      624   Pre-Validation\n",
       "3      OneKP    661    249   524      661   Pre-Validation\n",
       "4     PAFTOL   8339    361  6053     7307   Pre-Validation\n",
       "5     REFSEQ   5162    277  2011     5162   Pre-Validation\n",
       "6        SRA   1636    199  1309     1520   Pre-Validation\n",
       "7      Total  20045    417  8379    16996   Pre-Validation\n",
       "0        GAP    760    101   742      758  Post-validation\n",
       "1      INSDC   2840    176  1122     2840  Post-validation\n",
       "2     Lietal    623    290   515      623  Post-validation\n",
       "3      OneKP    657    247   520      657  Post-validation\n",
       "4     PAFTOL   8194    358  5987     7211  Post-validation\n",
       "5     REFSEQ   5143    275  2001     5143  Post-validation\n",
       "6        SRA   1608    198  1291     1495  Post-validation\n",
       "7      Total  19825    417  8309    16855  Post-validation\n",
       "0        GAP    345     67   340      345             Tree\n",
       "1      INSDC   2809    174  1118     2809             Tree\n",
       "2     Lietal    617    287   510      617             Tree\n",
       "3      OneKP    653    246   516      653             Tree\n",
       "4     PAFTOL   3490    259  2728     3157             Tree\n",
       "5     REFSEQ   5134    273  1995     5134             Tree\n",
       "6        SRA   1005    133   832      957             Tree\n",
       "7      Total  14053    412  5490    12561             Tree"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ini = pre_val.groupby('DataSource').agg({'Sample':'count','family':'nunique','genus':'nunique','sci_name':'nunique'}).reset_index()\n",
    "ini = pd.concat([ini,stats_total(pre_val)],ignore_index=True)\n",
    "ini['stage'] = 'Pre-Validation'\n",
    "\n",
    "pre = post_val.groupby('DataSource').agg({'Sample':'count','family':'nunique','genus':'nunique','sci_name':'nunique'}).reset_index()\n",
    "pre = pd.concat([pre,stats_total(post_val)],ignore_index=True)\n",
    "pre['stage'] = 'Post-validation'\n",
    "\n",
    "final = final_df.groupby('DataSource').agg({'Sample':'count','family':'nunique','genus':'nunique','sci_name':'nunique'}).reset_index()\n",
    "final = pd.concat([final,stats_total(final_df)],ignore_index=True)\n",
    "final['stage'] = 'Tree'\n",
    "stats_taxo = pd.concat([ini, pre, final])\n",
    "stats_taxo = stats_taxo[stats_taxo.DataSource!='Gymno']\n",
    "stats_taxo#.set_index # .stack().to_frame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "literary-wrist",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th>stage</th>\n",
       "      <th>level_0</th>\n",
       "      <th>DataSource</th>\n",
       "      <th colspan=\"4\" halign=\"left\">Post-validation</th>\n",
       "      <th colspan=\"4\" halign=\"left\">Pre-Validation</th>\n",
       "      <th colspan=\"4\" halign=\"left\">Tree</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>sci_name</th>\n",
       "      <th>Sample</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>sci_name</th>\n",
       "      <th>Sample</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>sci_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>N</td>\n",
       "      <td>GAP</td>\n",
       "      <td>760</td>\n",
       "      <td>101</td>\n",
       "      <td>742</td>\n",
       "      <td>758</td>\n",
       "      <td>772</td>\n",
       "      <td>102</td>\n",
       "      <td>754</td>\n",
       "      <td>770</td>\n",
       "      <td>345</td>\n",
       "      <td>67</td>\n",
       "      <td>340</td>\n",
       "      <td>345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>N</td>\n",
       "      <td>INSDC</td>\n",
       "      <td>2840</td>\n",
       "      <td>176</td>\n",
       "      <td>1122</td>\n",
       "      <td>2840</td>\n",
       "      <td>2851</td>\n",
       "      <td>179</td>\n",
       "      <td>1129</td>\n",
       "      <td>2851</td>\n",
       "      <td>2809</td>\n",
       "      <td>174</td>\n",
       "      <td>1118</td>\n",
       "      <td>2809</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>N</td>\n",
       "      <td>Lietal</td>\n",
       "      <td>623</td>\n",
       "      <td>290</td>\n",
       "      <td>515</td>\n",
       "      <td>623</td>\n",
       "      <td>624</td>\n",
       "      <td>290</td>\n",
       "      <td>516</td>\n",
       "      <td>624</td>\n",
       "      <td>617</td>\n",
       "      <td>287</td>\n",
       "      <td>510</td>\n",
       "      <td>617</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>N</td>\n",
       "      <td>OneKP</td>\n",
       "      <td>657</td>\n",
       "      <td>247</td>\n",
       "      <td>520</td>\n",
       "      <td>657</td>\n",
       "      <td>661</td>\n",
       "      <td>249</td>\n",
       "      <td>524</td>\n",
       "      <td>661</td>\n",
       "      <td>653</td>\n",
       "      <td>246</td>\n",
       "      <td>516</td>\n",
       "      <td>653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>N</td>\n",
       "      <td>PAFTOL</td>\n",
       "      <td>8194</td>\n",
       "      <td>358</td>\n",
       "      <td>5987</td>\n",
       "      <td>7211</td>\n",
       "      <td>8339</td>\n",
       "      <td>361</td>\n",
       "      <td>6053</td>\n",
       "      <td>7307</td>\n",
       "      <td>3490</td>\n",
       "      <td>259</td>\n",
       "      <td>2728</td>\n",
       "      <td>3157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>N</td>\n",
       "      <td>REFSEQ</td>\n",
       "      <td>5143</td>\n",
       "      <td>275</td>\n",
       "      <td>2001</td>\n",
       "      <td>5143</td>\n",
       "      <td>5162</td>\n",
       "      <td>277</td>\n",
       "      <td>2011</td>\n",
       "      <td>5162</td>\n",
       "      <td>5134</td>\n",
       "      <td>273</td>\n",
       "      <td>1995</td>\n",
       "      <td>5134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>N</td>\n",
       "      <td>SRA</td>\n",
       "      <td>1608</td>\n",
       "      <td>198</td>\n",
       "      <td>1291</td>\n",
       "      <td>1495</td>\n",
       "      <td>1636</td>\n",
       "      <td>199</td>\n",
       "      <td>1309</td>\n",
       "      <td>1520</td>\n",
       "      <td>1005</td>\n",
       "      <td>133</td>\n",
       "      <td>832</td>\n",
       "      <td>957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>N</td>\n",
       "      <td>Total</td>\n",
       "      <td>19825</td>\n",
       "      <td>417</td>\n",
       "      <td>8309</td>\n",
       "      <td>16855</td>\n",
       "      <td>20045</td>\n",
       "      <td>417</td>\n",
       "      <td>8379</td>\n",
       "      <td>16996</td>\n",
       "      <td>14053</td>\n",
       "      <td>412</td>\n",
       "      <td>5490</td>\n",
       "      <td>12561</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "stage level_0 DataSource Post-validation                       Pre-Validation  \\\n",
       "                                  Sample family genus sci_name         Sample   \n",
       "0           N        GAP             760    101   742      758            772   \n",
       "1           N      INSDC            2840    176  1122     2840           2851   \n",
       "2           N     Lietal             623    290   515      623            624   \n",
       "3           N      OneKP             657    247   520      657            661   \n",
       "4           N     PAFTOL            8194    358  5987     7211           8339   \n",
       "5           N     REFSEQ            5143    275  2001     5143           5162   \n",
       "6           N        SRA            1608    198  1291     1495           1636   \n",
       "7           N      Total           19825    417  8309    16855          20045   \n",
       "\n",
       "stage                         Tree                        \n",
       "      family genus sci_name Sample family genus sci_name  \n",
       "0        102   754      770    345     67   340      345  \n",
       "1        179  1129     2851   2809    174  1118     2809  \n",
       "2        290   516      624    617    287   510      617  \n",
       "3        249   524      661    653    246   516      653  \n",
       "4        361  6053     7307   3490    259  2728     3157  \n",
       "5        277  2011     5162   5134    273  1995     5134  \n",
       "6        199  1309     1520   1005    133   832      957  \n",
       "7        417  8379    16996  14053    412  5490    12561  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stats_taxo = pd.concat([ini.set_index(['stage','DataSource']).stack().to_frame().rename(columns={0:'N'}), \n",
    "                        pre.set_index(['stage','DataSource']).stack().to_frame().rename(columns={0:'N'}), \n",
    "                        final.set_index(['stage','DataSource']).stack().to_frame().rename(columns={0:'N'})])\n",
    "stats_taxo = stats_taxo.unstack(level=-2).transpose().reset_index()\n",
    "stats_taxo.to_clipboard()\n",
    "stats_taxo"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
