{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import SeqIO; from Bio.Seq import Seq; from Bio.SeqRecord import SeqRecord; from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "import os\n",
    "import re"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Refs\n",
    "* https://biopython.org/wiki/SeqRecord\n",
    "* https://ena-docs.readthedocs.io/en/latest/faq/locus_tags.html\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# latest paftol export containing ENA sample accessions\n",
    "pft_file = '../PAFTOL_DB/2022-03-14_paftol_export_stats.csv'\n",
    "cptree_file = 'cpTree_v7/cpTree_v7_tree_clean.csv'\n",
    "org_dir = 'C:/Data/PAFTOL/Organelles/'\n",
    "\n",
    "# DataSource = 'GAP' \n",
    "# ENA_project_number = 'PRJEB49212'\n",
    "DataSource = 'PAFTOL' \n",
    "ENA_project_number = 'PRJEB51960' #https://www.ncbi.nlm.nih.gov/bioproject/588607\n",
    "project_locus_tag = 'PRJEB51960'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3451 ['PAFTOL_004799', 'PAFTOL_000992']\n"
     ]
    }
   ],
   "source": [
    "## All Validated Samples\n",
    "R = pd.read_csv(cptree_file)\n",
    "R = R[R.DataSource==DataSource]\n",
    "subset = R.Sample_Name.to_list()\n",
    "print(R.shape[0],subset[:2])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load data and subset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "pdir = DataSource + '/'\n",
    "sdir = 'ENA_submissions/' + pdir\n",
    "if not os.path.exists(sdir):\n",
    "    os.makedirs(sdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9715\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>idPaftol</th>\n",
       "      <th>idSequencing</th>\n",
       "      <th>Sample_Name</th>\n",
       "      <th>Project</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>Taxonomical_Notes</th>\n",
       "      <th>SumContigLength</th>\n",
       "      <th>...</th>\n",
       "      <th>Coverage_Kmer</th>\n",
       "      <th>Coverage_base</th>\n",
       "      <th>Run_Time</th>\n",
       "      <th>maxK</th>\n",
       "      <th>NRepeat_Pattern</th>\n",
       "      <th>NPath</th>\n",
       "      <th>Redo_FastPlast</th>\n",
       "      <th>error_pt</th>\n",
       "      <th>error_nr</th>\n",
       "      <th>SeqID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>961</td>\n",
       "      <td>PAFTOL_000961</td>\n",
       "      <td>Pilot</td>\n",
       "      <td>Ranunculales</td>\n",
       "      <td>Circaeasteraceae</td>\n",
       "      <td>Circaeaster</td>\n",
       "      <td>agrestis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>138795.0</td>\n",
       "      <td>...</td>\n",
       "      <td>59.4</td>\n",
       "      <td>90.7</td>\n",
       "      <td>329.43</td>\n",
       "      <td>105.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pis_000961</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4</td>\n",
       "      <td>2571</td>\n",
       "      <td>PAFTOL_002571</td>\n",
       "      <td>Connaraceae</td>\n",
       "      <td>Oxalidales</td>\n",
       "      <td>Connaraceae</td>\n",
       "      <td>Rourea</td>\n",
       "      <td>minor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>213954.0</td>\n",
       "      <td>...</td>\n",
       "      <td>18.5</td>\n",
       "      <td>60.2</td>\n",
       "      <td>747.57</td>\n",
       "      <td>105.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pis_002571</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows Ã— 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   idPaftol  idSequencing    Sample_Name      Project         order  \\\n",
       "0         2           961  PAFTOL_000961        Pilot  Ranunculales   \n",
       "1         4          2571  PAFTOL_002571  Connaraceae    Oxalidales   \n",
       "\n",
       "             family        genus   species Taxonomical_Notes  SumContigLength  \\\n",
       "0  Circaeasteraceae  Circaeaster  agrestis               NaN         138795.0   \n",
       "1       Connaraceae       Rourea     minor               NaN         213954.0   \n",
       "\n",
       "   ... Coverage_Kmer  Coverage_base  Run_Time   maxK  NRepeat_Pattern  NPath  \\\n",
       "0  ...          59.4           90.7    329.43  105.0              NaN    NaN   \n",
       "1  ...          18.5           60.2    747.57  105.0              NaN    NaN   \n",
       "\n",
       "   Redo_FastPlast  error_pt  error_nr       SeqID  \n",
       "0           False       NaN       NaN  Pis_000961  \n",
       "1           False       NaN       NaN  Pis_002571  \n",
       "\n",
       "[2 rows x 42 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load GetOrg result table\n",
    "Org_df = pd.read_csv(org_dir + DataSource + '/' + DataSource + '_Organelle_Recovery.csv').astype({'idSequencing':'int'})\n",
    "if DataSource == 'PAFTOL':\n",
    "    Org_df['SeqID'] = Org_df.Sample_Name.str.replace('PAFTOL_','Pis_')\n",
    "elif DataSource in ['GAP']:\n",
    "    Org_df['SeqID'] = Org_df.Sample_Name\n",
    "elif DataSource in ['SRA']:\n",
    "    sra2_pt = pd.read_csv('cpSRA_GetOrg/cpSRA_GetOrg_Organelle_Recovery.csv')\n",
    "    Org_df = pd.concat([Org_df,sra2_pt],ignore_index=True)\n",
    "    Org_df = Org_df.groupby('Sample_Name').head(1)\n",
    "    Org_df['SeqID'] = Org_df.Sample_Name\n",
    "print(Org_df.shape[0])\n",
    "Org_df[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3451 3451\n"
     ]
    }
   ],
   "source": [
    "if subset is not None:\n",
    "    Org_df = Org_df[Org_df.Sample_Name.isin(subset)]\n",
    "    print(Org_df.shape[0],Org_df.Sample_Name.nunique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\ProgramData\\Anaconda3\\lib\\site-packages\\IPython\\core\\interactiveshell.py:3146: DtypeWarning: Columns (29,35,45,57,74,83) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_Name</th>\n",
       "      <th>idPaftol</th>\n",
       "      <th>DataSource</th>\n",
       "      <th>Project</th>\n",
       "      <th>Order</th>\n",
       "      <th>Family</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Species</th>\n",
       "      <th>Taxonomical_Notes</th>\n",
       "      <th>ENASampleNum</th>\n",
       "      <th>idSequencing</th>\n",
       "      <th>ExternalSequenceID</th>\n",
       "      <th>ENAExpNumber</th>\n",
       "      <th>ENARunNumber</th>\n",
       "      <th>NumReads</th>\n",
       "      <th>NumRecoveredGenes</th>\n",
       "      <th>SumContigLength</th>\n",
       "      <th>NCBI_TaxID</th>\n",
       "      <th>NCBI_sciname</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PAFTOL_000564</td>\n",
       "      <td>1</td>\n",
       "      <td>PAFTOL</td>\n",
       "      <td>Asteraceae</td>\n",
       "      <td>Asterales</td>\n",
       "      <td>Asteraceae</td>\n",
       "      <td>Gymnolaena</td>\n",
       "      <td>sp.</td>\n",
       "      <td>Initially recorded as Gymnolaena litoralis but...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>564</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PAFTOL_000961</td>\n",
       "      <td>2</td>\n",
       "      <td>PAFTOL</td>\n",
       "      <td>Pilot</td>\n",
       "      <td>Ranunculales</td>\n",
       "      <td>Circaeasteraceae</td>\n",
       "      <td>Circaeaster</td>\n",
       "      <td>agrestis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERS5501508</td>\n",
       "      <td>961</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERX4839538</td>\n",
       "      <td>ERR5033191</td>\n",
       "      <td>625228.0</td>\n",
       "      <td>332.0</td>\n",
       "      <td>138795.0</td>\n",
       "      <td>39288.0</td>\n",
       "      <td>Circaeaster agrestis</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Sample_Name  idPaftol DataSource     Project         Order  \\\n",
       "0  PAFTOL_000564         1     PAFTOL  Asteraceae     Asterales   \n",
       "1  PAFTOL_000961         2     PAFTOL       Pilot  Ranunculales   \n",
       "\n",
       "             Family        Genus   Species  \\\n",
       "0        Asteraceae   Gymnolaena       sp.   \n",
       "1  Circaeasteraceae  Circaeaster  agrestis   \n",
       "\n",
       "                                   Taxonomical_Notes ENASampleNum  \\\n",
       "0  Initially recorded as Gymnolaena litoralis but...          NaN   \n",
       "1                                                NaN   ERS5501508   \n",
       "\n",
       "   idSequencing ExternalSequenceID ENAExpNumber ENARunNumber  NumReads  \\\n",
       "0           564                NaN          NaN          NaN       NaN   \n",
       "1           961                NaN   ERX4839538   ERR5033191  625228.0   \n",
       "\n",
       "   NumRecoveredGenes  SumContigLength  NCBI_TaxID          NCBI_sciname  \n",
       "0                NaN              NaN         NaN                   NaN  \n",
       "1              332.0         138795.0     39288.0  Circaeaster agrestis  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load paftol export and merge ena sample IDs\n",
    "db = pd.read_csv(pft_file)\n",
    "db = db[db.idPaftol.notnull()].astype({'idPaftol':'int','idSequencing':'int'})\n",
    "db = db[['Sample_Name','idPaftol', 'DataSource', 'Project', 'Order', 'Family', 'Genus', 'Species', 'Taxonomical_Notes', \n",
    "         'ENASampleNum', 'idSequencing', 'ExternalSequenceID', 'ENAExpNumber',\n",
    "       'ENARunNumber', 'NumReads', 'NumRecoveredGenes', 'SumContigLength','NCBI_TaxID', 'NCBI_sciname']]\n",
    "#        'title', 'taxId','id', 'alias', 'firstCreated', 'firstPublic', 'releaseStatus', 'secondaryId', 'scientificName', 'commonName']]\n",
    "db[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'idPaftol': 0, 'idSequencing': 0, 'Sample_Name': 0, 'Project': 0, 'order': 0, 'family': 0, 'genus': 0, 'species': 0, 'Taxonomical_Notes': 2231, 'SumContigLength': 363, 'sci_name': 0, 'log_pt': 0, 'log_nr': 0, 'pt_recovered': 0, 'Nseq_pt': 0, 'Sum_len_pt': 0, 'Nanybase_pt': 0, 'nr_recovered': 0, 'Nseq_nr': 27, 'Sum_len_nr': 27, 'Nanybase_nr': 27, 'Completed_Reads': 0, 'Completed_Extending': 0, 'Completed_Assembly': 0, 'Completed_Slimming': 0, 'Completed_Output': 0, 'Warning_Look': 0, 'Warning_Multi_Structure': 0, 'Warning_Self_Loop': 0, 'Info_Large_Repeats': 0, 'Info_Disentangling_Failed': 0, 'Assembly': 6, 'Coverage_Kmer': 635, 'Coverage_base': 8, 'Run_Time': 5, 'maxK': 6, 'NRepeat_Pattern': 3380, 'NPath': 3422, 'Redo_FastPlast': 0, 'error_pt': 3450, 'error_nr': 3432, 'SeqID': 0, 'DataSource': 0, 'ENASampleNum': 47, 'ENAExpNumber': 0, 'ENARunNumber': 0, 'NCBI_TaxID': 172, 'NCBI_sciname': 172}\n",
      "3451 3451\n"
     ]
    }
   ],
   "source": [
    "Org_df = pd.merge(Org_df,\n",
    "         db[['Sample_Name','DataSource','ENASampleNum','ENAExpNumber','ENARunNumber','NCBI_TaxID','NCBI_sciname']],\n",
    "        how='left',on='Sample_Name').groupby('Sample_Name').head(1)\n",
    "print(Org_df.isna().sum().to_dict())\n",
    "print(Org_df.shape[0],Org_df.Sample_Name.nunique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "geseq_dir = sdir + 'fasta/'\n",
    "if not os.path.exists(geseq_dir):\n",
    "    os.makedirs(geseq_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Whole plastomes\n",
    "### Annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "whole_pt = Org_df[(Org_df.Assembly.isin(['circular genome','circular genome with gaps']) & (Org_df.Sum_len_pt>80000))]\n",
    "print(whole_pt.shape[0],'samples',whole_pt.genus.nunique(),'genera')\n",
    "whole_pt[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Output fasta files \n",
    "for idx, row in whole_pt.iterrows():\n",
    "    seqs = list(SeqIO.parse(org_dir + DataSource + '/' + 'fasta_pt/' + row.Sample_Name + '_pt.fasta',format='fasta'))\n",
    "    # Verif only 1 sequence\n",
    "    if len(seqs) > 1:\n",
    "        print('error, more than 1 sequence')\n",
    "        pass\n",
    "    seq = seqs[0]\n",
    "    print(row.Sample_Name, row.SeqID, row.family, row.sci_name,len(seq),seq.seq.count('N'))\n",
    "    seq.id = row.SeqID\n",
    "    seq.description = ''\n",
    "    SeqIO.write(seq, geseq_dir + row.SeqID + '-circular.fasta',format='fasta')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Submit files to GeSeq https://chlorobox.mpimp-golm.mpg.de/geseq.html and extract zip file in wdir/GeSeq_output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "whole_pt.to_csv(sdir + DataSource + '_wholePT.csv',index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Contigs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "references = pd.read_csv('References/NCBI_plastid_ref/NCBI_plastid_ref_list_SELECTED.csv')\n",
    "print(references.shape)\n",
    "references.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "incomplete_pt = Org_df[(Org_df.Assembly.isin(['circular genome','circular genome with gaps'])==False) &\n",
    "                   (Org_df.pt_recovered==True)]\n",
    "print(incomplete_pt.shape[0],'samples',incomplete_pt.genus.nunique(),'genera')\n",
    "incomplete_pt = incomplete_pt[incomplete_pt.Sum_len_pt<200000].set_index('Sample_Name')\n",
    "print(incomplete_pt.shape[0],'samples',incomplete_pt.genus.nunique(),'genera')\n",
    "### Random subset for test \n",
    "incomplete_pt = incomplete_pt.sample(n=100)\n",
    "print(incomplete_pt.shape[0],'samples',incomplete_pt.genus.nunique(),'genera')\n",
    "incomplete_pt[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "references[references.order=='Sapindales']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "order_dc = {'Gunnerales':'Dilleniales','Picramniales':'Sapindales'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "commands = []\n",
    "for Sample_Name, row in incomplete_pt.iterrows():\n",
    "    print(Sample_Name,row.order,row.family)\n",
    "    ### Identify best reference\n",
    "    # check if reference in same family\n",
    "    best_ref = references[references.family==row.family]\n",
    "    if best_ref.shape[0]<1:\n",
    "        # otherwise order\n",
    "        best_ref = references[references.order==row.order]\n",
    "        if best_ref.shape[0]<1:\n",
    "            #Otherwise use closest order\n",
    "            best_ref = references[references.order==order_dc[row.order]]\n",
    "    print('\\t Ref',best_ref.shape, end = ': ')\n",
    "    best_ref = best_ref.iloc[0]\n",
    "    print(best_ref.SeqID,best_ref.order,best_ref.family)\n",
    "    incomplete_pt.loc[Sample_Name,'Ref-SeqID'] = best_ref.SeqID\n",
    "    incomplete_pt.loc[Sample_Name,'Ref-order'] = best_ref.order\n",
    "    incomplete_pt.loc[Sample_Name,'Ref-family'] = best_ref.family\n",
    "    \n",
    "    ### launch blastn\n",
    "    # program_path = 'C:/Program Files/NCBI/blast-BLAST_VERSION+/bin/blastn.exe'\n",
    "    program_path = 'blastn'\n",
    "    ref_path = 'References/NCBI_plastid_ref/' + best_ref['SeqID'] + '.fasta'\n",
    "    contigs_path = '../Organelles/' + DataSource + '/fasta_pt/' + Sample_Name + '_pt.fasta'\n",
    "    out_path = 'ENA_submissions/' + DataSource + '/out_blast/'+ Sample_Name + '.blast'\n",
    "    command = program_path + ' -query ' + contigs_path + ' -subject ' + ref_path + \\\n",
    "                ' -outfmt \"6 qseqid sseqid pident length slen qlen mismatch gapopen qstart qend sstart send evalue bitscore\"' + \\\n",
    "                ' -out ' + out_path\n",
    "    commands.append(command)\n",
    "    os.system(command)\n",
    "    \n",
    "    ### read blast output\n",
    "    blast_output = pd.read_table(out_path, sep='\\t',header=None)\n",
    "    blast_output.columns = ['qseqid','sseqid','pident','length','slen','qlen','mismatch','gapopen','qstart','qend','sstart','send','evalue','bitscore']\n",
    "#     print('\\t',blast_output.shape[0],'matched for', blast_output.qseqid.nunique(),'qseqid',end='; ')\n",
    "    incomplete_pt.loc[Sample_Name,'Blast-N_qseqid_match'] = blast_output.qseqid.nunique()\n",
    "    # Keep best match (by bitscore) per queried contig\n",
    "    blast_output = blast_output.sort_values('bitscore',ascending=False).groupby('qseqid').head(1)\n",
    "    blast_output['oriented_sstart'] = blast_output[['sstart','send']].min(axis=1); blast_output['oriented_ssend'] = blast_output[['sstart','send']].max(axis=1)\n",
    "    blast_output['Need_RevComp'] = blast_output.sstart > blast_output.oriented_sstart\n",
    "    blast_output = blast_output.sort_values('oriented_sstart').set_index('qseqid')\n",
    "#     print('Contigs to reverse complement:',blast_output.Need_RevComp.value_counts().to_dict())\n",
    "    \n",
    "    \n",
    "    ### Load contigs and write concatenated fasta\n",
    "    gapN = 100\n",
    "    records = SeqIO.to_dict(SeqIO.parse(contigs_path, \"fasta\"))\n",
    "    print('\\t',blast_output.shape[0],'/',len(records),'contigs mapped to reference')\n",
    "    concat_seq = ''\n",
    "    for qseqid, blast_row in blast_output.iterrows():\n",
    "        if blast_row.Need_RevComp==True:\n",
    "            concat_seq += records[qseqid].seq.reverse_complement()\n",
    "        else:\n",
    "            concat_seq += records[qseqid].seq \n",
    "        concat_seq += Seq(gapN * 'N')\n",
    "    concat_seq = concat_seq[:-gapN] # Remove trailing Ns\n",
    "    print('\\t sum len concat:',blast_output.qlen.sum(),'/',int(row.Sum_len_pt),round(blast_output.qlen.sum()/int(row.Sum_len_pt)*100,1),'%')\n",
    "    incomplete_pt.loc[Sample_Name,'Concat-len'] = blast_output.qlen.sum()\n",
    "    incomplete_pt.loc[Sample_Name,'Concat-%_mapped'] = round(blast_output.qlen.sum()/int(row.Sum_len_pt)*100,1)\n",
    "    incomplete_pt.loc[Sample_Name,'Concat-len_with_Ns'] = len(concat_seq)\n",
    "    Concat_record = SeqRecord(\n",
    "        concat_seq,\n",
    "        id=row.SeqID + '_concat_pt',\n",
    "        name=Sample_Name + '_concat_pt',\n",
    "        description=''\n",
    "    )\n",
    "    SeqIO.write(Concat_record,'ENA_submissions/' + DataSource + '/fasta/' + row.SeqID + '-concat.fasta','fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "incomplete_pt.to_csv(sdir + DataSource + '_concatPT.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ENA submission\n",
    "https://ena-docs.readthedocs.io/en/latest/submit/fileprep/assembly.html?highlight=chloroplast#\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Leempoel K., Zuntini A. R., Bailey P., Eiserhardt W. L., Gallego B., Maurin O., Allnutt T., Bellot S., Clarkson J. J., Clements M., Cowan R. S., Crayn D. D., de LÃ­rio E., FranÃ§oso E., HÃ¶wener A., Hu A., Joyce E., Larson D., Li H., Li D., Liu J., Lum M., McGinnie C., McLay T., Nargar K., Perez Escobar O., Przelomska N., Roy S., Schmidt-Lebuhn A., Shah T., Simpson L., Yang J., Yi T., Forest F., Kersey P. J., Leitch I. J., Baker W. J.'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "paper_title = 'Building the largest, curated plastid genome dataset and phylogeny by leveraging nuclear-oriented sequence data'\n",
    "paper_authors = 'Leempoel K., Zuntini A. R., Bailey P., Eiserhardt W. L., Gallego B., Maurin O., Allnutt T., Bellot S., Clarkson J. J., ' + \\\n",
    "        'Clements M., Cowan R. S., Crayn D. D., de LÃ­rio E., FranÃ§oso E., HÃ¶wener A., Hu A., Joyce E., Larson D., Li H., Li D., Liu J., Lum M., McGinnie C., ' + \\\n",
    "        'McLay T., Nargar K., Perez Escobar O., Przelomska N., Roy S., Schmidt-Lebuhn A., Shah T., Simpson L., Yang J., Yi T., Forest F., Kersey P. J., ' + \\\n",
    "        'Leitch I. J., Baker W. J.'\n",
    "paper_journal = 'Unpublished'\n",
    "paper_id = 'NA'\n",
    "paper_authors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "100\n",
      "97\n",
      "93 {'scaffold': 93}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_Name</th>\n",
       "      <th>idPaftol</th>\n",
       "      <th>idSequencing</th>\n",
       "      <th>Project</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>Taxonomical_Notes</th>\n",
       "      <th>SumContigLength</th>\n",
       "      <th>...</th>\n",
       "      <th>ENARunNumber</th>\n",
       "      <th>NCBI_TaxID</th>\n",
       "      <th>NCBI_sciname</th>\n",
       "      <th>Ref-SeqID</th>\n",
       "      <th>Ref-order</th>\n",
       "      <th>Ref-family</th>\n",
       "      <th>Blast-N_qseqid_match</th>\n",
       "      <th>Concat-len</th>\n",
       "      <th>Concat-%_mapped</th>\n",
       "      <th>Concat-len_with_Ns</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PAFTOL_007905</td>\n",
       "      <td>4838</td>\n",
       "      <td>7905</td>\n",
       "      <td>Celastraceae</td>\n",
       "      <td>Celastrales</td>\n",
       "      <td>Celastraceae</td>\n",
       "      <td>Euonymus</td>\n",
       "      <td>australianus</td>\n",
       "      <td>Euonymus-Celastrus-Haydenoxylon clade</td>\n",
       "      <td>201111.0</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR7619260</td>\n",
       "      <td>1089391.0</td>\n",
       "      <td>Euonymus australianus</td>\n",
       "      <td>NC_045280.1</td>\n",
       "      <td>Celastrales</td>\n",
       "      <td>Celastraceae</td>\n",
       "      <td>30.0</td>\n",
       "      <td>24986.0</td>\n",
       "      <td>98.3</td>\n",
       "      <td>27886.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PAFTOL_014479</td>\n",
       "      <td>8655</td>\n",
       "      <td>14479</td>\n",
       "      <td>Asteraceae</td>\n",
       "      <td>Asterales</td>\n",
       "      <td>Asteraceae</td>\n",
       "      <td>Olgaea</td>\n",
       "      <td>leucophylla</td>\n",
       "      <td>NaN</td>\n",
       "      <td>182634.0</td>\n",
       "      <td>...</td>\n",
       "      <td>ERR7621142</td>\n",
       "      <td>591993.0</td>\n",
       "      <td>Olgaea leucophylla</td>\n",
       "      <td>NC_044728.1</td>\n",
       "      <td>Asterales</td>\n",
       "      <td>Asteraceae</td>\n",
       "      <td>30.0</td>\n",
       "      <td>36976.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>39876.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows Ã— 55 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Sample_Name  idPaftol  idSequencing       Project        order  \\\n",
       "0  PAFTOL_007905      4838          7905  Celastraceae  Celastrales   \n",
       "1  PAFTOL_014479      8655         14479    Asteraceae    Asterales   \n",
       "\n",
       "         family     genus       species  \\\n",
       "0  Celastraceae  Euonymus  australianus   \n",
       "1    Asteraceae    Olgaea   leucophylla   \n",
       "\n",
       "                       Taxonomical_Notes  SumContigLength  ... ENARunNumber  \\\n",
       "0  Euonymus-Celastrus-Haydenoxylon clade         201111.0  ...   ERR7619260   \n",
       "1                                    NaN         182634.0  ...   ERR7621142   \n",
       "\n",
       "   NCBI_TaxID           NCBI_sciname    Ref-SeqID    Ref-order    Ref-family  \\\n",
       "0   1089391.0  Euonymus australianus  NC_045280.1  Celastrales  Celastraceae   \n",
       "1    591993.0     Olgaea leucophylla  NC_044728.1    Asterales    Asteraceae   \n",
       "\n",
       "   Blast-N_qseqid_match  Concat-len  Concat-%_mapped  Concat-len_with_Ns  \n",
       "0                  30.0     24986.0             98.3             27886.0  \n",
       "1                  30.0     36976.0            100.0             39876.0  \n",
       "\n",
       "[2 rows x 55 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# submit_pt = pd.concat([pd.read_csv(sdir + DataSource + '_wholePT.csv'), pd.read_csv(sdir + DataSource + '_concatPT.csv')],ignore_index=True)\n",
    "# submit_pt = pd.read_csv(sdir + DataSource + '_wholePT.csv')\n",
    "submit_pt = pd.read_csv(sdir + DataSource + '_concatPT.csv')\n",
    "print(submit_pt.shape[0])\n",
    "submit_pt = submit_pt[(submit_pt.ENASampleNum.notnull()) & (submit_pt.ENARunNumber.notnull())]\n",
    "print(submit_pt.shape[0])\n",
    "submit_pt = submit_pt[submit_pt.NCBI_sciname.notnull()]\n",
    "print(submit_pt.shape[0],submit_pt.Assembly.value_counts().to_dict())\n",
    "submit_pt.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1678470, 4)\n",
      "(1172782, 4)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>len</th>\n",
       "      <th>Sample_Name</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GAP_026547-rrn23</td>\n",
       "      <td>2812</td>\n",
       "      <td>GAP_026547</td>\n",
       "      <td>rrn23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GAP_026547-rrn16</td>\n",
       "      <td>1491</td>\n",
       "      <td>GAP_026547</td>\n",
       "      <td>rrn16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              seqid   len Sample_Name   gene\n",
       "0  GAP_026547-rrn23  2812  GAP_026547  rrn23\n",
       "1  GAP_026547-rrn16  1491  GAP_026547  rrn16"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_seq = pd.read_csv('cpTree_v7/AllSamples_Allgenes_All_SeqTable.csv')\n",
    "print(all_seq.shape)\n",
    "genes_82 = pd.read_table('cp_genes_82.txt',header=None).rename(columns={0:'gene'})\n",
    "all_seq = all_seq[all_seq.gene.isin(genes_82.gene)]\n",
    "print(all_seq.shape)\n",
    "all_seq.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def list_genes(record, ls_types = ['gene','CDS','rRNA']):\n",
    "    # list gene names\n",
    "    feat_ls = []; idx = 0\n",
    "    for feature in record.features:\n",
    "        if feature.type in ls_types:\n",
    "            feat_ls.append(feature.qualifiers['gene'][0])\n",
    "        idx += 1\n",
    "    return list(set(feat_ls))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "439\n",
      "PAFTOL_007905 GeSeqJob-20220305-120228_Pis_007905_concat_pt_GenBank.gb:1 27886 30 / 16 187 % []\n",
      "PAFTOL_014479 GeSeqJob-20220305-120228_Pis_014479_concat_pt_GenBank.gb:1 39876 29 / 26 111 % []\n",
      "PAFTOL_006622 GeSeqJob-20220305-120228_Pis_006622_concat_pt_GenBank.gb:1 158632 77 / 77 100 % ['rps8']\n",
      "PAFTOL_010719 GeSeqJob-20220305-120228_Pis_010719_concat_pt_GenBank.gb:1 56570 32 / 31 103 % ['rpoB' 'rps8']\n",
      "PAFTOL_012011 GeSeqJob-20220305-120228_Pis_012011_concat_pt_GenBank.gb:1 127768 78 / 77 101 % []\n",
      "PAFTOL_004680 GeSeqJob-20220305-120228_Pis_004680_concat_pt_GenBank.gb:1 63389 42 / 35 120 % ['matK']\n",
      "PAFTOL_018841 GeSeqJob-20220305-120228_Pis_018841_concat_pt_GenBank.gb:1 58859 45 / 40 112 % ['rps3']\n",
      "PAFTOL_015265 GeSeqJob-20220305-120228_Pis_015265_concat_pt_GenBank.gb:1 76256 63 / 55 114 % []\n",
      "PAFTOL_000975 GeSeqJob-20220305-120228_Pis_000975_concat_pt_GenBank.gb:1 39785 28 / 17 164 % []\n",
      "PAFTOL_004459 GeSeqJob-20220305-120228_Pis_004459_concat_pt_GenBank.gb:1 89459 64 / 58 110 % ['psbA']\n",
      "PAFTOL_006015 GeSeqJob-20220305-120228_Pis_006015_concat_pt_GenBank.gb:1 51057 33 / 27 122 % []\n",
      "PAFTOL_011083 GeSeqJob-20220305-120228_Pis_011083_concat_pt_GenBank.gb:1 57118 49 / 31 158 % ['ndhG']\n",
      "PAFTOL_012079 GeSeqJob-20220305-120228_Pis_012079_concat_pt_GenBank.gb:1 158868 78 / 78 100 % []\n",
      "PAFTOL_000972 GeSeqJob-20220305-120228_Pis_000972_concat_pt_GenBank.gb:1 107373 66 / 52 126 % []\n",
      "PAFTOL_007253 GeSeqJob-20220305-120228_Pis_007253_concat_pt_GenBank.gb:1 113084 77 / 73 105 % []\n",
      "PAFTOL_003041 GeSeqJob-20220305-120228_Pis_003041_concat_pt_GenBank.gb:1 136390 78 / 78 100 % []\n",
      "PAFTOL_004743 GeSeqJob-20220305-120228_Pis_004743_concat_pt_GenBank.gb:1 159189 78 / 78 100 % []\n",
      "PAFTOL_015149 GeSeqJob-20220305-120228_Pis_015149_concat_pt_GenBank.gb:1 90674 69 / 64 107 % ['rps3' 'rps19']\n",
      "PAFTOL_015615 GeSeqJob-20220305-120228_Pis_015615_concat_pt_GenBank.gb:1 133179 77 / 75 102 % []\n",
      "PAFTOL_004371 GeSeqJob-20220305-120228_Pis_004371_concat_pt_GenBank.gb:1 94647 68 / 51 133 % []\n",
      "PAFTOL_007973 GeSeqJob-20220305-120228_Pis_007973_concat_pt_GenBank.gb:1 32397 22 / 17 129 % []\n",
      "PAFTOL_004477 GeSeqJob-20220305-120228_Pis_004477_concat_pt_GenBank.gb:1 44952 50 / 26 192 % []\n",
      "PAFTOL_029857 GeSeqJob-20220305-120228_Pis_029857_concat_pt_GenBank.gb:1 31862 20 / 15 133 % ['atpA']\n",
      "PAFTOL_004257 GeSeqJob-20220305-120228_Pis_004257_concat_pt_GenBank.gb:1 61698 43 / 32 134 % ['ndhB' 'rpl16' 'rps16']\n",
      "PAFTOL_021529 GeSeqJob-20220305-120228_Pis_021529_concat_pt_GenBank.gb:1 40359 47 / 29 162 % ['rpl14' 'rps18']\n",
      "PAFTOL_006111 GeSeqJob-20220305-120228_Pis_006111_concat_pt_GenBank.gb:1 129050 77 / 77 100 % []\n",
      "PAFTOL_010453 GeSeqJob-20220305-120228_Pis_010453_concat_pt_GenBank.gb:1 39215 27 / 22 122 % []\n",
      "PAFTOL_019329 GeSeqJob-20220305-120228_Pis_019329_concat_pt_GenBank.gb:1 76214 54 / 55 98 % ['ndhG' 'rps3' 'rpl22']\n",
      "PAFTOL_019293 GeSeqJob-20220305-120228_Pis_019293_concat_pt_GenBank.gb:1 111535 75 / 71 105 % []\n",
      "PAFTOL_016305 GeSeqJob-20220305-120228_Pis_016305_concat_pt_GenBank.gb:1 27702 33 / 18 183 % []\n",
      "PAFTOL_002628 GeSeqJob-20220305-120228_Pis_002628_concat_pt_GenBank.gb:1 125283 77 / 73 105 % []\n",
      "PAFTOL_011889 GeSeqJob-20220305-120228_Pis_011889_concat_pt_GenBank.gb:1 53264 40 / 31 129 % ['psbA']\n",
      "PAFTOL_012985 GeSeqJob-20220305-120228_Pis_012985_concat_pt_GenBank.gb:"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\ProgramData\\Anaconda3\\lib\\site-packages\\Bio\\GenBank\\__init__.py:352: BiopythonParserWarning: It appears that '159219..86471' is a feature that spans the origin, but the sequence topology is undefined. Skipping feature.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 164213 78 / 78 100 % []\n",
      "issue with  GeSeqJob-20220305-120228_Pis_012985_concat_pt_GenBank.gb\n",
      "PAFTOL_023435 GeSeqJob-20220305-120228_Pis_023435_concat_pt_GenBank.gb:1 74401 66 / 37 178 % []\n",
      "PAFTOL_014813 GeSeqJob-20220305-120228_Pis_014813_concat_pt_GenBank.gb:1 40013 38 / 24 158 % []\n",
      "PAFTOL_004390 GeSeqJob-20220305-120228_Pis_004390_concat_pt_GenBank.gb:1 70143 68 / 48 141 % []\n",
      "PAFTOL_008709 GeSeqJob-20220305-120228_Pis_008709_concat_pt_GenBank.gb:1 140960 75 / 76 98 % ['ndhA']\n",
      "issue with  GeSeqJob-20220305-120228_Pis_008709_concat_pt_GenBank.gb\n",
      "PAFTOL_010063 GeSeqJob-20220305-120228_Pis_010063_concat_pt_GenBank.gb:1 55109 54 / 43 125 %"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\ProgramData\\Anaconda3\\lib\\site-packages\\Bio\\GenBank\\__init__.py:352: BiopythonParserWarning: It appears that '130170..1841' is a feature that spans the origin, but the sequence topology is undefined. Skipping feature.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " []\n",
      "PAFTOL_006252 GeSeqJob-20220305-120228_Pis_006252_concat_pt_GenBank.gb:1 46378 26 / 27 96 % ['rpoA' 'rps3' 'atpE' 'rpl22']\n",
      "PAFTOL_011879 GeSeqJob-20220305-120228_Pis_011879_concat_pt_GenBank.gb:1 107978 75 / 72 104 % []\n",
      "PAFTOL_006957 GeSeqJob-20220305-120228_Pis_006957_concat_pt_GenBank.gb:1 23966 16 / 11 145 % []\n",
      "PAFTOL_004696 GeSeqJob-20220305-120228_Pis_004696_concat_pt_GenBank.gb:1 132490 75 / 71 105 % []\n",
      "PAFTOL_014801 GeSeqJob-20220305-120228_Pis_014801_concat_pt_GenBank.gb:1 36829 20 / 19 105 % ['rps11']\n",
      "PAFTOL_015539 GeSeqJob-20220305-120228_Pis_015539_concat_pt_GenBank.gb:1 117894 78 / 74 105 % []\n",
      "PAFTOL_008650 GeSeqJob-20220305-120228_Pis_008650_concat_pt_GenBank.gb:1 42928 32 / 26 123 % []\n",
      "PAFTOL_007942 GeSeqJob-20220305-120228_Pis_007942_concat_pt_GenBank.gb:1 36090 20 / 19 105 % []\n",
      "PAFTOL_005730 GeSeqJob-20220305-120228_Pis_005730_concat_pt_GenBank.gb:1 55075 62 / 32 193 % []\n",
      "PAFTOL_008711 GeSeqJob-20220305-120228_Pis_008711_concat_pt_GenBank.gb:1 40958 37 / 21 176 % []\n",
      "PAFTOL_005727 GeSeqJob-20220305-120228_Pis_005727_concat_pt_GenBank.gb:1 125590 75 / 72 104 % []\n",
      "PAFTOL_006960 GeSeqJob-20220305-120228_Pis_006960_concat_pt_GenBank.gb:1 52611 38 / 33 115 % []\n",
      "PAFTOL_014513 GeSeqJob-20220305-120228_Pis_014513_concat_pt_GenBank.gb:1 42659 41 / 17 241 % ['rps11']\n",
      "PAFTOL_006135 GeSeqJob-20220305-120228_Pis_006135_concat_pt_GenBank.gb:1 74580 60 / 46 130 % []\n",
      "PAFTOL_002655 GeSeqJob-20220305-120228_Pis_002655_concat_pt_GenBank.gb:1 55551 39 / 31 125 % []\n",
      "PAFTOL_019233 GeSeqJob-20220305-120228_Pis_019233_concat_pt_GenBank.gb:1 125563 77 / 76 101 % []\n",
      "PAFTOL_015763 GeSeqJob-20220305-120228_Pis_015763_concat_pt_GenBank.gb:1 158635 78 / 78 100 % []\n",
      "issue with  GeSeqJob-20220305-120228_Pis_015763_concat_pt_GenBank.gb\n",
      "PAFTOL_007313 GeSeqJob-20220305-120228_Pis_007313_concat_pt_GenBank.gb:"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\ProgramData\\Anaconda3\\lib\\site-packages\\Bio\\GenBank\\__init__.py:352: BiopythonParserWarning: It appears that '156258..85559' is a feature that spans the origin, but the sequence topology is undefined. Skipping feature.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 116079 77 / 75 102 % []\n",
      "PAFTOL_001072 GeSeqJob-20220305-120228_Pis_001072_concat_pt_GenBank.gb:1 55418 37 / 28 132 % []\n",
      "PAFTOL_006292 GeSeqJob-20220305-120228_Pis_006292_concat_pt_GenBank.gb:1 124260 77 / 74 104 % []\n",
      "PAFTOL_015271 GeSeqJob-20220305-120228_Pis_015271_concat_pt_GenBank.gb:1 63837 54 / 44 122 % ['cemA']\n",
      "PAFTOL_006145 GeSeqJob-20220305-120228_Pis_006145_concat_pt_GenBank.gb:1 117306 73 / 70 104 % []\n",
      "PAFTOL_010261 GeSeqJob-20220305-120228_Pis_010261_concat_pt_GenBank.gb:1 127697 78 / 77 101 % []\n",
      "PAFTOL_015595 GeSeqJob-20220305-120228_Pis_015595_concat_pt_GenBank.gb:1 127330 75 / 72 104 % []\n",
      "PAFTOL_005892 GeSeqJob-20220305-120228_Pis_005892_concat_pt_GenBank.gb:1 39093 47 / 28 167 % ['rps19' 'rps15' 'rps16' 'petN']\n",
      "PAFTOL_006892 GeSeqJob-20220305-120228_Pis_006892_concat_pt_GenBank.gb:1 132519 76 / 72 105 % []\n",
      "PAFTOL_008832 GeSeqJob-20220305-120228_Pis_008832_concat_pt_GenBank.gb:1 99735 65 / 60 108 % []\n",
      "PAFTOL_026123 GeSeqJob-20220305-120228_Pis_026123_concat_pt_GenBank.gb:1 47727 36 / 29 124 % []\n",
      "PAFTOL_010597 GeSeqJob-20220305-120228_Pis_010597_concat_pt_GenBank.gb:1 75894 49 / 48 102 % []\n",
      "PAFTOL_006700 GeSeqJob-20220305-120228_Pis_006700_concat_pt_GenBank.gb:1 66171 58 / 49 118 % []\n",
      "PAFTOL_008745 GeSeqJob-20220305-120228_Pis_008745_concat_pt_GenBank.gb:1 131878 78 / 78 100 % []\n",
      "PAFTOL_011288 GeSeqJob-20220305-120228_Pis_011288_concat_pt_GenBank.gb:1 108929 77 / 75 102 % []\n",
      "PAFTOL_016791 GeSeqJob-20220305-120228_Pis_016791_concat_pt_GenBank.gb:1 39578 27 / 24 112 % []\n",
      "PAFTOL_006376 GeSeqJob-20220305-120228_Pis_006376_concat_pt_GenBank.gb:1 73230 68 / 47 144 % []\n",
      "PAFTOL_016651 GeSeqJob-20220305-120228_Pis_016651_concat_pt_GenBank.gb:1 90353 61 / 47 129 % []\n",
      "PAFTOL_008344 GeSeqJob-20220305-120228_Pis_008344_concat_pt_GenBank.gb:1 38618 39 / 18 216 % []\n",
      "PAFTOL_007444 GeSeqJob-20220305-120228_Pis_007444_concat_pt_GenBank.gb:1 128904 77 / 75 102 % ['ndhB']\n",
      "PAFTOL_008504 GeSeqJob-20220305-120228_Pis_008504_concat_pt_GenBank.gb:1 53264 56 / 38 147 % ['rps16']\n",
      "PAFTOL_010553 GeSeqJob-20220305-120228_Pis_010553_concat_pt_GenBank.gb:1 78739 69 / 58 118 % []\n",
      "PAFTOL_015261 GeSeqJob-20220305-120228_Pis_015261_concat_pt_GenBank.gb:1 45371 27 / 23 117 % []\n",
      "PAFTOL_005317 GeSeqJob-20220305-120228_Pis_005317_concat_pt_GenBank.gb:1 24492 21 / 14 150 % []\n",
      "PAFTOL_005677 GeSeqJob-20220305-120228_Pis_005677_concat_pt_GenBank.gb:1 50991 31 / 28 110 % ['ndhI' 'psbK']\n",
      "PAFTOL_026053 GeSeqJob-20220305-120228_Pis_026053_concat_pt_GenBank.gb:1 114559 72 / 69 104 % []\n",
      "PAFTOL_016303 GeSeqJob-20220305-120228_Pis_016303_concat_pt_GenBank.gb:1 47067 52 / 31 167 % []\n",
      "PAFTOL_001347 GeSeqJob-20220305-120228_Pis_001347_concat_pt_GenBank.gb:1 131495 78 / 77 101 % []\n",
      "PAFTOL_011485 GeSeqJob-20220305-120228_Pis_011485_concat_pt_GenBank.gb:1 85832 65 / 61 106 % []\n",
      "PAFTOL_010109 GeSeqJob-20220305-120228_Pis_010109_concat_pt_GenBank.gb:1 92674 72 / 66 109 % []\n",
      "PAFTOL_004103 GeSeqJob-20220305-120228_Pis_004103_concat_pt_GenBank.gb:1 123477 78 / 75 104 % []\n",
      "PAFTOL_023513 GeSeqJob-20220305-120228_Pis_023513_concat_pt_GenBank.gb:1 125576 70 / 66 106 % []\n",
      "PAFTOL_015563 GeSeqJob-20220305-120228_Pis_015563_concat_pt_GenBank.gb:1 133919 75 / 75 100 % []\n",
      "issue with  GeSeqJob-20220305-120228_Pis_015563_concat_pt_GenBank.gb\n",
      "PAFTOL_006522 GeSeqJob-20220305-120228_Pis_006522_concat_pt_GenBank.gb:1 146883 77 / 76 101 % ['rpl36']\n",
      "PAFTOL_011251 GeSeqJob-20220305-120228_Pis_011251_concat_pt_GenBank.gb:"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\ProgramData\\Anaconda3\\lib\\site-packages\\Bio\\GenBank\\__init__.py:1177: BiopythonParserWarning: Should have at least 2 parts for compound location\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 100706 73 / 65 112 % []\n",
      "PAFTOL_016943 GeSeqJob-20220305-120228_Pis_016943_concat_pt_GenBank.gb:1 83361 63 / 59 106 % ['matK']\n",
      "PAFTOL_006738 GeSeqJob-20220305-120228_Pis_006738_concat_pt_GenBank.gb:1 45754 39 / 28 139 % []\n",
      "PAFTOL_013403 GeSeqJob-20220305-120228_Pis_013403_concat_pt_GenBank.gb:1 114540 63 / 62 101 % []\n",
      "issue with  GeSeqJob-20220305-120228_Pis_013403_concat_pt_GenBank.gb\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\ProgramData\\Anaconda3\\lib\\site-packages\\Bio\\GenBank\\__init__.py:1265: BiopythonParserWarning: Couldn't parse feature location: 'join(113277..113667,join(114322..114540,1..155))'\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "# Rename and modify genbank files, convert to embl flat file\n",
    "if not os.path.exists(sdir + '/Genbank_files/'):\n",
    "    os.makedirs(sdir + '/Genbank_files/')\n",
    "gb_files = [file for file in os.listdir(sdir + 'geseq/') if file.endswith('.gb')]\n",
    "print(len(gb_files))\n",
    "for idx, row in submit_pt.iterrows():\n",
    "    idx_file = [i for i, x in enumerate([row.SeqID in file for file in gb_files]) if x]\n",
    "    print(row.Sample_Name,gb_files[idx_file[0]],end=':')\n",
    "    records = list(SeqIO.parse(sdir + 'geseq/' + gb_files[idx_file[0]], \"genbank\"))\n",
    "    record = records[0]\n",
    "    record_genes = list_genes(record); record_genes = [gene for gene in record_genes if gene in genes_82.gene.to_list()]\n",
    "    blast_genes = all_seq[all_seq.Sample_Name==row.Sample_Name]\n",
    "    print(len(records), len(records[0]), len(record_genes),'/',blast_genes.gene.nunique(),\n",
    "         int(len(record_genes)/blast_genes.gene.nunique()*100),'%',blast_genes[blast_genes.gene.isin(record_genes)==False].gene.unique())\n",
    "    try:\n",
    "        if row.Assembly in ['circular genome','circular genome with gaps']:\n",
    "            record.description = row.ENARunNumber + ' chloroplast, complete genome'\n",
    "        elif row.Assembly in ['scaffold']:\n",
    "            record.description = row.ENARunNumber + ' chloroplast, partial genome (concatenated contigs)'\n",
    "            record.annotations['topology'] = 'linear'\n",
    "        submit_pt.loc[idx,'ENA_SeqID'] = row.ENARunNumber + '_pt'\n",
    "        record.id = submit_pt.loc[idx,'ENA_SeqID']\n",
    "        record.name = submit_pt.loc[idx,'ENA_SeqID']\n",
    "        record.version = submit_pt.loc[idx,'ENA_SeqID'] + '.1'\n",
    "        record.annotations['source'] = row.ENASampleNum + ', ' + row.sci_name + ', isolate'\n",
    "        record.annotations['accessions'] = row.ENASampleNum\n",
    "        record.annotations['organism'] = row.NCBI_sciname\n",
    "        record.annotations[\"data_file_division\"]=\"PLN\";\n",
    "        record.annotations.pop('taxonomy')\n",
    "        record.annotations['references'][0].title = paper_title    \n",
    "        record.annotations['references'][0].authors = paper_authors    \n",
    "        record.annotations['references'][0].journal = paper_journal      \n",
    "        record.annotations['references'][0].pubmed_id = paper_id    \n",
    "        record.features[0].qualifiers['organism'] = row.NCBI_sciname\n",
    "        record.features[0].qualifiers['db_xref'] = ['taxon:' + str(int(row['NCBI_TaxID']))]\n",
    "        SeqIO.write(record,sdir + 'Genbank_files/' + row.SeqID + '_pt.gb',format='genbank')\n",
    "    except:\n",
    "        print('issue with ',gb_files[idx_file[0]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def del_features(record, feat_del_ls = ['translation','info','annotator']):\n",
    "    for feature in record.features:\n",
    "        for feat_to_del in feat_del_ls:\n",
    "            if feat_to_del in feature.qualifiers:\n",
    "                del feature.qualifiers[feat_to_del]\n",
    "    return record\n",
    "\n",
    "def del_fragment(record, ls_types = ['gene','CDS','rRNA']):\n",
    "    # list gene names\n",
    "    ls_feat = []\n",
    "    for feature in record.features:\n",
    "        if feature.type in ls_types:\n",
    "            ls_feat.append(feature.qualifiers['gene'][0])\n",
    "    ls_feat = [ifeat for ifeat in ls_feat if 'fragment' not in ifeat]\n",
    "\n",
    "    # remove fragment feature if full feature is present\n",
    "    rm_count = 0\n",
    "    for feature in record.features:\n",
    "        if feature.type in ls_types:\n",
    "            if 'fragment' in feature.qualifiers['gene'][0]:\n",
    "                if feature.qualifiers['gene'][0].split('-')[0] in ls_feat:\n",
    "                    rm_count += 1\n",
    "                    record.features.remove(feature)\n",
    "    print('removed',rm_count,'partial features')\n",
    "    return record\n",
    "\n",
    "def del_longfeat(record, ls_types = ['gene','CDS','rRNA'], len_thrs = 5000):\n",
    "    for feature in record.features:\n",
    "        if feature.type in ls_types:\n",
    "            if len(feature)>len_thrs:\n",
    "                print('removing ',feature.qualifiers['gene'],len(feature))\n",
    "                record.features.remove(feature)\n",
    "    return record\n",
    "\n",
    "def add_gap_features(record, gapN = 100):\n",
    "    # find gaps in record\n",
    "    gaps_idx = [m.start() for m in re.finditer(gapN * 'N', str(record.seq))]\n",
    "    print(len(gaps_idx),'gaps')\n",
    "    if len(gaps_idx)>0:\n",
    "        ## for each gap, add a sequence feature\n",
    "        for gap_pos in gaps_idx:\n",
    "            # find index of gap in features\n",
    "            features_start = [feat.location.start for feat in record.features]\n",
    "            gap_feature_idx = [i for i in range(len(features_start)) if features_start[i] > gap_pos]\n",
    "            \n",
    "            gap_feat = SeqFeature(FeatureLocation(gap_pos,gap_pos+gapN), strand=1, type='assembly_gap')\n",
    "            gap_feat.qualifiers['gap_type']='within scaffold'; \n",
    "            gap_feat.qualifiers['estimated_length']='unknown'\n",
    "            gap_feat.qualifiers['linkage_evidence']='unspecified'\n",
    "            if len(gap_feature_idx)>0:\n",
    "                gap_feature_idx = gap_feature_idx[0]\n",
    "                record.features.insert(gap_feature_idx, gap_feat)\n",
    "            elif len(gap_feature_idx)==0: #if position is further than any other feature, append at the end of feature list\n",
    "                record.features.append(gap_feat)\n",
    "    return record\n",
    "\n",
    "# https://ena-docs.readthedocs.io/en/latest/faq/locus_tags.html\n",
    "def add_locus_tag(record, locus_tag, ls_types = ['gene','CDS','rRNA','tRNA','intron','exon']):\n",
    "    locus_tags_genes = {}; idx = 1\n",
    "    for feature in record.features:\n",
    "        if feature.type in ls_types:\n",
    "            if feature.qualifiers['gene'][0] not in locus_tags_genes:\n",
    "                locus_tags_genes[feature.qualifiers['gene'][0]] = locus_tag + '_' + str(idx)\n",
    "                idx += 1\n",
    "            feature.qualifiers['locus_tag']=locus_tags_genes[feature.qualifiers['gene'][0]]\n",
    "    return record\n",
    "\n",
    "def gzip_file(in_file):\n",
    "    import gzip\n",
    "    import shutil\n",
    "    f_gz = in_file + '.gz'\n",
    "    with open(in_file, 'rb') as f_in:\n",
    "        with gzip.open(f_gz, 'wb') as f_out:\n",
    "            shutil.copyfileobj(f_in, f_out)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pis_007905 ERR7619260_pt > 27886\n",
      "removed 0 partial features\n",
      "removed 0 partial features\n",
      "29 gaps\n",
      "Pis_014479 ERR7621142_pt > 39876\n",
      "removed 0 partial features\n",
      "removed 0 partial features\n",
      "29 gaps\n",
      "Pis_006622 ERR7618796_pt > 158632\n",
      "removed 2 partial features\n",
      "removed 2 partial features\n",
      "removing  ['ycf2'] 56526\n",
      "removing  ['ndhB'] 49707\n",
      "removing  ['trnI-GAU'] 50686\n",
      "removing  ['trnA-UGC'] 50527\n",
      "removing  ['rrn23'] 49689\n",
      "8 gaps\n",
      "Pis_010719 ERR7620148_pt > 56570\n",
      "removed 0 partial features\n",
      "removed 0 partial features\n",
      "25 gaps\n",
      "Pis_012011 ERR7620451_pt > 127768\n",
      "removed 1 partial features\n",
      "removed 1 partial features\n",
      "7 gaps\n",
      "Pis_004680 ERR5033456_pt > 63389\n",
      "removed 1 partial features\n",
      "removed 0 partial features\n",
      "46 gaps\n",
      "Pis_018841 ERR7621734_pt > 58859\n",
      "removed 1 partial features\n",
      "removed 0 partial features\n",
      "43 gaps\n",
      "Pis_015265 ERR7621379_pt > 76256\n",
      "removed 1 partial features\n",
      "removed 0 partial features\n",
      "42 gaps\n",
      "Pis_000975 ERR4180144_pt > 39785\n",
      "removed 0 partial features\n",
      "removed 0 partial features\n",
      "17 gaps\n",
      "Pis_004459 ERR4009378_pt > 89459\n",
      "removed 0 partial features\n",
      "removed 0 partial features\n",
      "46 gaps\n",
      "Pis_006015 ERR5034809_pt > 51057\n",
      "removed 0 partial features\n",
      "removed 0 partial features\n",
      "40 gaps\n",
      "Pis_011083 ERR7620245_pt > 57118\n",
      "removed 0 partial features\n",
      "removed 0 partial features\n",
      "38 gaps\n",
      "Pis_012079 ERR7620467_pt > 158868\n",
      "removed 2 partial features\n",
      "removed 2 partial features\n",
      "2 gaps\n",
      "Pis_000972 ERR4180000_pt > 107373\n",
      "removed 1 partial features\n",
      "removed 0 partial features\n",
      "31 gaps\n",
      "Pis_007253 ERR7619149_pt > 113084\n",
      "removed 1 partial features\n",
      "removed 1 partial features\n",
      "7 gaps\n",
      "Pis_003041 ERR5033201_pt > 136390\n",
      "removed 5 partial features\n",
      "removed 3 partial features\n",
      "11 gaps\n",
      "Pis_004743 ERR5033487_pt > 159189\n",
      "removed 1 partial features\n",
      "removed 1 partial features\n",
      "1 gaps\n",
      "Pis_015149 ERR9230181_pt > 90674\n",
      "removed 4 partial features\n",
      "removed 3 partial features\n",
      "42 gaps\n",
      "Pis_015615 ERR9230065_pt > 133179\n",
      "removed 12 partial features\n",
      "removed 12 partial features\n",
      "11 gaps\n",
      "Pis_004371 ERR9229812_pt > 94647\n",
      "removed 3 partial features\n",
      "removed 2 partial features\n",
      "removing  ['trnK-UUU'] 92137\n",
      "36 gaps\n",
      "Pis_007973 ERR7619284_pt > 32397\n",
      "removed 0 partial features\n",
      "removed 0 partial features\n",
      "removing  ['atpF'] 30317\n",
      "31 gaps\n",
      "Pis_004477 ERR4009387_pt > 44952\n",
      "removed 1 partial features\n",
      "removed 1 partial features\n",
      "44 gaps\n",
      "Pis_029857 ERR9230283_pt > 31862\n",
      "removed 1 partial features\n",
      "removed 1 partial features\n",
      "28 gaps\n",
      "Pis_004257 ERR5033391_pt > 61698\n",
      "removed 0 partial features\n",
      "removed 0 partial features\n",
      "41 gaps\n",
      "Pis_021529 ERR7622080_pt > 40359\n",
      "removed 4 partial features\n",
      "removed 4 partial features\n",
      "37 gaps\n",
      "Pis_006111 ERR4180107_pt > 129050\n",
      "removed 0 partial features\n",
      "removed 0 partial features\n",
      "16 gaps\n",
      "Pis_010453 ERR7619986_pt > 39215\n",
      "removed 1 partial features\n",
      "removed 1 partial features\n",
      "32 gaps\n",
      "Pis_019329 ERR7621926_pt > 76214\n",
      "removed 1 partial features\n",
      "removed 1 partial features\n",
      "52 gaps\n",
      "Pis_019293 ERR7621913_pt > 111535\n",
      "removed 0 partial features\n",
      "removed 0 partial features\n",
      "39 gaps\n",
      "Pis_016305 ERR7621540_pt > 27702\n",
      "removed 1 partial features\n",
      "removed 1 partial features\n",
      "29 gaps\n",
      "Pis_002628 ERR7617970_pt > 125283\n",
      "removed 3 partial features\n",
      "removed 2 partial features\n",
      "removing  ['rps12'] 17205\n",
      "4 gaps\n",
      "Pis_011889 ERR7620418_pt > 53264\n",
      "removed 0 partial features\n",
      "removed 0 partial features\n",
      "34 gaps\n",
      "Pis_012985 ERR9229779_pt > issue Pis_012985\n",
      "Pis_023435 ERR9229778_pt > 74401\n",
      "removed 0 partial features\n",
      "removed 0 partial features\n",
      "55 gaps\n",
      "Pis_014813 ERR7621203_pt > 40013\n",
      "removed 2 partial features\n",
      "removed 0 partial features\n",
      "removing  ['trnK-UUU'] 37510\n",
      "29 gaps\n",
      "Pis_004390 ERR5033418_pt > 70143\n",
      "removed 0 partial features\n",
      "removed 0 partial features\n",
      "56 gaps\n",
      "Pis_008709 ERR7619660_pt > issue Pis_008709\n",
      "Pis_010063 ERR5970564_pt > 55109\n",
      "removed 0 partial features\n",
      "removed 0 partial features\n",
      "58 gaps\n",
      "Pis_006252 ERR5034821_pt > 46378\n",
      "removed 1 partial features\n",
      "removed 1 partial features\n",
      "44 gaps\n",
      "Pis_011879 ERR7620413_pt > 107978\n",
      "removed 2 partial features\n",
      "removed 2 partial features\n",
      "34 gaps\n",
      "Pis_006957 ERR7619054_pt > 23966\n",
      "removed 0 partial features\n",
      "removed 0 partial features\n",
      "7 gaps\n",
      "Pis_004696 ERR5033465_pt > 132490\n",
      "removed 4 partial features\n",
      "removed 4 partial features\n",
      "16 gaps\n",
      "Pis_014801 ERR7621197_pt > 36829\n",
      "removed 1 partial features\n",
      "removed 1 partial features\n",
      "18 gaps\n",
      "Pis_015539 ERR9230036_pt > 117894\n",
      "removed 4 partial features\n",
      "removed 3 partial features\n",
      "9 gaps\n",
      "Pis_008650 ERR7619624_pt > 42928\n",
      "removed 0 partial features\n",
      "removed 0 partial features\n",
      "25 gaps\n",
      "Pis_007942 ERR7619461_pt > 36090\n",
      "removed 1 partial features\n",
      "removed 1 partial features\n",
      "7 gaps\n",
      "Pis_005730 ERR5033763_pt > 55075\n",
      "removed 0 partial features\n",
      "removed 0 partial features\n",
      "50 gaps\n",
      "Pis_008711 ERR7619662_pt > 40958\n",
      "removed 0 partial features\n",
      "removed 0 partial features\n",
      "27 gaps\n",
      "Pis_005727 ERR7618467_pt > 125590\n",
      "removed 3 partial features\n",
      "removed 3 partial features\n",
      "13 gaps\n",
      "Pis_006960 ERR5034088_pt > 52611\n",
      "removed 0 partial features\n",
      "removed 0 partial features\n",
      "22 gaps\n",
      "Pis_014513 ERR7621158_pt > 42659\n",
      "removed 1 partial features\n",
      "removed 1 partial features\n",
      "35 gaps\n",
      "Pis_006135 ERR9229895_pt > 74580\n",
      "removed 0 partial features\n",
      "removed 0 partial features\n",
      "46 gaps\n",
      "Pis_002655 ERR7618030_pt > 55551\n",
      "removed 1 partial features\n",
      "removed 1 partial features\n",
      "22 gaps\n",
      "Pis_019233 ERR7621895_pt > 125563\n",
      "removed 0 partial features\n",
      "removed 0 partial features\n",
      "9 gaps\n",
      "Pis_015763 ERR9230114_pt > issue Pis_015763\n",
      "Pis_007313 ERR7619182_pt > 116079\n",
      "removed 1 partial features\n",
      "removed 0 partial features\n",
      "36 gaps\n",
      "Pis_001072 ERR5033229_pt > 55418\n",
      "removed 1 partial features\n",
      "removed 1 partial features\n",
      "17 gaps\n",
      "Pis_006292 ERR5034827_pt > 124260\n",
      "removed 1 partial features\n",
      "removed 1 partial features\n",
      "17 gaps\n",
      "Pis_015271 ERR7621381_pt > 63837\n",
      "removed 1 partial features\n",
      "removed 1 partial features\n",
      "39 gaps\n",
      "Pis_006145 ERR9229900_pt > 117306\n",
      "removed 0 partial features\n",
      "removed 0 partial features\n",
      "34 gaps\n",
      "Pis_010261 ERR5034522_pt > 127697\n",
      "removed 2 partial features\n",
      "removed 2 partial features\n",
      "7 gaps\n",
      "Pis_015595 ERR9230056_pt > 127330\n",
      "removed 15 partial features\n",
      "removed 12 partial features\n",
      "12 gaps\n",
      "Pis_005892 ERR5033812_pt > 39093\n",
      "removed 0 partial features\n",
      "removed 0 partial features\n",
      "44 gaps\n",
      "Pis_006892 ERR7619009_pt > 132519\n",
      "removed 1 partial features\n",
      "removed 1 partial features\n",
      "23 gaps\n",
      "Pis_008832 ERR5034427_pt > 99735\n",
      "removed 0 partial features\n",
      "removed 0 partial features\n",
      "26 gaps\n",
      "Pis_026123 ERR7622528_pt > 47727\n",
      "removed 0 partial features\n",
      "removed 0 partial features\n",
      "44 gaps\n",
      "Pis_010597 ERR7620100_pt > 75894\n",
      "removed 1 partial features\n",
      "removed 1 partial features\n",
      "1 gaps\n",
      "Pis_006700 ERR7618821_pt > 66171\n",
      "removed 0 partial features\n",
      "removed 0 partial features\n",
      "62 gaps\n",
      "Pis_008745 ERR5034409_pt > 131878\n",
      "removed 0 partial features\n",
      "removed 0 partial features\n",
      "6 gaps\n",
      "Pis_011288 ERR5034333_pt > 108929\n",
      "removed 0 partial features\n",
      "removed 0 partial features\n",
      "30 gaps\n",
      "Pis_016791 ERR7621612_pt > 39578\n",
      "removed 2 partial features\n",
      "removed 1 partial features\n",
      "15 gaps\n",
      "Pis_006376 ERR7618650_pt > 73230\n",
      "removed 0 partial features\n",
      "removed 0 partial features\n",
      "52 gaps\n",
      "Pis_016651 ERR7621559_pt > 90353\n",
      "removed 1 partial features\n",
      "removed 1 partial features\n",
      "58 gaps\n",
      "Pis_008344 ERR7619354_pt > 38618\n",
      "removed 0 partial features\n",
      "removed 0 partial features\n",
      "34 gaps\n",
      "Pis_007444 ERR9229784_pt > 128904\n",
      "removed 0 partial features\n",
      "removed 0 partial features\n",
      "removing  ['clpP1'] 63787\n",
      "9 gaps\n",
      "Pis_008504 ERR5034961_pt > 53264\n",
      "removed 2 partial features\n",
      "removed 2 partial features\n",
      "40 gaps\n",
      "Pis_010553 ERR7620083_pt > 78739\n",
      "removed 1 partial features\n",
      "removed 1 partial features\n",
      "49 gaps\n",
      "Pis_015261 ERR7621377_pt > 45371\n",
      "removed 0 partial features\n",
      "removed 0 partial features\n",
      "32 gaps\n",
      "Pis_005317 ERR7618337_pt > 24492\n",
      "removed 0 partial features\n",
      "removed 0 partial features\n",
      "9 gaps\n",
      "Pis_005677 ERR7618431_pt > 50991\n",
      "removed 0 partial features\n",
      "removed 0 partial features\n",
      "43 gaps\n",
      "Pis_026053 ERR7622505_pt > 114559\n",
      "removed 0 partial features\n",
      "removed 0 partial features\n",
      "28 gaps\n",
      "Pis_016303 ERR7621539_pt > 47067\n",
      "removed 0 partial features\n",
      "removed 0 partial features\n",
      "38 gaps\n",
      "Pis_001347 ERR4180145_pt > 131495\n",
      "removed 1 partial features\n",
      "removed 1 partial features\n",
      "9 gaps\n",
      "Pis_011485 ERR7620304_pt > 85832\n",
      "removed 2 partial features\n",
      "removed 1 partial features\n",
      "40 gaps\n",
      "Pis_010109 ERR7619941_pt > 92674\n",
      "removed 1 partial features\n",
      "removed 1 partial features\n",
      "46 gaps\n",
      "Pis_004103 ERR5034666_pt > 123477\n",
      "removed 2 partial features\n",
      "removed 2 partial features\n",
      "25 gaps\n",
      "Pis_023513 ERR9230211_pt > 125576\n",
      "removed 1 partial features\n",
      "removed 1 partial features\n",
      "39 gaps\n",
      "Pis_015563 ERR9230044_pt > issue Pis_015563\n",
      "Pis_006522 ERR7618747_pt > 146883\n",
      "removed 1 partial features\n",
      "removed 1 partial features\n",
      "22 gaps\n",
      "Pis_011251 ERR9229930_pt > 100706\n",
      "removed 2 partial features\n",
      "removed 2 partial features\n",
      "45 gaps\n",
      "Pis_016943 ERR7621644_pt > 83361\n",
      "removed 0 partial features\n",
      "removed 0 partial features\n",
      "42 gaps\n",
      "Pis_006738 ERR7618852_pt > 45754\n",
      "removed 0 partial features\n",
      "removed 0 partial features\n",
      "26 gaps\n",
      "Pis_013403 ERR7620893_pt > issue Pis_013403\n"
     ]
    }
   ],
   "source": [
    "# ENA files\n",
    "if not os.path.exists(sdir + '/EMBL_files/'):\n",
    "    os.makedirs(sdir + '/EMBL_files/')\n",
    "commands = []\n",
    "for idx, row in submit_pt.iterrows():\n",
    "    ### Convert .gb to EMBL flat files\n",
    "    file_gb = sdir + 'Genbank_files/' + row.SeqID + '_pt.gb'\n",
    "    print(row.SeqID,row.ENA_SeqID,end=' > ')\n",
    "    try:\n",
    "        record = list(SeqIO.parse(file_gb, \"genbank\"))[0]\n",
    "        print(len(record))\n",
    "\n",
    "        # Deleted features preventing ena submission\n",
    "        record_clean = del_features(record = record, feat_del_ls = ['translation','info','annotator'])\n",
    "        # Delete fragmentary annotations (e.g. rr16-fragment)\n",
    "        record_clean = del_fragment(record = record_clean); record_clean = del_fragment(record = record_clean);\n",
    "        # Delete abnormaly long genes\n",
    "        record_clean = del_longfeat(record = record_clean,len_thrs=10000);\n",
    "        # Add gap features for partial plastomes. Recognises gaps of 100 Ns by default\n",
    "        record_clean = add_gap_features(record = record_clean)\n",
    "        # Add locus tag if sequence has gaps\n",
    "        if 'assembly_gap' in [feature.type for feature in record.features]:\n",
    "              record_clean = add_locus_tag(record = record_clean, locus_tag = project_locus_tag)\n",
    "\n",
    "        file_embl = sdir + '/EMBL_files/' + row.ENA_SeqID + '.embl'\n",
    "        SeqIO.write(record_clean, file_gb, \"genbank\")\n",
    "        SeqIO.write(record_clean, file_embl, \"embl\")\n",
    "        gzip_file(file_embl)\n",
    "\n",
    "        ### Chromosome list file\n",
    "        # https://ena-docs.readthedocs.io/en/latest/submit/fileprep/assembly.html?highlight=chloroplast#chromosome-list-file\n",
    "        if row.Assembly in ['circular genome','circular genome with gaps']:\n",
    "            list_file=row.ENA_SeqID + '\tPT\tCircular-Chromosome\tChloroplast'       \n",
    "        elif row.Assembly in ['scaffold']:\n",
    "            list_file=row.ENA_SeqID + '\tPT\tLinear-Chromosome\tChloroplast'\n",
    "        with open(file_embl.replace('.embl','_CHR.txt'), 'wb') as f:\n",
    "            f.write(list_file.encode(\"ascii\"))\n",
    "        gzip_file(file_embl.replace('.embl','_CHR.txt'))\n",
    "\n",
    "        ### Manifest file \n",
    "        # https://ena-docs.readthedocs.io/en/latest/submit/assembly/genome.html \n",
    "        manifest_file = file_embl.replace('.embl','_manifest.txt')\n",
    "        if row.Assembly in ['circular genome','circular genome with gaps']:\n",
    "            tmp_description=' chloroplast, complete genome' \n",
    "        elif row.Assembly in ['scaffold']:\n",
    "            tmp_description=' chloroplast, partial genome (concatenated contigs)' \n",
    "        manifest={\n",
    "            'STUDY':ENA_project_number,\n",
    "            'SAMPLE':row.ENASampleNum,\n",
    "            'ASSEMBLYNAME':row.ENA_SeqID,\n",
    "            'ASSEMBLY_TYPE':'clone or isolate',\n",
    "            'COVERAGE':row.Coverage_base,\n",
    "            'PROGRAM':'GetOrganelle v1.7.5 +  GeSeq 2.03',\n",
    "            'PLATFORM':'ILLUMINA',\n",
    "            'MOLECULETYPE':'genomic DNA',\n",
    "            'FLATFILE': row.ENA_SeqID + '.embl.gz',\n",
    "            'CHROMOSOME_LIST': row.ENA_SeqID + '_CHR.txt.gz',\n",
    "            'RUN_REF': row.ENARunNumber,\n",
    "            'DESCRIPTION': row.NCBI_sciname + tmp_description\n",
    "        }\n",
    "#         if row.Assembly in ['scaffold']:\n",
    "#             manifest['PARTIAL']='TRUE'\n",
    "        manifest = pd.DataFrame.from_dict(manifest,orient='index')\n",
    "        manifest.to_csv(manifest_file,header=None,sep='\\t')\n",
    "\n",
    "        # Submission command\n",
    "        commands.append('java -jar C:/Data/PAFTOL/ENA_submissions/webin-cli-4.3.0.jar ' + \n",
    "              '-username Webin-52995 -passwordFile C:/Data/PAFTOL/ENA_submissions/ena_pwd.txt -context genome -manifest ' +\n",
    "              row.ENA_SeqID + '_manifest.txt' + ' -validate') \n",
    "    except:\n",
    "        print('issue',row.SeqID)\n",
    "pd.DataFrame(commands).to_csv(sdir + '/ENA_Submission_Commands_' + DataSource + '.txt',index=False,header=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "cd C:\\Data\\PAFTOL\\cpDNA\\ENA_submissions\\DataSource\\EMBL_files\n",
    "\n",
    "then launch commands"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
