{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "measured-trade",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import SeqIO, AlignIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import generic_dna, Gapped\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "seventh-gentleman",
   "metadata": {},
   "source": [
    "https://www.biostars.org/p/430724/\n",
    "\n",
    "https://github.com/michaelgruenstaeudl/annonex2embl\n",
    "\n",
    "https://biopython.org/docs/1.75/api/Bio.Seq.html\n",
    "\n",
    "https://biopython.org/docs/1.75/api/Bio.SeqRecord.html\n",
    "\n",
    "https://ena-docs.readthedocs.io/en/latest/submit/assembly/genome.html \n",
    "\n",
    "https://www.programcreek.com/python/example/114216/Bio.SeqFeature.SeqFeature"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dangerous-stockholm",
   "metadata": {},
   "source": [
    "# Load samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "fluid-poker",
   "metadata": {},
   "outputs": [],
   "source": [
    "wdir = 'cpTree_v6'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "pursuant-rally",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16682 {'REFSEQ': 6020, 'INSDC': 3948, 'PAFTOL': 3535, 'SRA': 1612, 'OneKP': 665, 'Lietal': 549, 'GAP': 343, 'Gymno': 10}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_Name</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>sci_name</th>\n",
       "      <th>DataSource</th>\n",
       "      <th>length</th>\n",
       "      <th>order</th>\n",
       "      <th>class</th>\n",
       "      <th>family</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GAP_026547</td>\n",
       "      <td>Albidella</td>\n",
       "      <td>oligococca</td>\n",
       "      <td>Albidella oligococca</td>\n",
       "      <td>GAP</td>\n",
       "      <td>59917.0</td>\n",
       "      <td>Alismatales</td>\n",
       "      <td>monocots</td>\n",
       "      <td>Alismataceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GAP_026933</td>\n",
       "      <td>Stuckenia</td>\n",
       "      <td>pectinata</td>\n",
       "      <td>Stuckenia pectinata</td>\n",
       "      <td>GAP</td>\n",
       "      <td>31766.0</td>\n",
       "      <td>Alismatales</td>\n",
       "      <td>monocots</td>\n",
       "      <td>Potamogetonaceae</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Sample_Name      genus     species              sci_name DataSource  \\\n",
       "0  GAP_026547  Albidella  oligococca  Albidella oligococca        GAP   \n",
       "1  GAP_026933  Stuckenia   pectinata   Stuckenia pectinata        GAP   \n",
       "\n",
       "    length        order     class            family  \n",
       "0  59917.0  Alismatales  monocots      Alismataceae  \n",
       "1  31766.0  Alismatales  monocots  Potamogetonaceae  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# List of samples\n",
    "cp_df = pd.read_csv(wdir + '/' + wdir + '_tree_clean.csv').drop(columns=['nentries','TaxID','mol_type','Label'])\n",
    "print(cp_df.shape[0],cp_df.DataSource.value_counts().to_dict())\n",
    "cp_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "nasty-storage",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "17187 Index(['idSpecimen', 'idPaftol', 'DataSource', 'Project', 'Order', 'Family',\n",
      "       'Genus', 'Species', 'Taxonomical_Notes', 'KewID', 'Collector',\n",
      "       'CollectorNo', 'VoucherNo', 'BankID', 'MuseumID', 'MuseumName',\n",
      "       'MuseumBarcode', 'HerbcatURL', 'LCD', 'MSB', 'SourceSpecimen',\n",
      "       'country', 'country_code', 'SpecimenReference', 'Gymnosperm',\n",
      "       'idSample', 'MaterialSource', 'AgeOfMaterial', 'Action',\n",
      "       'Sample_Description', 'ExtractionType', 'DNAVolume', 'Quality',\n",
      "       'SampleConcentration', 'NewSampleConcentration', 'GelImage',\n",
      "       'SampleTapeStation', 'Compliant', 'SecENASampleNum', 'idLibrary',\n",
      "       'GenerateLibrary', 'Sonication', 'LibConcentration', 'LibTapeStation',\n",
      "       'Plate No', 'LibQuality', 'RemainingVolume', 'Library_Description',\n",
      "       'Indexes', 'IndexNameFwd', 'SeqFwdPlatform1', 'SeqFwdPlatform2',\n",
      "       'IndexNameRv', 'SeqRv', 'Library_Status', 'Coordinate', 'idSequencing',\n",
      "       'ExternalSequenceID', 'HasDuplicate', 'IsMerged', 'Loc', 'Platform',\n",
      "       'SequencingRun', 'Sequence_Status', 'R1FastqFile', 'R2FastqFile',\n",
      "       'HybridisationPool', 'Blacklisted', 'BlacklistedReason',\n",
      "       'BlacklistedPerson', 'SequencingStrategy', 'ENAExpNumber',\n",
      "       'ENARunNumber', 'NumReads', 'NumRecoveredGenes', 'SumContigLength',\n",
      "       'sci_name', 'Sample', 'ENASampleNum', 'num_seqs', 'sum_len', 'min_len',\n",
      "       'max_len', 'NCBI_TaxID', 'NCBI_sciname'],\n",
      "      dtype='object')\n",
      "14268\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\ProgramData\\Anaconda3\\lib\\site-packages\\IPython\\core\\interactiveshell.py:3147: DtypeWarning: Columns (29,35,45,57,68,77) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n",
      "C:\\Users\\kle10kg\\AppData\\Roaming\\Python\\Python37\\site-packages\\ipykernel_launcher.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>idSpecimen</th>\n",
       "      <th>idPaftol</th>\n",
       "      <th>DataSource</th>\n",
       "      <th>Project</th>\n",
       "      <th>Order</th>\n",
       "      <th>Family</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Species</th>\n",
       "      <th>Taxonomical_Notes</th>\n",
       "      <th>KewID</th>\n",
       "      <th>...</th>\n",
       "      <th>sci_name</th>\n",
       "      <th>Sample</th>\n",
       "      <th>ENASampleNum</th>\n",
       "      <th>num_seqs</th>\n",
       "      <th>sum_len</th>\n",
       "      <th>min_len</th>\n",
       "      <th>max_len</th>\n",
       "      <th>NCBI_TaxID</th>\n",
       "      <th>NCBI_sciname</th>\n",
       "      <th>Sample_Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>PAFTOL</td>\n",
       "      <td>Asteraceae</td>\n",
       "      <td>Asterales</td>\n",
       "      <td>Asteraceae</td>\n",
       "      <td>Gymnolaena</td>\n",
       "      <td>sp.</td>\n",
       "      <td>Initially recorded as Gymnolaena litoralis but...</td>\n",
       "      <td>295689-2</td>\n",
       "      <td>...</td>\n",
       "      <td>Gymnolaena sp. - Jan-93 (K)</td>\n",
       "      <td>564</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PAFTOL_000564</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>PAFTOL</td>\n",
       "      <td>Pilot</td>\n",
       "      <td>Ranunculales</td>\n",
       "      <td>Circaeasteraceae</td>\n",
       "      <td>Circaeaster</td>\n",
       "      <td>agrestis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>167828-1</td>\n",
       "      <td>...</td>\n",
       "      <td>Circaeaster agrestis</td>\n",
       "      <td>961</td>\n",
       "      <td>ERS5501508</td>\n",
       "      <td>332.0</td>\n",
       "      <td>138795.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>1,485</td>\n",
       "      <td>39288.0</td>\n",
       "      <td>Circaeaster agrestis</td>\n",
       "      <td>PAFTOL_000961</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows Ã— 86 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   idSpecimen  idPaftol DataSource     Project         Order  \\\n",
       "0           1         1     PAFTOL  Asteraceae     Asterales   \n",
       "1           2         2     PAFTOL       Pilot  Ranunculales   \n",
       "\n",
       "             Family        Genus   Species  \\\n",
       "0        Asteraceae   Gymnolaena       sp.   \n",
       "1  Circaeasteraceae  Circaeaster  agrestis   \n",
       "\n",
       "                                   Taxonomical_Notes     KewID  ...  \\\n",
       "0  Initially recorded as Gymnolaena litoralis but...  295689-2  ...   \n",
       "1                                                NaN  167828-1  ...   \n",
       "\n",
       "                      sci_name Sample ENASampleNum  num_seqs   sum_len  \\\n",
       "0  Gymnolaena sp. - Jan-93 (K)    564          NaN       NaN       NaN   \n",
       "1         Circaeaster agrestis    961   ERS5501508     332.0  138795.0   \n",
       "\n",
       "  min_len max_len  NCBI_TaxID          NCBI_sciname    Sample_Name  \n",
       "0     NaN     NaN         NaN                   NaN  PAFTOL_000564  \n",
       "1    72.0   1,485     39288.0  Circaeaster agrestis  PAFTOL_000961  \n",
       "\n",
       "[2 rows x 86 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Latest DB\n",
    "db = pd.read_csv('C:/Data/PAFTOL/PAFTOL_DB/2022-01-31_paftol_export_stats.csv')\n",
    "print(db.shape[0],db.columns)\n",
    "# PAFTOL samples only\n",
    "db_A353 = db[db.DataSource.isin(['GAP','PAFTOL'])]\n",
    "db_A353['Sample_Name']= db_A353.DataSource + '_' + db_A353.idSequencing.astype(int).astype('str').str.zfill(6)\n",
    "print(db_A353.shape[0])\n",
    "db_A353.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "plain-pavilion",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16682 {'Sample_Name': 0, 'family': 0, 'genus': 0, 'species': 0, 'sci_name': 0, 'DataSource': 0, 'idPaftol': 12804, 'KewID': 12812, 'idSequencing': 12804, 'ENASampleNum': 13413, 'ENAExpNumber': 13374, 'ENARunNumber': 13374, 'NCBI_TaxID': 12992, 'NCBI_sciname': 12992}\n",
      "3308\n",
      "3165\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_Name</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>sci_name</th>\n",
       "      <th>DataSource</th>\n",
       "      <th>idPaftol</th>\n",
       "      <th>KewID</th>\n",
       "      <th>idSequencing</th>\n",
       "      <th>ENASampleNum</th>\n",
       "      <th>ENAExpNumber</th>\n",
       "      <th>ENARunNumber</th>\n",
       "      <th>NCBI_TaxID</th>\n",
       "      <th>NCBI_sciname</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GAP_026547</td>\n",
       "      <td>Alismataceae</td>\n",
       "      <td>Albidella</td>\n",
       "      <td>oligococca</td>\n",
       "      <td>Albidella oligococca</td>\n",
       "      <td>GAP</td>\n",
       "      <td>12214.0</td>\n",
       "      <td>77163628-1</td>\n",
       "      <td>26547.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERX7344886</td>\n",
       "      <td>ERR7776839</td>\n",
       "      <td>2890917</td>\n",
       "      <td>Albidella oligococca</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GAP_026933</td>\n",
       "      <td>Potamogetonaceae</td>\n",
       "      <td>Stuckenia</td>\n",
       "      <td>pectinata</td>\n",
       "      <td>Stuckenia pectinata</td>\n",
       "      <td>GAP</td>\n",
       "      <td>12456.0</td>\n",
       "      <td>77099639-1</td>\n",
       "      <td>26933.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERX7170367</td>\n",
       "      <td>ERR7599787</td>\n",
       "      <td>55444</td>\n",
       "      <td>Stuckenia pectinata</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Sample_Name            family      genus     species              sci_name  \\\n",
       "0  GAP_026547      Alismataceae  Albidella  oligococca  Albidella oligococca   \n",
       "1  GAP_026933  Potamogetonaceae  Stuckenia   pectinata   Stuckenia pectinata   \n",
       "\n",
       "  DataSource  idPaftol       KewID  idSequencing ENASampleNum ENAExpNumber  \\\n",
       "0        GAP   12214.0  77163628-1       26547.0          NaN   ERX7344886   \n",
       "1        GAP   12456.0  77099639-1       26933.0          NaN   ERX7170367   \n",
       "\n",
       "  ENARunNumber  NCBI_TaxID          NCBI_sciname  \n",
       "0   ERR7776839     2890917  Albidella oligococca  \n",
       "1   ERR7599787       55444   Stuckenia pectinata  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merge dataframes\n",
    "## Q. Add voucher info?\n",
    "cp_df = pd.merge(cp_df[['Sample_Name', 'family', 'genus', 'species', 'sci_name', 'DataSource']],\n",
    "                  db_A353[['Sample_Name','idPaftol','KewID','idSequencing','ENASampleNum','ENAExpNumber','ENARunNumber','NCBI_TaxID','NCBI_sciname']],\n",
    "                  how='left',on='Sample_Name')\n",
    "print(cp_df.shape[0], cp_df.isna().sum().to_dict())\n",
    "A353_df = cp_df[cp_df.ENARunNumber.notnull()]\n",
    "print(A353_df.shape[0])\n",
    "A353_df = A353_df[A353_df.NCBI_TaxID.notnull()].astype({'NCBI_TaxID':'int'})\n",
    "print(A353_df.shape[0])\n",
    "A353_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "affected-induction",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3164\n",
      "200\n"
     ]
    }
   ],
   "source": [
    "#tmp \n",
    "removed_taxa = ['Murraya exotica']\n",
    "A353_df = A353_df[A353_df.sci_name.isin(removed_taxa)==False]\n",
    "print(A353_df.shape[0])\n",
    "A353_df = A353_df[:200]\n",
    "print(A353_df.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "massive-belgium",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "79 {'Gene': 0, 'Product': 75, 'type': 75}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Product</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rbcL</td>\n",
       "      <td>ribulose 1,5-bisphosphate carboxylase/oxygenas...</td>\n",
       "      <td>cds</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rrn16</td>\n",
       "      <td>16S ribosomal RNA</td>\n",
       "      <td>rRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>accD</td>\n",
       "      <td>AccD</td>\n",
       "      <td>cds</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>matK</td>\n",
       "      <td>matK</td>\n",
       "      <td>cds</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Gene                                            Product  type\n",
       "0   rbcL  ribulose 1,5-bisphosphate carboxylase/oxygenas...   cds\n",
       "1  rrn16                                  16S ribosomal RNA  rRNA\n",
       "2   accD                                               AccD   cds\n",
       "3   matK                                               matK   cds"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# List of genes\n",
    "genes_df = pd.read_csv('cp_genes_all.csv')\n",
    "print(genes_df.shape[0],genes_df.isna().sum().to_dict())\n",
    "# tmp \n",
    "genes_df = genes_df[:4]\n",
    "genes_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "developmental-freeze",
   "metadata": {},
   "source": [
    "# Processing Alignment and generation of EMBL file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "rural-lingerie",
   "metadata": {},
   "outputs": [],
   "source": [
    "def gzip_file(in_file):\n",
    "    import gzip\n",
    "    import shutil\n",
    "    f_gz = in_file + '.gz'\n",
    "    with open(in_file, 'rb') as f_in:\n",
    "        with gzip.open(f_gz, 'wb') as f_out:\n",
    "            shutil.copyfileobj(f_in, f_out)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "progressive-sixth",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--aatgca\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Seq('atgca')"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def cut_to_codon(record):\n",
    "    # Get index of first nucleotide\n",
    "    pos_ini = len(record.seq) - len(record.seq.lstrip('-')) \n",
    "    # length of leading gap divisible by 3\n",
    "    to_del = ((pos_ini // 3) * 3)\n",
    "    # Update record without leading gap divisible by 3\n",
    "    record.seq = record.seq[to_del:]\n",
    "    # Update position of first nucleotide\n",
    "    pos_ini = len(record.seq) - len(record.seq.lstrip('-'))\n",
    "    # No change if first position\n",
    "    if pos_ini==0:\n",
    "        record.seq = record.seq\n",
    "    # Cut partial codon otherwise\n",
    "    elif pos_ini==1:\n",
    "        record.seq = record.seq[3:]\n",
    "    elif pos_ini==2: \n",
    "        record.seq = record.seq[3:]\n",
    "    return record\n",
    "    \n",
    "rec_test = SeqRecord(Seq('--aatgca'), id='test_seq')\n",
    "print(rec_test.seq)\n",
    "new_rec = cut_to_codon(record=rec_test)\n",
    "new_rec.seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "suitable-shareware",
   "metadata": {},
   "outputs": [],
   "source": [
    "def cut_seq(records,cut):\n",
    "    for record in records:\n",
    "        record.seq = record.seq[cut:]\n",
    "    return records\n",
    "# print(len(alignment[0].seq))\n",
    "# tmp = cut_seq(records = alignment[:2], cut = 1)\n",
    "# print(len(tmp[0].seq))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "quick-connectivity",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "rbcL 14847\n",
      "record_ls: 14847 cp_df: 16682 record_df: 16682\n",
      "coverage: {0.1: 97.8, 0.5: 99.9, 0.9: 100.0}\n",
      "Leadgap refseq: {0.1: 0.0, 0.5: 0.0, 0.9: 0.0}\n",
      "pident refseq: {0.1: 94.19, 0.5: 97.07, 0.9: 99.65}\n",
      "N records: 136\n",
      "java -jar C:/Data/PAFTOL/Organelles/ENA_submissions/webin-cli-4.3.0.jar -username Webin-52995 -passwordFile C:/Data/PAFTOL/Organelles/ENA_submissions/ena_pwd.txt -context sequence -manifest rbcL_A353_manifest.txt -validate\n",
      " \n",
      "rrn16 15731\n",
      "record_ls: 15731 cp_df: 16682 record_df: 16682\n",
      "coverage: {0.1: 100.0, 0.5: 100.0, 0.9: 100.0}\n",
      "Leadgap refseq: {0.1: 0.0, 0.5: 0.0, 0.9: 0.0}\n",
      "pident refseq: {0.1: 99.19, 0.5: 99.87, 0.9: 100.0}\n",
      "N records: 198\n",
      "java -jar C:/Data/PAFTOL/Organelles/ENA_submissions/webin-cli-4.3.0.jar -username Webin-52995 -passwordFile C:/Data/PAFTOL/Organelles/ENA_submissions/ena_pwd.txt -context sequence -manifest rrn16_A353_manifest.txt -validate\n",
      " \n",
      "accD 11135\n",
      "record_ls: 11135 cp_df: 16682 record_df: 16682\n",
      "coverage: {0.1: 69.7, 0.5: 99.9, 0.9: 100.0}\n",
      "Leadgap refseq: {0.1: 63.0, 0.5: 63.0, 0.9: 63.0}\n",
      "pident refseq: {0.1: 90.81, 0.5: 96.025, 0.9: 99.79999999999998}\n",
      "N records: 90\n",
      "java -jar C:/Data/PAFTOL/Organelles/ENA_submissions/webin-cli-4.3.0.jar -username Webin-52995 -passwordFile C:/Data/PAFTOL/Organelles/ENA_submissions/ena_pwd.txt -context sequence -manifest accD_A353_manifest.txt -validate\n",
      " \n",
      "matK 13992\n",
      "record_ls: 13992 cp_df: 16682 record_df: 16682\n",
      "coverage: {0.1: 98.4, 0.5: 100.0, 0.9: 100.0}\n",
      "Leadgap refseq: {0.1: 51.0, 0.5: 68.0, 0.9: 68.0}\n",
      "pident refseq: {0.1: 86.83, 0.5: 94.81, 0.9: 99.47}\n",
      "N records: 73\n",
      "java -jar C:/Data/PAFTOL/Organelles/ENA_submissions/webin-cli-4.3.0.jar -username Webin-52995 -passwordFile C:/Data/PAFTOL/Organelles/ENA_submissions/ena_pwd.txt -context sequence -manifest matK_A353_manifest.txt -validate\n",
      " \n"
     ]
    }
   ],
   "source": [
    "for idx, row in genes_df.iterrows():\n",
    "    ## Load alignment\n",
    "    alignment = AlignIO.read(open(wdir + '/Aligned_genes/' + row.Gene + '_aln.fasta'), \"fasta\", alphabet=Gapped(generic_dna))\n",
    "    print(row.Gene, len(alignment))\n",
    "    \n",
    "    ## Read stats from aligned sequences\n",
    "    record_ls = []; \n",
    "    for record in alignment:\n",
    "        try:\n",
    "            record_ls.append({'Sample_Name':record.id,'coverage':record.description.split('reference_coverage:')[1].split(',')[0],\n",
    "                              'Leadgap':len(record.seq) - len(record.seq.lstrip('-')),'pident':record.description.split('pident:')[1].split(',')[0]})\n",
    "        except:\n",
    "            record_ls.append({'Sample_Name':record.id})\n",
    "            \n",
    "    ## Merge alignment df with samples df      \n",
    "    record_df = pd.merge(cp_df,pd.DataFrame(record_ls).astype({'coverage':'float','pident':'float'}),how='left',on='Sample_Name')\n",
    "    print('record_ls:',len(record_ls),'cp_df:',cp_df.shape[0],'record_df:',record_df.shape[0])\n",
    "    \n",
    "    ## find first refseq with 100% coverage\n",
    "    print('coverage:',record_df.coverage.quantile([.1,.5,.9]).to_dict())\n",
    "    ref100 = record_df[(record_df.DataSource=='REFSEQ') & (record_df.coverage==100.0)]\n",
    "    print('Leadgap refseq:',ref100.Leadgap.quantile([.1,.5,.9]).to_dict())\n",
    "    print('pident refseq:',ref100.pident.quantile([.1,.5,.9]).to_dict())\n",
    "    \n",
    "    ## Cut leading gap\n",
    "    alignment = cut_seq(records = alignment, cut = int(ref100.Leadgap.median()))\n",
    "    \n",
    "    ## Create embl records\n",
    "    record_export = []\n",
    "    for record in alignment:\n",
    "        if record.id in A353_df.Sample_Name.to_list():\n",
    "            rec_sample_dc = record_df[record_df.Sample_Name==record.id].to_dict('records')[0]\n",
    "            new_description = rec_sample_dc['NCBI_sciname'] + ' ' + row.Product\n",
    "            if rec_sample_dc['coverage']==100:\n",
    "                new_description = new_description + ', complete sequence'\n",
    "            else:\n",
    "                new_description += ', partial sequence'\n",
    "            new_description += ', isolate ' + rec_sample_dc['ENARunNumber']\n",
    "            rec = SeqRecord(record.seq, id=rec_sample_dc['ENARunNumber'] + '-' + row.Gene, name=rec_sample_dc['Sample_Name'],\n",
    "                       description= new_description, dbxrefs=['taxon:' + str(int(rec_sample_dc['NCBI_TaxID']))]) #.ungap(\"-\")\n",
    "            rec = SeqRecord(cut_to_codon(record=record).seq.ungap('-'), id=rec_sample_dc['ENARunNumber'] + '-' + row.Gene, name=rec_sample_dc['Sample_Name'],\n",
    "                       description= new_description, dbxrefs=['taxon:' + str(int(rec_sample_dc['NCBI_TaxID']))]) #.ungap(\"-\")\n",
    "            \n",
    "            rec.annotations[\"molecule_type\"]=\"genomic DNA\";\n",
    "            rec.annotations[\"topology\"]=\"linear\";\n",
    "            rec.annotations[\"organism\"]=rec_sample_dc['NCBI_sciname'];\n",
    "            rec.annotations[\"taxon\"]=str(int(rec_sample_dc['NCBI_TaxID']));\n",
    "            rec.annotations[\"data_file_division\"]=\"PLN\";\n",
    "            rec.features.append(SeqFeature(FeatureLocation(0, len(record.seq.ungap('-'))),\n",
    "                                           type = \"source\", \n",
    "                                           qualifiers  = {'isolate':rec_sample_dc['ENARunNumber'],\n",
    "                                                          'organism':rec_sample_dc['NCBI_sciname'],\n",
    "                                                          'organelle':\"plastid\",\n",
    "                                                          'mol_type':'genomic DNA',\n",
    "                                                         'db_xref': ['taxon:' + str(int(rec_sample_dc['NCBI_TaxID']))] }\n",
    "                                          ))\n",
    "\n",
    "            rec.features.append(SeqFeature(FeatureLocation(0, len(record.seq.ungap('-'))),\n",
    "                                           type = \"gene\", \n",
    "                                           qualifiers  = {'gene':row.Gene}))\n",
    "            if row['type'] == 'cds':\n",
    "                rec.features.append(SeqFeature(FeatureLocation(0, len(record.seq.ungap('-'))),\n",
    "                                           type = \"cds\", \n",
    "                                           qualifiers  = {'gene':row.Gene,\n",
    "                                                          'codon_start':1,\n",
    "                                                          'transl_table':11,\n",
    "                                                          'product':row.Product}))\n",
    "            elif row['type'] == 'rRNA':\n",
    "                rec.features.append(SeqFeature(FeatureLocation(0, len(record.seq.ungap('-'))),\n",
    "                                           type = \"rRNA\", \n",
    "                                           qualifiers  = {'gene':row.Gene,\n",
    "                                                          'product':row.Product}))\n",
    "            record_export.append(rec)\n",
    "    print('N records:',len(record_export))\n",
    "    \n",
    "    ## Write embl flat file, metadata table,  and submission commands\n",
    "    SeqIO.write(record_export, wdir + '/ENA_gene_submission/' + row.Gene + '_A353.embl', \"embl\")\n",
    "    \n",
    "    ## Convert .gb to EMBL flat files\n",
    "    gzip_file(wdir + '/ENA_gene_submission/' + row.Gene + '_A353.embl')\n",
    "    \n",
    "    ## Manifest file \n",
    "    manifest={\n",
    "        'STUDY':'PRJEB35285', #https://www.ncbi.nlm.nih.gov/bioproject/588607\n",
    "        'Name':row.Gene,\n",
    "        'FLATFILE': row.Gene + '_A353.embl.gz',\n",
    "        }\n",
    "    manifest = pd.DataFrame.from_dict(manifest,orient='index')\n",
    "    manifest.to_csv(wdir + '/ENA_gene_submission/' + row.Gene + '_A353_manifest.txt',header=None,sep='\\t')\n",
    "\n",
    "    ## Submission command\n",
    "    print('java -jar C:/Data/PAFTOL/Organelles/ENA_submissions/webin-cli-4.3.0.jar -username Webin-52995' \\\n",
    "          + ' -passwordFile C:/Data/PAFTOL/Organelles/ENA_submissions/ena_pwd.txt -context sequence -manifest ' \\\n",
    "          + row.Gene + '_A353_manifest.txt' + ' -validate')\n",
    "    print(' ')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
