{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b22e25bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import SeqIO; from Bio.Seq import Seq; from Bio.SeqRecord import SeqRecord; from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "import os\n",
    "import re\n",
    "#### TESTING"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d1bb769",
   "metadata": {},
   "source": [
    "### Refs\n",
    "* https://biopython.org/wiki/SeqRecord\n",
    "* https://ena-docs.readthedocs.io/en/latest/faq/locus_tags.html\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6fbf0f58",
   "metadata": {},
   "outputs": [],
   "source": [
    "# latest paftol export containing ENA sample accessions\n",
    "###pft_file = '../../paftol_export/paftol_export_files/2022-06-06_paftol_export.csv'\n",
    "### Paul B. - NB the June version of the paftol export table verions doesn't have these keys in the index:\n",
    "### Sample_Name, NCBI_sciname and sci_name, NCBI_TaxID\n",
    "### It is because Kevin's table has these extra columns added from elsewhere hence 'stats' in ../paftol_export_stats.csv\n",
    "### pft_file = '../../PAFTOL_DB/2022-08-04_paftol_export_stats.csv' - created updated file here:\n",
    "pft_file = '../../paftol_export/paftol_export_files/2022-08-04_paftol_export_stats.csv'\n",
    "### 13.7.2023 - used the above file except for some small exceptions just below:\n",
    "###pft_file = 'ENA_submissions/PAFTOL/paftol_export_8_missing_samples_ONLY.csv'\n",
    "cptree_file = 'cpTree_v7/cpTree_v7_tree_clean.csv'\n",
    "org_dir = 'Organelles/'\n",
    " \n",
    "# ENA_project_number = 'PRJEB49212'\n",
    "DataSource = 'SRA' # PAFTOL, SRA, GAP\n",
    "if DataSource == 'PAFTOL':\n",
    "    ENA_project_number = 'PRJEB51960' #https://www.ncbi.nlm.nih.gov/bioproject/588607\n",
    "    project_locus_tag = 'PRJEB51960'\n",
    "elif DataSource == 'SRA':\n",
    "    ENA_project_number = 'PRJEB55358' #https://www.ncbi.nlm.nih.gov/bioproject/588607\n",
    "    project_locus_tag = 'PRJEB55358'    \n",
    "elif DataSource == 'GAP':\n",
    "    ENA_project_number = 'PRJEB55354' #https://www.ncbi.nlm.nih.gov/bioproject/588607\n",
    "    project_locus_tag = 'PRJEB55354'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4694d45c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1611 ['SRR9611034', 'SRR8655284']\n"
     ]
    }
   ],
   "source": [
    "## All Validated Samples\n",
    "R = pd.read_csv(cptree_file)\n",
    "R = R[R.DataSource==DataSource]\n",
    "subset = R.Sample_Name.to_list()\n",
    "print(R.shape[0],subset[:2])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0edb6727",
   "metadata": {},
   "source": [
    "## Load data and subset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "06f2bfb6",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdir = DataSource + '/'\n",
    "sdir = 'ENA_submissions/' + pdir\n",
    "if not os.path.exists(sdir):\n",
    "    os.makedirs(sdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "bb568bd9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1455\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>idPaftol</th>\n",
       "      <th>idSequencing</th>\n",
       "      <th>Sample_Name</th>\n",
       "      <th>Project</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>SumContigLength</th>\n",
       "      <th>sum_len</th>\n",
       "      <th>...</th>\n",
       "      <th>Assembly</th>\n",
       "      <th>Coverage_Kmer</th>\n",
       "      <th>Coverage_base</th>\n",
       "      <th>maxK</th>\n",
       "      <th>NRepeat_Pattern</th>\n",
       "      <th>NPath</th>\n",
       "      <th>Redo_FastPlast</th>\n",
       "      <th>error_pt</th>\n",
       "      <th>error_nr</th>\n",
       "      <th>SeqID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1376</th>\n",
       "      <td>14611.0</td>\n",
       "      <td>28675</td>\n",
       "      <td>GAP_028675</td>\n",
       "      <td>GAP</td>\n",
       "      <td>Poales</td>\n",
       "      <td>Poaceae</td>\n",
       "      <td>Tripogonella</td>\n",
       "      <td>loliiformis</td>\n",
       "      <td>240504.0</td>\n",
       "      <td>240504.0</td>\n",
       "      <td>...</td>\n",
       "      <td>scaffold</td>\n",
       "      <td>25.6</td>\n",
       "      <td>76.1</td>\n",
       "      <td>85.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>1000.0</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GAP_028675</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      idPaftol  idSequencing Sample_Name Project   order   family  \\\n",
       "1376   14611.0         28675  GAP_028675     GAP  Poales  Poaceae   \n",
       "\n",
       "             genus      species  SumContigLength   sum_len  ...  Assembly  \\\n",
       "1376  Tripogonella  loliiformis         240504.0  240504.0  ...  scaffold   \n",
       "\n",
       "      Coverage_Kmer  Coverage_base  maxK  NRepeat_Pattern   NPath  \\\n",
       "1376           25.6           76.1  85.0             10.0  1000.0   \n",
       "\n",
       "      Redo_FastPlast  error_pt  error_nr       SeqID  \n",
       "1376            True       NaN       NaN  GAP_028675  \n",
       "\n",
       "[1 rows x 41 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load GetOrg result table\n",
    "Org_df = pd.read_csv(org_dir + DataSource + '/' + DataSource + '_Organelle_Recovery.csv').astype({'idSequencing':'int'})\n",
    "if DataSource == 'PAFTOL':\n",
    "    Org_df['SeqID'] = Org_df.Sample_Name.str.replace('PAFTOL_','Pis_')\n",
    "elif DataSource in ['GAP']:\n",
    "    Org_df['SeqID'] = Org_df.Sample_Name\n",
    "elif DataSource in ['SRA']:\n",
    "    sra2_pt = pd.read_csv('cpSRA_GetOrg/cpSRA_GetOrg_Organelle_Recovery.csv')\n",
    "    Org_df = pd.concat([Org_df,sra2_pt],ignore_index=True)\n",
    "    Org_df = Org_df.groupby('Sample_Name').head(1)\n",
    "    Org_df['SeqID'] = Org_df.Sample_Name\n",
    "print(Org_df.shape[0])\n",
    "Org_df[:2]\n",
    "###Org_df[ 'order' ]\n",
    "###Org_df[ 'Sample_Name' ]\n",
    "###Org_df[ Org_df.Sample_Name == 'SRR5237176']### Paul B - present\n",
    "Org_df[ Org_df.Sample_Name == 'GAP_028675']### Paul B - present"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "518fc2cf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "344 344\n"
     ]
    }
   ],
   "source": [
    "if subset is not None:\n",
    "    Org_df = Org_df[Org_df.Sample_Name.isin(subset)]\n",
    "    print(Org_df.shape[0],Org_df.Sample_Name.nunique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "82f53296",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/80/92djrly528vg3zy0d_trd0300000gp/T/ipykernel_28053/798644719.py:2: DtypeWarning: Columns (29,35,44,56,71,83) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  db = pd.read_csv(pft_file)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_Name</th>\n",
       "      <th>idPaftol</th>\n",
       "      <th>DataSource</th>\n",
       "      <th>Project</th>\n",
       "      <th>Order</th>\n",
       "      <th>Family</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Species</th>\n",
       "      <th>Taxonomical_Notes</th>\n",
       "      <th>ENASampleNum</th>\n",
       "      <th>idSequencing</th>\n",
       "      <th>ExternalSequenceID</th>\n",
       "      <th>ENAExpNumber</th>\n",
       "      <th>ENARunNumber</th>\n",
       "      <th>NumReads</th>\n",
       "      <th>NumRecoveredGenes</th>\n",
       "      <th>SumContigLength</th>\n",
       "      <th>NCBI_TaxID</th>\n",
       "      <th>NCBI_sciname</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PAFTOL_000564</td>\n",
       "      <td>1</td>\n",
       "      <td>PAFTOL</td>\n",
       "      <td>Asteraceae</td>\n",
       "      <td>Asterales</td>\n",
       "      <td>Asteraceae</td>\n",
       "      <td>Gymnolaena</td>\n",
       "      <td>sp.</td>\n",
       "      <td>Initially recorded as Gymnolaena litoralis but...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>564</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PAFTOL_000961</td>\n",
       "      <td>2</td>\n",
       "      <td>PAFTOL</td>\n",
       "      <td>Pilot</td>\n",
       "      <td>Ranunculales</td>\n",
       "      <td>Circaeasteraceae</td>\n",
       "      <td>Circaeaster</td>\n",
       "      <td>agrestis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERS5501508</td>\n",
       "      <td>961</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERX4839538</td>\n",
       "      <td>ERR5033191</td>\n",
       "      <td>625228.0</td>\n",
       "      <td>332.0</td>\n",
       "      <td>138795.0</td>\n",
       "      <td>39288.0</td>\n",
       "      <td>Circaeaster agrestis</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Sample_Name  idPaftol DataSource     Project         Order  \\\n",
       "0  PAFTOL_000564         1     PAFTOL  Asteraceae     Asterales   \n",
       "1  PAFTOL_000961         2     PAFTOL       Pilot  Ranunculales   \n",
       "\n",
       "             Family        Genus   Species  \\\n",
       "0        Asteraceae   Gymnolaena       sp.   \n",
       "1  Circaeasteraceae  Circaeaster  agrestis   \n",
       "\n",
       "                                   Taxonomical_Notes ENASampleNum  \\\n",
       "0  Initially recorded as Gymnolaena litoralis but...          NaN   \n",
       "1                                                NaN   ERS5501508   \n",
       "\n",
       "   idSequencing ExternalSequenceID ENAExpNumber ENARunNumber  NumReads  \\\n",
       "0           564                NaN          NaN          NaN       NaN   \n",
       "1           961                NaN   ERX4839538   ERR5033191  625228.0   \n",
       "\n",
       "   NumRecoveredGenes  SumContigLength  NCBI_TaxID          NCBI_sciname  \n",
       "0                NaN              NaN         NaN                   NaN  \n",
       "1              332.0         138795.0     39288.0  Circaeaster agrestis  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load paftol export and merge ena sample IDs\n",
    "db = pd.read_csv(pft_file)\n",
    "db = db[db.idPaftol.notnull()].astype({'idPaftol':'int','idSequencing':'int'})\n",
    "db = db[['Sample_Name','idPaftol', 'DataSource', 'Project', 'Order', 'Family', 'Genus', 'Species', 'Taxonomical_Notes', \n",
    "         'ENASampleNum', 'idSequencing', 'ExternalSequenceID', 'ENAExpNumber',\n",
    "         'ENARunNumber', 'NumReads', 'NumRecoveredGenes', 'SumContigLength','NCBI_TaxID', 'NCBI_sciname']]\n",
    "#        'title', 'taxId','id', 'alias', 'firstCreated', 'firstPublic', 'releaseStatus', 'secondaryId', 'scientificName', 'commonName']]\n",
    "### Paul B. - temp alteration to the above lines to process a few samples:\n",
    "###db = db[['Sample_Name','DataSource','ENASampleNum', 'ENAExpNumber','ENARunNumber','NCBI_TaxID','NCBI_sciname']]\n",
    "db[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "25069c62",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'idPaftol': 0, 'idSequencing': 0, 'Sample_Name': 0, 'Project': 0, 'order': 0, 'family': 0, 'genus': 0, 'species': 0, 'SumContigLength': 45, 'sum_len': 45, 'sci_name': 0, 'log_pt': 0, 'log_nr': 0, 'pt_recovered': 0, 'Nseq_pt': 0, 'Sum_len_pt': 0, 'Nanybase_pt': 0, 'nr_recovered': 0, 'Nseq_nr': 12, 'Sum_len_nr': 12, 'Nanybase_nr': 12, 'Completed_Reads': 0, 'Completed_Extending': 0, 'Completed_Assembly': 0, 'Completed_Slimming': 0, 'Completed_Output': 0, 'Warning_Look': 0, 'Warning_Multi_Structure': 0, 'Warning_Self_Loop': 0, 'Info_Large_Repeats': 0, 'Info_Disentangling_Failed': 0, 'Assembly': 0, 'Coverage_Kmer': 42, 'Coverage_base': 1, 'maxK': 0, 'NRepeat_Pattern': 338, 'NPath': 340, 'Redo_FastPlast': 0, 'error_pt': 344, 'error_nr': 333, 'SeqID': 0, 'DataSource': 0, 'ENASampleNum': 8, 'ENAExpNumber': 8, 'ENARunNumber': 8, 'NCBI_TaxID': 11, 'NCBI_sciname': 11}\n",
      "344 344\n"
     ]
    }
   ],
   "source": [
    "Org_df = pd.merge(Org_df,\n",
    "         db[['Sample_Name','DataSource','ENASampleNum','ENAExpNumber','ENARunNumber','NCBI_TaxID','NCBI_sciname']],\n",
    "        how='left',on='Sample_Name').groupby('Sample_Name').head(1)\n",
    "print(Org_df.isna().sum().to_dict())\n",
    "print(Org_df.shape[0],Org_df.Sample_Name.nunique())\n",
    "###Org_df[ Org_df.Sample_Name == 'SRR5237176']### Paul B - present - OK\n",
    "###Org_df[ Org_df.Sample_Name == 'GAP_026371' ] ### Paul B - present\n",
    "###Org_df[ Org_df.Sample_Name == 'GAP_028675']### Paul B - present"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9df0de12",
   "metadata": {},
   "outputs": [],
   "source": [
    "geseq_dir = sdir + 'fasta/'\n",
    "if not os.path.exists(geseq_dir):\n",
    "    os.makedirs(geseq_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46f2a546",
   "metadata": {},
   "source": [
    "# Whole plastomes\n",
    "### Annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "500a892c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "330 samples 304 genera\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_Name</th>\n",
       "      <th>idSequencing</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>SumContigLength</th>\n",
       "      <th>sum_len</th>\n",
       "      <th>sci_name</th>\n",
       "      <th>log_pt</th>\n",
       "      <th>...</th>\n",
       "      <th>Redo_FastPlast</th>\n",
       "      <th>error_pt</th>\n",
       "      <th>error_nr</th>\n",
       "      <th>SeqID</th>\n",
       "      <th>DataSource</th>\n",
       "      <th>ENASampleNum</th>\n",
       "      <th>ENAExpNumber</th>\n",
       "      <th>ENARunNumber</th>\n",
       "      <th>NCBI_TaxID</th>\n",
       "      <th>NCBI_sciname</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR12649607</td>\n",
       "      <td>19711.0</td>\n",
       "      <td>Malvales</td>\n",
       "      <td>Malvaceae</td>\n",
       "      <td>Abutilon</td>\n",
       "      <td>theophrasti</td>\n",
       "      <td>175491.0</td>\n",
       "      <td>175491.0</td>\n",
       "      <td>Abutilon theophrasti</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SRR12649607</td>\n",
       "      <td>SRA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SRX9130922</td>\n",
       "      <td>SRR12649607</td>\n",
       "      <td>3631.0</td>\n",
       "      <td>Abutilon theophrasti</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR11342831</td>\n",
       "      <td>19713.0</td>\n",
       "      <td>Pandanales</td>\n",
       "      <td>Velloziaceae</td>\n",
       "      <td>Acanthochlamys</td>\n",
       "      <td>bracteata</td>\n",
       "      <td>7875.0</td>\n",
       "      <td>7875.0</td>\n",
       "      <td>Acanthochlamys bracteata</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SRR11342831</td>\n",
       "      <td>SRA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SRX7945768</td>\n",
       "      <td>SRR11342831</td>\n",
       "      <td>145231.0</td>\n",
       "      <td>Acanthochlamys bracteata</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 46 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Sample_Name  idSequencing       order        family           genus  \\\n",
       "2  SRR12649607       19711.0    Malvales     Malvaceae        Abutilon   \n",
       "3  SRR11342831       19713.0  Pandanales  Velloziaceae  Acanthochlamys   \n",
       "\n",
       "       species  SumContigLength   sum_len                  sci_name  log_pt  \\\n",
       "2  theophrasti         175491.0  175491.0      Abutilon theophrasti    True   \n",
       "3    bracteata           7875.0    7875.0  Acanthochlamys bracteata    True   \n",
       "\n",
       "   ...  Redo_FastPlast  error_pt  error_nr        SeqID  DataSource  \\\n",
       "2  ...           False       NaN       NaN  SRR12649607         SRA   \n",
       "3  ...           False       NaN       NaN  SRR11342831         SRA   \n",
       "\n",
       "   ENASampleNum  ENAExpNumber  ENARunNumber  NCBI_TaxID  \\\n",
       "2           NaN    SRX9130922   SRR12649607      3631.0   \n",
       "3           NaN    SRX7945768   SRR11342831    145231.0   \n",
       "\n",
       "               NCBI_sciname  \n",
       "2      Abutilon theophrasti  \n",
       "3  Acanthochlamys bracteata  \n",
       "\n",
       "[2 rows x 46 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "whole_pt = Org_df[(Org_df.Assembly.isin(['circular genome','circular genome with gaps']) & (Org_df.Sum_len_pt>80000))]\n",
    "print(whole_pt.shape[0],'samples',whole_pt.genus.nunique(),'genera')\n",
    "whole_pt[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "07f68652",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRR12649607 SRR12649607 Malvaceae Abutilon theophrasti 160446 0\n",
      "SRR11342831 SRR11342831 Velloziaceae Acanthochlamys bracteata 153642 10\n",
      "ERR4210270 ERR4210270 Brassicaceae Pseudocamelina glaucophylla 153384 77\n",
      "SRR3478564 SRR3478564 Dioscoreaceae Tacca chantrieri 163007 0\n",
      "SRR6425650 SRR6425650 Apocynaceae Alafia barteri 155272 0\n",
      "ERR2789774 ERR2789774 Brassicaceae Ballantinia pumilio 154346 0\n",
      "SRR1145773 SRR1145773 Rosaceae Bencomia exstipulata 153731 0\n",
      "SRR10679069 SRR10679069 Lamiaceae Caryopteris trichosphaera 151382 0\n",
      "SRR6425648 SRR6425648 Apocynaceae Cascabela thevetia 155029 0\n",
      "SRR13385162 SRR13385162 Poaceae Chimonobambusa sichuanensis 139602 0\n",
      "SRR2155068 SRR2155068 Asteraceae Conoclinium coelestinum 151335 0\n",
      "ERR4210279 ERR4210279 Brassicaceae Delpinophytum patagonicum 154932 0\n",
      "SRR9309785 SRR9309785 Phrymaceae Diplacus longiflorus 153798 0\n",
      "SRR8666566 SRR8666566 Fabaceae Gastrolobium grandiflorum 152366 10\n",
      "SRR13089735 SRR13089735 Gentianaceae Gentianopsis paludosa 151569 0\n",
      "SRR12649932 SRR12649932 Theaceae Gordonia alpestris 157211 0\n",
      "SRR9164790 SRR9164790 Poaceae Guadua angustifolia 135336 0\n",
      "SRR12360733 SRR12360733 Poaceae Heteropogon contortus 134885 0\n",
      "ERR4208165 ERR4208165 Brassicaceae Hornungia petraea 150506 0\n",
      "SRR6425645 SRR6425645 Apocynaceae Hunteria zeylanica 154654 0\n",
      "SRR13374389 SRR13374389 Poaceae Indocalamus tongchuensis 139777 0\n",
      "SRR6425614 SRR6425614 Apocynaceae Laxoplumeria baehniana 154673 0\n",
      "SRR6425619 SRR6425619 Apocynaceae Mesechites trifidus 153557 0\n",
      "SRR1145774 SRR1145774 Phrymaceae Mimulus ringens 153955 0\n",
      "ERR4210227 ERR4210227 Brassicaceae Murbeckiella boryi 154774 0\n",
      "SRR8247328 SRR8247328 Oleaceae Nestegis sandwicensis 155557 0\n",
      "SRR8247331 SRR8247331 Oleaceae Notelaea longifolia 155063 0\n",
      "ERR4210238 ERR4210238 Brassicaceae Ochthodium aegyptiacum 154281 0\n",
      "SRR12282384 SRR12282384 Salicaceae Olmediella betschleriana 158630 0\n",
      "SRR8302701 SRR8302701 Betulaceae Ostrya chinensis 159585 0\n",
      "SRR8302720 SRR8302720 Betulaceae Ostryopsis nobilis 159159 0\n",
      "ERR4210200 ERR4210200 Brassicaceae Pennellia longifolia 155152 0\n",
      "ERR4205155 ERR4205155 Brassicaceae Polyctenium fremontii 154693 0\n",
      "SRR12360708 SRR12360708 Poaceae Pseudanthistiria umbellata 137900 0\n",
      "ERR4208775 ERR4208775 Brassicaceae Robeschia schimperi 154443 0\n",
      "SRR12619666 SRR12619666 Lardizabalaceae Sargentodoxa cuneata 158094 0\n",
      "ERR3770473 ERR3770473 Asteraceae Schlagintweitia huteri 153086 10\n",
      "SRR10679081 SRR10679081 Lamiaceae Schnabelia oligophylla 156768 0\n",
      "SRR8666287 SRR8666287 Malvaceae Seringia nephrosperma 158731 0\n",
      "SRR13070755 SRR13070755 Rosaceae Sibbaldianthe bifurca 156734 0\n",
      "SRR12853157 SRR12853157 Gentianaceae Sinoswertia tetraptera 152787 0\n",
      "ERR4210201 ERR4210201 Brassicaceae Sphaerocardamum divaricatum 154923 0\n",
      "SRR12717620 SRR12717620 Menispermaceae Stephania tetrandra 159974 0\n",
      "SRR13197184 SRR13197184 Saxifragaceae Tanakaea radicans 155265 0\n",
      "SRR6425641 SRR6425641 Apocynaceae Tassadia propinqua 155264 0\n",
      "ERR4210253 ERR4210253 Brassicaceae Thelypodium laciniatum 153521 0\n",
      "ERR4205217 ERR4205217 Brassicaceae Vella lucentina 153668 0\n",
      "SRR8666347 SRR8666347 Poaceae Xerochloa barbata 138872 10\n",
      "SRR13179903 SRR13179903 Fabaceae Aganope dinghuensis 156789 0\n",
      "SRR13192465 SRR13192465 Rosaceae Dasiphora parvifolia 152898 0\n",
      "SRR11792100 SRR11792100 Poaceae Melinis ambigua 139797 0\n",
      "SRR934047 SRR934047 Apocynaceae Orthosia scoparia 158597 0\n",
      "SRR12599358 SRR12599358 Elaeocarpaceae Sloanea sinensis 157546 0\n",
      "SRR11653895 SRR11653895 Capparaceae Cadaba natalensis 155836 0\n",
      "SRR8208354 SRR8208354 Asteraceae Klasea centauroides 153070 0\n",
      "SRR13480990 SRR13480990 Asteraceae Leontodon longirostris 152407 0\n",
      "SRR12360688 SRR12360688 Poaceae Schizachyrium delavayi 139855 0\n",
      "SRR12762343 SRR12762343 Loranthaceae Scurrula chingii 122764 0\n",
      "SRR10137792 SRR10137792 Cucurbitaceae Indofevillea khasiana 159179 0\n",
      "SRR7121689 SRR7121689 Asteraceae Pseudognaphalium affine 153112 0\n",
      "ERR4205121 ERR4205121 Brassicaceae Alyssopsis mollis 154837 0\n",
      "SRR10121863 SRR10121863 Lauraceae Neocinnamomum delavayi 165807 0\n",
      "SRR7121884 SRR7121884 Lamiaceae Pseudocaryopteris paniculata 151954 0\n",
      "SRR7121569 SRR7121569 Poaceae Neyraudia reynaudiana 134950 0\n",
      "SRR7121896 SRR7121896 Orchidaceae Zeuxine strateumatica 141436 0\n",
      "SRR7121625 SRR7121625 Hamamelidaceae Mytilaria laosensis 159911 0\n",
      "DRR082674 DRR082674 Orobanchaceae Melampyrum roseum 143896 0\n",
      "SRR576532 SRR576532 Francoaceae Melianthus villosus 156509 0\n",
      "SRR7121584 SRR7121584 Asteraceae Youngia japonica 152583 0\n",
      "SRR7121822 SRR7121822 Fabaceae Phyllodium pulchellum 149685 0\n",
      "SRR9313657 SRR9313657 Juglandaceae Platycarya strobilacea 158710 0\n",
      "SRR7122037 SRR7122037 Rosaceae Agrimonia pilosa 155127 0\n",
      "SRR7121950 SRR7121950 Plantaginaceae Mecardonia procumbens 154636 0\n",
      "SRR6940071 SRR6940071 Lamiaceae Westringia fruticosa 152883 0\n",
      "SRR7122069 SRR7122069 Loranthaceae Macrosolen cochinchinensis 129339 0\n",
      "SRR5880481 SRR5880481 Pentaphylacaceae Anneslea fragrans 157831 0\n",
      "SRR13189613 SRR13189613 Apiaceae Bupleurum yinchowense 155851 0\n",
      "ERR4204116 ERR4204116 Brassicaceae Clypeola jonthlaspi 153857 0\n",
      "SRR7121958 SRR7121958 Lamiaceae Colebrookea oppositifolia 152100 0\n",
      "ERR4210289 ERR4210289 Brassicaceae Subularia aquatica 152430 0\n",
      "SRR10121868 SRR10121868 Lauraceae Syndiclis chinensis 158528 0\n",
      "SRR7121916 SRR7121916 Colchicaceae Disporum calcaratum 156692 0\n",
      "SRR7121699 SRR7121699 Bignoniaceae Tecoma stans 164538 0\n",
      "SRR7121570 SRR7121570 Malvaceae Urena procumbens 161541 0\n",
      "SRR12599428 SRR12599428 Elaeocarpaceae Crinodendron patagua 159400 0\n",
      "SRR7121559 SRR7121559 Asteraceae Crassocephalum crepidioides 150596 0\n",
      "ERR4207766 ERR4207766 Brassicaceae Diptychocarpus strictus 159917 100\n",
      "SRR13257968 SRR13257968 Araliaceae Dendropanax trifidus 156505 0\n",
      "SRR10947082 SRR10947082 Basellaceae Ullucus tuberosus 155761 0\n",
      "SRR12982332 SRR12982332 Polygonaceae Calligonum bakuense 162128 0\n",
      "SRR7121849 SRR7121849 Fabaceae Millettia pachyloba 152365 0\n",
      "SRR7121658 SRR7121658 Apocynaceae Gomphocarpus fruticosus 162628 0\n",
      "SRR7121839 SRR7121839 Asteraceae Gnaphalium polycaulon 152379 0\n",
      "SRR7191154 SRR7191154 Asteraceae Blumea balsamifera 151176 0\n",
      "SRR7121869 SRR7121869 Fabaceae Flemingia kweichowensis 153208 0\n",
      "SRR12282398 SRR12282398 Salicaceae Azara integrifolia 158291 0\n",
      "SRR7121989 SRR7121989 Araliaceae Brassaiopsis ciliata 156479 0\n",
      "SRR7121948 SRR7121948 Phyllanthaceae Bridelia tomentosa 168345 0\n",
      "SRR7121459 SRR7121459 Fabaceae Cassia fistula 161671 0\n",
      "SRR7121552 SRR7121552 Araliaceae Trevesia palmata 156479 0\n",
      "SRR7122024 SRR7122024 Asteraceae Centratherum punctatum 152739 0\n",
      "SRR10376715 SRR10376715 Hydrocharitaceae Thalassia testudinum 178730 0\n",
      "SRR11994239 SRR11994239 Boraginaceae Amsinckia grandiflora 149506 0\n",
      "SRR13311451 SRR13311451 Actinidiaceae Clematoclethra scandens 157325 0\n",
      "SRR11305036 SRR11305036 Asteraceae Vernonia galamensis 153130 0\n",
      "SRR7121964 SRR7121964 Lamiaceae Pseudocaryopteris paniculata 151953 0\n",
      "SRR7122162 SRR7122162 Fabaceae Flemingia stricta 153323 0\n",
      "SRR7121564 SRR7121564 Asteraceae Blumea oxyodonta 151053 0\n",
      "SRR7121578 SRR7121578 Asteraceae Ageratum houstonianum 151541 0\n",
      "SRR1801313 SRR1801313 Brassicaceae Cakile maritima 153724 0\n",
      "SRR7121609 SRR7121609 Asteraceae Pseudognaphalium affine 153140 0\n",
      "SRR7121883 SRR7121883 Lamiaceae Gmelina arborea 154369 0\n",
      "SRR7121864 SRR7121864 Malvaceae Sterculia monosperma 161099 0\n",
      "SRR7122099 SRR7122099 Malvaceae Pterospermum truncatolobatum 162314 0\n",
      "SRR12880891 SRR12880891 Fabaceae Thermopsis turkestanica 153455 0\n",
      "SRR10121867 SRR10121867 Lauraceae Caryodaphnopsis tonkinensis 148829 0\n",
      "SRR10076754 SRR10076754 Amaranthaceae Salicornia brachiata 153248 0\n",
      "SRR10121860 SRR10121860 Lauraceae Actinodaphne lecomtei 154219 0\n",
      "SRR8208363 SRR8208363 Poaceae Cleistogenes squarrosa 134482 0\n",
      "SRR4128964 SRR4128964 Brassicaceae Descurainia sophioides 153224 0\n",
      "SRR12000841 SRR12000841 Columelliaceae Columellia oblonga 153070 0\n",
      "ERR4566420 ERR4566420 Lamiaceae Tectona grandis 153944 0\n",
      "SRR8146324 SRR8146324 Asteraceae Saussurea polylepis 152523 0\n",
      "SRR6048019 SRR6048019 Apiaceae Bupleurum latissimum 155622 0\n",
      "SRR8899842 SRR8899842 Dioscoreaceae Tacca leontopetaloides 162477 0\n",
      "SRR12360712 SRR12360712 Poaceae Iseilema membranaceum 139582 0\n",
      "SRR13249324 SRR13249324 Fagaceae Castanopsis faberi 160672 0\n",
      "ERR873461 ERR873461 Bignoniaceae Tanaecium tetragonolobum 159020 0\n",
      "SRR7267932 SRR7267932 Anacardiaceae Sclerocarya birrea 162528 0\n",
      "SRR5674478 SRR5674478 Cannabaceae Trema orientale 157141 0\n",
      "SRR8907157 SRR8907157 Caryocaraceae Caryocar brasiliense 165798 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRR5265130 SRR5265130 Lamiaceae Tectona grandis 153928 0\n",
      "SRR6940074 SRR6940074 Mazaceae Lancea tibetica 154098 0\n",
      "SRR14626653 SRR14626653 Asteraceae Adenostemma lavenia 150063 0\n",
      "SRR13700329 SRR13700329 Bromeliaceae Aechmea bromeliifolia 159807 0\n",
      "ERR5439620 ERR5439620 Poaceae Ampelocalamus scandens 139609 10\n",
      "ERR5439634 ERR5439634 Poaceae Chimonocalamus pallens 139683 0\n",
      "SRR7121716 SRR7121716 Asteraceae Cyathocline purpurea 151086 0\n",
      "SRR11788051 SRR11788051 Apocynaceae Cynanchum wilfordii 161234 0\n",
      "ERR5439644 ERR5439644 Poaceae Fargesia denudata 139776 10\n",
      "SRR13364360 SRR13364360 Asteraceae Gymnanthemum amygdalinum 153133 0\n",
      "SRR14240564 SRR14240564 Lamiaceae Hanceola exserta 153296 0\n",
      "SRR13985460 SRR13985460 Araliaceae Heptapleurum heptaphyllum 156527 0\n",
      "SRR14328332 SRR14328332 Lamiaceae Lagopsis supina 151728 0\n",
      "DRR151838 DRR151838 Meliaceae Lansium domesticum 159567 0\n",
      "SRR13284716 SRR13284716 Asteraceae Neopallasia pectinata 152938 0\n",
      "ERR2990308 ERR2990308 Brassicaceae Neslia paniculata 154936 0\n",
      "ERR5439657 ERR5439657 Poaceae Oldeania alpina 139716 10\n",
      "SRR13805712 SRR13805712 Amaranthaceae Oreosalsola abrotanoides 151622 0\n",
      "ERR5439658 ERR5439658 Poaceae Otatea acuminata 136359 0\n",
      "SRR13945346 SRR13945346 Fabaceae Pachyrhizus erosus 151947 0\n",
      "SRR13945344 SRR13945344 Fabaceae Pachyrhizus tuberosus 151857 0\n",
      "SRR14547473 SRR14547473 Fabaceae Phyllolobium chinense 124026 0\n",
      "SRR13700324 SRR13700324 Bromeliaceae Racinaea ropalocarpa 157046 0\n",
      "SRR14561442 SRR14561442 Apiaceae Semenovia thomsonii 142281 0\n",
      "SRR14085851 SRR14085851 Typhaceae Sparganium erectum 161800 0\n",
      "DRR270591 DRR270591 Lardizabalaceae Stauntonia hexaphylla 157882 0\n",
      "SRR7121672 SRR7121672 Asteraceae Tarlmounia elliptica 152801 0\n",
      "SRR14294472 SRR14294472 Caprifoliaceae Kolkwitzia amabilis 160951 0\n",
      "SRR14294450 SRR14294450 Caprifoliaceae Morina longifolia 171926 0\n",
      "SRR14570887 SRR14570887 Hydrangeaceae Fendlerella utahensis 156409 0\n",
      "SRR14570832 SRR14570832 Loasaceae Nasa urens 158300 0\n",
      "SRR13390755 SRR13390755 Orchidaceae Acineta mireyae 155948 0\n",
      "SRR13390752 SRR13390752 Orchidaceae Calopogon barbatus 148830 0\n",
      "SRR11184855 SRR11184855 Salvadoraceae Salvadora oleoides 154025 0\n",
      "SRR14570883 SRR14570883 Hydrangeaceae Deutzia pilosa 157944 0\n",
      "SRR8666408 SRR8666408 Asteraceae Pterocaulon intermedium 152203 0\n",
      "SRR8666676 SRR8666676 Asteraceae Rhodanthe margarethae 153214 0\n",
      "ERR3089129 ERR3089129 Asteraceae Tripleurospermum maritimum 149958 0\n",
      "SRR13823287 SRR13823287 Asteraceae Crassocephalum crepidioides 150596 0\n",
      "SRR14710253 SRR14710253 Asteraceae Acmella paniculata 152117 0\n",
      "SRR14401136 SRR14401136 Asteraceae Blumea balsamifera 151163 0\n",
      "SRR8666495 SRR8666495 Asteraceae Calocephalus beardii 153003 0\n",
      "ERR4993550 ERR4993550 Asteraceae Andryala laevitomentosa 153255 0\n",
      "ERR4993548 ERR4993548 Asteraceae Andryala glandulosa 152819 0\n",
      "SRR13528379 SRR13528379 Asteraceae Sinosenecio jishouensis 151257 0\n",
      "ERR4993551 ERR4993551 Asteraceae Andryala pinnatifida 152651 0\n",
      "SRR9118977 SRR9118977 Asteraceae Shangwua jacea 152808 0\n",
      "SRR9119151 SRR9119151 Asteraceae Amphoricarpos autariatus 152620 0\n",
      "SRR11926452 SRR11926452 Asteraceae Volutaria canariensis 152542 0\n",
      "SRR11926453 SRR11926453 Asteraceae Tricholepis gedrosiaca 152895 0\n",
      "SRR11926454 SRR11926454 Asteraceae Staehelina petiolata 152195 0\n",
      "SRR11926509 SRR11926509 Asteraceae Karvandarina cartilaginea 152577 0\n",
      "SRR11926503 SRR11926503 Asteraceae Plectocephalus cachinalensis 152840 0\n",
      "SRR11926455 SRR11926455 Asteraceae Hirtellina lobelii 152671 0\n",
      "SRR11926504 SRR11926504 Asteraceae Onopordum nervosum 153232 0\n",
      "SRR9119028 SRR9119028 Asteraceae Heterocypsela andersonii 152346 0\n",
      "SRR9119088 SRR9119088 Asteraceae Allocephalus gamolepis 152508 0\n",
      "SRR10249442 SRR10249442 Apiaceae Laserpitium latifolium 158171 0\n",
      "SRR14574492 SRR14574492 Amaryllidaceae Pancratium maritimum 160123 0\n",
      "SRR14574493 SRR14574493 Amaryllidaceae Nerine sarniensis 158312 0\n",
      "SRR14574491 SRR14574491 Amaryllidaceae Strumaria truncata 157566 0\n",
      "SRR14574494 SRR14574494 Amaryllidaceae Lapiedra martinezii 159022 0\n",
      "SRR14574495 SRR14574495 Amaryllidaceae Acis autumnalis 157839 0\n",
      "SRR14516337 SRR14516337 Rubiaceae Cinchona officinalis 156984 0\n",
      "SRR14827348 SRR14827348 Gentianaceae Veratrilla baillonii 153998 0\n",
      "SRR8208362 SRR8208362 Polygonaceae Koenigia divaricata 159124 0\n",
      "SRR12935153 SRR12935153 Cactaceae Neolloydia conoidea 106936 0\n",
      "SRR12935124 SRR12935124 Cactaceae Turbinicarpus alonsoi 117805 0\n",
      "SRR12935163 SRR12935163 Cactaceae Turbinicarpus viereckii 111153 0\n",
      "SRR12935139 SRR12935139 Cactaceae Strombocactus disciformis 119523 0\n",
      "SRR5137213 SRR5137213 Cactaceae Stenocereus thurberi 126070 0\n",
      "SRR12935149 SRR12935149 Cactaceae Acharagma roseanum 110518 0\n",
      "SRR12935099 SRR12935099 Cactaceae Coryphantha cornifera 111930 0\n",
      "SRR12935103 SRR12935103 Cactaceae Mammillaria albicans 99867 0\n",
      "SRR12935162 SRR12935162 Cactaceae Acharagma roseanum 111774 0\n",
      "SRR14237214 SRR14237214 Solanaceae Tubocapsicum anomalum 155802 0\n",
      "SRR5010168 SRR5010168 Orchidaceae Neuwiedia zollingeri 162174 0\n",
      "SRR10120773 SRR10120773 Bromeliaceae Vriesea simulans 157804 0\n",
      "SRR10120830 SRR10120830 Bromeliaceae Vriesea roethii 157754 0\n",
      "SRR13700300 SRR13700300 Bromeliaceae Hechtia glomerata 160056 0\n",
      "SRR9846926 SRR9846926 Bromeliaceae Fascicularia bicolor 159323 0\n",
      "SRR9846925 SRR9846925 Bromeliaceae Fascicularia bicolor 159262 0\n",
      "SRR10120812 SRR10120812 Bromeliaceae Vriesea billbergioides 157822 0\n",
      "SRR8666182 SRR8666182 Malvaceae Waltheria indica 160112 0\n",
      "SRR8666755 SRR8666755 Malvaceae Waltheria virgata 159206 0\n",
      "SRR8666786 SRR8666786 Malvaceae Triumfetta maconochieana 160075 0\n",
      "SRR8666699 SRR8666699 Malvaceae Brachychiton gregorii 161145 0\n",
      "SRR6195038 SRR6195038 Malvaceae Kokia drynarioides 160615 0\n",
      "SRR8666534 SRR8666534 Amaranthaceae Tecticornia indica 152787 0\n",
      "SRR12951210 SRR12951210 Amaranthaceae Caroxylon passerinum 150950 0\n",
      "SRR8666406 SRR8666406 Amaranthaceae Tecticornia halocnemoides 152560 0\n",
      "ERR5439656 ERR5439656 Poaceae Oxytenanthera abyssinica 139486 0\n",
      "ERR5439635 ERR5439635 Poaceae Schizostachyum pergracile 138434 0\n",
      "SRR12888163 SRR12888163 Poaceae Trikeraia hookeri 137695 0\n",
      "ERR5439638 ERR5439638 Poaceae Dinochloa malayana 139305 0\n",
      "SRR8666403 SRR8666403 Poaceae Diplachne fusca 135432 0\n",
      "SRR12360744 SRR12360744 Poaceae Agenium leptocladum 137726 0\n",
      "ERR5439636 ERR5439636 Poaceae Cephalostachyum scandens 139207 0\n",
      "ERR5439653 ERR5439653 Poaceae Chloothamnus elatus 139586 0\n",
      "SRR11576619 SRR11576619 Poaceae Patzkea paniculata 135400 0\n",
      "SRR11576632 SRR11576632 Poaceae Pseudobromus africanus 135432 0\n",
      "ERR5439640 ERR5439640 Poaceae Dinochloa malayana 139461 0\n",
      "SRR13171520 SRR13171520 Poaceae Harpachne harpachnoides 133830 0\n",
      "SRR12846062 SRR12846062 Urticaceae Procris crenata 154124 0\n",
      "SRR14663578 SRR14663578 Fabaceae Flemingia macrophylla 152988 0\n",
      "SRR8666583 SRR8666583 Fabaceae Alysicarpus muelleri 149638 0\n",
      "ERR964506 ERR964506 Fabaceae Eperua falcata 158698 0\n",
      "SRR12136574 SRR12136574 Fabaceae Calophaca sinica 130565 0\n",
      "SRR12136578 SRR12136578 Fabaceae Colutea nepalensis 124868 0\n",
      "SRR12136586 SRR12136586 Fabaceae Phyllolobium camptodontum 124765 0\n",
      "SRR12136572 SRR12136572 Fabaceae Sulla coronaria 122781 0\n",
      "SRR13825327 SRR13825327 Verbenaceae Duranta erecta 165780 0\n",
      "SRR9620534 SRR9620534 Plantaginaceae Littorella uniflora 130759 0\n",
      "SRR6940037 SRR6940037 Lamiaceae Prostanthera lasianthos 151279 0\n",
      "SRR6940060 SRR6940060 Lamiaceae Pseudodictamnus mediterraneus 151677 0\n",
      "SRR6940032 SRR6940032 Lamiaceae Phlomis fruticosa 151638 0\n",
      "SRR6940053 SRR6940053 Lamiaceae Monarda didyma 151851 0\n",
      "SRR6940081 SRR6940081 Lamiaceae Petraeovitex bambusetorum 152961 0\n",
      "SRR10679080 SRR10679080 Lamiaceae Rotheca serrata 150371 0\n",
      "SRR14322715 SRR14322715 Lamiaceae Rotheca myricoides 150596 0\n",
      "SRR6940077 SRR6940077 Lamiaceae Glechoma hederacea 151719 0\n",
      "SRR6940057 SRR6940057 Lamiaceae Collinsonia canadensis 152362 0\n",
      "SRR6940047 SRR6940047 Lamiaceae Lycopus americanus 152108 0\n",
      "SRR13857830 SRR13857830 Lamiaceae Platostoma palustre 152534 0\n",
      "SRR10060812 SRR10060812 Convolvulaceae Lepistemon owariensis 154948 0\n",
      "SRR6940061 SRR6940061 Orobanchaceae Aureolaria pectinata 152723 0\n",
      "SRR12518848 SRR12518848 Orobanchaceae Rhinanthus major 145367 0\n",
      "SRR14373633 SRR14373633 Apocynaceae Tabernaemontana divaricata 157954 0\n",
      "SRR934040 SRR934040 Apocynaceae Marsdenia astephanoides 157108 0\n",
      "SRR934043 SRR934043 Apocynaceae Eustegia minuta 162781 0\n",
      "ERR3011879 ERR3011879 Myrtaceae Rhodamnia argentea 159040 0\n",
      "SRR9089205 SRR9089205 Arecaceae Geonoma undata 157459 0\n",
      "SRR5534848 SRR5534848 Arecaceae Howea forsteriana 158040 0\n",
      "SRR11428446 SRR11428446 Bignoniaceae Jacaranda copaia 153916 0\n",
      "SRR8655290 SRR8655290 Araceae Montrichardia arborescens 166640 0\n",
      "SRR8655284 SRR8655284 Araceae Aglaonema costatum 165343 0\n",
      "SRR14784941 SRR14784941 Araceae Cryptocoryne elliptica 159968 0\n",
      "SRR14368083 SRR14368083 Proteaceae Hicksbeachia pinnatifolia 159759 0\n",
      "SRR14368082 SRR14368082 Proteaceae Lasjia hildebrandii 159532 0\n",
      "SRR13717521 SRR13717521 Rutaceae Orixa japonica 158622 0\n",
      "SRR2153911 SRR2153911 Calyceraceae Nastanthus patagonicus 152277 0\n",
      "SRR14242008 SRR14242008 Crassulaceae Meterostachys sikokianus 150775 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRR7901696 SRR7901696 Crassulaceae Pistorinia breviflora 145822 0\n",
      "SRR10121871 SRR10121871 Lauraceae Alseodaphnopsis hainanensis 152828 0\n",
      "SRR10121870 SRR10121870 Lauraceae Alseodaphnopsis petiolaris 154159 0\n",
      "SRR11342808 SRR11342808 Velloziaceae Barbaceniopsis castillonii 155669 0\n",
      "SRR5574039 SRR5574039 Annonaceae Klarobelia inundata 158399 0\n",
      "SRR5574040 SRR5574040 Annonaceae Mosannona costaricensis 159774 0\n",
      "SRR5574037 SRR5574037 Annonaceae Mosannona discolor 159850 0\n",
      "SRR5574038 SRR5574038 Annonaceae Onychopetalum periquino 160413 0\n",
      "SRR5574035 SRR5574035 Annonaceae Oxandra asbeckii 171100 0\n",
      "SRR5574036 SRR5574036 Annonaceae Oxandra polyantha 171139 0\n",
      "SRR5574041 SRR5574041 Annonaceae Xylopia peruviana 183356 0\n",
      "SRR11319310 SRR11319310 Brassicaceae Hirschfeldia incana 152980 0\n",
      "SRR10669182 SRR10669182 Brassicaceae Camelina sativa 153079 0\n",
      "ERR2560442 ERR2560442 Brassicaceae Erucastrum elatum 153094 0\n",
      "ERR2559998 ERR2559998 Brassicaceae Odontarrhena argentea 153749 0\n",
      "SRR1801304 SRR1801304 Brassicaceae Eruca vesicaria 153393 0\n",
      "SRR11653903 SRR11653903 Brassicaceae Psychine stylosa 153356 0\n",
      "SRR11653853 SRR11653853 Brassicaceae Leavenworthia uniflora 154891 0\n",
      "SRR8528391 SRR8528391 Brassicaceae Andrzeiowskia cardamine 154773 0\n",
      "SRR12599429 SRR12599429 Elaeocarpaceae Vallea stipularis 158456 0\n",
      "SRR8666637 SRR8666637 Sapindaceae Atalaya hemiglauca 160827 0\n",
      "SRR8844649 SRR8844649 Tetracarpaeaceae Tetracarpaea tasmannica 157788 0\n",
      "SRR7901273 SRR7901273 Saxifragaceae Jepsonia parryi 153056 0\n",
      "SRR7901268 SRR7901268 Saxifragaceae Darmera peltata 154923 0\n",
      "SRR5891624 SRR5891624 Vitaceae Causonis japonica 162501 0\n",
      "SRR15498137 SRR15498137 Cymodoceaceae Syringodium isoetifolium 159333 0\n",
      "SRR16114382 SRR16114382 Malvaceae Kosteletzkya pentacarpos 161777 0\n",
      "SRR15214094 SRR15214094 Menispermaceae Fibraurea recisa 161849 0\n",
      "SRR15130292 SRR15130292 Fabaceae Cheniella didyma 157186 0\n",
      "SRR16469679 SRR16469679 Hydrocharitaceae Hydrocharis morsus-ranae 153881 0\n",
      "ERR2990243 ERR2990243 Brassicaceae Diplotaxis tenuifolia 153467 0\n",
      "SRR15255748 SRR15255748 Tamaricaceae Myrtama elegans 155245 0\n",
      "SRR16953423 SRR16953423 Fabaceae Ammodendron bifolium 154533 0\n",
      "SRR8931123 SRR8931123 Lamiaceae Hyssopus officinalis 149817 0\n",
      "SRR11516645 SRR11516645 Verbenaceae Petrea volubilis 153368 0\n",
      "SRR15069871 SRR15069871 Plantaginaceae Gratiola ramosa 152643 0\n",
      "SRR12558543 SRR12558543 Plantaginaceae Picrorhiza kurroa 152662 0\n",
      "SRR15494298 SRR15494298 Poaceae Peridictyon sanctum 136914 0\n",
      "SRR15316867 SRR15316867 Poaceae Achnatherum pekinense 137837 0\n",
      "SRR15241102 SRR15241102 Gesneriaceae Henckelia pumila 153551 0\n",
      "SRR15904137 SRR15904137 Orchidaceae Hancockia uniflora 157833 0\n",
      "SRR13838408 SRR13838408 Bromeliaceae Quesnelia humilis 159215 0\n",
      "SRR13838382 SRR13838382 Bromeliaceae Hohenbergiopsis guatemalensis 159918 0\n",
      "SRR13838390 SRR13838390 Bromeliaceae Portea petropolitana 159792 0\n",
      "SRR13838373 SRR13838373 Bromeliaceae Nidularium fulgens 159613 0\n",
      "SRR13838400 SRR13838400 Bromeliaceae Orthophytum glabrum 159050 0\n",
      "SRR13838389 SRR13838389 Bromeliaceae Araeococcus micranthus 159481 0\n",
      "SRR13838403 SRR13838403 Bromeliaceae Androlepis fragrans 159781 0\n",
      "SRR13838393 SRR13838393 Bromeliaceae Cryptanthus bromelioides 159244 0\n",
      "SRR13838413 SRR13838413 Bromeliaceae Billbergia zebrina 159085 0\n",
      "SRR13838385 SRR13838385 Bromeliaceae Ronnbergia allenii 159492 0\n",
      "SRR13838421 SRR13838421 Bromeliaceae Sincoraea ulei 159851 0\n",
      "SRR9208067 SRR9208067 Asteraceae Elephantopus mollis 152195 0\n",
      "SRR12716187 SRR12716187 Asteraceae Leptorhynchos baileyi 152971 0\n",
      "SRR14572204 SRR14572204 Asteraceae Emilia abyssinica 151128 0\n",
      "SRR14572170 SRR14572170 Asteraceae Adenostyles alliariae 150428 0\n",
      "SRR13010887 SRR13010887 Asteraceae Perityle incana 151938 0\n"
     ]
    }
   ],
   "source": [
    "# Output fasta files \n",
    "for idx, row in whole_pt.iterrows():\n",
    "    seqs = list(SeqIO.parse(org_dir + DataSource + '/' + 'fasta_pt/' + row.Sample_Name + '_pt.fasta',format='fasta'))\n",
    "    # Verif only 1 sequence\n",
    "    if len(seqs) > 1:\n",
    "        print('error, more than 1 sequence')\n",
    "        pass\n",
    "    seq = seqs[0]\n",
    "    print(row.Sample_Name, row.SeqID, row.family, row.sci_name,len(seq),seq.seq.count('N'))\n",
    "    seq.id = row.SeqID\n",
    "    seq.description = ''\n",
    "    SeqIO.write(seq, geseq_dir + row.SeqID + '-circular.fasta',format='fasta')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ca72dd7",
   "metadata": {},
   "source": [
    "Submit files to GeSeq https://chlorobox.mpimp-golm.mpg.de/geseq.html and extract zip file in wdir/GeSeq_output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f30a3250",
   "metadata": {},
   "outputs": [],
   "source": [
    "whole_pt.to_csv(sdir + DataSource + '_wholePT.csv',index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cae22399",
   "metadata": {},
   "source": [
    "# Contigs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "8003ba28",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(124, 24)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SeqID</th>\n",
       "      <th>Ini_Species</th>\n",
       "      <th>TaxID</th>\n",
       "      <th>mol_type</th>\n",
       "      <th>length</th>\n",
       "      <th>Ini_sci_name</th>\n",
       "      <th>sci_name_query</th>\n",
       "      <th>Similar_match</th>\n",
       "      <th>Duplicates</th>\n",
       "      <th>kew_id</th>\n",
       "      <th>...</th>\n",
       "      <th>sci_name</th>\n",
       "      <th>authors</th>\n",
       "      <th>rank</th>\n",
       "      <th>taxonomic_status</th>\n",
       "      <th>Ini_kew_id</th>\n",
       "      <th>Ini_taxonomic_status</th>\n",
       "      <th>Duplicate_type</th>\n",
       "      <th>Nsp</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_045294.1</td>\n",
       "      <td>Acorus tatarinowii</td>\n",
       "      <td>123564</td>\n",
       "      <td>genomic DNA</td>\n",
       "      <td>153296</td>\n",
       "      <td>Acorus tatarinowii</td>\n",
       "      <td>Acorus tatarinowii</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>77188885-1</td>\n",
       "      <td>...</td>\n",
       "      <td>Acorus calamus var. angustatus</td>\n",
       "      <td>Besser</td>\n",
       "      <td>VARIETY</td>\n",
       "      <td>Accepted</td>\n",
       "      <td>84028-1</td>\n",
       "      <td>Synonym</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>Acorales</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_053912.1</td>\n",
       "      <td>Aponogeton lakhonensis</td>\n",
       "      <td>217757</td>\n",
       "      <td>genomic DNA</td>\n",
       "      <td>154860</td>\n",
       "      <td>Aponogeton lakhonensis</td>\n",
       "      <td>Aponogeton lakhonensis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>82954-1</td>\n",
       "      <td>...</td>\n",
       "      <td>Aponogeton lakhonensis</td>\n",
       "      <td>A.Camus</td>\n",
       "      <td>SPECIES</td>\n",
       "      <td>Accepted</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>62</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>Alismatales</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         SeqID             Ini_Species   TaxID     mol_type  length  \\\n",
       "0  NC_045294.1      Acorus tatarinowii  123564  genomic DNA  153296   \n",
       "1  NC_053912.1  Aponogeton lakhonensis  217757  genomic DNA  154860   \n",
       "\n",
       "             Ini_sci_name          sci_name_query Similar_match  Duplicates  \\\n",
       "0      Acorus tatarinowii      Acorus tatarinowii           NaN       False   \n",
       "1  Aponogeton lakhonensis  Aponogeton lakhonensis           NaN       False   \n",
       "\n",
       "       kew_id  ...                        sci_name  authors     rank  \\\n",
       "0  77188885-1  ...  Acorus calamus var. angustatus   Besser  VARIETY   \n",
       "1     82954-1  ...          Aponogeton lakhonensis  A.Camus  SPECIES   \n",
       "\n",
       "  taxonomic_status Ini_kew_id Ini_taxonomic_status Duplicate_type Nsp  \\\n",
       "0         Accepted    84028-1              Synonym            NaN   2   \n",
       "1         Accepted        NaN                  NaN            NaN  62   \n",
       "\n",
       "           class        order  \n",
       "0  Magnoliopsida     Acorales  \n",
       "1  Magnoliopsida  Alismatales  \n",
       "\n",
       "[2 rows x 24 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "references = pd.read_csv('References/NCBI_plastid_ref/NCBI_plastid_ref_list_SELECTED.csv')\n",
    "print(references.shape)\n",
    "references.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "2d5087aa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "327 samples 322 genera\n",
      "327 samples 322 genera\n",
      "327 samples 322 genera\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>idPaftol</th>\n",
       "      <th>idSequencing</th>\n",
       "      <th>Project</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>SumContigLength</th>\n",
       "      <th>sum_len</th>\n",
       "      <th>sci_name</th>\n",
       "      <th>...</th>\n",
       "      <th>Redo_FastPlast</th>\n",
       "      <th>error_pt</th>\n",
       "      <th>error_nr</th>\n",
       "      <th>SeqID</th>\n",
       "      <th>DataSource</th>\n",
       "      <th>ENASampleNum</th>\n",
       "      <th>ENAExpNumber</th>\n",
       "      <th>ENARunNumber</th>\n",
       "      <th>NCBI_TaxID</th>\n",
       "      <th>NCBI_sciname</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GAP_026365</th>\n",
       "      <td>12049.0</td>\n",
       "      <td>26365</td>\n",
       "      <td>GAP</td>\n",
       "      <td>Poales</td>\n",
       "      <td>Poaceae</td>\n",
       "      <td>Astrebla</td>\n",
       "      <td>lappacea</td>\n",
       "      <td>238665.0</td>\n",
       "      <td>238665.0</td>\n",
       "      <td>Astrebla lappacea</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GAP_026365</td>\n",
       "      <td>GAP</td>\n",
       "      <td>ERS9105966</td>\n",
       "      <td>ERX7170297</td>\n",
       "      <td>ERR7599717</td>\n",
       "      <td>66004.0</td>\n",
       "      <td>Astrebla lappacea</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GAP_026369</th>\n",
       "      <td>12052.0</td>\n",
       "      <td>26369</td>\n",
       "      <td>GAP</td>\n",
       "      <td>Poales</td>\n",
       "      <td>Poaceae</td>\n",
       "      <td>Phragmites</td>\n",
       "      <td>australis</td>\n",
       "      <td>240276.0</td>\n",
       "      <td>240276.0</td>\n",
       "      <td>Phragmites australis</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GAP_026369</td>\n",
       "      <td>GAP</td>\n",
       "      <td>ERS9105985</td>\n",
       "      <td>ERX7170316</td>\n",
       "      <td>ERR7599736</td>\n",
       "      <td>29695.0</td>\n",
       "      <td>Phragmites australis</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 46 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             idPaftol  idSequencing Project   order   family       genus  \\\n",
       "Sample_Name                                                                \n",
       "GAP_026365    12049.0         26365     GAP  Poales  Poaceae    Astrebla   \n",
       "GAP_026369    12052.0         26369     GAP  Poales  Poaceae  Phragmites   \n",
       "\n",
       "               species  SumContigLength   sum_len              sci_name  ...  \\\n",
       "Sample_Name                                                              ...   \n",
       "GAP_026365    lappacea         238665.0  238665.0     Astrebla lappacea  ...   \n",
       "GAP_026369   australis         240276.0  240276.0  Phragmites australis  ...   \n",
       "\n",
       "             Redo_FastPlast  error_pt  error_nr       SeqID  DataSource  \\\n",
       "Sample_Name                                                               \n",
       "GAP_026365             True       NaN       NaN  GAP_026365         GAP   \n",
       "GAP_026369             True       NaN       NaN  GAP_026369         GAP   \n",
       "\n",
       "             ENASampleNum  ENAExpNumber  ENARunNumber  NCBI_TaxID  \\\n",
       "Sample_Name                                                         \n",
       "GAP_026365     ERS9105966    ERX7170297    ERR7599717     66004.0   \n",
       "GAP_026369     ERS9105985    ERX7170316    ERR7599736     29695.0   \n",
       "\n",
       "                     NCBI_sciname  \n",
       "Sample_Name                        \n",
       "GAP_026365      Astrebla lappacea  \n",
       "GAP_026369   Phragmites australis  \n",
       "\n",
       "[2 rows x 46 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "incomplete_pt = Org_df[(Org_df.Assembly.isin(['circular genome','circular genome with gaps'])==False) &\n",
    "                   (Org_df.pt_recovered==True)]\n",
    "print(incomplete_pt.shape[0],'samples',incomplete_pt.genus.nunique(),'genera')\n",
    "incomplete_pt = incomplete_pt[incomplete_pt.Sum_len_pt<200000].set_index('Sample_Name')\n",
    "print(incomplete_pt.shape[0],'samples',incomplete_pt.genus.nunique(),'genera')\n",
    "### Random subset for test \n",
    "### Paul B. - changed from random sampling: incomplete_pt = incomplete_pt.sample(n=500)\n",
    "###incomplete_pt = incomplete_pt.head(500)\n",
    "print(incomplete_pt.shape[0],'samples',incomplete_pt.genus.nunique(),'genera')\n",
    "incomplete_pt[:2]\n",
    "#incomplete_pt[ incomplete_pt.SeqID == 'SRR9611034' ] ### PaulB. - this works\n",
    "###incomplete_pt[ incomplete_pt.Sample_Name == 'SRR9611034' ] ### PaulB. - but have lost access to Sample_Name - I think it has become the index.\n",
    "###incomplete_pt.loc['Sample_Name','SeqID']### Accessing the index with a string value fails though I think this is what you need to do.\n",
    "###Org_df[ Org_df.Sample_Name == 'SRR8666819']### Paul B - still present at this point - OK\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "998fa690",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SeqID</th>\n",
       "      <th>Ini_Species</th>\n",
       "      <th>TaxID</th>\n",
       "      <th>mol_type</th>\n",
       "      <th>length</th>\n",
       "      <th>Ini_sci_name</th>\n",
       "      <th>sci_name_query</th>\n",
       "      <th>Similar_match</th>\n",
       "      <th>Duplicates</th>\n",
       "      <th>kew_id</th>\n",
       "      <th>...</th>\n",
       "      <th>sci_name</th>\n",
       "      <th>authors</th>\n",
       "      <th>rank</th>\n",
       "      <th>taxonomic_status</th>\n",
       "      <th>Ini_kew_id</th>\n",
       "      <th>Ini_taxonomic_status</th>\n",
       "      <th>Duplicate_type</th>\n",
       "      <th>Nsp</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>NC_054306.1</td>\n",
       "      <td>Nardostachys jatamansi</td>\n",
       "      <td>179860</td>\n",
       "      <td>genomic DNA</td>\n",
       "      <td>155268</td>\n",
       "      <td>Nardostachys jatamansi</td>\n",
       "      <td>Nardostachys jatamansi</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>859418-1</td>\n",
       "      <td>...</td>\n",
       "      <td>Nardostachys jatamansi</td>\n",
       "      <td>(D.Don) DC.</td>\n",
       "      <td>SPECIES</td>\n",
       "      <td>Accepted</td>\n",
       "      <td>60457704-2</td>\n",
       "      <td>Synonym</td>\n",
       "      <td>Same_Taxon</td>\n",
       "      <td>1</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>Dipsacales</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>NC_033878.1</td>\n",
       "      <td>Sambucus williamsii</td>\n",
       "      <td>180062</td>\n",
       "      <td>genomic DNA</td>\n",
       "      <td>158305</td>\n",
       "      <td>Sambucus williamsii</td>\n",
       "      <td>Sambucus williamsii</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>149409-1</td>\n",
       "      <td>...</td>\n",
       "      <td>Sambucus williamsii</td>\n",
       "      <td>Hance</td>\n",
       "      <td>SPECIES</td>\n",
       "      <td>Accepted</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>25</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>Dipsacales</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          SeqID             Ini_Species   TaxID     mol_type  length  \\\n",
       "43  NC_054306.1  Nardostachys jatamansi  179860  genomic DNA  155268   \n",
       "44  NC_033878.1     Sambucus williamsii  180062  genomic DNA  158305   \n",
       "\n",
       "              Ini_sci_name          sci_name_query Similar_match  Duplicates  \\\n",
       "43  Nardostachys jatamansi  Nardostachys jatamansi           NaN        True   \n",
       "44     Sambucus williamsii     Sambucus williamsii           NaN       False   \n",
       "\n",
       "      kew_id  ...                sci_name      authors     rank  \\\n",
       "43  859418-1  ...  Nardostachys jatamansi  (D.Don) DC.  SPECIES   \n",
       "44  149409-1  ...     Sambucus williamsii        Hance  SPECIES   \n",
       "\n",
       "   taxonomic_status  Ini_kew_id Ini_taxonomic_status Duplicate_type Nsp  \\\n",
       "43         Accepted  60457704-2              Synonym     Same_Taxon   1   \n",
       "44         Accepted         NaN                  NaN            NaN  25   \n",
       "\n",
       "            class       order  \n",
       "43  Magnoliopsida  Dipsacales  \n",
       "44  Magnoliopsida  Dipsacales  \n",
       "\n",
       "[2 rows x 24 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "###references[references.order=='Sapindales']\n",
    "references[references.order=='Dipsacales']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "e92ce601",
   "metadata": {},
   "outputs": [],
   "source": [
    "###order_dc = {'Gunnerales':'Dilleniales','Picramniales':'Sapindales'}\n",
    "### Paul B. - added other orders - references.order might not have the order for the current sample (row.order)\n",
    "### Can test in 'references' above for an existing order  \n",
    "order_dc = {'Gunnerales':'Dilleniales','Picramniales':'Sapindales', 'Vahliales':'Lamiales', \\\n",
    "'Metteniusales':'Icacinales', 'Escalloniales':'Dipsacales'}\n",
    "# key=order not in reference set; value=nearest order in species tree "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "6a9c3625",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GAP_026365 Poales Poaceae\n",
      "\t Ref (1, 24): NC_035050.1 Poales Poaceae\n",
      "\t 14 / 14 contigs mapped to reference\n",
      "\t sum len concat: 109428 / 109428 100.0 %\n",
      "GAP_026369 Poales Poaceae\n",
      "\t Ref (1, 24): NC_035050.1 Poales Poaceae\n",
      "\t 9 / 9 contigs mapped to reference\n",
      "\t sum len concat: 116810 / 116810 100.0 %\n",
      "GAP_026371 Poales Poaceae\n",
      "\t Ref (1, 24): NC_035050.1 Poales Poaceae\n",
      "\t 8 / 8 contigs mapped to reference\n",
      "\t sum len concat: 216528 / 216528 100.0 %\n",
      "GAP_026381 Boraginales Boraginaceae\n",
      "\t Ref (1, 24): NC_053780.1 Boraginales Boraginaceae\n",
      "\t 32 / 32 contigs mapped to reference\n",
      "\t sum len concat: 90101 / 90101 100.0 %\n",
      "GAP_026383 Brassicales Brassicaceae\n",
      "\t Ref (1, 24): NC_049578.1 Brassicales Brassicaceae\n",
      "\t 2 / 2 contigs mapped to reference\n",
      "\t sum len concat: 155050 / 155050 100.0 %\n",
      "GAP_026385 Boraginales Boraginaceae\n",
      "\t Ref (1, 24): NC_053780.1 Boraginales Boraginaceae\n",
      "\t 10 / 10 contigs mapped to reference\n",
      "\t sum len concat: 143733 / 143733 100.0 %\n",
      "GAP_026391 Brassicales Brassicaceae\n",
      "\t Ref (1, 24): NC_049578.1 Brassicales Brassicaceae\n",
      "\t 3 / 3 contigs mapped to reference\n",
      "\t sum len concat: 155041 / 155041 100.0 %\n",
      "GAP_022035 Poales Poaceae\n",
      "\t Ref (1, 24): NC_035050.1 Poales Poaceae\n",
      "\t 5 / 5 contigs mapped to reference\n",
      "\t sum len concat: 38355 / 38355 100.0 %\n",
      "GAP_026393 Caryophyllales Nyctaginaceae\n",
      "\t Ref (2, 24): NC_046575.1 Caryophyllales Amaranthaceae\n",
      "\t 24 / 24 contigs mapped to reference\n",
      "\t sum len concat: 115569 / 115569 100.0 %\n",
      "GAP_022039 Ericales Ericaceae\n",
      "\t Ref (2, 24): NC_041127.1 Ericales Styracaceae\n",
      "\t 43 / 46 contigs mapped to reference\n",
      "\t sum len concat: 31333 / 32653 96.0 %\n",
      "GAP_026395 Brassicales Brassicaceae\n",
      "\t Ref (1, 24): NC_049578.1 Brassicales Brassicaceae\n",
      "\t 57 / 57 contigs mapped to reference\n",
      "\t sum len concat: 60302 / 60302 100.0 %\n",
      "GAP_026397 Brassicales Brassicaceae\n",
      "\t Ref (1, 24): NC_049578.1 Brassicales Brassicaceae\n",
      "\t 40 / 40 contigs mapped to reference\n",
      "\t sum len concat: 110210 / 110210 100.0 %\n",
      "GAP_026399 Poales Poaceae\n",
      "\t Ref (1, 24): NC_035050.1 Poales Poaceae\n",
      "\t 14 / 14 contigs mapped to reference\n",
      "\t sum len concat: 31641 / 31641 100.0 %\n",
      "GAP_022049 Malpighiales Rhizophoraceae\n",
      "\t Ref (2, 24): NC_030601.1 Malpighiales Erythroxylaceae\n",
      "\t 9 / 10 contigs mapped to reference\n",
      "\t sum len concat: 36273 / 37020 98.0 %\n",
      "GAP_026407 Gentianales Rubiaceae\n",
      "\t Ref (2, 24): NC_054359.1 Gentianales Gentianaceae\n",
      "\t 8 / 8 contigs mapped to reference\n",
      "\t sum len concat: 124286 / 124286 100.0 %\n",
      "GAP_026409 Lamiales Plantaginaceae\n",
      "\t Ref (2, 24): NC_053620.1 Lamiales Lamiaceae\n",
      "\t 9 / 9 contigs mapped to reference\n",
      "\t sum len concat: 129956 / 129956 100.0 %\n",
      "GAP_026411 Brassicales Brassicaceae\n",
      "\t Ref (1, 24): NC_049578.1 Brassicales Brassicaceae\n",
      "\t 9 / 9 contigs mapped to reference\n",
      "\t sum len concat: 47323 / 47323 100.0 %\n",
      "GAP_026413 Brassicales Brassicaceae\n",
      "\t Ref (1, 24): NC_049578.1 Brassicales Brassicaceae\n",
      "\t 22 / 22 contigs mapped to reference\n",
      "\t sum len concat: 118992 / 118992 100.0 %\n",
      "GAP_026415 Gentianales Loganiaceae\n",
      "\t Ref (1, 24): NC_050922.1 Gentianales Loganiaceae\n",
      "\t 48 / 48 contigs mapped to reference\n",
      "\t sum len concat: 78111 / 78111 100.0 %\n",
      "GAP_022079 Ericales Ericaceae\n",
      "\t Ref (2, 24): NC_041127.1 Ericales Styracaceae\n",
      "\t 17 / 17 contigs mapped to reference\n",
      "\t sum len concat: 24426 / 24426 100.0 %\n",
      "GAP_026417 Poales Poaceae\n",
      "\t Ref (1, 24): NC_035050.1 Poales Poaceae\n",
      "\t 5 / 5 contigs mapped to reference\n",
      "\t sum len concat: 119790 / 119790 100.0 %\n",
      "GAP_026419 Brassicales Brassicaceae\n",
      "\t Ref (1, 24): NC_049578.1 Brassicales Brassicaceae\n",
      "\t 2 / 2 contigs mapped to reference\n",
      "\t sum len concat: 154242 / 154242 100.0 %\n",
      "GAP_026421 Commelinales Commelinaceae\n",
      "\t Ref (2, 24): NC_040160.1 Commelinales Haemodoraceae\n",
      "\t 17 / 17 contigs mapped to reference\n",
      "\t sum len concat: 156482 / 156482 100.0 %\n",
      "GAP_022091 Malpighiales Salicaceae\n",
      "\t Ref (1, 24): NC_045235.1 Malpighiales Salicaceae\n",
      "\t 10 / 10 contigs mapped to reference\n",
      "\t sum len concat: 34079 / 34079 100.0 %\n",
      "GAP_026423 Poales Poaceae\n",
      "\t Ref (1, 24): NC_035050.1 Poales Poaceae\n",
      "\t 25 / 25 contigs mapped to reference\n",
      "\t sum len concat: 105583 / 105583 100.0 %\n",
      "GAP_026425 Poales Poaceae\n",
      "\t Ref (1, 24): NC_035050.1 Poales Poaceae\n",
      "\t 12 / 12 contigs mapped to reference\n",
      "\t sum len concat: 132518 / 132518 100.0 %\n",
      "GAP_026429 Caryophyllales Caryophyllaceae\n",
      "\t Ref (2, 24): NC_046575.1 Caryophyllales Amaranthaceae\n",
      "\t 17 / 18 contigs mapped to reference\n",
      "\t sum len concat: 110992 / 111455 99.6 %\n",
      "GAP_022103 Ericales Ericaceae\n",
      "\t Ref (2, 24): NC_041127.1 Ericales Styracaceae\n",
      "\t 15 / 15 contigs mapped to reference\n",
      "\t sum len concat: 67077 / 67077 100.0 %\n",
      "GAP_022105 Malpighiales Picrodendraceae\n",
      "\t Ref (2, 24): NC_030601.1 Malpighiales Erythroxylaceae\n",
      "\t 16 / 16 contigs mapped to reference\n",
      "\t sum len concat: 110842 / 110842 100.0 %\n",
      "GAP_022109 Gentianales Apocynaceae\n",
      "\t Ref (2, 24): NC_054359.1 Gentianales Gentianaceae\n",
      "\t 27 / 29 contigs mapped to reference\n",
      "\t sum len concat: 34432 / 35310 97.5 %\n",
      "GAP_026433 Gentianales Rubiaceae\n",
      "\t Ref (2, 24): NC_054359.1 Gentianales Gentianaceae\n",
      "\t 20 / 21 contigs mapped to reference\n",
      "\t sum len concat: 117915 / 118431 99.6 %\n",
      "GAP_022113 Malpighiales Picrodendraceae\n",
      "\t Ref (2, 24): NC_030601.1 Malpighiales Erythroxylaceae\n",
      "\t 9 / 10 contigs mapped to reference\n",
      "\t sum len concat: 117947 / 123328 95.6 %\n",
      "GAP_022117 Ranunculales Menispermaceae\n",
      "\t Ref (2, 24): NC_029427.1 Ranunculales Lardizabalaceae\n",
      "\t 28 / 28 contigs mapped to reference\n",
      "\t sum len concat: 107702 / 107702 100.0 %\n",
      "GAP_026435 Brassicales Brassicaceae\n",
      "\t Ref (1, 24): NC_049578.1 Brassicales Brassicaceae\n",
      "\t 5 / 5 contigs mapped to reference\n",
      "\t sum len concat: 155044 / 155044 100.0 %\n",
      "GAP_022127 Gentianales Apocynaceae\n",
      "\t Ref (2, 24): NC_054359.1 Gentianales Gentianaceae\n",
      "\t 12 / 12 contigs mapped to reference\n",
      "\t sum len concat: 136275 / 136275 100.0 %\n",
      "GAP_022129 Gentianales Rubiaceae\n",
      "\t Ref (2, 24): NC_054359.1 Gentianales Gentianaceae\n",
      "\t 32 / 32 contigs mapped to reference\n",
      "\t sum len concat: 105748 / 105748 100.0 %\n",
      "GAP_022131 Ericales Sapotaceae\n",
      "\t Ref (2, 24): NC_041127.1 Ericales Styracaceae\n",
      "\t 13 / 13 contigs mapped to reference\n",
      "\t sum len concat: 25222 / 25222 100.0 %\n",
      "GAP_022133 Dilleniales Dilleniaceae\n",
      "\t Ref (1, 24): NC_042740.1 Dilleniales Dilleniaceae\n",
      "\t 47 / 47 contigs mapped to reference\n",
      "\t sum len concat: 75322 / 75322 100.0 %\n",
      "GAP_026439 Brassicales Brassicaceae\n",
      "\t Ref (1, 24): NC_049578.1 Brassicales Brassicaceae\n",
      "\t 34 / 34 contigs mapped to reference\n",
      "\t sum len concat: 82417 / 82417 100.0 %\n",
      "GAP_026441 Ericales Ericaceae\n",
      "\t Ref (2, 24): NC_041127.1 Ericales Styracaceae\n",
      "\t 18 / 18 contigs mapped to reference\n",
      "\t sum len concat: 31267 / 31267 100.0 %\n",
      "GAP_022141 Gentianales Rubiaceae\n",
      "\t Ref (2, 24): NC_054359.1 Gentianales Gentianaceae\n",
      "\t 46 / 53 contigs mapped to reference\n",
      "\t sum len concat: 68010 / 70681 96.2 %\n",
      "GAP_026443 Caryophyllales Caryophyllaceae\n",
      "\t Ref (2, 24): NC_046575.1 Caryophyllales Amaranthaceae\n",
      "\t 31 / 32 contigs mapped to reference\n",
      "\t sum len concat: 109545 / 110026 99.6 %\n",
      "GAP_026447 Lamiales Orobanchaceae\n",
      "\t Ref (2, 24): NC_053620.1 Lamiales Lamiaceae\n",
      "\t 9 / 9 contigs mapped to reference\n",
      "\t sum len concat: 119906 / 119906 100.0 %\n",
      "GAP_026449 Poales Restionaceae\n",
      "\t Ref (2, 24): NC_042211.1 Poales Eriocaulaceae\n",
      "\t 15 / 16 contigs mapped to reference\n",
      "\t sum len concat: 37504 / 39015 96.1 %\n",
      "GAP_026451 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 48 / 53 contigs mapped to reference\n",
      "\t sum len concat: 48281 / 50090 96.4 %\n",
      "GAP_022155 Myrtales Myrtaceae\n",
      "\t Ref (1, 24): NC_039395.1 Myrtales Myrtaceae\n",
      "\t 54 / 54 contigs mapped to reference\n",
      "\t sum len concat: 79194 / 79194 100.0 %\n",
      "GAP_022157 Poales Poaceae\n",
      "\t Ref (1, 24): NC_035050.1 Poales Poaceae\n",
      "\t 14 / 14 contigs mapped to reference\n",
      "\t sum len concat: 31375 / 31375 100.0 %\n",
      "GAP_026453 Lamiales Acanthaceae\n",
      "\t Ref (2, 24): NC_053620.1 Lamiales Lamiaceae\n",
      "\t 41 / 41 contigs mapped to reference\n",
      "\t sum len concat: 47471 / 47471 100.0 %\n",
      "GAP_026457 Caryophyllales Molluginaceae\n",
      "\t Ref (2, 24): NC_046575.1 Caryophyllales Amaranthaceae\n",
      "\t 38 / 38 contigs mapped to reference\n",
      "\t sum len concat: 107826 / 107826 100.0 %\n",
      "GAP_022171 Gentianales Apocynaceae\n",
      "\t Ref (2, 24): NC_054359.1 Gentianales Gentianaceae\n",
      "\t 30 / 33 contigs mapped to reference\n",
      "\t sum len concat: 80598 / 82484 97.7 %\n",
      "GAP_022173 Malpighiales Picrodendraceae\n",
      "\t Ref (2, 24): NC_030601.1 Malpighiales Erythroxylaceae\n",
      "\t 4 / 4 contigs mapped to reference\n",
      "\t sum len concat: 21760 / 21760 100.0 %\n",
      "GAP_022183 Lamiales Linderniaceae\n",
      "\t Ref (2, 24): NC_053620.1 Lamiales Lamiaceae\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\t 4 / 4 contigs mapped to reference\n",
      "\t sum len concat: 128090 / 128090 100.0 %\n",
      "GAP_026461 Caryophyllales Montiaceae\n",
      "\t Ref (2, 24): NC_046575.1 Caryophyllales Amaranthaceae\n",
      "\t 48 / 49 contigs mapped to reference\n",
      "\t sum len concat: 78649 / 79071 99.5 %\n",
      "GAP_026463 Poales Poaceae\n",
      "\t Ref (1, 24): NC_035050.1 Poales Poaceae\n",
      "\t 34 / 34 contigs mapped to reference\n",
      "\t sum len concat: 40930 / 40930 100.0 %\n",
      "GAP_026467 Solanales Convolvulaceae\n",
      "\t Ref (1, 24): NC_042940.1 Solanales Convolvulaceae\n",
      "\t 46 / 46 contigs mapped to reference\n",
      "\t sum len concat: 31228 / 31228 100.0 %\n",
      "GAP_026469 Brassicales Gyrostemonaceae\n",
      "\t Ref (2, 24): NC_049578.1 Brassicales Brassicaceae\n",
      "\t 24 / 24 contigs mapped to reference\n",
      "\t sum len concat: 34959 / 34959 100.0 %\n",
      "GAP_026475 Gentianales Gentianaceae\n",
      "\t Ref (1, 24): NC_054359.1 Gentianales Gentianaceae\n",
      "\t 39 / 39 contigs mapped to reference\n",
      "\t sum len concat: 74107 / 74107 100.0 %\n",
      "GAP_022217 Gentianales Rubiaceae\n",
      "\t Ref (2, 24): NC_054359.1 Gentianales Gentianaceae\n",
      "\t 27 / 29 contigs mapped to reference\n",
      "\t sum len concat: 87341 / 88088 99.2 %\n",
      "GAP_022219 Poales Poaceae\n",
      "\t Ref (1, 24): NC_035050.1 Poales Poaceae\n",
      "\t 32 / 32 contigs mapped to reference\n",
      "\t sum len concat: 31529 / 31529 100.0 %\n",
      "GAP_026483 Poales Restionaceae\n",
      "\t Ref (2, 24): NC_042211.1 Poales Eriocaulaceae\n",
      "\t 9 / 9 contigs mapped to reference\n",
      "\t sum len concat: 119219 / 119219 100.0 %\n",
      "GAP_026485 Lamiales Linderniaceae\n",
      "\t Ref (2, 24): NC_053620.1 Lamiales Lamiaceae\n",
      "\t 46 / 47 contigs mapped to reference\n",
      "\t sum len concat: 90007 / 90345 99.6 %\n",
      "GAP_026487 Poales Poaceae\n",
      "\t Ref (1, 24): NC_035050.1 Poales Poaceae\n",
      "\t 6 / 6 contigs mapped to reference\n",
      "\t sum len concat: 102283 / 102283 100.0 %\n",
      "GAP_026497 Lamiales Phrymaceae\n",
      "\t Ref (2, 24): NC_053620.1 Lamiales Lamiaceae\n",
      "\t 46 / 46 contigs mapped to reference\n",
      "\t sum len concat: 29628 / 29628 100.0 %\n",
      "GAP_026499 Asterales Asteraceae\n",
      "\t Ref (1, 24): NC_044728.1 Asterales Asteraceae\n",
      "\t 53 / 53 contigs mapped to reference\n",
      "\t sum len concat: 43091 / 43091 100.0 %\n",
      "GAP_026501 Brassicales Brassicaceae\n",
      "\t Ref (1, 24): NC_049578.1 Brassicales Brassicaceae\n",
      "\t 52 / 52 contigs mapped to reference\n",
      "\t sum len concat: 88533 / 88533 100.0 %\n",
      "GAP_026503 Arecales Dasypogonaceae\n",
      "\t Ref (1, 24): NC_029970.1 Arecales Dasypogonaceae\n",
      "\t 37 / 38 contigs mapped to reference\n",
      "\t sum len concat: 65277 / 67356 96.9 %\n",
      "GAP_026505 Lamiales Phrymaceae\n",
      "\t Ref (2, 24): NC_053620.1 Lamiales Lamiaceae\n",
      "\t 32 / 33 contigs mapped to reference\n",
      "\t sum len concat: 100897 / 104555 96.5 %\n",
      "GAP_026507 Poales Poaceae\n",
      "\t Ref (1, 24): NC_035050.1 Poales Poaceae\n",
      "\t 50 / 50 contigs mapped to reference\n",
      "\t sum len concat: 72081 / 72081 100.0 %\n",
      "GAP_026511 Poales Poaceae\n",
      "\t Ref (1, 24): NC_035050.1 Poales Poaceae\n",
      "\t 33 / 33 contigs mapped to reference\n",
      "\t sum len concat: 49454 / 49454 100.0 %\n",
      "GAP_026513 Lamiales Phrymaceae\n",
      "\t Ref (2, 24): NC_053620.1 Lamiales Lamiaceae\n",
      "\t 50 / 50 contigs mapped to reference\n",
      "\t sum len concat: 38956 / 38956 100.0 %\n",
      "GAP_026515 Lamiales Orobanchaceae\n",
      "\t Ref (2, 24): NC_053620.1 Lamiales Lamiaceae\n",
      "\t 7 / 7 contigs mapped to reference\n",
      "\t sum len concat: 124248 / 124248 100.0 %\n",
      "GAP_026517 Lamiales Orobanchaceae\n",
      "\t Ref (2, 24): NC_053620.1 Lamiales Lamiaceae\n",
      "\t 1 / 1 contigs mapped to reference\n",
      "\t sum len concat: 150611 / 150611 100.0 %\n",
      "GAP_026521 Malvales Malvaceae\n",
      "\t Ref (1, 24): NC_038057.1 Malvales Malvaceae\n",
      "\t 7 / 7 contigs mapped to reference\n",
      "\t sum len concat: 134411 / 134411 100.0 %\n",
      "GAP_022271 Ranunculales Menispermaceae\n",
      "\t Ref (2, 24): NC_029427.1 Ranunculales Lardizabalaceae\n",
      "\t 33 / 35 contigs mapped to reference\n",
      "\t sum len concat: 63479 / 64982 97.7 %\n",
      "GAP_026527 Malvales Malvaceae\n",
      "\t Ref (1, 24): NC_038057.1 Malvales Malvaceae\n",
      "\t 19 / 20 contigs mapped to reference\n",
      "\t sum len concat: 42019 / 43522 96.5 %\n",
      "GAP_026531 Malvales Malvaceae\n",
      "\t Ref (1, 24): NC_038057.1 Malvales Malvaceae\n",
      "\t 38 / 38 contigs mapped to reference\n",
      "\t sum len concat: 38292 / 38292 100.0 %\n",
      "GAP_022299 Malpighiales Rhizophoraceae\n",
      "\t Ref (2, 24): NC_030601.1 Malpighiales Erythroxylaceae\n",
      "\t 4 / 4 contigs mapped to reference\n",
      "\t sum len concat: 35061 / 35061 100.0 %\n",
      "GAP_022301 Malpighiales Picrodendraceae\n",
      "\t Ref (2, 24): NC_030601.1 Malpighiales Erythroxylaceae\n",
      "\t 44 / 50 contigs mapped to reference\n",
      "\t sum len concat: 34005 / 36321 93.6 %\n",
      "GAP_026537 Arecales Dasypogonaceae\n",
      "\t Ref (1, 24): NC_029970.1 Arecales Dasypogonaceae\n",
      "\t 10 / 10 contigs mapped to reference\n",
      "\t sum len concat: 124164 / 124164 100.0 %\n",
      "GAP_022311 Saxifragales Hamamelidaceae\n",
      "\t Ref (2, 24): NC_037885.1 Saxifragales Haloragaceae\n",
      "\t 47 / 49 contigs mapped to reference\n",
      "\t sum len concat: 84625 / 85884 98.5 %\n",
      "GAP_022313 Apiales Araliaceae\n",
      "\t Ref (1, 24): NC_022812.1 Apiales Araliaceae\n",
      "\t 7 / 7 contigs mapped to reference\n",
      "\t sum len concat: 36912 / 36912 100.0 %\n",
      "GAP_026543 Brassicales Brassicaceae\n",
      "\t Ref (1, 24): NC_049578.1 Brassicales Brassicaceae\n",
      "\t 23 / 23 contigs mapped to reference\n",
      "\t sum len concat: 119199 / 119199 100.0 %\n",
      "GAP_026547 Alismatales Alismataceae\n",
      "\t Ref (2, 24): NC_053912.1 Alismatales Aponogetonaceae\n",
      "\t 47 / 50 contigs mapped to reference\n",
      "\t sum len concat: 58688 / 59917 97.9 %\n",
      "GAP_022329 Gentianales Rubiaceae\n",
      "\t Ref (2, 24): NC_054359.1 Gentianales Gentianaceae\n",
      "\t 14 / 15 contigs mapped to reference\n",
      "\t sum len concat: 26355 / 27031 97.5 %\n",
      "GAP_026561 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 45 / 48 contigs mapped to reference\n",
      "\t sum len concat: 61833 / 62926 98.3 %\n",
      "GAP_022353 Apiales Araliaceae\n",
      "\t Ref (1, 24): NC_022812.1 Apiales Araliaceae\n",
      "\t 16 / 16 contigs mapped to reference\n",
      "\t sum len concat: 120594 / 120594 100.0 %\n",
      "GAP_022355 Solanales Convolvulaceae\n",
      "\t Ref (1, 24): NC_042940.1 Solanales Convolvulaceae\n",
      "\t 21 / 22 contigs mapped to reference\n",
      "\t sum len concat: 31456 / 31874 98.7 %\n",
      "GAP_026571 Malvales Malvaceae\n",
      "\t Ref (1, 24): NC_038057.1 Malvales Malvaceae\n",
      "\t 30 / 31 contigs mapped to reference\n",
      "\t sum len concat: 99332 / 100474 98.9 %\n",
      "GAP_026573 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 40 / 42 contigs mapped to reference\n",
      "\t sum len concat: 36671 / 38494 95.3 %\n",
      "GAP_026581 Poales Restionaceae\n",
      "\t Ref (2, 24): NC_042211.1 Poales Eriocaulaceae\n",
      "\t 11 / 13 contigs mapped to reference\n",
      "\t sum len concat: 35373 / 37278 94.9 %\n",
      "GAP_026587 Poales Restionaceae\n",
      "\t Ref (2, 24): NC_042211.1 Poales Eriocaulaceae\n",
      "\t 13 / 13 contigs mapped to reference\n",
      "\t sum len concat: 32031 / 32031 100.0 %\n",
      "GAP_026589 Poales Restionaceae\n",
      "\t Ref (2, 24): NC_042211.1 Poales Eriocaulaceae\n",
      "\t 11 / 14 contigs mapped to reference\n",
      "\t sum len concat: 26164 / 31426 83.3 %\n",
      "GAP_026597 Poales Restionaceae\n",
      "\t Ref (2, 24): NC_042211.1 Poales Eriocaulaceae\n",
      "\t 19 / 23 contigs mapped to reference\n",
      "\t sum len concat: 32008 / 38723 82.7 %\n",
      "GAP_026601 Poales Restionaceae\n",
      "\t Ref (2, 24): NC_042211.1 Poales Eriocaulaceae\n",
      "\t 25 / 33 contigs mapped to reference\n",
      "\t sum len concat: 28754 / 33728 85.3 %\n",
      "GAP_026611 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 33 / 39 contigs mapped to reference\n",
      "\t sum len concat: 24724 / 28484 86.8 %\n",
      "GAP_026619 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 43 / 45 contigs mapped to reference\n",
      "\t sum len concat: 90117 / 91461 98.5 %\n",
      "GAP_026621 Malpighiales Phyllanthaceae\n",
      "\t Ref (2, 24): NC_030601.1 Malpighiales Erythroxylaceae\n",
      "\t 39 / 44 contigs mapped to reference\n",
      "\t sum len concat: 88114 / 90425 97.4 %\n",
      "GAP_026631 Lamiales Acanthaceae\n",
      "\t Ref (2, 24): NC_053620.1 Lamiales Lamiaceae\n",
      "\t 48 / 49 contigs mapped to reference\n",
      "\t sum len concat: 47298 / 47562 99.4 %\n",
      "GAP_026633 Caryophyllales Caryophyllaceae\n",
      "\t Ref (2, 24): NC_046575.1 Caryophyllales Amaranthaceae\n",
      "\t 3 / 3 contigs mapped to reference\n",
      "\t sum len concat: 150624 / 150624 100.0 %\n",
      "GAP_026639 Asterales Goodeniaceae\n",
      "\t Ref (2, 24): NC_044728.1 Asterales Asteraceae\n",
      "\t 19 / 22 contigs mapped to reference\n",
      "\t sum len concat: 122899 / 123550 99.5 %\n",
      "GAP_026641 Poales Poaceae\n",
      "\t Ref (1, 24): NC_035050.1 Poales Poaceae\n",
      "\t 51 / 51 contigs mapped to reference\n",
      "\t sum len concat: 55854 / 55854 100.0 %\n",
      "GAP_026643 Fabales Polygalaceae\n",
      "\t Ref (1, 24): NC_052912.1 Fabales Polygalaceae\n",
      "\t 16 / 16 contigs mapped to reference\n",
      "\t sum len concat: 122966 / 122966 100.0 %\n",
      "GAP_026649 Asterales Goodeniaceae\n",
      "\t Ref (2, 24): NC_044728.1 Asterales Asteraceae\n",
      "\t 43 / 45 contigs mapped to reference\n",
      "\t sum len concat: 56940 / 57929 98.3 %\n",
      "GAP_026651 Poales Poaceae\n",
      "\t Ref (1, 24): NC_035050.1 Poales Poaceae\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\t 5 / 5 contigs mapped to reference\n",
      "\t sum len concat: 118503 / 118503 100.0 %\n",
      "GAP_026653 Poales Poaceae\n",
      "\t Ref (1, 24): NC_035050.1 Poales Poaceae\n",
      "\t 21 / 21 contigs mapped to reference\n",
      "\t sum len concat: 108615 / 108615 100.0 %\n",
      "GAP_026655 Lamiales Phrymaceae\n",
      "\t Ref (2, 24): NC_053620.1 Lamiales Lamiaceae\n",
      "\t 41 / 42 contigs mapped to reference\n",
      "\t sum len concat: 38631 / 38899 99.3 %\n",
      "GAP_026669 Caryophyllales Caryophyllaceae\n",
      "\t Ref (2, 24): NC_046575.1 Caryophyllales Amaranthaceae\n",
      "\t 3 / 3 contigs mapped to reference\n",
      "\t sum len concat: 146645 / 146645 100.0 %\n",
      "GAP_026671 Boraginales Boraginaceae\n",
      "\t Ref (1, 24): NC_053780.1 Boraginales Boraginaceae\n",
      "\t 42 / 42 contigs mapped to reference\n",
      "\t sum len concat: 93693 / 93693 100.0 %\n",
      "GAP_026673 Poales Poaceae\n",
      "\t Ref (1, 24): NC_035050.1 Poales Poaceae\n",
      "\t 1 / 1 contigs mapped to reference\n",
      "\t sum len concat: 23720 / 23720 100.0 %\n",
      "GAP_026681 Poales Restionaceae\n",
      "\t Ref (2, 24): NC_042211.1 Poales Eriocaulaceae\n",
      "\t 31 / 40 contigs mapped to reference\n",
      "\t sum len concat: 86872 / 96173 90.3 %\n",
      "GAP_022479 Lamiales Acanthaceae\n",
      "\t Ref (2, 24): NC_053620.1 Lamiales Lamiaceae\n",
      "\t 13 / 13 contigs mapped to reference\n",
      "\t sum len concat: 91422 / 91422 100.0 %\n",
      "GAP_026691 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 39 / 45 contigs mapped to reference\n",
      "\t sum len concat: 80970 / 83929 96.5 %\n",
      "GAP_026693 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 13 / 14 contigs mapped to reference\n",
      "\t sum len concat: 25437 / 27079 93.9 %\n",
      "GAP_026695 Caryophyllales Amaranthaceae\n",
      "\t Ref (1, 24): NC_046575.1 Caryophyllales Amaranthaceae\n",
      "\t 47 / 47 contigs mapped to reference\n",
      "\t sum len concat: 91188 / 91188 100.0 %\n",
      "GAP_026697 Brassicales Gyrostemonaceae\n",
      "\t Ref (2, 24): NC_049578.1 Brassicales Brassicaceae\n",
      "\t 24 / 24 contigs mapped to reference\n",
      "\t sum len concat: 47906 / 47906 100.0 %\n",
      "GAP_026699 Poales Poaceae\n",
      "\t Ref (1, 24): NC_035050.1 Poales Poaceae\n",
      "\t 26 / 26 contigs mapped to reference\n",
      "\t sum len concat: 47272 / 47272 100.0 %\n",
      "GAP_026701 Malvales Malvaceae\n",
      "\t Ref (1, 24): NC_038057.1 Malvales Malvaceae\n",
      "\t 4 / 4 contigs mapped to reference\n",
      "\t sum len concat: 158334 / 158334 100.0 %\n",
      "GAP_022501 Commelinales Philydraceae\n",
      "\t Ref (1, 24): NC_039424.1 Commelinales Philydraceae\n",
      "\t 30 / 30 contigs mapped to reference\n",
      "\t sum len concat: 110453 / 110453 100.0 %\n",
      "GAP_026707 Poales Restionaceae\n",
      "\t Ref (2, 24): NC_042211.1 Poales Eriocaulaceae\n",
      "\t 28 / 46 contigs mapped to reference\n",
      "\t sum len concat: 39483 / 51527 76.6 %\n",
      "GAP_026715 Malpighiales Phyllanthaceae\n",
      "\t Ref (2, 24): NC_030601.1 Malpighiales Erythroxylaceae\n",
      "\t 17 / 17 contigs mapped to reference\n",
      "\t sum len concat: 29324 / 29324 100.0 %\n",
      "GAP_022521 Malpighiales Salicaceae\n",
      "\t Ref (1, 24): NC_045235.1 Malpighiales Salicaceae\n",
      "\t 9 / 9 contigs mapped to reference\n",
      "\t sum len concat: 126988 / 126988 100.0 %\n",
      "GAP_026723 Caryophyllales Amaranthaceae\n",
      "\t Ref (1, 24): NC_046575.1 Caryophyllales Amaranthaceae\n",
      "\t 21 / 21 contigs mapped to reference\n",
      "\t sum len concat: 55612 / 55612 100.0 %\n",
      "GAP_026731 Proteales Proteaceae\n",
      "\t Ref (1, 24): NC_036416.1 Proteales Proteaceae\n",
      "\t 4 / 6 contigs mapped to reference\n",
      "\t sum len concat: 26177 / 29272 89.4 %\n",
      "GAP_026741 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 29 / 31 contigs mapped to reference\n",
      "\t sum len concat: 73983 / 76039 97.3 %\n",
      "GAP_026743 Lamiales Lamiaceae\n",
      "\t Ref (1, 24): NC_053620.1 Lamiales Lamiaceae\n",
      "\t 6 / 6 contigs mapped to reference\n",
      "\t sum len concat: 137879 / 137879 100.0 %\n",
      "GAP_026745 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 19 / 21 contigs mapped to reference\n",
      "\t sum len concat: 30035 / 30620 98.1 %\n",
      "GAP_026747 Malpighiales Phyllanthaceae\n",
      "\t Ref (2, 24): NC_030601.1 Malpighiales Erythroxylaceae\n",
      "\t 11 / 11 contigs mapped to reference\n",
      "\t sum len concat: 126919 / 126919 100.0 %\n",
      "GAP_026749 Apiales Pittosporaceae\n",
      "\t Ref (2, 24): NC_054296.1 Apiales Apiaceae\n",
      "\t 5 / 6 contigs mapped to reference\n",
      "\t sum len concat: 29154 / 31503 92.5 %\n",
      "GAP_026751 Malpighiales Phyllanthaceae\n",
      "\t Ref (2, 24): NC_030601.1 Malpighiales Erythroxylaceae\n",
      "\t 47 / 49 contigs mapped to reference\n",
      "\t sum len concat: 51038 / 51844 98.4 %\n",
      "GAP_022557 Gentianales Apocynaceae\n",
      "\t Ref (2, 24): NC_054359.1 Gentianales Gentianaceae\n",
      "\t 5 / 5 contigs mapped to reference\n",
      "\t sum len concat: 153092 / 153092 100.0 %\n",
      "GAP_026753 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 18 / 18 contigs mapped to reference\n",
      "\t sum len concat: 120756 / 120756 100.0 %\n",
      "GAP_026763 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 24 / 24 contigs mapped to reference\n",
      "\t sum len concat: 35352 / 35352 100.0 %\n",
      "GAP_026765 Malpighiales Phyllanthaceae\n",
      "\t Ref (2, 24): NC_030601.1 Malpighiales Erythroxylaceae\n",
      "\t 23 / 23 contigs mapped to reference\n",
      "\t sum len concat: 86563 / 86563 100.0 %\n",
      "GAP_026767 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 41 / 45 contigs mapped to reference\n",
      "\t sum len concat: 80362 / 82054 97.9 %\n",
      "GAP_026769 Lamiales Lamiaceae\n",
      "\t Ref (1, 24): NC_053620.1 Lamiales Lamiaceae\n",
      "\t 16 / 16 contigs mapped to reference\n",
      "\t sum len concat: 126093 / 126093 100.0 %\n",
      "GAP_026771 Myrtales Myrtaceae\n",
      "\t Ref (1, 24): NC_039395.1 Myrtales Myrtaceae\n",
      "\t 29 / 29 contigs mapped to reference\n",
      "\t sum len concat: 109490 / 109490 100.0 %\n",
      "GAP_026773 Asterales Stylidiaceae\n",
      "\t Ref (2, 24): NC_044728.1 Asterales Asteraceae\n",
      "\t 13 / 16 contigs mapped to reference\n",
      "\t sum len concat: 28844 / 30986 93.1 %\n",
      "GAP_026779 Asterales Stylidiaceae\n",
      "\t Ref (2, 24): NC_044728.1 Asterales Asteraceae\n",
      "\t 38 / 39 contigs mapped to reference\n",
      "\t sum len concat: 75696 / 76573 98.9 %\n",
      "GAP_022587 Poales Poaceae\n",
      "\t Ref (1, 24): NC_035050.1 Poales Poaceae\n",
      "\t 44 / 44 contigs mapped to reference\n",
      "\t sum len concat: 69625 / 69625 100.0 %\n",
      "GAP_022589 Ericales Ericaceae\n",
      "\t Ref (2, 24): NC_041127.1 Ericales Styracaceae\n",
      "\t 38 / 44 contigs mapped to reference\n",
      "\t sum len concat: 42824 / 45044 95.1 %\n",
      "GAP_026783 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 4 / 4 contigs mapped to reference\n",
      "\t sum len concat: 148546 / 148546 100.0 %\n",
      "GAP_026785 Malvales Malvaceae\n",
      "\t Ref (1, 24): NC_038057.1 Malvales Malvaceae\n",
      "\t 52 / 52 contigs mapped to reference\n",
      "\t sum len concat: 68486 / 68486 100.0 %\n",
      "GAP_026791 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 28 / 29 contigs mapped to reference\n",
      "\t sum len concat: 70540 / 72120 97.8 %\n",
      "GAP_026793 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 44 / 45 contigs mapped to reference\n",
      "\t sum len concat: 68767 / 69398 99.1 %\n",
      "GAP_026795 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 37 / 43 contigs mapped to reference\n",
      "\t sum len concat: 29818 / 31672 94.1 %\n",
      "GAP_022615 Fabales Polygalaceae\n",
      "\t Ref (1, 24): NC_052912.1 Fabales Polygalaceae\n",
      "\t 20 / 21 contigs mapped to reference\n",
      "\t sum len concat: 109509 / 111256 98.4 %\n",
      "GAP_022617 Gentianales Rubiaceae\n",
      "\t Ref (2, 24): NC_054359.1 Gentianales Gentianaceae\n",
      "\t 21 / 23 contigs mapped to reference\n",
      "\t sum len concat: 43010 / 43899 98.0 %\n",
      "GAP_026797 Brassicales Brassicaceae\n",
      "\t Ref (1, 24): NC_049578.1 Brassicales Brassicaceae\n",
      "\t 51 / 51 contigs mapped to reference\n",
      "\t sum len concat: 93415 / 93415 100.0 %\n",
      "GAP_026801 Lamiales Lamiaceae\n",
      "\t Ref (1, 24): NC_053620.1 Lamiales Lamiaceae\n",
      "\t 9 / 9 contigs mapped to reference\n",
      "\t sum len concat: 74084 / 74084 100.0 %\n",
      "GAP_026803 Lamiales Lamiaceae\n",
      "\t Ref (1, 24): NC_053620.1 Lamiales Lamiaceae\n",
      "\t 21 / 21 contigs mapped to reference\n",
      "\t sum len concat: 54541 / 54541 100.0 %\n",
      "GAP_026807 Caryophyllales Amaranthaceae\n",
      "\t Ref (1, 24): NC_046575.1 Caryophyllales Amaranthaceae\n",
      "\t 19 / 19 contigs mapped to reference\n",
      "\t sum len concat: 76444 / 76444 100.0 %\n",
      "GAP_026809 Malpighiales Violaceae\n",
      "\t Ref (2, 24): NC_030601.1 Malpighiales Erythroxylaceae\n",
      "\t 13 / 13 contigs mapped to reference\n",
      "\t sum len concat: 41271 / 41271 100.0 %\n",
      "GAP_026811 Malpighiales Phyllanthaceae\n",
      "\t Ref (2, 24): NC_030601.1 Malpighiales Erythroxylaceae\n",
      "\t 44 / 47 contigs mapped to reference\n",
      "\t sum len concat: 90522 / 92229 98.1 %\n",
      "GAP_026815 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 42 / 45 contigs mapped to reference\n",
      "\t sum len concat: 34963 / 35936 97.3 %\n",
      "GAP_026819 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 19 / 19 contigs mapped to reference\n",
      "\t sum len concat: 58480 / 58480 100.0 %\n",
      "GAP_026821 Malpighiales Phyllanthaceae\n",
      "\t Ref (2, 24): NC_030601.1 Malpighiales Erythroxylaceae\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\t 2 / 2 contigs mapped to reference\n",
      "\t sum len concat: 129417 / 129417 100.0 %\n",
      "GAP_026823 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 13 / 13 contigs mapped to reference\n",
      "\t sum len concat: 132519 / 132519 100.0 %\n",
      "GAP_026829 Proteales Proteaceae\n",
      "\t Ref (1, 24): NC_036416.1 Proteales Proteaceae\n",
      "\t 10 / 11 contigs mapped to reference\n",
      "\t sum len concat: 51819 / 52529 98.6 %\n",
      "GAP_026833 Proteales Proteaceae\n",
      "\t Ref (1, 24): NC_036416.1 Proteales Proteaceae\n",
      "\t 8 / 9 contigs mapped to reference\n",
      "\t sum len concat: 26076 / 26969 96.7 %\n",
      "GAP_026835 Proteales Proteaceae\n",
      "\t Ref (1, 24): NC_036416.1 Proteales Proteaceae\n",
      "\t 5 / 7 contigs mapped to reference\n",
      "\t sum len concat: 32614 / 34171 95.4 %\n",
      "GAP_026845 Proteales Proteaceae\n",
      "\t Ref (1, 24): NC_036416.1 Proteales Proteaceae\n",
      "\t 24 / 26 contigs mapped to reference\n",
      "\t sum len concat: 45195 / 46627 96.9 %\n",
      "GAP_026847 Proteales Proteaceae\n",
      "\t Ref (1, 24): NC_036416.1 Proteales Proteaceae\n",
      "\t 14 / 16 contigs mapped to reference\n",
      "\t sum len concat: 130575 / 130768 99.9 %\n",
      "GAP_026855 Proteales Proteaceae\n",
      "\t Ref (1, 24): NC_036416.1 Proteales Proteaceae\n",
      "\t 12 / 13 contigs mapped to reference\n",
      "\t sum len concat: 28796 / 29756 96.8 %\n",
      "GAP_026857 Proteales Proteaceae\n",
      "\t Ref (1, 24): NC_036416.1 Proteales Proteaceae\n",
      "\t 37 / 38 contigs mapped to reference\n",
      "\t sum len concat: 56115 / 57106 98.3 %\n",
      "GAP_026861 Proteales Proteaceae\n",
      "\t Ref (1, 24): NC_036416.1 Proteales Proteaceae\n",
      "\t 24 / 24 contigs mapped to reference\n",
      "\t sum len concat: 127809 / 127809 100.0 %\n",
      "GAP_026863 Proteales Proteaceae\n",
      "\t Ref (1, 24): NC_036416.1 Proteales Proteaceae\n",
      "\t 24 / 26 contigs mapped to reference\n",
      "\t sum len concat: 118756 / 120468 98.6 %\n",
      "GAP_026865 Proteales Proteaceae\n",
      "\t Ref (1, 24): NC_036416.1 Proteales Proteaceae\n",
      "\t 14 / 14 contigs mapped to reference\n",
      "\t sum len concat: 125622 / 125622 100.0 %\n",
      "GAP_026869 Proteales Proteaceae\n",
      "\t Ref (1, 24): NC_036416.1 Proteales Proteaceae\n",
      "\t 22 / 22 contigs mapped to reference\n",
      "\t sum len concat: 119683 / 119683 100.0 %\n",
      "GAP_026871 Proteales Proteaceae\n",
      "\t Ref (1, 24): NC_036416.1 Proteales Proteaceae\n",
      "\t 9 / 9 contigs mapped to reference\n",
      "\t sum len concat: 42794 / 42794 100.0 %\n",
      "GAP_022703 Gentianales Apocynaceae\n",
      "\t Ref (2, 24): NC_054359.1 Gentianales Gentianaceae\n",
      "\t 33 / 34 contigs mapped to reference\n",
      "\t sum len concat: 41600 / 42038 99.0 %\n",
      "GAP_022705 Fabales Surianaceae\n",
      "\t Ref (2, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 10 / 11 contigs mapped to reference\n",
      "\t sum len concat: 28924 / 30074 96.2 %\n",
      "GAP_022709 Fabales Surianaceae\n",
      "\t Ref (2, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 8 / 10 contigs mapped to reference\n",
      "\t sum len concat: 32289 / 33899 95.3 %\n",
      "GAP_022711 Asterales Argophyllaceae\n",
      "\t Ref (2, 24): NC_044728.1 Asterales Asteraceae\n",
      "\t 15 / 15 contigs mapped to reference\n",
      "\t sum len concat: 48355 / 48355 100.0 %\n",
      "GAP_022717 Ranunculales Menispermaceae\n",
      "\t Ref (2, 24): NC_029427.1 Ranunculales Lardizabalaceae\n",
      "\t 7 / 7 contigs mapped to reference\n",
      "\t sum len concat: 34952 / 34952 100.0 %\n",
      "GAP_022721 Gentianales Apocynaceae\n",
      "\t Ref (2, 24): NC_054359.1 Gentianales Gentianaceae\n",
      "\t 24 / 24 contigs mapped to reference\n",
      "\t sum len concat: 122041 / 122041 100.0 %\n",
      "GAP_022725 Gentianales Apocynaceae\n",
      "\t Ref (2, 24): NC_054359.1 Gentianales Gentianaceae\n",
      "\t 44 / 45 contigs mapped to reference\n",
      "\t sum len concat: 93773 / 94470 99.3 %\n",
      "GAP_022727 Asterales Rousseaceae\n",
      "\t Ref (2, 24): NC_044728.1 Asterales Asteraceae\n",
      "\t 14 / 16 contigs mapped to reference\n",
      "\t sum len concat: 47898 / 48129 99.5 %\n",
      "GAP_022733 Caryophyllales Petiveriaceae\n",
      "\t Ref (2, 24): NC_046575.1 Caryophyllales Amaranthaceae\n",
      "\t 22 / 22 contigs mapped to reference\n",
      "\t sum len concat: 45685 / 45685 100.0 %\n",
      "GAP_022735 Gentianales Apocynaceae\n",
      "\t Ref (2, 24): NC_054359.1 Gentianales Gentianaceae\n",
      "\t 29 / 32 contigs mapped to reference\n",
      "\t sum len concat: 41206 / 42338 97.3 %\n",
      "GAP_022737 Gentianales Rubiaceae\n",
      "\t Ref (2, 24): NC_054359.1 Gentianales Gentianaceae\n",
      "\t 25 / 26 contigs mapped to reference\n",
      "\t sum len concat: 70356 / 70772 99.4 %\n",
      "GAP_022739 Gentianales Rubiaceae\n",
      "\t Ref (2, 24): NC_054359.1 Gentianales Gentianaceae\n",
      "\t 23 / 24 contigs mapped to reference\n",
      "\t sum len concat: 50978 / 51426 99.1 %\n",
      "GAP_022741 Gentianales Apocynaceae\n",
      "\t Ref (2, 24): NC_054359.1 Gentianales Gentianaceae\n",
      "\t 23 / 25 contigs mapped to reference\n",
      "\t sum len concat: 53837 / 54814 98.2 %\n",
      "GAP_022743 Poales Poaceae\n",
      "\t Ref (1, 24): NC_035050.1 Poales Poaceae\n",
      "\t 19 / 19 contigs mapped to reference\n",
      "\t sum len concat: 30658 / 30658 100.0 %\n",
      "GAP_026879 Caryophyllales Amaranthaceae\n",
      "\t Ref (1, 24): NC_046575.1 Caryophyllales Amaranthaceae\n",
      "\t 33 / 34 contigs mapped to reference\n",
      "\t sum len concat: 29477 / 29774 99.0 %\n",
      "GAP_026881 Caryophyllales Amaranthaceae\n",
      "\t Ref (1, 24): NC_046575.1 Caryophyllales Amaranthaceae\n",
      "\t 20 / 20 contigs mapped to reference\n",
      "\t sum len concat: 46176 / 46176 100.0 %\n",
      "GAP_026883 Caryophyllales Amaranthaceae\n",
      "\t Ref (1, 24): NC_046575.1 Caryophyllales Amaranthaceae\n",
      "\t 37 / 37 contigs mapped to reference\n",
      "\t sum len concat: 32456 / 32456 100.0 %\n",
      "GAP_026885 Caryophyllales Amaranthaceae\n",
      "\t Ref (1, 24): NC_046575.1 Caryophyllales Amaranthaceae\n",
      "\t 37 / 37 contigs mapped to reference\n",
      "\t sum len concat: 39868 / 39868 100.0 %\n",
      "GAP_026889 Caryophyllales Amaranthaceae\n",
      "\t Ref (1, 24): NC_046575.1 Caryophyllales Amaranthaceae\n",
      "\t 35 / 35 contigs mapped to reference\n",
      "\t sum len concat: 43919 / 43919 100.0 %\n",
      "GAP_026891 Caryophyllales Amaranthaceae\n",
      "\t Ref (1, 24): NC_046575.1 Caryophyllales Amaranthaceae\n",
      "\t 35 / 36 contigs mapped to reference\n",
      "\t sum len concat: 32405 / 32665 99.2 %\n",
      "GAP_026895 Poales Poaceae\n",
      "\t Ref (1, 24): NC_035050.1 Poales Poaceae\n",
      "\t 37 / 37 contigs mapped to reference\n",
      "\t sum len concat: 50035 / 50035 100.0 %\n",
      "GAP_026897 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 9 / 10 contigs mapped to reference\n",
      "\t sum len concat: 121159 / 121829 99.5 %\n",
      "GAP_026899 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 48 / 56 contigs mapped to reference\n",
      "\t sum len concat: 32600 / 36633 89.0 %\n",
      "GAP_026901 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 44 / 48 contigs mapped to reference\n",
      "\t sum len concat: 66905 / 68883 97.1 %\n",
      "GAP_026907 Malpighiales Picrodendraceae\n",
      "\t Ref (2, 24): NC_030601.1 Malpighiales Erythroxylaceae\n",
      "\t 10 / 13 contigs mapped to reference\n",
      "\t sum len concat: 24567 / 26921 91.3 %\n",
      "GAP_026913 Apiales Pittosporaceae\n",
      "\t Ref (2, 24): NC_054296.1 Apiales Apiaceae\n",
      "\t 16 / 17 contigs mapped to reference\n",
      "\t sum len concat: 78507 / 79637 98.6 %\n",
      "GAP_026923 Caryophyllales Amaranthaceae\n",
      "\t Ref (1, 24): NC_046575.1 Caryophyllales Amaranthaceae\n",
      "\t 35 / 35 contigs mapped to reference\n",
      "\t sum len concat: 47927 / 47927 100.0 %\n",
      "GAP_026925 Saxifragales Haloragaceae\n",
      "\t Ref (1, 24): NC_037885.1 Saxifragales Haloragaceae\n",
      "\t 5 / 5 contigs mapped to reference\n",
      "\t sum len concat: 43081 / 43081 100.0 %\n",
      "GAP_026929 Asterales Menyanthaceae\n",
      "\t Ref (1, 24): NC_041481.1 Asterales Menyanthaceae\n",
      "\t 17 / 17 contigs mapped to reference\n",
      "\t sum len concat: 61882 / 61882 100.0 %\n",
      "GAP_026931 Fabales Polygalaceae\n",
      "\t Ref (1, 24): NC_052912.1 Fabales Polygalaceae\n",
      "\t 6 / 6 contigs mapped to reference\n",
      "\t sum len concat: 34127 / 34127 100.0 %\n",
      "GAP_026933 Alismatales Potamogetonaceae\n",
      "\t Ref (2, 24): NC_053912.1 Alismatales Aponogetonaceae\n",
      "\t 10 / 10 contigs mapped to reference\n",
      "\t sum len concat: 31766 / 31766 100.0 %\n",
      "GAP_022817 Apiales Araliaceae\n",
      "\t Ref (1, 24): NC_022812.1 Apiales Araliaceae\n",
      "\t 8 / 8 contigs mapped to reference\n",
      "\t sum len concat: 35875 / 35875 100.0 %\n",
      "GAP_022819 Gentianales Rubiaceae\n",
      "\t Ref (2, 24): NC_054359.1 Gentianales Gentianaceae\n",
      "\t 18 / 18 contigs mapped to reference\n",
      "\t sum len concat: 42316 / 42316 100.0 %\n",
      "GAP_022821 Ericales Sapotaceae\n",
      "\t Ref (2, 24): NC_041127.1 Ericales Styracaceae\n",
      "\t 37 / 37 contigs mapped to reference\n",
      "\t sum len concat: 108888 / 108888 100.0 %\n",
      "GAP_026939 Lamiales Lamiaceae\n",
      "\t Ref (1, 24): NC_053620.1 Lamiales Lamiaceae\n",
      "\t 12 / 12 contigs mapped to reference\n",
      "\t sum len concat: 103007 / 103007 100.0 %\n",
      "GAP_026945 Caryophyllales Amaranthaceae\n",
      "\t Ref (1, 24): NC_046575.1 Caryophyllales Amaranthaceae\n",
      "\t 52 / 52 contigs mapped to reference\n",
      "\t sum len concat: 83391 / 83391 100.0 %\n",
      "GAP_026951 Apiales Pittosporaceae\n",
      "\t Ref (2, 24): NC_054296.1 Apiales Apiaceae\n",
      "\t 11 / 13 contigs mapped to reference\n",
      "\t sum len concat: 42189 / 44745 94.3 %\n",
      "GAP_027737 Asparagales Orchidaceae\n",
      "\t Ref (1, 24): NC_045400.1 Asparagales Orchidaceae\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\t 25 / 25 contigs mapped to reference\n",
      "\t sum len concat: 39561 / 39561 100.0 %\n",
      "GAP_027739 Asparagales Orchidaceae\n",
      "\t Ref (1, 24): NC_045400.1 Asparagales Orchidaceae\n",
      "\t 39 / 39 contigs mapped to reference\n",
      "\t sum len concat: 95933 / 95933 100.0 %\n",
      "GAP_027761 Asparagales Orchidaceae\n",
      "\t Ref (1, 24): NC_045400.1 Asparagales Orchidaceae\n",
      "\t 15 / 15 contigs mapped to reference\n",
      "\t sum len concat: 34951 / 34951 100.0 %\n",
      "GAP_027767 Asparagales Orchidaceae\n",
      "\t Ref (1, 24): NC_045400.1 Asparagales Orchidaceae\n",
      "\t 35 / 35 contigs mapped to reference\n",
      "\t sum len concat: 92520 / 92520 100.0 %\n",
      "GAP_027777 Asparagales Orchidaceae\n",
      "\t Ref (1, 24): NC_045400.1 Asparagales Orchidaceae\n",
      "\t 39 / 39 contigs mapped to reference\n",
      "\t sum len concat: 37595 / 37595 100.0 %\n",
      "GAP_027781 Asparagales Orchidaceae\n",
      "\t Ref (1, 24): NC_045400.1 Asparagales Orchidaceae\n",
      "\t 27 / 27 contigs mapped to reference\n",
      "\t sum len concat: 31308 / 31308 100.0 %\n",
      "GAP_027783 Asparagales Orchidaceae\n",
      "\t Ref (1, 24): NC_045400.1 Asparagales Orchidaceae\n",
      "\t 11 / 12 contigs mapped to reference\n",
      "\t sum len concat: 23306 / 23939 97.4 %\n",
      "GAP_027789 Asparagales Orchidaceae\n",
      "\t Ref (1, 24): NC_045400.1 Asparagales Orchidaceae\n",
      "\t 19 / 19 contigs mapped to reference\n",
      "\t sum len concat: 25413 / 25413 100.0 %\n",
      "GAP_027793 Asparagales Orchidaceae\n",
      "\t Ref (1, 24): NC_045400.1 Asparagales Orchidaceae\n",
      "\t 16 / 16 contigs mapped to reference\n",
      "\t sum len concat: 32279 / 32279 100.0 %\n",
      "GAP_027797 Asparagales Orchidaceae\n",
      "\t Ref (1, 24): NC_045400.1 Asparagales Orchidaceae\n",
      "\t 17 / 17 contigs mapped to reference\n",
      "\t sum len concat: 27045 / 27045 100.0 %\n",
      "GAP_027803 Asparagales Orchidaceae\n",
      "\t Ref (1, 24): NC_045400.1 Asparagales Orchidaceae\n",
      "\t 35 / 35 contigs mapped to reference\n",
      "\t sum len concat: 41215 / 41215 100.0 %\n",
      "GAP_027805 Asparagales Orchidaceae\n",
      "\t Ref (1, 24): NC_045400.1 Asparagales Orchidaceae\n",
      "\t 27 / 29 contigs mapped to reference\n",
      "\t sum len concat: 24086 / 24886 96.8 %\n",
      "GAP_027823 Asparagales Orchidaceae\n",
      "\t Ref (1, 24): NC_045400.1 Asparagales Orchidaceae\n",
      "\t 10 / 10 contigs mapped to reference\n",
      "\t sum len concat: 50898 / 50898 100.0 %\n",
      "GAP_027825 Asparagales Orchidaceae\n",
      "\t Ref (1, 24): NC_045400.1 Asparagales Orchidaceae\n",
      "\t 26 / 26 contigs mapped to reference\n",
      "\t sum len concat: 36532 / 36532 100.0 %\n",
      "GAP_027827 Asparagales Orchidaceae\n",
      "\t Ref (1, 24): NC_045400.1 Asparagales Orchidaceae\n",
      "\t 29 / 29 contigs mapped to reference\n",
      "\t sum len concat: 38605 / 38605 100.0 %\n",
      "GAP_027933 Asterales Asteraceae\n",
      "\t Ref (1, 24): NC_044728.1 Asterales Asteraceae\n",
      "\t 14 / 14 contigs mapped to reference\n",
      "\t sum len concat: 44521 / 44521 100.0 %\n",
      "GAP_027941 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 30 / 30 contigs mapped to reference\n",
      "\t sum len concat: 42152 / 42152 100.0 %\n",
      "GAP_027943 Asterales Asteraceae\n",
      "\t Ref (1, 24): NC_044728.1 Asterales Asteraceae\n",
      "\t 13 / 13 contigs mapped to reference\n",
      "\t sum len concat: 28485 / 28485 100.0 %\n",
      "GAP_027955 Asterales Asteraceae\n",
      "\t Ref (1, 24): NC_044728.1 Asterales Asteraceae\n",
      "\t 33 / 33 contigs mapped to reference\n",
      "\t sum len concat: 62407 / 62407 100.0 %\n",
      "GAP_027963 Asterales Asteraceae\n",
      "\t Ref (1, 24): NC_044728.1 Asterales Asteraceae\n",
      "\t 10 / 10 contigs mapped to reference\n",
      "\t sum len concat: 27805 / 27805 100.0 %\n",
      "GAP_027975 Asterales Asteraceae\n",
      "\t Ref (1, 24): NC_044728.1 Asterales Asteraceae\n",
      "\t 19 / 19 contigs mapped to reference\n",
      "\t sum len concat: 33883 / 33883 100.0 %\n",
      "GAP_028017 Asterales Asteraceae\n",
      "\t Ref (1, 24): NC_044728.1 Asterales Asteraceae\n",
      "\t 23 / 23 contigs mapped to reference\n",
      "\t sum len concat: 41547 / 41547 100.0 %\n",
      "GAP_028025 Asterales Asteraceae\n",
      "\t Ref (1, 24): NC_044728.1 Asterales Asteraceae\n",
      "\t 41 / 41 contigs mapped to reference\n",
      "\t sum len concat: 70396 / 70396 100.0 %\n",
      "GAP_028033 Caryophyllales Amaranthaceae\n",
      "\t Ref (1, 24): NC_046575.1 Caryophyllales Amaranthaceae\n",
      "\t 50 / 51 contigs mapped to reference\n",
      "\t sum len concat: 83932 / 84661 99.1 %\n",
      "GAP_028037 Apiales Apiaceae\n",
      "\t Ref (1, 24): NC_054296.1 Apiales Apiaceae\n",
      "\t 40 / 40 contigs mapped to reference\n",
      "\t sum len concat: 39929 / 39929 100.0 %\n",
      "GAP_028039 Apiales Araliaceae\n",
      "\t Ref (1, 24): NC_022812.1 Apiales Araliaceae\n",
      "\t 15 / 15 contigs mapped to reference\n",
      "\t sum len concat: 123441 / 123441 100.0 %\n",
      "GAP_028041 Apiales Araliaceae\n",
      "\t Ref (1, 24): NC_022812.1 Apiales Araliaceae\n",
      "\t 8 / 8 contigs mapped to reference\n",
      "\t sum len concat: 32857 / 32857 100.0 %\n",
      "GAP_028045 Brassicales Brassicaceae\n",
      "\t Ref (1, 24): NC_049578.1 Brassicales Brassicaceae\n",
      "\t 15 / 15 contigs mapped to reference\n",
      "\t sum len concat: 47921 / 47921 100.0 %\n",
      "GAP_028053 Solanales Convolvulaceae\n",
      "\t Ref (1, 24): NC_042940.1 Solanales Convolvulaceae\n",
      "\t 12 / 12 contigs mapped to reference\n",
      "\t sum len concat: 28251 / 28251 100.0 %\n",
      "GAP_028055 Solanales Convolvulaceae\n",
      "\t Ref (1, 24): NC_042940.1 Solanales Convolvulaceae\n",
      "\t 14 / 14 contigs mapped to reference\n",
      "\t sum len concat: 55734 / 55734 100.0 %\n",
      "GAP_028057 Solanales Convolvulaceae\n",
      "\t Ref (1, 24): NC_042940.1 Solanales Convolvulaceae\n",
      "\t 16 / 16 contigs mapped to reference\n",
      "\t sum len concat: 33306 / 33306 100.0 %\n",
      "GAP_028059 Solanales Convolvulaceae\n",
      "\t Ref (1, 24): NC_042940.1 Solanales Convolvulaceae\n",
      "\t 9 / 9 contigs mapped to reference\n",
      "\t sum len concat: 129376 / 129376 100.0 %\n",
      "GAP_028075 Poales Poaceae\n",
      "\t Ref (1, 24): NC_035050.1 Poales Poaceae\n",
      "\t 32 / 32 contigs mapped to reference\n",
      "\t sum len concat: 41834 / 41834 100.0 %\n",
      "GAP_028085 Poales Poaceae\n",
      "\t Ref (1, 24): NC_035050.1 Poales Poaceae\n",
      "\t 20 / 20 contigs mapped to reference\n",
      "\t sum len concat: 61987 / 61987 100.0 %\n",
      "GAP_028093 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 19 / 21 contigs mapped to reference\n",
      "\t sum len concat: 37573 / 38788 96.9 %\n",
      "GAP_028097 Poales Poaceae\n",
      "\t Ref (1, 24): NC_035050.1 Poales Poaceae\n",
      "\t 36 / 36 contigs mapped to reference\n",
      "\t sum len concat: 41321 / 41321 100.0 %\n",
      "GAP_028099 Poales Poaceae\n",
      "\t Ref (1, 24): NC_035050.1 Poales Poaceae\n",
      "\t 29 / 29 contigs mapped to reference\n",
      "\t sum len concat: 39073 / 39073 100.0 %\n",
      "GAP_028113 Poales Poaceae\n",
      "\t Ref (1, 24): NC_035050.1 Poales Poaceae\n",
      "\t 35 / 35 contigs mapped to reference\n",
      "\t sum len concat: 35161 / 35161 100.0 %\n",
      "GAP_028119 Sapindales Rutaceae\n",
      "\t Ref (1, 24): NC_024929.1 Sapindales Rutaceae\n",
      "\t 10 / 10 contigs mapped to reference\n",
      "\t sum len concat: 42102 / 42102 100.0 %\n",
      "GAP_028121 Sapindales Rutaceae\n",
      "\t Ref (1, 24): NC_024929.1 Sapindales Rutaceae\n",
      "\t 23 / 23 contigs mapped to reference\n",
      "\t sum len concat: 74511 / 74511 100.0 %\n",
      "GAP_028127 Sapindales Rutaceae\n",
      "\t Ref (1, 24): NC_024929.1 Sapindales Rutaceae\n",
      "\t 39 / 39 contigs mapped to reference\n",
      "\t sum len concat: 59532 / 59532 100.0 %\n",
      "GAP_028129 Sapindales Rutaceae\n",
      "\t Ref (1, 24): NC_024929.1 Sapindales Rutaceae\n",
      "\t 24 / 24 contigs mapped to reference\n",
      "\t sum len concat: 118563 / 118563 100.0 %\n",
      "GAP_028141 Malvales Malvaceae\n",
      "\t Ref (1, 24): NC_038057.1 Malvales Malvaceae\n",
      "\t 16 / 16 contigs mapped to reference\n",
      "\t sum len concat: 28311 / 28311 100.0 %\n",
      "GAP_028143 Malvales Malvaceae\n",
      "\t Ref (1, 24): NC_038057.1 Malvales Malvaceae\n",
      "\t 16 / 16 contigs mapped to reference\n",
      "\t sum len concat: 127222 / 127222 100.0 %\n",
      "GAP_028147 Malvales Malvaceae\n",
      "\t Ref (1, 24): NC_038057.1 Malvales Malvaceae\n",
      "\t 22 / 22 contigs mapped to reference\n",
      "\t sum len concat: 124221 / 124221 100.0 %\n",
      "GAP_028151 Lamiales Lamiaceae\n",
      "\t Ref (1, 24): NC_053620.1 Lamiales Lamiaceae\n",
      "\t 37 / 37 contigs mapped to reference\n",
      "\t sum len concat: 75296 / 75296 100.0 %\n",
      "GAP_028153 Malvales Malvaceae\n",
      "\t Ref (1, 24): NC_038057.1 Malvales Malvaceae\n",
      "\t 23 / 23 contigs mapped to reference\n",
      "\t sum len concat: 124205 / 124205 100.0 %\n",
      "GAP_028157 Malvales Malvaceae\n",
      "\t Ref (1, 24): NC_038057.1 Malvales Malvaceae\n",
      "\t 5 / 5 contigs mapped to reference\n",
      "\t sum len concat: 133698 / 133698 100.0 %\n",
      "GAP_028159 Malvales Malvaceae\n",
      "\t Ref (1, 24): NC_038057.1 Malvales Malvaceae\n",
      "\t 10 / 10 contigs mapped to reference\n",
      "\t sum len concat: 129629 / 129629 100.0 %\n",
      "GAP_028161 Malvales Malvaceae\n",
      "\t Ref (1, 24): NC_038057.1 Malvales Malvaceae\n",
      "\t 25 / 25 contigs mapped to reference\n",
      "\t sum len concat: 99902 / 99902 100.0 %\n",
      "GAP_028165 Malvales Malvaceae\n",
      "\t Ref (1, 24): NC_038057.1 Malvales Malvaceae\n",
      "\t 39 / 39 contigs mapped to reference\n",
      "\t sum len concat: 102598 / 102598 100.0 %\n",
      "GAP_028167 Malvales Malvaceae\n",
      "\t Ref (1, 24): NC_038057.1 Malvales Malvaceae\n",
      "\t 21 / 21 contigs mapped to reference\n",
      "\t sum len concat: 126157 / 126157 100.0 %\n",
      "GAP_028169 Malvales Malvaceae\n",
      "\t Ref (1, 24): NC_038057.1 Malvales Malvaceae\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\t 3 / 3 contigs mapped to reference\n",
      "\t sum len concat: 155732 / 155732 100.0 %\n",
      "GAP_028175 Malvales Malvaceae\n",
      "\t Ref (1, 24): NC_038057.1 Malvales Malvaceae\n",
      "\t 17 / 17 contigs mapped to reference\n",
      "\t sum len concat: 129866 / 129866 100.0 %\n",
      "GAP_028177 Malvales Malvaceae\n",
      "\t Ref (1, 24): NC_038057.1 Malvales Malvaceae\n",
      "\t 11 / 11 contigs mapped to reference\n",
      "\t sum len concat: 33803 / 33803 100.0 %\n",
      "GAP_028179 Malvales Malvaceae\n",
      "\t Ref (1, 24): NC_038057.1 Malvales Malvaceae\n",
      "\t 55 / 55 contigs mapped to reference\n",
      "\t sum len concat: 49566 / 49566 100.0 %\n",
      "GAP_028181 Malvales Malvaceae\n",
      "\t Ref (1, 24): NC_038057.1 Malvales Malvaceae\n",
      "\t 51 / 51 contigs mapped to reference\n",
      "\t sum len concat: 84695 / 84695 100.0 %\n",
      "GAP_028187 Santalales Loranthaceae\n",
      "\t Ref (1, 24): NC_040862.1 Santalales Loranthaceae\n",
      "\t 29 / 29 contigs mapped to reference\n",
      "\t sum len concat: 84106 / 84106 100.0 %\n",
      "GAP_028189 Apiales Araliaceae\n",
      "\t Ref (1, 24): NC_022812.1 Apiales Araliaceae\n",
      "\t 5 / 5 contigs mapped to reference\n",
      "\t sum len concat: 132367 / 132367 100.0 %\n",
      "GAP_028199 Poales Restionaceae\n",
      "\t Ref (2, 24): NC_042211.1 Poales Eriocaulaceae\n",
      "\t 17 / 25 contigs mapped to reference\n",
      "\t sum len concat: 19374 / 23358 82.9 %\n",
      "GAP_028203 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 29 / 30 contigs mapped to reference\n",
      "\t sum len concat: 121820 / 122584 99.4 %\n",
      "GAP_028205 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 38 / 39 contigs mapped to reference\n",
      "\t sum len concat: 104984 / 105259 99.7 %\n",
      "GAP_028209 Sapindales Rutaceae\n",
      "\t Ref (1, 24): NC_024929.1 Sapindales Rutaceae\n",
      "\t 6 / 6 contigs mapped to reference\n",
      "\t sum len concat: 129496 / 129496 100.0 %\n",
      "GAP_028211 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 14 / 14 contigs mapped to reference\n",
      "\t sum len concat: 132210 / 132210 100.0 %\n",
      "GAP_028213 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 24 / 26 contigs mapped to reference\n",
      "\t sum len concat: 119870 / 121348 98.8 %\n",
      "GAP_028215 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 25 / 30 contigs mapped to reference\n",
      "\t sum len concat: 51266 / 53630 95.6 %\n",
      "GAP_028217 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 35 / 36 contigs mapped to reference\n",
      "\t sum len concat: 97164 / 97918 99.2 %\n",
      "GAP_028219 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 44 / 47 contigs mapped to reference\n",
      "\t sum len concat: 53398 / 54869 97.3 %\n",
      "GAP_028221 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 21 / 22 contigs mapped to reference\n",
      "\t sum len concat: 97945 / 98683 99.3 %\n",
      "GAP_028231 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 28 / 28 contigs mapped to reference\n",
      "\t sum len concat: 37724 / 37724 100.0 %\n",
      "GAP_028233 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 3 / 3 contigs mapped to reference\n",
      "\t sum len concat: 150208 / 150208 100.0 %\n",
      "GAP_028235 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 5 / 5 contigs mapped to reference\n",
      "\t sum len concat: 144487 / 144487 100.0 %\n",
      "GAP_028237 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 26 / 30 contigs mapped to reference\n",
      "\t sum len concat: 22557 / 23770 94.9 %\n",
      "GAP_028239 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 17 / 18 contigs mapped to reference\n",
      "\t sum len concat: 43973 / 44422 99.0 %\n",
      "GAP_028241 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 42 / 45 contigs mapped to reference\n",
      "\t sum len concat: 67156 / 68791 97.6 %\n",
      "GAP_028245 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 43 / 47 contigs mapped to reference\n",
      "\t sum len concat: 59030 / 61645 95.8 %\n",
      "GAP_028247 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 49 / 50 contigs mapped to reference\n",
      "\t sum len concat: 62684 / 63034 99.4 %\n",
      "GAP_028253 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 46 / 47 contigs mapped to reference\n",
      "\t sum len concat: 83646 / 84090 99.5 %\n",
      "GAP_028255 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 9 / 9 contigs mapped to reference\n",
      "\t sum len concat: 138802 / 138802 100.0 %\n",
      "GAP_028257 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 20 / 22 contigs mapped to reference\n",
      "\t sum len concat: 119940 / 121361 98.8 %\n",
      "GAP_028259 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 28 / 28 contigs mapped to reference\n",
      "\t sum len concat: 111317 / 111317 100.0 %\n",
      "GAP_028261 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 19 / 24 contigs mapped to reference\n",
      "\t sum len concat: 24136 / 27720 87.1 %\n",
      "GAP_028263 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 12 / 12 contigs mapped to reference\n",
      "\t sum len concat: 123588 / 123588 100.0 %\n",
      "GAP_028265 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 31 / 33 contigs mapped to reference\n",
      "\t sum len concat: 70081 / 70755 99.0 %\n",
      "GAP_028267 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 35 / 38 contigs mapped to reference\n",
      "\t sum len concat: 43702 / 45174 96.7 %\n",
      "GAP_028275 Malvales Malvaceae\n",
      "\t Ref (1, 24): NC_038057.1 Malvales Malvaceae\n",
      "\t 59 / 59 contigs mapped to reference\n",
      "\t sum len concat: 74643 / 74643 100.0 %\n",
      "GAP_028279 Brassicales Brassicaceae\n",
      "\t Ref (1, 24): NC_049578.1 Brassicales Brassicaceae\n",
      "\t 58 / 58 contigs mapped to reference\n",
      "\t sum len concat: 65251 / 65251 100.0 %\n",
      "GAP_028287 Asparagales Asparagaceae\n",
      "\t Ref (1, 24): NC_053784.1 Asparagales Asparagaceae\n",
      "\t 16 / 16 contigs mapped to reference\n",
      "\t sum len concat: 50788 / 50788 100.0 %\n",
      "GAP_028299 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 19 / 22 contigs mapped to reference\n",
      "\t sum len concat: 42817 / 44288 96.7 %\n",
      "GAP_028303 Asparagales Asparagaceae\n",
      "\t Ref (1, 24): NC_053784.1 Asparagales Asparagaceae\n",
      "\t 29 / 30 contigs mapped to reference\n",
      "\t sum len concat: 31739 / 32135 98.8 %\n",
      "GAP_028309 Rosales Rhamnaceae\n",
      "\t Ref (1, 24): NC_037151.1 Rosales Rhamnaceae\n",
      "\t 30 / 30 contigs mapped to reference\n",
      "\t sum len concat: 68209 / 68209 100.0 %\n",
      "GAP_028315 Poales Poaceae\n",
      "\t Ref (1, 24): NC_035050.1 Poales Poaceae\n",
      "\t 36 / 36 contigs mapped to reference\n",
      "\t sum len concat: 36267 / 36267 100.0 %\n",
      "GAP_028317 Poales Poaceae\n",
      "\t Ref (1, 24): NC_035050.1 Poales Poaceae\n",
      "\t 24 / 24 contigs mapped to reference\n",
      "\t sum len concat: 26790 / 26790 100.0 %\n",
      "GAP_028319 Rosales Rhamnaceae\n",
      "\t Ref (1, 24): NC_037151.1 Rosales Rhamnaceae\n",
      "\t 19 / 19 contigs mapped to reference\n",
      "\t sum len concat: 34931 / 34931 100.0 %\n",
      "GAP_028323 Brassicales Brassicaceae\n",
      "\t Ref (1, 24): NC_049578.1 Brassicales Brassicaceae\n",
      "\t 18 / 18 contigs mapped to reference\n",
      "\t sum len concat: 121276 / 121276 100.0 %\n",
      "GAP_028331 Liliales Colchicaceae\n",
      "\t Ref (2, 24): NC_024728.1 Liliales Liliaceae\n",
      "\t 23 / 24 contigs mapped to reference\n",
      "\t sum len concat: 57450 / 57852 99.3 %\n",
      "GAP_028523 Asparagales Hypoxidaceae\n",
      "\t Ref (2, 24): NC_053784.1 Asparagales Asparagaceae\n",
      "\t 35 / 35 contigs mapped to reference\n",
      "\t sum len concat: 33691 / 33691 100.0 %\n",
      "GAP_028527 Poales Poaceae\n",
      "\t Ref (1, 24): NC_035050.1 Poales Poaceae\n",
      "\t 45 / 45 contigs mapped to reference\n",
      "\t sum len concat: 61660 / 61660 100.0 %\n",
      "GAP_028545 Santalales Opiliaceae\n",
      "\t Ref (2, 24): NC_040862.1 Santalales Loranthaceae\n",
      "\t 27 / 34 contigs mapped to reference\n",
      "\t sum len concat: 34377 / 38062 90.3 %\n",
      "GAP_028547 Vitales Vitaceae\n",
      "\t Ref (1, 24): NC_039803.1 Vitales Vitaceae\n",
      "\t 32 / 32 contigs mapped to reference\n",
      "\t sum len concat: 38576 / 38576 100.0 %\n",
      "GAP_028661 Poales Poaceae\n",
      "\t Ref (1, 24): NC_035050.1 Poales Poaceae\n",
      "\t 30 / 30 contigs mapped to reference\n",
      "\t sum len concat: 97175 / 97175 100.0 %\n",
      "GAP_028671 Poales Poaceae\n",
      "\t Ref (1, 24): NC_035050.1 Poales Poaceae\n",
      "\t 35 / 35 contigs mapped to reference\n",
      "\t sum len concat: 103093 / 103093 100.0 %\n",
      "GAP_028675 Poales Poaceae\n",
      "\t Ref (1, 24): NC_035050.1 Poales Poaceae\n",
      "\t 5 / 5 contigs mapped to reference\n",
      "\t sum len concat: 394410 / 394410 100.0 %\n",
      "GAP_028677 Poales Poaceae\n",
      "\t Ref (1, 24): NC_035050.1 Poales Poaceae\n",
      "\t 8 / 8 contigs mapped to reference\n",
      "\t sum len concat: 36660 / 36660 100.0 %\n",
      "GAP_028683 Poales Poaceae\n",
      "\t Ref (1, 24): NC_035050.1 Poales Poaceae\n",
      "\t 16 / 16 contigs mapped to reference\n",
      "\t sum len concat: 149867 / 149867 100.0 %\n",
      "GAP_028685 Poales Poaceae\n",
      "\t Ref (1, 24): NC_035050.1 Poales Poaceae\n",
      "\t 52 / 52 contigs mapped to reference\n",
      "\t sum len concat: 54548 / 54548 100.0 %\n",
      "GAP_028689 Poales Poaceae\n",
      "\t Ref (1, 24): NC_035050.1 Poales Poaceae\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\t 9 / 9 contigs mapped to reference\n",
      "\t sum len concat: 114731 / 114731 100.0 %\n",
      "GAP_028691 Poales Poaceae\n",
      "\t Ref (1, 24): NC_035050.1 Poales Poaceae\n",
      "\t 42 / 42 contigs mapped to reference\n",
      "\t sum len concat: 33418 / 33418 100.0 %\n",
      "GAP_028695 Poales Poaceae\n",
      "\t Ref (1, 24): NC_035050.1 Poales Poaceae\n",
      "\t 13 / 13 contigs mapped to reference\n",
      "\t sum len concat: 79108 / 79108 100.0 %\n",
      "GAP_028697 Poales Poaceae\n",
      "\t Ref (1, 24): NC_035050.1 Poales Poaceae\n",
      "\t 20 / 20 contigs mapped to reference\n",
      "\t sum len concat: 26494 / 26494 100.0 %\n",
      "GAP_028599 Poales Poaceae\n",
      "\t Ref (1, 24): NC_035050.1 Poales Poaceae\n",
      "\t 8 / 8 contigs mapped to reference\n",
      "\t sum len concat: 29869 / 29869 100.0 %\n",
      "GAP_028605 Poales Poaceae\n",
      "\t Ref (1, 24): NC_035050.1 Poales Poaceae\n",
      "\t 8 / 8 contigs mapped to reference\n",
      "\t sum len concat: 33139 / 33139 100.0 %\n",
      "GAP_028611 Asparagales Boryaceae\n",
      "\t Ref (2, 24): NC_053784.1 Asparagales Asparagaceae\n",
      "\t 29 / 29 contigs mapped to reference\n",
      "\t sum len concat: 116549 / 116549 100.0 %\n",
      "GAP_029179 Asparagales Orchidaceae\n",
      "\t Ref (1, 24): NC_045400.1 Asparagales Orchidaceae\n",
      "\t 15 / 15 contigs mapped to reference\n",
      "\t sum len concat: 27564 / 27564 100.0 %\n",
      "GAP_029183 Asparagales Orchidaceae\n",
      "\t Ref (1, 24): NC_045400.1 Asparagales Orchidaceae\n",
      "\t 34 / 34 contigs mapped to reference\n",
      "\t sum len concat: 40030 / 40030 100.0 %\n",
      "GAP_029185 Asparagales Orchidaceae\n",
      "\t Ref (1, 24): NC_045400.1 Asparagales Orchidaceae\n",
      "\t 16 / 16 contigs mapped to reference\n",
      "\t sum len concat: 26318 / 26318 100.0 %\n",
      "GAP_029191 Asparagales Orchidaceae\n",
      "\t Ref (1, 24): NC_045400.1 Asparagales Orchidaceae\n",
      "\t 20 / 20 contigs mapped to reference\n",
      "\t sum len concat: 27372 / 27372 100.0 %\n",
      "GAP_029197 Asparagales Orchidaceae\n",
      "\t Ref (1, 24): NC_045400.1 Asparagales Orchidaceae\n",
      "\t 16 / 16 contigs mapped to reference\n",
      "\t sum len concat: 34063 / 34063 100.0 %\n",
      "GAP_029205 Asparagales Orchidaceae\n",
      "\t Ref (1, 24): NC_045400.1 Asparagales Orchidaceae\n",
      "\t 37 / 37 contigs mapped to reference\n",
      "\t sum len concat: 79199 / 79199 100.0 %\n",
      "GAP_029215 Asparagales Orchidaceae\n",
      "\t Ref (1, 24): NC_045400.1 Asparagales Orchidaceae\n",
      "\t 25 / 25 contigs mapped to reference\n",
      "\t sum len concat: 34918 / 34918 100.0 %\n",
      "GAP_029227 Asparagales Orchidaceae\n",
      "\t Ref (1, 24): NC_045400.1 Asparagales Orchidaceae\n",
      "\t 29 / 29 contigs mapped to reference\n",
      "\t sum len concat: 39793 / 39793 100.0 %\n"
     ]
    }
   ],
   "source": [
    "if not os.path.exists(sdir + '/out_blast/'):\n",
    "    os.makedirs(sdir + '/out_blast/')\n",
    "commands = []\n",
    "for Sample_Name, row in incomplete_pt.iterrows():\n",
    "    print(Sample_Name,row.order,row.family)\n",
    "    ### Identify best reference\n",
    "    # check if reference in same family\n",
    "    best_ref = references[references.family==row.family]\n",
    "    if best_ref.shape[0]<1:\n",
    "        # otherwise order\n",
    "        best_ref = references[references.order==row.order]\n",
    "        if best_ref.shape[0]<1:\n",
    "            #Otherwise use closest order\n",
    "            ### Paul B. 'Vahliales' doesn't exist in key row.order so adding if conditional\n",
    "            if order_dc[row.order]:\n",
    "                best_ref = references[references.order==order_dc[row.order]]\n",
    "            else: \n",
    "                print('WARNING: order does not exist for Sample_Name')\n",
    "    print('\\t Ref',best_ref.shape, end = ': ')\n",
    "    best_ref = best_ref.iloc[0]\n",
    "    print(best_ref.SeqID,best_ref.order,best_ref.family)\n",
    "    incomplete_pt.loc[Sample_Name,'Ref-SeqID'] = best_ref.SeqID\n",
    "    incomplete_pt.loc[Sample_Name,'Ref-order'] = best_ref.order\n",
    "    incomplete_pt.loc[Sample_Name,'Ref-family'] = best_ref.family\n",
    "    \n",
    "    ### launch blastn\n",
    "    # program_path = 'C:/Program Files/NCBI/blast-BLAST_VERSION+/bin/blastn.exe'\n",
    "    program_path = 'blastn'\n",
    "    ref_path = 'References/NCBI_plastid_ref/' + best_ref['SeqID'] + '.fasta'\n",
    "    ### Paul B. contigs_path = '../Organelles/' + DataSource + '/fasta_pt/' + Sample_Name + '_pt.fasta'\n",
    "    contigs_path = org_dir + DataSource + '/fasta_pt/' + Sample_Name + '_pt.fasta'\n",
    "    out_path = 'ENA_submissions/' + DataSource + '/out_blast/'+ Sample_Name + '.blast'\n",
    "    command = program_path + ' -query ' + contigs_path + ' -subject ' + ref_path + \\\n",
    "                ' -outfmt \"6 qseqid sseqid pident length slen qlen mismatch gapopen qstart qend sstart send evalue bitscore\"' + \\\n",
    "                ' -out ' + out_path\n",
    "    commands.append(command)\n",
    "    os.system(command)\n",
    "    \n",
    "    ### read blast output\n",
    "    blast_output = pd.read_table(out_path, sep='\\t',header=None)\n",
    "    blast_output.columns = ['qseqid','sseqid','pident','length','slen','qlen','mismatch','gapopen','qstart','qend','sstart','send','evalue','bitscore']\n",
    "#     print('\\t',blast_output.shape[0],'matched for', blast_output.qseqid.nunique(),'qseqid',end='; ')\n",
    "    incomplete_pt.loc[Sample_Name,'Blast-N_qseqid_match'] = blast_output.qseqid.nunique()\n",
    "    # Keep best match (by bitscore) per queried contig\n",
    "    blast_output = blast_output.sort_values('bitscore',ascending=False).groupby('qseqid').head(1)\n",
    "    blast_output['oriented_sstart'] = blast_output[['sstart','send']].min(axis=1); blast_output['oriented_ssend'] = blast_output[['sstart','send']].max(axis=1)\n",
    "    blast_output['Need_RevComp'] = blast_output.sstart > blast_output.oriented_sstart\n",
    "    blast_output = blast_output.sort_values('oriented_sstart').set_index('qseqid')\n",
    "#     print('Contigs to reverse complement:',blast_output.Need_RevComp.value_counts().to_dict())\n",
    "    \n",
    "    \n",
    "    ### Load contigs and write concatenated fasta\n",
    "    gapN = 100\n",
    "    records = SeqIO.to_dict(SeqIO.parse(contigs_path, \"fasta\"))\n",
    "    print('\\t',blast_output.shape[0],'/',len(records),'contigs mapped to reference')\n",
    "    ### Paul B - Q: what happens if contig can't be placed in the order due to no blast hit?\n",
    "    concat_seq = ''\n",
    "    for qseqid, blast_row in blast_output.iterrows():\n",
    "        if blast_row.Need_RevComp==True:\n",
    "            concat_seq += records[qseqid].seq.reverse_complement()\n",
    "        else:\n",
    "            concat_seq += records[qseqid].seq \n",
    "        concat_seq += Seq(gapN * 'N')\n",
    "    concat_seq = concat_seq[:-gapN] # Remove trailing Ns\n",
    "    print('\\t sum len concat:',blast_output.qlen.sum(),'/',int(row.Sum_len_pt),round(blast_output.qlen.sum()/int(row.Sum_len_pt)*100,1),'%')\n",
    "    incomplete_pt.loc[Sample_Name,'Concat-len'] = blast_output.qlen.sum()\n",
    "    incomplete_pt.loc[Sample_Name,'Concat-%_mapped'] = round(blast_output.qlen.sum()/int(row.Sum_len_pt)*100,1)\n",
    "    incomplete_pt.loc[Sample_Name,'Concat-len_with_Ns'] = len(concat_seq)\n",
    "    Concat_record = SeqRecord(\n",
    "        concat_seq,\n",
    "        id=row.SeqID + '_concat_pt',\n",
    "        name=Sample_Name + '_concat_pt',\n",
    "        description=''\n",
    "    )\n",
    "    SeqIO.write(Concat_record,'ENA_submissions/' + DataSource + '/fasta/' + row.SeqID + '-concat.fasta','fasta')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34c3689b",
   "metadata": {},
   "source": [
    "* The above files are written to the /fasta/ directory.\n",
    "* Submit files for the concatenated/partial plastids to GeSeq https://chlorobox.mpimp-golm.mpg.de/geseq.html\n",
    "    * If only wanting to select a subset of the data, can just limit the number of files uploaded. The rest of the pipeline will just process those samples \n",
    "* Extract zip file into the /geseq/ folder\n",
    "* In code below, geseq Genbank files are modified and written to /Genbank_files folder\n",
    "* Then those modifed Genbank files are converted to EMBL format and written to the /EMBL_files folder \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "4537803d",
   "metadata": {},
   "outputs": [],
   "source": [
    "incomplete_pt.to_csv(sdir + DataSource + '_concatPT_test.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b3312d68",
   "metadata": {},
   "source": [
    "### ENA submission\n",
    "https://ena-docs.readthedocs.io/en/latest/submit/fileprep/assembly.html?highlight=chloroplast#\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "fe4cfafa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Leempoel K., Zuntini A. R., Bailey P., Eiserhardt W. L., Gallego B., Maurin O., Allnutt T., Bellot S., Clarkson J. J., Clements M., Cowan R. S., Crayn D. D., de Lírio E., Françoso E., Höwener A., Hu A., Joyce E., Larson D., Li H., Li D., Liu J., Lum M., McGinnie C., McLay T., Nargar K., Perez Escobar O., Przelomska N., Roy S., Schmidt-Lebuhn A., Shah T., Simpson L., Yang J., Yi T., Forest F., Kersey P. J., Leitch I. J., Baker W. J.'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "paper_title = 'Building the largest, curated plastid genome dataset and phylogeny by leveraging nuclear-oriented sequence data'\n",
    "paper_authors = 'Leempoel K., Zuntini A. R., Bailey P., Eiserhardt W. L., Gallego B., Maurin O., Allnutt T., Bellot S., Clarkson J. J., ' + \\\n",
    "        'Clements M., Cowan R. S., Crayn D. D., de Lírio E., Françoso E., Höwener A., Hu A., Joyce E., Larson D., Li H., Li D., Liu J., Lum M., McGinnie C., ' + \\\n",
    "        'McLay T., Nargar K., Perez Escobar O., Przelomska N., Roy S., Schmidt-Lebuhn A., Shah T., Simpson L., Yang J., Yi T., Forest F., Kersey P. J., ' + \\\n",
    "        'Leitch I. J., Baker W. J.'\n",
    "paper_journal = 'Unpublished'\n",
    "paper_id = 'NA'\n",
    "paper_authors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ab5b44e6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "31\n",
      "31\n",
      "31 {'scaffold': 31}\n",
      "31\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_Name</th>\n",
       "      <th>idSequencing</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>SumContigLength</th>\n",
       "      <th>sum_len</th>\n",
       "      <th>sci_name</th>\n",
       "      <th>log_pt</th>\n",
       "      <th>...</th>\n",
       "      <th>Ref-family</th>\n",
       "      <th>Blast-N_qseqid_match</th>\n",
       "      <th>Concat-len</th>\n",
       "      <th>Concat-%_mapped</th>\n",
       "      <th>Concat-len_with_Ns</th>\n",
       "      <th>ENASampleNum</th>\n",
       "      <th>ENAExpNumber</th>\n",
       "      <th>ENARunNumber</th>\n",
       "      <th>NCBI_TaxID</th>\n",
       "      <th>NCBI_sciname</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ERR4180132</td>\n",
       "      <td>21989.0</td>\n",
       "      <td>Myrtales</td>\n",
       "      <td>Penaeaceae</td>\n",
       "      <td>Penaea</td>\n",
       "      <td>petraea</td>\n",
       "      <td>126591.0</td>\n",
       "      <td>126591.0</td>\n",
       "      <td>Penaea petraea</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERS4591104</td>\n",
       "      <td>ERX4143556</td>\n",
       "      <td>ERR4180132</td>\n",
       "      <td>186937</td>\n",
       "      <td>Sonderothamnus petraeus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ERR420512</td>\n",
       "      <td>23727.0</td>\n",
       "      <td>Fabales</td>\n",
       "      <td>Fabaceae</td>\n",
       "      <td>Senegalia</td>\n",
       "      <td>senegal</td>\n",
       "      <td>134874.0</td>\n",
       "      <td>134874.0</td>\n",
       "      <td>Senegalia senegal</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERS399685</td>\n",
       "      <td>ERX386805</td>\n",
       "      <td>ERR420512</td>\n",
       "      <td>138043</td>\n",
       "      <td>Senegalia senegal</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 53 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Sample_Name  idSequencing     order      family      genus  species  \\\n",
       "0  ERR4180132       21989.0  Myrtales  Penaeaceae     Penaea  petraea   \n",
       "1   ERR420512       23727.0   Fabales    Fabaceae  Senegalia  senegal   \n",
       "\n",
       "   SumContigLength   sum_len           sci_name  log_pt  ...  Ref-family  \\\n",
       "0         126591.0  126591.0     Penaea petraea    True  ...         NaN   \n",
       "1         134874.0  134874.0  Senegalia senegal    True  ...         NaN   \n",
       "\n",
       "   Blast-N_qseqid_match  Concat-len  Concat-%_mapped  Concat-len_with_Ns  \\\n",
       "0                   NaN         NaN              NaN                 NaN   \n",
       "1                   NaN         NaN              NaN                 NaN   \n",
       "\n",
       "   ENASampleNum  ENAExpNumber  ENARunNumber  NCBI_TaxID  \\\n",
       "0    ERS4591104    ERX4143556    ERR4180132      186937   \n",
       "1     ERS399685     ERX386805     ERR420512      138043   \n",
       "\n",
       "              NCBI_sciname  \n",
       "0  Sonderothamnus petraeus  \n",
       "1        Senegalia senegal  \n",
       "\n",
       "[2 rows x 53 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Paul B. - I think from this point, only cells 1 - 4 are necessary to run the cells below \"ENA submission\" \n",
    "### title. In this cell you need to alter these variables to process either whole plastomes OR partial plastomes\n",
    "### - I don't think you can do both together\n",
    "### ALSO on repeating run you MUST run both main cells again not just the final cell that creates the EMBL file, \n",
    "### otherwise the assembly_gap features end up being put on the record twice!!!\n",
    "### Also, still need the /geseq/ dir.\n",
    "\n",
    "### Paul B. - actually I think this line concatenates both circular and concat. samples so they could be\n",
    "### processed together\n",
    "#submit_pt = pd.concat([pd.read_csv(sdir + DataSource + '_wholePT.csv'), pd.read_csv(sdir + DataSource + '_concatPT.csv')],ignore_index=True)\n",
    "### 12.7.2023 - Trying this now:\n",
    "#submit_pt = pd.concat([pd.read_csv(sdir + DataSource + '_wholePT.29_still_to_do.csv'), pd.read_csv(sdir + DataSource + '_concatPT.68_still_to_do.csv')],ignore_index=True)\n",
    "### NB - it is possible but would need to combine the geseq* dirs together so whole and concat annos were in same dir. \n",
    "submit_pt = pd.read_csv(sdir + DataSource + '_concatPT.31_still_to_do.csv')\n",
    "#submit_pt = pd.read_csv(sdir + DataSource + '_concatPT.csv')\n",
    "print(submit_pt.shape[0])\n",
    "submit_pt = submit_pt[(submit_pt.ENASampleNum.notnull()) & (submit_pt.ENARunNumber.notnull())]\n",
    "print(submit_pt.shape[0])\n",
    "submit_pt = submit_pt[ submit_pt.NCBI_sciname.notnull()]\n",
    "print(submit_pt.shape[0], submit_pt.Assembly.value_counts().to_dict())\n",
    "submit_pt = submit_pt[submit_pt.NCBI_TaxID.notnull()]\n",
    "print(submit_pt.shape[0])\n",
    "submit_pt.head(2)\n",
    "\n",
    "# In a separate run, selecting samples that have no NCBI_sciname or NCBI_TaxID:\n",
    "# (NB - comment out the lines below to process the main data set above)\n",
    "#submit_pt = pd.read_csv(sdir + DataSource + '_concatPT.csv')\n",
    "#submit_pt = submit_pt[submit_pt.NCBI_sciname.isnull() | submit_pt.NCBI_TaxID.isnull()]\n",
    "#print(submit_pt.shape[0])\n",
    "#submit_pt.to_csv(sdir + DataSource + '_concatPT_no_sciname_or_taxid.csv')\n",
    "# Now alter the file to add NCBI_TaxID=2738895 and NCBI_sciname=spermatophyte_root to each row,\n",
    "# then upload again into submit_pt:\n",
    "#submit_pt = pd.read_csv(sdir + DataSource + '_concatPT_sciname_spermatophyte_root_taxid_2738895.csv')\n",
    "# Also have 55 samples that failed validation due to no taxon_id in the ENA accession from the main batch:\n",
    "#submit_pt = pd.read_csv(sdir + DataSource + '_concatPT_sciname_spermatophyte_root_taxid_2738895_plus_55_samples.csv')\n",
    "#print(submit_pt.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a93cf327",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SeqID</th>\n",
       "      <th>NCBI_TaxID</th>\n",
       "      <th>NCBI_sciname</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ERR4180132</td>\n",
       "      <td>186937</td>\n",
       "      <td>Sonderothamnus petraeus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ERR420512</td>\n",
       "      <td>138043</td>\n",
       "      <td>Senegalia senegal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ERR420513</td>\n",
       "      <td>162639</td>\n",
       "      <td>Acrocarpus fraxinifolius</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ERR420518</td>\n",
       "      <td>54878</td>\n",
       "      <td>Enterolobium cyclocarpum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ERR420521</td>\n",
       "      <td>185774</td>\n",
       "      <td>Jacaranda mimosifolia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ERR4604253</td>\n",
       "      <td>2065072</td>\n",
       "      <td>Ruilopezia viridis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SRR11342819</td>\n",
       "      <td>2707190</td>\n",
       "      <td>Barbacenia involucrata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SRR12664414</td>\n",
       "      <td>2746967</td>\n",
       "      <td>Cryptocoryne hudoroi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SRR13390744</td>\n",
       "      <td>2778903</td>\n",
       "      <td>Tolumnia bahamensis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>SRR14294427</td>\n",
       "      <td>134525</td>\n",
       "      <td>Triosteum himalayanum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>SRR1524841</td>\n",
       "      <td>97729</td>\n",
       "      <td>Boesenbergia rotunda</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>SRR1524842</td>\n",
       "      <td>97729</td>\n",
       "      <td>Boesenbergia rotunda</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>SRR2482964</td>\n",
       "      <td>34335</td>\n",
       "      <td>Bensoniella oregona</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>SRR3542939</td>\n",
       "      <td>1611645</td>\n",
       "      <td>Leptosiphon rosaceus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>SRR3734429</td>\n",
       "      <td>652145</td>\n",
       "      <td>Camassia quamash subsp. maxima</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>SRR5237180</td>\n",
       "      <td>128004</td>\n",
       "      <td>Ursinia anthemoides</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>SRR576525</td>\n",
       "      <td>337364</td>\n",
       "      <td>Erodium chrysanthum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>SRR6425615</td>\n",
       "      <td>2066940</td>\n",
       "      <td>Landolphia dawei</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>SRR6425630</td>\n",
       "      <td>403085</td>\n",
       "      <td>Dyera costulata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>SRR6425632</td>\n",
       "      <td>141613</td>\n",
       "      <td>Strempeliopsis strempelioides</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>SRR6787480</td>\n",
       "      <td>117594</td>\n",
       "      <td>Acleisanthes chenopodioides</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>SRR799414</td>\n",
       "      <td>24962</td>\n",
       "      <td>Castanospermum australe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>SRR8432301</td>\n",
       "      <td>403052</td>\n",
       "      <td>Dicerandra linearifolia var. robustior</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>SRR8666244</td>\n",
       "      <td>162878</td>\n",
       "      <td>Petalostylis labicheoides</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>SRR9611034</td>\n",
       "      <td>41401</td>\n",
       "      <td>Phryma leptostachya</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>SRR5237176</td>\n",
       "      <td>56526</td>\n",
       "      <td>Gazania krebsiana</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>SRR8666819</td>\n",
       "      <td>2529040</td>\n",
       "      <td>Pterocaulon serrulatum var. velutinum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>ERR4993553</td>\n",
       "      <td>221192</td>\n",
       "      <td>Pilosella onegensis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>SRR9119184</td>\n",
       "      <td>1569433</td>\n",
       "      <td>Paralychnophora harleyi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>SRR9119016</td>\n",
       "      <td>2576593</td>\n",
       "      <td>Stilpnopappus tomentosus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>SRR9118957</td>\n",
       "      <td>125658</td>\n",
       "      <td>Argyroglottis turbinata</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          SeqID  NCBI_TaxID                            NCBI_sciname\n",
       "0    ERR4180132      186937                 Sonderothamnus petraeus\n",
       "1     ERR420512      138043                       Senegalia senegal\n",
       "2     ERR420513      162639                Acrocarpus fraxinifolius\n",
       "3     ERR420518       54878                Enterolobium cyclocarpum\n",
       "4     ERR420521      185774                   Jacaranda mimosifolia\n",
       "5    ERR4604253     2065072                      Ruilopezia viridis\n",
       "6   SRR11342819     2707190                  Barbacenia involucrata\n",
       "7   SRR12664414     2746967                    Cryptocoryne hudoroi\n",
       "8   SRR13390744     2778903                     Tolumnia bahamensis\n",
       "9   SRR14294427      134525                   Triosteum himalayanum\n",
       "10   SRR1524841       97729                    Boesenbergia rotunda\n",
       "11   SRR1524842       97729                    Boesenbergia rotunda\n",
       "12   SRR2482964       34335                     Bensoniella oregona\n",
       "13   SRR3542939     1611645                    Leptosiphon rosaceus\n",
       "14   SRR3734429      652145          Camassia quamash subsp. maxima\n",
       "15   SRR5237180      128004                     Ursinia anthemoides\n",
       "16    SRR576525      337364                     Erodium chrysanthum\n",
       "17   SRR6425615     2066940                        Landolphia dawei\n",
       "18   SRR6425630      403085                         Dyera costulata\n",
       "19   SRR6425632      141613           Strempeliopsis strempelioides\n",
       "20   SRR6787480      117594             Acleisanthes chenopodioides\n",
       "21    SRR799414       24962                 Castanospermum australe\n",
       "22   SRR8432301      403052  Dicerandra linearifolia var. robustior\n",
       "23   SRR8666244      162878               Petalostylis labicheoides\n",
       "24   SRR9611034       41401                     Phryma leptostachya\n",
       "25   SRR5237176       56526                       Gazania krebsiana\n",
       "26   SRR8666819     2529040   Pterocaulon serrulatum var. velutinum\n",
       "27   ERR4993553      221192                     Pilosella onegensis\n",
       "28   SRR9119184     1569433                 Paralychnophora harleyi\n",
       "29   SRR9119016     2576593                Stilpnopappus tomentosus\n",
       "30   SRR9118957      125658                 Argyroglottis turbinata"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### submit_pt = submit_pt[submit_pt.ENARunNumber=='ERR7619260']\n",
    "###submit_pt = submit_pt[ submit_pt.ENARunNumber=='ERR5010320' ]\n",
    "len(submit_pt)\n",
    "submit_pt.columns\n",
    "submit_pt[ ['SeqID','NCBI_TaxID', 'NCBI_sciname'] ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "7f190258",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1678470, 4)\n",
      "(1172782, 4)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>len</th>\n",
       "      <th>Sample_Name</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GAP_026547-rrn23</td>\n",
       "      <td>2812</td>\n",
       "      <td>GAP_026547</td>\n",
       "      <td>rrn23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GAP_026547-rrn16</td>\n",
       "      <td>1491</td>\n",
       "      <td>GAP_026547</td>\n",
       "      <td>rrn16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              seqid   len Sample_Name   gene\n",
       "0  GAP_026547-rrn23  2812  GAP_026547  rrn23\n",
       "1  GAP_026547-rrn16  1491  GAP_026547  rrn16"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_seq = pd.read_csv('cpTree_v7/AllSamples_Allgenes_All_SeqTable.csv')\n",
    "print(all_seq.shape)\n",
    "genes_82 = pd.read_table('cp_genes_82.txt',header=None).rename(columns={0:'gene'})\n",
    "all_seq = all_seq[all_seq.gene.isin(genes_82.gene)]\n",
    "print(all_seq.shape)\n",
    "all_seq.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ed8db436",
   "metadata": {},
   "outputs": [],
   "source": [
    "def list_genes(record, ls_types = ['gene','CDS','rRNA']):\n",
    "    # list gene names\n",
    "    feat_ls = []; idx = 0\n",
    "    for feature in record.features:\n",
    "        if feature.type in ls_types:\n",
    "            feat_ls.append(feature.qualifiers['gene'][0])\n",
    "        idx += 1\n",
    "    return list(set(feat_ls))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "c050751f",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1281\n",
      "\n",
      "\n",
      "List of missing Genbank files that need to be run through GeSeq again:\n"
     ]
    }
   ],
   "source": [
    "### Paul B. - modified this cell to add a dictionary to assess samples not in gb_files.\n",
    "### NB - check /geseq/ first - I've turned it into a symlink that points either to whole or concat geseq directories  \n",
    "\n",
    "\n",
    "# Rename and modify genbank files in this cell (then convert to embl flat file in downstream cell)\n",
    "## Filename example: GeSeqJob-20220601-70838_Pis_010569_concat_pt_GenBank.gb\n",
    "if not os.path.exists(sdir + '/Genbank_files/'):\n",
    "    os.makedirs(sdir + '/Genbank_files/')\n",
    "gb_files = [file for file in os.listdir(sdir + 'geseq/') if file.endswith('.gb')]\n",
    "\n",
    "gbFilesDict = {} # Records each file in a dictionary key so that a list can be made of samples not in gb_files         \n",
    "for file in os.listdir(sdir + 'geseq/'):\n",
    "    if file.endswith('.gb'):\n",
    "        fields = file.split('_')\n",
    "        ### Paul B. - added conditional to get for SRA data where the SeqID comprises only the first field\n",
    "        if DataSource == 'SRA':\n",
    "            key = fields[1]\n",
    "        else:\n",
    "            key = fields[1] + '_' + fields[2]\n",
    "        gbFilesDict[key] = file\n",
    "        ###print(gbFilesDict[key])\n",
    "gbFilesAbsent = [] # Records the Genbank files absent from GeSeq analysis - trying to re-run these in GeSeq\n",
    "print(len(gb_files))\n",
    "for idx, row in submit_pt.iterrows():\n",
    "    #print(idx, ' ', row.SeqID)\n",
    "    #indexStored = '' ### Paul B. altered code here so as not to use list comprehension - wasn't working for me\n",
    "    ### Paul B. - removed loop to replace with key search: for index, file in enumerate(gb_files):\n",
    "    if row.SeqID in gbFilesDict:\n",
    "        ###print('Outer: ', idx, ' ', row.SeqID, '; inner: ', index, ' ', file)\n",
    "        ###if row.SeqID in file:\n",
    "        #print('Outer: ', idx, ' ', row.SeqID, '; inner: ', index, ' ', gbFilesDict[row.SeqID])\n",
    "            ###print(index, row.SeqID)\n",
    "        #indexStored = int(index)\n",
    "        ###print(row.Sample_Name,gbFilesDict[row.SeqID])\n",
    "    ### Original  code:\n",
    "    ### idx_file = [i for i, x in enumerate([row.SeqID in file for file in gb_files]) if x]\n",
    "    ### print(row.Sample_Name,gb_files[index], end=':')\n",
    "    ### records = list(SeqIO.parse(sdir + 'geseq/' + gb_files[idx_file[0]], \"genbank\"))\n",
    "        records = list(SeqIO.parse(sdir + 'geseq/' + gbFilesDict[row.SeqID], \"genbank\"))\n",
    "        ### Paul B. - I don't think the next 3 lines are doing anything other than be involved in the print \n",
    "        record = records[0]\n",
    "        record_genes = list_genes(record); record_genes = [gene for gene in record_genes if gene in genes_82.gene.to_list()]\n",
    "        blast_genes = all_seq[all_seq.Sample_Name==row.Sample_Name]\n",
    "        #print(len(records), len(records[0]), len(record_genes),'/',blast_genes.gene.nunique(),\n",
    "        #    int(len(record_genes)/blast_genes.gene.nunique()*100),'%',blast_genes[blast_genes.gene.isin(record_genes)==False].gene.unique())\n",
    "        try:\n",
    "            if row.Assembly in ['circular genome','circular genome with gaps']:\n",
    "                ### 23.11.2022 - not submitting SRA or GAP data via TPA now\n",
    "                ###if DataSource == 'SRA' or DataSource == 'GAP':\n",
    "                ###    record.description = 'TPA_asm: ' + row.ENARunNumber + ' chloroplast, complete genome'\n",
    "                ###else:\n",
    "                record.description = row.ENARunNumber + ' chloroplast, complete genome'  \n",
    "            elif row.Assembly in ['scaffold']:\n",
    "                ###if DataSource == 'SRA' or DataSource == 'GAP':\n",
    "                ###    record.description = 'TPA_asm: ' + row.ENARunNumber + ' chloroplast, partial genome (concatenated contigs)'\n",
    "                ###else:\n",
    "                record.description = row.ENARunNumber + ' chloroplast, partial genome (concatenated contigs)'\n",
    "                record.annotations['topology'] = 'linear'\n",
    "            ### Paul B. - pushed back indentation level to also print circular samples\n",
    "            submit_pt.loc[idx,'ENA_SeqID'] = row.ENARunNumber + '_pt'\n",
    "            record.id = submit_pt.loc[idx,'ENA_SeqID']\n",
    "            record.name = submit_pt.loc[idx,'ENA_SeqID']\n",
    "            record.version = submit_pt.loc[idx,'ENA_SeqID'] + '.1'\n",
    "            record.annotations['source'] = row.ENASampleNum + ', ' + row.sci_name + ', isolate'\n",
    "            record.annotations['accessions'] = row.ENASampleNum\n",
    "            record.annotations['organism'] = row.NCBI_sciname\n",
    "            record.annotations[\"data_file_division\"]=\"PLN\";\n",
    "            record.annotations.pop('taxonomy')\n",
    "            record.annotations['references'][0].title = paper_title \n",
    "            record.annotations['references'][0].authors = paper_authors    \n",
    "            record.annotations['references'][0].journal = paper_journal      \n",
    "            record.annotations['references'][0].pubmed_id = paper_id    \n",
    "            record.features[0].qualifiers['organism'] = row.NCBI_sciname\n",
    "            record.features[0].qualifiers['db_xref'] = ['taxon:' + str(int(row['NCBI_TaxID']))]\n",
    "            #print(record)\n",
    "            SeqIO.write(record,sdir + 'Genbank_files/' + row.SeqID + '_pt.gb',format='genbank')\n",
    "            \n",
    "        except:\n",
    "            ### Paul B. - changed: print('issue with ',gb_files[idx_file[0]])\n",
    "            print('issue with ',gbFilesDict[row.SeqID])\n",
    "    else:\n",
    "        print('WARNING: no geseq file in submit_pt table for: ', row.SeqID)\n",
    "        gbFilesAbsent.append(row.SeqID)\n",
    "print('\\n\\nList of missing Genbank files that need to be run through GeSeq again:')\n",
    "for sample in gbFilesAbsent: \n",
    "        print(sample)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "a7f2ab7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def del_features(record, feat_del_ls = ['translation','info','annotator']):\n",
    "    for feature in record.features:\n",
    "        for feat_to_del in feat_del_ls:\n",
    "            if feat_to_del in feature.qualifiers:\n",
    "                del feature.qualifiers[feat_to_del]  ### Paul B - NOT removing from the record object - don't understand but seems to work!\n",
    "    return record\n",
    "\n",
    "def del_fragment(record, ls_types = ['gene','CDS','rRNA','intron']): ### Paul B.: added introns there are some intron fragments when BLAST and Chloe are used together (only for rps12 though); tRNA appears not to be an issue\n",
    "    # list gene names\n",
    "    ls_feat = []\n",
    "    for feature in record.features:\n",
    "        if feature.type in ls_types:\n",
    "            ls_feat.append(feature.qualifiers['gene'][0])\n",
    "    ls_feat = [ifeat for ifeat in ls_feat if 'fragment' not in ifeat]\n",
    "\n",
    "    # remove fragment feature if full feature is present\n",
    "    rm_count = 0\n",
    "    for feature in record.features:\n",
    "        if feature.type in ls_types:\n",
    "            if 'fragment' in feature.qualifiers['gene'][0]:\n",
    "                if feature.qualifiers['gene'][0].split('-')[0] in ls_feat:\n",
    "                    rm_count += 1\n",
    "                    record.features.remove(feature)   ### Paul B - removing from the record object - OK\n",
    "    print('removed',rm_count,'partial features')\n",
    "    return record\n",
    "\n",
    "def del_longfeat(record, ls_types = ['gene','CDS','rRNA'], len_thrs = 5000):  ### Paul B. - longest plastid gene I found was 6,791 in sample 000259 \n",
    "    for feature in record.features:\n",
    "        if feature.type in ls_types:\n",
    "            if len(feature)>len_thrs:\n",
    "                print('removing long feature ',feature.qualifiers['gene'],len(feature))\n",
    "                record.features.remove(feature)      ### Paul B - removing from the record object - OK\n",
    "    return record\n",
    "\n",
    "def add_gap_features(record, gapN = 100):\n",
    "    # find gaps in record\n",
    "    gaps_idx = [m.start() for m in re.finditer(gapN * 'N', str(record.seq))]\n",
    "    print(len(gaps_idx),'gaps')\n",
    "    if len(gaps_idx)>0:\n",
    "        ## for each gap, add a sequence feature\n",
    "        for gap_pos in gaps_idx:\n",
    "            # find index of gap in features\n",
    "            features_start = [feat.location.start for feat in record.features]\n",
    "            gap_feature_idx = [i for i in range(len(features_start)) if features_start[i] > gap_pos]\n",
    "            \n",
    "            gap_feat = SeqFeature(FeatureLocation(gap_pos,gap_pos+gapN), strand=1, type='assembly_gap')\n",
    "            gap_feat.qualifiers['gap_type']='within scaffold'; \n",
    "            gap_feat.qualifiers['estimated_length']='unknown'\n",
    "            gap_feat.qualifiers['linkage_evidence']='unspecified'\n",
    "            if len(gap_feature_idx)>0:\n",
    "                gap_feature_idx = gap_feature_idx[0]\n",
    "                record.features.insert(gap_feature_idx, gap_feat)\n",
    "            elif len(gap_feature_idx)==0: #if position is further than any other feature, append at the end of feature list\n",
    "                record.features.append(gap_feat)\n",
    "    return record\n",
    "\n",
    "# https://ena-docs.readthedocs.io/en/latest/faq/locus_tags.html\n",
    "def add_locus_tag(record, locus_tag, ls_types = ['gene','CDS','rRNA','tRNA','intron','exon']):\n",
    "    locus_tags_genes = {}; idx = 1\n",
    "    for feature in record.features:\n",
    "        if feature.type in ls_types:\n",
    "            if feature.qualifiers['gene'][0] not in locus_tags_genes:\n",
    "                locus_tags_genes[feature.qualifiers['gene'][0]] = locus_tag + '_' + str(idx)\n",
    "                idx += 1\n",
    "            feature.qualifiers['locus_tag']=locus_tags_genes[feature.qualifiers['gene'][0]]\n",
    "            ### Paul B. Don't understand how changing contents of feature can change the\n",
    "            ### contents of record but it does seem to.\n",
    "    return record\n",
    "\n",
    "def add_locus_tag_v1(record, locus_tag, ls_types = ['gene','CDS','rRNA','tRNA','intron','exon']):\n",
    "    ''' Paul B. - extended the existing add_locus_tag() function so locus tags could be \n",
    "        assigned to gene features along the genome rather than per gene name.\n",
    "        \n",
    "        NBNB - NOT removing from the record object - don't understand but seems to work!\n",
    "    '''\n",
    "    locus_tags_genes = {}; idx = 0 # I think it's incorrect logic to track genes, need to create a new locus tag ID for EACH gene feature\n",
    "    geneName = ''  # Holds a gene name until all connected gene features have been seen\n",
    "    idxErr = 1000\n",
    "    geneNameErr = '' # Holds a gene name where a feature is defined before the 'gene' feature until \n",
    "                     # all connected gene features have been seen \n",
    "    print('Number of feats', len(record.features))\n",
    "    for feature in record.features:   # Returns an array of features - therefore can assume order is as along the genome\n",
    "        if feature.type == 'gene': # Removed these features, they are both associated with a gene \n",
    "                                   # feature:  or feature.type == 'rRNA' or feature.type == 'tRNA':\n",
    "            idx += 1 # locus tag counter\n",
    "            #print('New gene, rRNA or tRNA: index ', idx, feature.qualifiers['gene'][0]) # NB - this qualifier key value is an array!\n",
    "            #locus_tags_genes[feature.qualifiers['gene'][0]] = locus_tag + '_' + str(idx)\n",
    "            locus_tags_genes[feature.qualifiers['gene'][0]] = str(idx)\n",
    "            geneName = feature.qualifiers['gene'][0] # NB - the gene name does appear to be house in an array - weird, unless it enables you to have multiple names\n",
    "        if feature.type in ls_types:\n",
    "            if feature.qualifiers['gene'][0] != geneName:\n",
    "                # Need to check whether next feature belongs to the current gene name.\n",
    "                # There are examples where this is not the case!\n",
    "                # Q1: could you ever have a feature of gene before it's defined? - yes see below \n",
    "                print('WARNING: current feature NOT part of current gene!!!', feature.qualifiers['gene'][0])\n",
    "                # In which case, try to use the locus tag of the correct gene if it already exists in \n",
    "                # an earlier part of the genome, hopefully closeby:\n",
    "                if feature.qualifiers['gene'][0] in locus_tags_genes:\n",
    "                    # Removed the 'warn' for final run:  feature.qualifiers['locus_tag'] = locus_tag + '_warn' + locus_tags_genes[feature.qualifiers['gene'][0]]\n",
    "                    feature.qualifiers['locus_tag'] = locus_tag + '_' + locus_tags_genes[feature.qualifiers['gene'][0]]\n",
    "                else:\n",
    "                    print(\"ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\")\n",
    "                    # Need to force create a new locus tag for gene that doesn't exist (yet)\n",
    "                    if feature.qualifiers['gene'][0] != geneNameErr:\n",
    "                        idxErr += 1 \n",
    "                        # Removed the '_1err' for final run:  feature.qualifiers['locus_tag'] = locus_tag + '_1err' + str(idxErr)\n",
    "                        feature.qualifiers['locus_tag'] = locus_tag + '_' + str(idxErr)\n",
    "                        geneNameErr = feature.qualifiers['gene'][0]\n",
    "                        # idxErr will get incemented every time a new gene is hit but this \n",
    "                        # shouldn't be very likely and will cease as soon as a gene feat is found.\n",
    "                    else:\n",
    "                        # Removed the '_2err' for final run: feature.qualifiers['locus_tag'] = locus_tag + '_2err' + str(idxErr)  # locus_tags_genes[feature.qualifiers['gene'][0]] will not exist here yet! \n",
    "                        feature.qualifiers['locus_tag'] = locus_tag + '_' + str(idxErr)\n",
    "                    ### Other ideas:\n",
    "                    ### Could also assess overlap in coordinates to confirm the location of each gene feature\n",
    "                    ### but this is more complex - look at my set analysis - could add to locus_tags_genes if feature has nesting coords\n",
    "                    ### locus_tags_genes {gene}{coordSet and locusTagID}\n",
    "                    ### Not sure what the best key would be as I've seen a gene with no 'gene' feature!\n",
    "            else:\n",
    "                feature.qualifiers['locus_tag'] = locus_tag + '_' + str(idx)\n",
    "    return record\n",
    "\n",
    "def gzip_file(in_file):\n",
    "    import gzip\n",
    "    import shutil\n",
    "    f_gz = in_file + '.gz'\n",
    "    with open(in_file, 'rb') as f_in:\n",
    "        with gzip.open(f_gz, 'wb') as f_out:\n",
    "            shutil.copyfileobj(f_in, f_out)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "e27bc72c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ERR4180132 ERR4180132_pt > 32036\n",
      "removed 0 partial features\n",
      "32 gaps\n",
      "Number of feats 127\n",
      "ERR420512 ERR420512_pt > 39515\n",
      "removed 1 partial features\n",
      "14 gaps\n",
      "Number of feats 106\n",
      "WARNING: current feature NOT part of current gene!!! rps12-fragment\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "ERR420513 ERR420513_pt > 28388\n",
      "removed 0 partial features\n",
      "30 gaps\n",
      "Number of feats 117\n",
      "ERR420518 ERR420518_pt > 64889\n",
      "removed 0 partial features\n",
      "23 gaps\n",
      "Number of feats 161\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "ERR420521 ERR420521_pt > 52371\n",
      "removed 2 partial features\n",
      "21 gaps\n",
      "Number of feats 150\n",
      "WARNING: current feature NOT part of current gene!!! rps12-fragment\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! rps12-fragment\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "ERR4604253 ERR4604253_pt > 151715\n",
      "removed 2 partial features\n",
      "removing long feature  ['rps12'] 124213\n",
      "0 gaps\n",
      "Number of feats 329\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! ycf1-fragment\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! rps19-fragment\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "SRR11342819 SRR11342819_pt > 75521\n",
      "removed 0 partial features\n",
      "56 gaps\n",
      "Number of feats 249\n",
      "SRR12664414 SRR12664414_pt > 133785\n",
      "removed 0 partial features\n",
      "10 gaps\n",
      "Number of feats 275\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "SRR13390744 SRR13390744_pt > 142634\n",
      "removed 1 partial features\n",
      "removing long feature  ['trnA-UGC'] 52692\n",
      "removing long feature  ['trnI-GAU'] 53705\n",
      "6 gaps\n",
      "Number of feats 286\n",
      "WARNING: current feature NOT part of current gene!!! rps16-fragment\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! trnA-UGC\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! trnA-UGC\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! trnA-UGC\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! trnA-UGC\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! trnI-GAU\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! trnI-GAU\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! trnI-GAU\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! trnI-GAU\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "SRR14294427 SRR14294427_pt > 111883\n",
      "removed 1 partial features\n",
      "41 gaps\n",
      "Number of feats 305\n",
      "WARNING: current feature NOT part of current gene!!! rps18-fragment\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "SRR1524841 SRR1524841_pt > 17604\n",
      "removed 0 partial features\n",
      "23 gaps\n",
      "Number of feats 82\n",
      "SRR1524842 SRR1524842_pt > 23494\n",
      "removed 0 partial features\n",
      "25 gaps\n",
      "Number of feats 90\n",
      "SRR2482964 SRR2482964_pt > 122358\n",
      "removed 0 partial features\n",
      "14 gaps\n",
      "Number of feats 274\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "SRR3542939 SRR3542939_pt > 18911\n",
      "removed 2 partial features\n",
      "10 gaps\n",
      "Number of feats 60\n",
      "WARNING: current feature NOT part of current gene!!! psbA-fragment\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! ccsA-fragment\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "SRR3734429 SRR3734429_pt > 156233\n",
      "removed 1 partial features\n",
      "7 gaps\n",
      "Number of feats 295\n",
      "WARNING: current feature NOT part of current gene!!! ycf1-fragment\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "SRR5237180 SRR5237180_pt > 94316\n",
      "removed 3 partial features\n",
      "4 gaps\n",
      "Number of feats 194\n",
      "WARNING: current feature NOT part of current gene!!! rps19-fragment\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! clpP1-fragment\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! ycf2-fragment\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "SRR576525 SRR576525_pt > 113987\n",
      "removed 0 partial features\n",
      "removing long feature  ['ndhA'] 19211\n",
      "removing long feature  ['rps12'] 24957\n",
      "10 gaps\n",
      "Number of feats 269\n",
      "WARNING: current feature NOT part of current gene!!! ndhA\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! ndhA\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! ndhA\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! ndhA\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! ndhA\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! ndhA\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! ndhA\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! ndhA\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! ndhA\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! ndhA\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! ndhA\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! ndhA\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! ndhA\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! ndhA\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! ndhA\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! ndhA\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! ndhA\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! ndhA\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! ndhA\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! ndhA\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! ndhA\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! ndhA\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRR6425615 SRR6425615_pt > 155587\n",
      "removed 2 partial features\n",
      "3 gaps\n",
      "Number of feats 337\n",
      "WARNING: current feature NOT part of current gene!!! rps19-fragment\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "WARNING: current feature NOT part of current gene!!! ycf1-fragment\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "SRR6425630 SRR6425630_pt > 129587\n",
      "removed 1 partial features\n",
      "2 gaps\n",
      "Number of feats 282\n",
      "WARNING: current feature NOT part of current gene!!! rps19-fragment\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "SRR6425632 SRR6425632_pt > 154463\n",
      "removed 2 partial features\n",
      "0 gaps\n",
      "Number of feats 334\n",
      "WARNING: current feature NOT part of current gene!!! rps19-fragment\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "WARNING: current feature NOT part of current gene!!! ycf1-fragment\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "SRR6787480 SRR6787480_pt > 44480\n",
      "removed 0 partial features\n",
      "15 gaps\n",
      "Number of feats 112\n",
      "SRR799414 SRR799414_pt > 35591\n",
      "removed 0 partial features\n",
      "37 gaps\n",
      "Number of feats 137\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "SRR8432301 SRR8432301_pt > 152019\n",
      "removed 2 partial features\n",
      "removing long feature  ['trnK-UUU'] 151573\n",
      "0 gaps\n",
      "Number of feats 332\n",
      "WARNING: current feature NOT part of current gene!!! trnK-UUU\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! trnK-UUU\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! trnK-UUU\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! rps19-fragment\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "WARNING: current feature NOT part of current gene!!! ycf1-fragment\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "SRR8666244 SRR8666244_pt > 158100\n",
      "removed 2 partial features\n",
      "removing long feature  ['trnI-GAU'] 81699\n",
      "removing long feature  ['ycf2'] 88554\n",
      "2 gaps\n",
      "Number of feats 292\n",
      "WARNING: current feature NOT part of current gene!!! trnI-GAU\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! trnI-GAU\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! trnI-GAU\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! trnI-GAU\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! ycf2\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! ycf2\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! ycf2\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! ycf2\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! ycf1-fragment\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! rps19-fragment\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "SRR9611034 SRR9611034_pt > 17281\n",
      "removed 0 partial features\n",
      "6 gaps\n",
      "Number of feats 52\n",
      "SRR5237176 SRR5237176_pt > 87878\n",
      "removed 3 partial features\n",
      "14 gaps\n",
      "Number of feats 214\n",
      "WARNING: current feature NOT part of current gene!!! rps19-fragment\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! rps12-fragment\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! rps19-fragment\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "SRR8666819 SRR8666819_pt > 152423\n",
      "removed 3 partial features\n",
      "removing long feature  ['rps12'] 124614\n",
      "0 gaps\n",
      "Number of feats 326\n",
      "WARNING: current feature NOT part of current gene!!! rps19-fragment\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! ycf1-fragment\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! rps19-fragment\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "ERR4993553 ERR4993553_pt > 128649\n",
      "removed 1 partial features\n",
      "2 gaps\n",
      "Number of feats 280\n",
      "WARNING: current feature NOT part of current gene!!! rps19-fragment\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "SRR9119184 SRR9119184_pt > 152594\n",
      "removed 2 partial features\n",
      "4 gaps\n",
      "Number of feats 334\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "WARNING: current feature NOT part of current gene!!! ycf1-fragment\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "WARNING: current feature NOT part of current gene!!! rps19-fragment\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRR9119016 SRR9119016_pt > 128291\n",
      "removed 2 partial features\n",
      "2 gaps\n",
      "Number of feats 281\n",
      "WARNING: current feature NOT part of current gene!!! rps19-fragment\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "WARNING: current feature NOT part of current gene!!! rps12\n",
      "WARNING: current feature NOT part of current gene!!! ycf1-fragment\n",
      "ERROR: can't assign a locus tag ID for a gene that hasn't been defined along the sequence yet!\n",
      "SRR9118957 SRR9118957_pt > 27587\n",
      "removed 0 partial features\n",
      "9 gaps\n",
      "Number of feats 88\n"
     ]
    }
   ],
   "source": [
    "# Now ready to modify ENA files and delete certain features.\n",
    "### Paul B - also the /Genbank_files/ are being modified again here and printed to same file, \n",
    "### so for troubleshooting this cell will need to re-run the above cell that produces the \n",
    "### /Genbank_files/ for the erste time \n",
    "if not os.path.exists(sdir + '/EMBL_files/'):\n",
    "    os.makedirs(sdir + '/EMBL_files/')\n",
    "commands = []\n",
    "commands.append('#!/bin/bash') ### Paul B. added shebang line \n",
    "for idx, row in submit_pt.iterrows():  ### Paul B. - NB - if not processing all fasta files, the except clause will be triggered - OK \n",
    "    ### Convert .gb to EMBL flat files\n",
    "    file_gb = sdir + 'Genbank_files/' + row.SeqID + '_pt.gb'\n",
    "    print(row.SeqID,row.ENA_SeqID,end=' > ')\n",
    "    try:\n",
    "        record = list(SeqIO.parse(file_gb, \"genbank\"))[0]\n",
    "        print(len(record))\n",
    "\n",
    "        # Deleted features preventing ena submission\n",
    "        record_clean = del_features(record = record, feat_del_ls = ['translation','info','annotator'])\n",
    "        # Delete fragmentary annotations (e.g. rr16-fragment)\n",
    "        record_clean = del_fragment(record = record_clean); ### Paul B.: removed this duplicate call to del_fragment(): record_clean = del_fragment(record = record_clean);\n",
    "        # Delete abnormaly long genes\n",
    "        record_clean = del_longfeat(record = record_clean,len_thrs=10000);\n",
    "        # Add gap features for partial plastomes. Recognises gaps of 100 Ns by default\n",
    "        record_clean = add_gap_features(record = record_clean)\n",
    "        # Add locus tag if sequence has gaps   ### Paul B. - why do you need a locus tag for gaps, they aren't genes?\n",
    "        ### Paul B. removed: if 'assembly_gap' in [feature.type for feature in record.features]:\n",
    "            ### Paul B. arhh I see, it's just testing in general whether any feature type\n",
    "            ### is of the concat type, then add_locus_tag() is triggered to apply a locus_tag ID\n",
    "            ### to each gene locus - OK - however as it is, the circular sequences would not have a\n",
    "            ### locus tag applied - so I think the conditional can be removed so function is applied to\n",
    "            ### both concat and circular genomes - done \n",
    "            ### NB - I think it could also be used in line above to test whether gap features have already been added!\n",
    "        record_clean = add_locus_tag_v1(record = record_clean, locus_tag = project_locus_tag)\n",
    "\n",
    "        file_embl = sdir + '/EMBL_files/' + row.ENA_SeqID + '.embl'\n",
    "        SeqIO.write(record_clean, file_gb, \"genbank\")\n",
    "        ### Paul B. NBNB - the /Genbank_files/ are also being modified here \n",
    "        ### and overwritten to same file!! again! Didn't notice this line at first\n",
    "        ###exit()   ### Paul B.\n",
    "        SeqIO.write(record_clean, file_embl, \"embl\")\n",
    "        gzip_file(file_embl)\n",
    "\n",
    "        ### Chromosome list file\n",
    "        # https://ena-docs.readthedocs.io/en/latest/submit/fileprep/assembly.html?highlight=chloroplast#chromosome-list-file\n",
    "        if row.Assembly in ['circular genome','circular genome with gaps']:\n",
    "            list_file=row.ENA_SeqID + '\tPT\tCircular-Chromosome\tChloroplast'       \n",
    "        elif row.Assembly in ['scaffold']:\n",
    "            list_file=row.ENA_SeqID + '\tPT\tLinear-Chromosome\tChloroplast'\n",
    "        with open(file_embl.replace('.embl','_CHR.txt'), 'wb') as f:\n",
    "            f.write(list_file.encode(\"ascii\"))\n",
    "        gzip_file(file_embl.replace('.embl','_CHR.txt'))\n",
    "\n",
    "        ### Manifest file \n",
    "        # https://ena-docs.readthedocs.io/en/latest/submit/assembly/genome.html \n",
    "        manifest_file = file_embl.replace('.embl','_manifest.txt')\n",
    "        if row.Assembly in ['circular genome','circular genome with gaps']:\n",
    "            tmp_description=' chloroplast, complete genome' \n",
    "        elif row.Assembly in ['scaffold']:\n",
    "            tmp_description=' chloroplast, partial genome (concatenated contigs)'\n",
    "        manifest={\n",
    "            'STUDY':ENA_project_number,\n",
    "            'SAMPLE':row.ENASampleNum,\n",
    "            'ASSEMBLYNAME':row.ENA_SeqID,\n",
    "            'ASSEMBLY_TYPE':'clone or isolate',\n",
    "            'COVERAGE':row.Coverage_base,\n",
    "            'PROGRAM':'GetOrganelle v1.7.5 +  GeSeq 2.03',\n",
    "            'PLATFORM':'ILLUMINA',\n",
    "            'MOLECULETYPE':'genomic DNA',\n",
    "            'FLATFILE': row.ENA_SeqID + '.embl.gz',\n",
    "            'CHROMOSOME_LIST': row.ENA_SeqID + '_CHR.txt.gz',\n",
    "            'RUN_REF': row.ENARunNumber,\n",
    "            'DESCRIPTION': row.NCBI_sciname + tmp_description,\n",
    "        }\n",
    "        # Need to submit SRA and GAP data to ENA under third party annotation (TPA):\n",
    "        ### 23.11.2022 - not submitting SRA or GAP data via TPA now\n",
    "        ### if DataSource == 'SRA' or DataSource == 'GAP':\n",
    "        ###    manifest['TPA'] = 'true'\n",
    "        ###    manifest['DESCRIPTION'] = 'TPA_asm: ' + row.NCBI_sciname + tmp_description\n",
    "#         if row.Assembly in ['scaffold']:\n",
    "#             manifest['PARTIAL']='TRUE'\n",
    "        manifest = pd.DataFrame.from_dict(manifest,orient='index')\n",
    "        manifest.to_csv(manifest_file,header=None,sep='\\t')\n",
    "\n",
    "        # Submission command\n",
    "        commands.append('java -jar $WEBIN ' + \n",
    "              '-username Webin-52995 -passwordFile ../../../../ENA_submissions/ena_pwd.txt -context genome -manifest ' +\n",
    "              row.ENA_SeqID + '_manifest.txt' + ' -validate') \n",
    "    except:\n",
    "        print('issue modifying Genbank file or making embl file',row.SeqID)\n",
    "pd.DataFrame(commands).to_csv(sdir + '/ENA_Submission_Commands_' + DataSource + '.sh',index=False,header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "91eba799",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Sample_Name', 'idPaftol', 'idSequencing', 'Project', 'order', 'family',\n",
       "       'genus', 'species', 'Taxonomical_Notes', 'SumContigLength', 'sci_name',\n",
       "       'log_pt', 'log_nr', 'pt_recovered', 'Nseq_pt', 'Sum_len_pt',\n",
       "       'Nanybase_pt', 'nr_recovered', 'Nseq_nr', 'Sum_len_nr', 'Nanybase_nr',\n",
       "       'Completed_Reads', 'Completed_Extending', 'Completed_Assembly',\n",
       "       'Completed_Slimming', 'Completed_Output', 'Warning_Look',\n",
       "       'Warning_Multi_Structure', 'Warning_Self_Loop', 'Info_Large_Repeats',\n",
       "       'Info_Disentangling_Failed', 'Assembly', 'Coverage_Kmer',\n",
       "       'Coverage_base', 'Run_Time', 'maxK', 'NRepeat_Pattern', 'NPath',\n",
       "       'Redo_FastPlast', 'error_pt', 'error_nr', 'SeqID', 'DataSource',\n",
       "       'ENASampleNum', 'ENAExpNumber', 'ENARunNumber', 'NCBI_TaxID',\n",
       "       'NCBI_sciname', 'Ref-SeqID', 'Ref-order', 'Ref-family',\n",
       "       'Blast-N_qseqid_match', 'Concat-len', 'Concat-%_mapped',\n",
       "       'Concat-len_with_Ns', 'ENA_SeqID'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "submit_pt.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26b6cbef",
   "metadata": {},
   "source": [
    "* The ENA_Submission_Commands_PAFTOL.txt file contains a WEBIN-CLI command for each sample annotation.\n",
    "* Command assumes that you move into the /EMBL_files/ folder to run it.\n",
    "* Launch command is now a shell script e.g.:\n",
    "    * cd /EMBL_files\n",
    "    * chmod 744 ../ENA_Submission_Commands_PAFTOL.sh\n",
    "    * ../ENA_Submission_Commands_${DataSource}.sh > ENA_Submission_Commands_${DataSource}.log 2>&1 &\n",
    "* Can validate samples (-validate flag) on Macbook but need to use JHI cluster to submit samples (-submit flag)\n",
    "* Should also use Slurm on the JHI cluster to submit:\n",
    "    * sbatch --wrap \"./ENA_Submission_Commands_${DataSource}_-submit.sh > ./ENA_Submission_Commands_${DataSource}_-submit.log 2>&1 \""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50fe6922",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
