{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# latest paftol export containing ENA sample accessions\n",
    "# https://www.ebi.ac.uk/ena/submit/report/swagger-ui.html#/samples/findAllUsingGET_13\n",
    "pft_file = '../PAFTOL_DB/2022-03-14_paftol_export_stats.csv'\n",
    "cptree_file = 'cpTree_v7/cpTree_v7_tree_clean.csv'\n",
    "DataSource = 'PAFTOL' \n",
    "org_dir = 'C:/Data/PAFTOL/Organelles/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3451 ['PAFTOL_004799', 'PAFTOL_000992']\n"
     ]
    }
   ],
   "source": [
    "## All Validated Samples\n",
    "R = pd.read_csv(cptree_file)\n",
    "R = R[R.DataSource==DataSource]\n",
    "subset = R.Sample_Name.to_list()\n",
    "print(R.shape[0],subset[:2])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load data and subset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "pdir = DataSource + '/'\n",
    "sdir = 'ENA_submissions/' + pdir\n",
    "if not os.path.exists(sdir):\n",
    "    os.makedirs(sdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9715\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>idPaftol</th>\n",
       "      <th>idSequencing</th>\n",
       "      <th>Sample_Name</th>\n",
       "      <th>Project</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>Taxonomical_Notes</th>\n",
       "      <th>SumContigLength</th>\n",
       "      <th>...</th>\n",
       "      <th>Coverage_Kmer</th>\n",
       "      <th>Coverage_base</th>\n",
       "      <th>Run_Time</th>\n",
       "      <th>maxK</th>\n",
       "      <th>NRepeat_Pattern</th>\n",
       "      <th>NPath</th>\n",
       "      <th>Redo_FastPlast</th>\n",
       "      <th>error_pt</th>\n",
       "      <th>error_nr</th>\n",
       "      <th>Geseq_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>961</td>\n",
       "      <td>PAFTOL_000961</td>\n",
       "      <td>Pilot</td>\n",
       "      <td>Ranunculales</td>\n",
       "      <td>Circaeasteraceae</td>\n",
       "      <td>Circaeaster</td>\n",
       "      <td>agrestis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>138795.0</td>\n",
       "      <td>...</td>\n",
       "      <td>59.4</td>\n",
       "      <td>90.7</td>\n",
       "      <td>329.43</td>\n",
       "      <td>105.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pis_000961</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4</td>\n",
       "      <td>2571</td>\n",
       "      <td>PAFTOL_002571</td>\n",
       "      <td>Connaraceae</td>\n",
       "      <td>Oxalidales</td>\n",
       "      <td>Connaraceae</td>\n",
       "      <td>Rourea</td>\n",
       "      <td>minor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>213954.0</td>\n",
       "      <td>...</td>\n",
       "      <td>18.5</td>\n",
       "      <td>60.2</td>\n",
       "      <td>747.57</td>\n",
       "      <td>105.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pis_002571</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows Ã— 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   idPaftol  idSequencing    Sample_Name      Project         order  \\\n",
       "0         2           961  PAFTOL_000961        Pilot  Ranunculales   \n",
       "1         4          2571  PAFTOL_002571  Connaraceae    Oxalidales   \n",
       "\n",
       "             family        genus   species Taxonomical_Notes  SumContigLength  \\\n",
       "0  Circaeasteraceae  Circaeaster  agrestis               NaN         138795.0   \n",
       "1       Connaraceae       Rourea     minor               NaN         213954.0   \n",
       "\n",
       "   ... Coverage_Kmer  Coverage_base  Run_Time   maxK  NRepeat_Pattern  NPath  \\\n",
       "0  ...          59.4           90.7    329.43  105.0              NaN    NaN   \n",
       "1  ...          18.5           60.2    747.57  105.0              NaN    NaN   \n",
       "\n",
       "   Redo_FastPlast  error_pt  error_nr  Geseq_name  \n",
       "0           False       NaN       NaN  Pis_000961  \n",
       "1           False       NaN       NaN  Pis_002571  \n",
       "\n",
       "[2 rows x 42 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load GetOrg result table\n",
    "Org_df = pd.read_csv(org_dir + DataSource + '/' + DataSource + '_Organelle_Recovery.csv').astype({'idSequencing':'int'})\n",
    "if DataSource == 'PAFTOL':\n",
    "    Org_df['Geseq_name'] = Org_df.Sample_Name.str.replace('PAFTOL_','Pis_')\n",
    "elif DataSource in ['GAP']:\n",
    "    Org_df['Geseq_name'] = Org_df.Sample_Name\n",
    "elif DataSource in ['SRA']:\n",
    "    sra2_pt = pd.read_csv('cpSRA_GetOrg/cpSRA_GetOrg_Organelle_Recovery.csv')\n",
    "    Org_df = pd.concat([Org_df,sra2_pt],ignore_index=True)\n",
    "    Org_df = Org_df.groupby('Sample_Name').head(1)\n",
    "    Org_df['Geseq_name'] = Org_df.Sample_Name\n",
    "print(Org_df.shape[0])\n",
    "Org_df[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3451 3451\n"
     ]
    }
   ],
   "source": [
    "if subset is not None:\n",
    "    Org_df = Org_df[Org_df.Sample_Name.isin(subset)]\n",
    "    print(Org_df.shape[0],Org_df.Sample_Name.nunique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_Name</th>\n",
       "      <th>idPaftol</th>\n",
       "      <th>DataSource</th>\n",
       "      <th>Project</th>\n",
       "      <th>Order</th>\n",
       "      <th>Family</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Species</th>\n",
       "      <th>Taxonomical_Notes</th>\n",
       "      <th>ENASampleNum</th>\n",
       "      <th>idSequencing</th>\n",
       "      <th>ExternalSequenceID</th>\n",
       "      <th>ENAExpNumber</th>\n",
       "      <th>ENARunNumber</th>\n",
       "      <th>NumReads</th>\n",
       "      <th>NumRecoveredGenes</th>\n",
       "      <th>SumContigLength</th>\n",
       "      <th>NCBI_TaxID</th>\n",
       "      <th>NCBI_sciname</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PAFTOL_000564</td>\n",
       "      <td>1</td>\n",
       "      <td>PAFTOL</td>\n",
       "      <td>Asteraceae</td>\n",
       "      <td>Asterales</td>\n",
       "      <td>Asteraceae</td>\n",
       "      <td>Gymnolaena</td>\n",
       "      <td>sp.</td>\n",
       "      <td>Initially recorded as Gymnolaena litoralis but...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>564</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PAFTOL_000961</td>\n",
       "      <td>2</td>\n",
       "      <td>PAFTOL</td>\n",
       "      <td>Pilot</td>\n",
       "      <td>Ranunculales</td>\n",
       "      <td>Circaeasteraceae</td>\n",
       "      <td>Circaeaster</td>\n",
       "      <td>agrestis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERS5501508</td>\n",
       "      <td>961</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERX4839538</td>\n",
       "      <td>ERR5033191</td>\n",
       "      <td>625228.0</td>\n",
       "      <td>332.0</td>\n",
       "      <td>138795.0</td>\n",
       "      <td>39288.0</td>\n",
       "      <td>Circaeaster agrestis</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Sample_Name  idPaftol DataSource     Project         Order  \\\n",
       "0  PAFTOL_000564         1     PAFTOL  Asteraceae     Asterales   \n",
       "1  PAFTOL_000961         2     PAFTOL       Pilot  Ranunculales   \n",
       "\n",
       "             Family        Genus   Species  \\\n",
       "0        Asteraceae   Gymnolaena       sp.   \n",
       "1  Circaeasteraceae  Circaeaster  agrestis   \n",
       "\n",
       "                                   Taxonomical_Notes ENASampleNum  \\\n",
       "0  Initially recorded as Gymnolaena litoralis but...          NaN   \n",
       "1                                                NaN   ERS5501508   \n",
       "\n",
       "   idSequencing ExternalSequenceID ENAExpNumber ENARunNumber  NumReads  \\\n",
       "0           564                NaN          NaN          NaN       NaN   \n",
       "1           961                NaN   ERX4839538   ERR5033191  625228.0   \n",
       "\n",
       "   NumRecoveredGenes  SumContigLength  NCBI_TaxID          NCBI_sciname  \n",
       "0                NaN              NaN         NaN                   NaN  \n",
       "1              332.0         138795.0     39288.0  Circaeaster agrestis  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load paftol export and merge ena sample IDs\n",
    "db = pd.read_csv(pft_file)\n",
    "db = db[db.idPaftol.notnull()].astype({'idPaftol':'int','idSequencing':'int'})\n",
    "db = db[['Sample_Name','idPaftol', 'DataSource', 'Project', 'Order', 'Family', 'Genus', 'Species', 'Taxonomical_Notes', \n",
    "         'ENASampleNum', 'idSequencing', 'ExternalSequenceID', 'ENAExpNumber',\n",
    "       'ENARunNumber', 'NumReads', 'NumRecoveredGenes', 'SumContigLength','NCBI_TaxID', 'NCBI_sciname']]\n",
    "#        'title', 'taxId','id', 'alias', 'firstCreated', 'firstPublic', 'releaseStatus', 'secondaryId', 'scientificName', 'commonName']]\n",
    "db[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'idPaftol': 0, 'idSequencing': 0, 'Sample_Name': 0, 'Project': 0, 'order': 0, 'family': 0, 'genus': 0, 'species': 0, 'Taxonomical_Notes': 2231, 'SumContigLength': 363, 'sci_name': 0, 'log_pt': 0, 'log_nr': 0, 'pt_recovered': 0, 'Nseq_pt': 0, 'Sum_len_pt': 0, 'Nanybase_pt': 0, 'nr_recovered': 0, 'Nseq_nr': 27, 'Sum_len_nr': 27, 'Nanybase_nr': 27, 'Completed_Reads': 0, 'Completed_Extending': 0, 'Completed_Assembly': 0, 'Completed_Slimming': 0, 'Completed_Output': 0, 'Warning_Look': 0, 'Warning_Multi_Structure': 0, 'Warning_Self_Loop': 0, 'Info_Large_Repeats': 0, 'Info_Disentangling_Failed': 0, 'Assembly': 6, 'Coverage_Kmer': 635, 'Coverage_base': 8, 'Run_Time': 5, 'maxK': 6, 'NRepeat_Pattern': 3380, 'NPath': 3422, 'Redo_FastPlast': 0, 'error_pt': 3450, 'error_nr': 3432, 'Geseq_name': 0, 'DataSource': 0, 'ENASampleNum': 47, 'ENAExpNumber': 0, 'ENARunNumber': 0, 'NCBI_TaxID': 172, 'NCBI_sciname': 172}\n",
      "3451 3451\n"
     ]
    }
   ],
   "source": [
    "Org_df = pd.merge(Org_df,\n",
    "         db[['Sample_Name','DataSource','ENASampleNum','ENAExpNumber','ENARunNumber','NCBI_TaxID','NCBI_sciname']],\n",
    "        how='left',on='Sample_Name').groupby('Sample_Name').head(1)\n",
    "print(Org_df.isna().sum().to_dict())\n",
    "print(Org_df.shape[0],Org_df.Sample_Name.nunique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "geseq_dir = sdir + 'fasta/'\n",
    "if not os.path.exists(geseq_dir):\n",
    "    os.makedirs(geseq_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Distribution of contigs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "Seq_ls = []\n",
    "for idx, row in Org_df.iterrows():\n",
    "    for record in SeqIO.parse(org_dir + DataSource + '/' + 'fasta_pt/' + row.Sample_Name + '_pt.fasta', \"fasta\"):\n",
    "        Seq_ls.append({'Sample_Name':row.Sample_Name,'SeqID':record.id,'SeqLen':len(record.seq)})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15178 1612\n",
      "{0.1: 469.0, 0.5: 2199.0, 0.9: 20300.4} 0.7155751745948082\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYQAAAEICAYAAABfz4NwAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8vihELAAAACXBIWXMAAAsTAAALEwEAmpwYAAAU2UlEQVR4nO3dfbBc9X3f8fenko0VbFwRzK0qaSycKg88TBxzg0nseu4McZFxpmIypiMPMXJCRh2KEydDpyOSzjj9Qy1OS2tDAhnFdhCJClawM9KEobFGyU2ahAcLh1gIWUEGFWQUZMdPyHUJot/+sT/Vy30Q0u7V3rur92tmZ89+z/md/X3vXulzzzm796aqkCTpH833BCRJC4OBIEkCDARJUmMgSJIAA0GS1BgIkiTAQJAkNQaCznhJ3pnkr5J8K8nXk/xlkh/vc58fTPIXczVHaRAWz/cEpPmU5Bzgj4AbgG3Aa4F/Drw4n/OS5oNHCDrT/SBAVd1TVS9X1Xer6nNV9UWAJD+fZF+SbyT54yRvPj4wybuTfKkdWfxmkj9L8guv9oRJfjjJznY0sj/Jv+pad1eS30pyf5IXkjyc5AdOR+PSVAaCznR/C7ycZEuS9yRZenxFkquBXwV+BngT8D+Be9q684DPAP8eOA/4MvCOV3uyJGcDO4H/DpwPvB+4I8lFXZu9H/gPwFLgALCpvxalk2Mg6IxWVd8G3gkU8DvAV5PsSDIG/GvgP1XVvqo6BvxH4K3tKOEq4Imquq+qXgI+BvzdSTzlTwMHq+p3q+pYVX2BTrC8r2ubz1bVI+05twJvnZNmpVdhIOiM1/7D/2BVrQAuBv4pnf/g3wx8PMk3k3wT+DoQYHnb5tmufVT34xN4M/D24/ts+70W+Cdd23QHy/8GXt9ja9Ip8aKy1KWqvpTkLjpHB88Cm6pq69TtkqwGVnY9TvfjE3gW+LOqevfczFiaOx4h6IzWLvDelGRFe7ySzjn8h4DfBm4+fn4/yRuTXNOG3g9clORnkiwGfolX/pTfhuR13Tc672j6wSQfSPKadvvxJD8ygHalEzIQdKZ7AXg78HCS79AJgseBm6rqD4GPAvcm+Xarvwegqr4GXAPcAvw9sBr4yyn7/knguzPc/gWwDniOzumhjwJnnb4WpZMT/0CONDeSTAK/X1WfmO+5SL3wCEGSBBgIkqTGU0aSJMAjBElSM7SfQzjvvPNq1apVPY39zne+w9lnnz23E1ogRrk3GO3+7G04DVtvjz766Neq6k0zrRvaQFi1ahW7d+/uaezk5CQTExNzO6EFYpR7g9Huz96G07D1luR/zbbOU0aSJMBAkCQ1BoIkCTAQJEmNgSBJAgwESVJjIEiSAANBktQYCJIkYIg/qdyPPV/5Fh/ceP+0+sFb3jsPs5GkhcEjBEkSYCBIkhoDQZIEGAiSpMZAkCQBBoIkqTEQJEmAgSBJagwESRJgIEiSGgNBkgQYCJKkxkCQJAEGgiSpMRAkSYCBIElqDARJEmAgSJKaVw2EJJ9KciTJ4121c5PsTPJku1/ate7mJAeS7E9yZVf90iR72rrbkqTVz0ry6VZ/OMmqOe5RknQSTuYI4S5gzZTaRmBXVa0GdrXHJLkQWAdc1MbckWRRG3MnsAFY3W7H93k98I2q+mfAfwM+2mszkqTevWogVNWfA1+fUl4LbGnLW4Cru+r3VtWLVfU0cAC4LMky4JyqerCqCrh7ypjj+7oPuOL40YMkaXAW9zhurKoOA1TV4STnt/py4KGu7Q612ktteWr9+Jhn276OJfkW8P3A16Y+aZINdI4yGBsbY3JysrfJL4GbLjk2rd7r/haSo0ePjkQfsxnl/uxtOI1Sb70Gwmxm+sm+TlA/0ZjpxarNwGaA8fHxmpiY6GGKcPvW7dy6Z3rrB6/tbX8LyeTkJL1+XYbBKPdnb8NplHrr9V1Gz7fTQLT7I61+CFjZtd0K4LlWXzFD/RVjkiwG3sj0U1SSpNOs10DYAaxvy+uB7V31de2dQxfQuXj8SDu99EKSy9v1geumjDm+r/cBf9KuM0iSBuhVTxkluQeYAM5Lcgj4CHALsC3J9cAzwDUAVbU3yTbgCeAYcGNVvdx2dQOddywtAR5oN4BPAr+X5ACdI4N1c9KZJOmUvGogVNX7Z1l1xSzbbwI2zVDfDVw8Q/3/0AJFkjR//KSyJAkwECRJjYEgSQIMBElSYyBIkgADQZLUGAiSJMBAkCQ1BoIkCTAQJEmNgSBJAgwESVJjIEiSAANBktQYCJIkwECQJDUGgiQJMBAkSY2BIEkCDARJUmMgSJIAA0GS1BgIkiTAQJAkNQaCJAkwECRJjYEgSQIMBElS01cgJPmVJHuTPJ7kniSvS3Jukp1Jnmz3S7u2vznJgST7k1zZVb80yZ627rYk6WdekqRT13MgJFkO/BIwXlUXA4uAdcBGYFdVrQZ2tcckubCtvwhYA9yRZFHb3Z3ABmB1u63pdV6SpN70e8poMbAkyWLg+4DngLXAlrZ+C3B1W14L3FtVL1bV08AB4LIky4BzqurBqirg7q4xkqQBWdzrwKr6SpL/AjwDfBf4XFV9LslYVR1u2xxOcn4bshx4qGsXh1rtpbY8tT5Nkg10jiQYGxtjcnKyp7mPLYGbLjk2rd7r/haSo0ePjkQfsxnl/uxtOI1Sbz0HQrs2sBa4APgm8AdJfvZEQ2ao1Qnq04tVm4HNAOPj4zUxMXEKM/6e27du59Y901s/eG1v+1tIJicn6fXrMgxGuT97G06j1Fs/p4x+Cni6qr5aVS8BnwV+Eni+nQai3R9p2x8CVnaNX0HnFNOhtjy1LkkaoH4C4Rng8iTf194VdAWwD9gBrG/brAe2t+UdwLokZyW5gM7F40fa6aUXklze9nNd1xhJ0oD0cw3h4ST3AV8AjgF/Ted0zuuBbUmupxMa17Tt9ybZBjzRtr+xql5uu7sBuAtYAjzQbpKkAeo5EACq6iPAR6aUX6RztDDT9puATTPUdwMX9zMXSVJ//KSyJAkwECRJjYEgSQIMBElSYyBIkgADQZLUGAiSJMBAkCQ1BoIkCTAQJEmNgSBJAgwESVJjIEiSAANBktQYCJIkwECQJDUGgiQJMBAkSY2BIEkCDARJUmMgSJIAA0GS1BgIkiTAQJAkNQaCJAkwECRJjYEgSQIMBElS01cgJPnHSe5L8qUk+5L8RJJzk+xM8mS7X9q1/c1JDiTZn+TKrvqlSfa0dbclST/zkiSdun6PED4O/I+q+mHgR4F9wEZgV1WtBna1xyS5EFgHXASsAe5Isqjt505gA7C63db0OS9J0inqORCSnAO8C/gkQFX9Q1V9E1gLbGmbbQGubstrgXur6sWqeho4AFyWZBlwTlU9WFUF3N01RpI0IIv7GPsW4KvA7yb5UeBR4MPAWFUdBqiqw0nOb9svBx7qGn+o1V5qy1Pr0yTZQOdIgrGxMSYnJ3ua+NgSuOmSY9Pqve5vITl69OhI9DGbUe7P3obTKPXWTyAsBt4G/GJVPZzk47TTQ7OY6bpAnaA+vVi1GdgMMD4+XhMTE6c04eNu37qdW/dMb/3gtb3tbyGZnJyk16/LMBjl/uxtOI1Sb/1cQzgEHKqqh9vj++gExPPtNBDt/kjX9iu7xq8Anmv1FTPUJUkD1HMgVNXfAc8m+aFWugJ4AtgBrG+19cD2trwDWJfkrCQX0Ll4/Eg7vfRCksvbu4uu6xojSRqQfk4ZAfwisDXJa4GngJ+jEzLbklwPPANcA1BVe5NsoxMax4Abq+rltp8bgLuAJcAD7SZJGqC+AqGqHgPGZ1h1xSzbbwI2zVDfDVzcz1wkSf3xk8qSJMBAkCQ1BoIkCTAQJEmNgSBJAgwESVJjIEiSAANBktQYCJIkwECQJDUGgiQJMBAkSY2BIEkCDARJUmMgSJIAA0GS1BgIkiTAQJAkNQaCJAkwECRJjYEgSQIMBElSYyBIkgADQZLUGAiSJMBAkCQ1BoIkCTAQJElN34GQZFGSv07yR+3xuUl2Jnmy3S/t2vbmJAeS7E9yZVf90iR72rrbkqTfeUmSTs1cHCF8GNjX9XgjsKuqVgO72mOSXAisAy4C1gB3JFnUxtwJbABWt9uaOZiXJOkU9BUISVYA7wU+0VVeC2xpy1uAq7vq91bVi1X1NHAAuCzJMuCcqnqwqgq4u2uMJGlAFvc5/mPAvwPe0FUbq6rDAFV1OMn5rb4ceKhru0Ot9lJbnlqfJskGOkcSjI2NMTk52dOkx5bATZccm1bvdX8LydGjR0eij9mMcn/2NpxGqbeeAyHJTwNHqurRJBMnM2SGWp2gPr1YtRnYDDA+Pl4TEyfztNPdvnU7t+6Z3vrBa3vb30IyOTlJr1+XYTDK/dnbcBql3vo5QngH8C+TXAW8Djgnye8DzydZ1o4OlgFH2vaHgJVd41cAz7X6ihnqkqQB6vkaQlXdXFUrqmoVnYvFf1JVPwvsANa3zdYD29vyDmBdkrOSXEDn4vEj7fTSC0kub+8uuq5rjCRpQPq9hjCTW4BtSa4HngGuAaiqvUm2AU8Ax4Abq+rlNuYG4C5gCfBAu0mSBmhOAqGqJoHJtvz3wBWzbLcJ2DRDfTdw8VzMRZLUGz+pLEkCDARJUmMgSJIAA0GS1BgIkiTAQJAkNQaCJAkwECRJjYEgSQIMBElSYyBIkgADQZLUGAiSJMBAkCQ1BoIkCTAQJEmNgSBJAgwESVJjIEiSAANBktQYCJIkwECQJDUGgiQJMBAkSY2BIEkCDARJUmMgSJIAA0GS1PQcCElWJvnTJPuS7E3y4VY/N8nOJE+2+6VdY25OciDJ/iRXdtUvTbKnrbstSfprS5J0qvo5QjgG3FRVPwJcDtyY5EJgI7CrqlYDu9pj2rp1wEXAGuCOJIvavu4ENgCr221NH/OSJPWg50CoqsNV9YW2/AKwD1gOrAW2tM22AFe35bXAvVX1YlU9DRwALkuyDDinqh6sqgLu7hojSRqQxXOxkySrgB8DHgbGquowdEIjyflts+XAQ13DDrXaS215an2m59lA50iCsbExJicne5rv2BK46ZJj0+q97m8hOXr06Ej0MZtR7s/ehtMo9dZ3ICR5PfAZ4Jer6tsnOP0/04o6QX16sWozsBlgfHy8JiYmTnm+ALdv3c6te6a3fvDa3va3kExOTtLr12UYjHJ/9jacRqm3vt5llOQ1dMJga1V9tpWfb6eBaPdHWv0QsLJr+ArguVZfMUNdkjRA/bzLKMAngX1V9V+7Vu0A1rfl9cD2rvq6JGcluYDOxeNH2umlF5Jc3vZ5XdcYSdKA9HPK6B3AB4A9SR5rtV8FbgG2JbkeeAa4BqCq9ibZBjxB5x1KN1bVy23cDcBdwBLggXaTJA1Qz4FQVX/BzOf/Aa6YZcwmYNMM9d3Axb3ORZLUPz+pLEkC5uhtp6Ni1cb7Z6wfvOW9A56JJA2eRwiSJMBAkCQ1BoIkCTAQJEmNgSBJAgwESVJjIEiSAANBktQYCJIkwECQJDUGgiQJMBAkSY2BIEkCDARJUmMgSJIAA0GS1BgIkiTAQJAkNf4JzZMw25/WBP+8pqTR4RGCJAkwECRJjYEgSQIMBElS40XlPs12wdmLzZKGjYEgLQCrNt7PTZcc44NTfsDwBwsNkoFwmnjkoLng95EGacEEQpI1wMeBRcAnquqWeZ7SaeE/cEkL1YIIhCSLgN8C3g0cAj6fZEdVPTG/MxucE334bSYGiKS5tiACAbgMOFBVTwEkuRdYC5wxgXCqZguQmc5D96KXwJmvUJuvo65T7ReGv2eNtlTVfM+BJO8D1lTVL7THHwDeXlUfmrLdBmBDe/hDwP4en/I84Gs9jl3oRrk3GO3+7G04DVtvb66qN820YqEcIWSG2rSkqqrNwOa+nyzZXVXj/e5nIRrl3mC0+7O34TRKvS2UD6YdAlZ2PV4BPDdPc5GkM9JCCYTPA6uTXJDktcA6YMc8z0mSzigL4pRRVR1L8iHgj+m87fRTVbX3ND5l36edFrBR7g1Guz97G04j09uCuKgsSZp/C+WUkSRpnhkIkiTgDAyEJGuS7E9yIMnG+Z7PbJIcTLInyWNJdrfauUl2Jnmy3S/t2v7m1tP+JFd21S9t+zmQ5LYkafWzkny61R9Osuo09/OpJEeSPN5VG0g/Sda353gyyfoB9fbrSb7SXr/Hklw1bL0lWZnkT5PsS7I3yYdbfVRet9n6G/rXrmdVdcbc6Fyw/jLwFuC1wN8AF873vGaZ60HgvCm13wA2tuWNwEfb8oWtl7OAC1qPi9q6R4CfoPNZjweA97T6vwF+uy2vAz59mvt5F/A24PFB9gOcCzzV7pe25aUD6O3XgX87w7ZD0xuwDHhbW34D8Ldt/qPyus3W39C/dr3ezrQjhP//KzKq6h+A478iY1isBba05S3A1V31e6vqxap6GjgAXJZkGXBOVT1Yne/Cu6eMOb6v+4Arjv9UczpU1Z8DX59SHkQ/VwI7q+rrVfUNYCewZgC9zWZoequqw1X1hbb8ArAPWM7ovG6z9TeboeqvF2daICwHnu16fIgTfwPMpwI+l+TRdH5lB8BYVR2GzjczcH6rz9bX8rY8tf6KMVV1DPgW8P2noY8TGUQ/8/mafyjJF9sppeOnVYayt3aq48eAhxnB121KfzBCr92pONMC4aR+RcYC8Y6qehvwHuDGJO86wbaz9XWifhfy12Iu+5mvPu8EfgB4K3AYuLXVh663JK8HPgP8clV9+0SbzjKfBdsbzNjfyLx2p+pMC4Sh+RUZVfVcuz8C/CGd013Pt8NT2v2RtvlsfR1qy1PrrxiTZDHwRk7+tMdcGUQ/8/KaV9XzVfVyVf1f4HfovH6vmOeU+SzI3pK8hs5/llur6rOtPDKv20z9jcpr15P5vogxyBudT2Y/ReeC0PGLyhfN97xmmOfZwBu6lv+KzvnF/8wrL+b9Rlu+iFde7HqK713s+jxwOd+72HVVq9/IKy92bRtAX6t45YXX094PnYt2T9O5cLe0LZ87gN6WdS3/Cp1zz0PVW5vH3cDHptRH4nU7QX9D/9r1/DWZ7wkMvGG4is67Cb4M/Np8z2eWOb6lfeP9DbD3+DzpnHvcBTzZ7s/tGvNrraf9tHc4tPo48Hhb95t879PprwP+gM6FsUeAt5zmnu6hc/j9Ep2fjq4fVD/Az7f6AeDnBtTb7wF7gC/S+b1cy4atN+CddE5jfBF4rN2uGqHXbbb+hv616/Xmr66QJAFn3jUESdIsDARJEmAgSJIaA0GSBBgIkqTGQJAkAQaCJKn5f+dAszSpo8NPAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "Seq_df = pd.DataFrame(Seq_ls)\n",
    "print(Seq_df.shape[0], Seq_df.Sample_Name.nunique())\n",
    "print(Seq_df.SeqLen.quantile([.1,.5,.9]).round(1).to_dict(),(Seq_df.SeqLen<5000).sum()/Seq_df.shape[0])\n",
    "Seq_df.head(2)\n",
    "Seq_df.hist(bins=50);"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Whole plastomes\n",
    "### Annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "337 samples 280 genera\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>idPaftol</th>\n",
       "      <th>idSequencing</th>\n",
       "      <th>Sample_Name</th>\n",
       "      <th>Project</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>Taxonomical_Notes</th>\n",
       "      <th>SumContigLength</th>\n",
       "      <th>...</th>\n",
       "      <th>Redo_FastPlast</th>\n",
       "      <th>error_pt</th>\n",
       "      <th>error_nr</th>\n",
       "      <th>Geseq_name</th>\n",
       "      <th>DataSource</th>\n",
       "      <th>ENASampleNum</th>\n",
       "      <th>ENAExpNumber</th>\n",
       "      <th>ENARunNumber</th>\n",
       "      <th>NCBI_TaxID</th>\n",
       "      <th>NCBI_sciname</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>961</td>\n",
       "      <td>PAFTOL_000961</td>\n",
       "      <td>Pilot</td>\n",
       "      <td>Ranunculales</td>\n",
       "      <td>Circaeasteraceae</td>\n",
       "      <td>Circaeaster</td>\n",
       "      <td>agrestis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>138795.0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pis_000961</td>\n",
       "      <td>PAFTOL</td>\n",
       "      <td>ERS5501508</td>\n",
       "      <td>ERX4839538</td>\n",
       "      <td>ERR5033191</td>\n",
       "      <td>39288.0</td>\n",
       "      <td>Circaeaster agrestis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>11</td>\n",
       "      <td>12989</td>\n",
       "      <td>PAFTOL_012989</td>\n",
       "      <td>Achariaceae</td>\n",
       "      <td>Malpighiales</td>\n",
       "      <td>Achariaceae</td>\n",
       "      <td>Gynocardia</td>\n",
       "      <td>odorata</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pis_012989</td>\n",
       "      <td>PAFTOL</td>\n",
       "      <td>ERS8701094</td>\n",
       "      <td>ERX8772557</td>\n",
       "      <td>ERR9229774</td>\n",
       "      <td>124848.0</td>\n",
       "      <td>Gynocardia odorata</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows Ã— 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   idPaftol  idSequencing    Sample_Name      Project         order  \\\n",
       "0         2           961  PAFTOL_000961        Pilot  Ranunculales   \n",
       "6        11         12989  PAFTOL_012989  Achariaceae  Malpighiales   \n",
       "\n",
       "             family        genus   species Taxonomical_Notes  SumContigLength  \\\n",
       "0  Circaeasteraceae  Circaeaster  agrestis               NaN         138795.0   \n",
       "6       Achariaceae   Gynocardia   odorata               NaN              NaN   \n",
       "\n",
       "   ... Redo_FastPlast  error_pt  error_nr  Geseq_name  DataSource  \\\n",
       "0  ...          False       NaN       NaN  Pis_000961      PAFTOL   \n",
       "6  ...          False       NaN       NaN  Pis_012989      PAFTOL   \n",
       "\n",
       "   ENASampleNum  ENAExpNumber  ENARunNumber  NCBI_TaxID          NCBI_sciname  \n",
       "0    ERS5501508    ERX4839538    ERR5033191     39288.0  Circaeaster agrestis  \n",
       "6    ERS8701094    ERX8772557    ERR9229774    124848.0    Gynocardia odorata  \n",
       "\n",
       "[2 rows x 48 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "whole_pt = Org_df[(Org_df.Assembly.isin(['circular genome','circular genome with gaps']) & (Org_df.Sum_len_pt>80000))]\n",
    "print(whole_pt.shape[0],'samples',whole_pt.genus.nunique(),'genera')\n",
    "whole_pt[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Output fasta files \n",
    "for idx, row in whole_pt.iterrows():\n",
    "    seqs = list(SeqIO.parse(org_dir + DataSource + '/' + 'fasta_pt/' + row.Sample_Name + '_pt.fasta',format='fasta'))\n",
    "    # Verif only 1 sequence\n",
    "    if len(seqs) > 1:\n",
    "        print('error, more than 1 sequence')\n",
    "        pass\n",
    "    seq = seqs[0]\n",
    "    print(row.Sample_Name, row.Geseq_name, row.family, row.sci_name,len(seq),seq.seq.count('N'))\n",
    "    seq.id = row.Geseq_name\n",
    "    seq.description = ''\n",
    "    SeqIO.write(seq, geseq_dir + row.Geseq_name + '-circular.fasta',format='fasta')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Submit files to GeSeq https://chlorobox.mpimp-golm.mpg.de/geseq.html and extract zip file in wdir/GeSeq_output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "whole_pt.to_csv(sdir + DataSource + '_wholePT.csv',index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Contigs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1275 samples 1071 genera\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Assembly</th>\n",
       "      <th>Completed_Assembly</th>\n",
       "      <th>Completed_Extending</th>\n",
       "      <th>Completed_Output</th>\n",
       "      <th>Completed_Reads</th>\n",
       "      <th>Completed_Slimming</th>\n",
       "      <th>Coverage_Kmer</th>\n",
       "      <th>Coverage_base</th>\n",
       "      <th>Info_Disentangling_Failed</th>\n",
       "      <th>Info_Large_Repeats</th>\n",
       "      <th>...</th>\n",
       "      <th>sci_name</th>\n",
       "      <th>species</th>\n",
       "      <th>sum_len</th>\n",
       "      <th>Geseq_name</th>\n",
       "      <th>DataSource</th>\n",
       "      <th>ENASampleNum</th>\n",
       "      <th>ENAExpNumber</th>\n",
       "      <th>ENARunNumber</th>\n",
       "      <th>NCBI_TaxID</th>\n",
       "      <th>NCBI_sciname</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>scaffold</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>153.2</td>\n",
       "      <td>492.1</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>Abelmoschus esculentus</td>\n",
       "      <td>esculentus</td>\n",
       "      <td>260382.0</td>\n",
       "      <td>SRR5819803</td>\n",
       "      <td>SRA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SRR5819803</td>\n",
       "      <td>455045.0</td>\n",
       "      <td>Abelmoschus esculentus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>scaffold</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>99.4</td>\n",
       "      <td>569.5</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>Abronia maritima</td>\n",
       "      <td>maritima</td>\n",
       "      <td>258912.0</td>\n",
       "      <td>SRR6435332</td>\n",
       "      <td>SRA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SRR6435332</td>\n",
       "      <td>2067806.0</td>\n",
       "      <td>Abronia maritima</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows Ã— 46 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Assembly  Completed_Assembly  Completed_Extending  Completed_Output  \\\n",
       "0  scaffold                True                 True              True   \n",
       "1  scaffold                True                 True              True   \n",
       "\n",
       "   Completed_Reads  Completed_Slimming  Coverage_Kmer  Coverage_base  \\\n",
       "0             True                True          153.2          492.1   \n",
       "1            False                True           99.4          569.5   \n",
       "\n",
       "   Info_Disentangling_Failed  Info_Large_Repeats  ...                sci_name  \\\n",
       "0                       True               False  ...  Abelmoschus esculentus   \n",
       "1                       True               False  ...        Abronia maritima   \n",
       "\n",
       "      species   sum_len  Geseq_name  DataSource  ENASampleNum  ENAExpNumber  \\\n",
       "0  esculentus  260382.0  SRR5819803         SRA           NaN           NaN   \n",
       "1    maritima  258912.0  SRR6435332         SRA           NaN           NaN   \n",
       "\n",
       "   ENARunNumber NCBI_TaxID            NCBI_sciname  \n",
       "0    SRR5819803   455045.0  Abelmoschus esculentus  \n",
       "1    SRR6435332  2067806.0        Abronia maritima  \n",
       "\n",
       "[2 rows x 46 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "contigs_pt = Org_df[(Org_df.Assembly.isin(['circular genome','circular genome with gaps'])==False) &\n",
    "                   (Org_df.pt_recovered==True)]\n",
    "print(contigs_pt.shape[0],'samples',contigs_pt.genus.nunique(),'genera')\n",
    "contigs_pt[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Output fasta files\n",
    "min_len = 50000\n",
    "contigs_pt['Np_contigs'] = 0; contigs_pt['Sump_contigs'] = 0;\n",
    "for idx, row in contigs_pt.iterrows():\n",
    "    contigx = 1\n",
    "    for seq in SeqIO.parse(org_dir + DataSource + '/' + 'fasta_pt/' + row.Sample_Name + '_pt.fasta',format='fasta'):\n",
    "        # if longer than\n",
    "        if len(seq) > min_len:\n",
    "            if '--' in seq.id:\n",
    "                seq.description = '' # seq.id.split('--')[0] + ' contig_info:' + seq.id.split('--')[1] + ', length:' + str(len(seq.seq))\n",
    "                seq.id = row.Geseq_name + '-' + str(contigx)\n",
    "            elif 'circular' in seq.id:\n",
    "                seq.description = 'circular' + ' contig_info:' + seq.id.split('-')[0] + ', length:' + str(len(seq.seq))\n",
    "                seq.id = row.Geseq_name + '-' + 'circular'\n",
    "            SeqIO.write(seq, geseq_dir + seq.id + '.fasta',format='fasta')\n",
    "            contigs_pt.loc[idx,'Np_contigs'] += 1; contigs_pt.loc[idx,'Sump_contigs'] += len(seq)\n",
    "            contigx += 1\n",
    "    if contigs_pt.loc[idx,'Np_contigs']>0:\n",
    "        print(row.Sample_Name, row.Geseq_name, row.family, row.sci_name,row.Sum_len_pt,\n",
    "              contigs_pt.loc[idx,'Np_contigs'],contigs_pt.loc[idx,'Sump_contigs'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "488 456 samples 392 genera\n"
     ]
    }
   ],
   "source": [
    "contigs_pt = contigs_pt[contigs_pt.Np_contigs>0]\n",
    "print(contigs_pt.Np_contigs.sum(), contigs_pt.shape[0],'samples',contigs_pt.genus.nunique(),'genera')\n",
    "contigs_pt.to_csv(sdir + DataSource + '_contigsPT.csv',index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ENA submission\n",
    "https://ena-docs.readthedocs.io/en/latest/submit/fileprep/assembly.html?highlight=chloroplast#\n",
    "\n",
    "cd C:\\Data\\PAFTOL\\cpDNA\\ENA_submissions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "paper_title = 'Building the largest, curated plastid genome dataset and phylogeny by leveraging nuclear-oriented sequence data'\n",
    "paper_authors = 'KEVIN LEEMPOEL1, PAUL BAILEY1, ELAINE FRANCOSO1, BERTA GALLEGO1, De-Zhu Li2,3,4, Hong-Tao Li2,3, OLIVIER MAURIN1, SHYAMALI ROY1,' +  \\\n",
    "            'Jun-Bo Yang2,3, Ting-Shuang Yi2,3, ALEXANDRE R. ZUNTINI1, WOLF L. EISERHARDT1,5,' +  \\\n",
    "            'PAUL J. KERSEY1, ILIA J. LEITCH1, FELIX FOREST1 AND WILLIAM J. BAKER1'\n",
    "paper_journal = 'Unpublished'\n",
    "paper_id = 'NA'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "337\n",
      "331\n",
      "317\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>idPaftol</th>\n",
       "      <th>idSequencing</th>\n",
       "      <th>Sample_Name</th>\n",
       "      <th>Project</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>Taxonomical_Notes</th>\n",
       "      <th>SumContigLength</th>\n",
       "      <th>...</th>\n",
       "      <th>Redo_FastPlast</th>\n",
       "      <th>error_pt</th>\n",
       "      <th>error_nr</th>\n",
       "      <th>Geseq_name</th>\n",
       "      <th>DataSource</th>\n",
       "      <th>ENASampleNum</th>\n",
       "      <th>ENAExpNumber</th>\n",
       "      <th>ENARunNumber</th>\n",
       "      <th>NCBI_TaxID</th>\n",
       "      <th>NCBI_sciname</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>961</td>\n",
       "      <td>PAFTOL_000961</td>\n",
       "      <td>Pilot</td>\n",
       "      <td>Ranunculales</td>\n",
       "      <td>Circaeasteraceae</td>\n",
       "      <td>Circaeaster</td>\n",
       "      <td>agrestis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>138795.0</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pis_000961</td>\n",
       "      <td>PAFTOL</td>\n",
       "      <td>ERS5501508</td>\n",
       "      <td>ERX4839538</td>\n",
       "      <td>ERR5033191</td>\n",
       "      <td>39288.0</td>\n",
       "      <td>Circaeaster agrestis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>11</td>\n",
       "      <td>12989</td>\n",
       "      <td>PAFTOL_012989</td>\n",
       "      <td>Achariaceae</td>\n",
       "      <td>Malpighiales</td>\n",
       "      <td>Achariaceae</td>\n",
       "      <td>Gynocardia</td>\n",
       "      <td>odorata</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pis_012989</td>\n",
       "      <td>PAFTOL</td>\n",
       "      <td>ERS8701094</td>\n",
       "      <td>ERX8772557</td>\n",
       "      <td>ERR9229774</td>\n",
       "      <td>124848.0</td>\n",
       "      <td>Gynocardia odorata</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows Ã— 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   idPaftol  idSequencing    Sample_Name      Project         order  \\\n",
       "0         2           961  PAFTOL_000961        Pilot  Ranunculales   \n",
       "1        11         12989  PAFTOL_012989  Achariaceae  Malpighiales   \n",
       "\n",
       "             family        genus   species Taxonomical_Notes  SumContigLength  \\\n",
       "0  Circaeasteraceae  Circaeaster  agrestis               NaN         138795.0   \n",
       "1       Achariaceae   Gynocardia   odorata               NaN              NaN   \n",
       "\n",
       "   ... Redo_FastPlast  error_pt  error_nr  Geseq_name  DataSource  \\\n",
       "0  ...          False       NaN       NaN  Pis_000961      PAFTOL   \n",
       "1  ...          False       NaN       NaN  Pis_012989      PAFTOL   \n",
       "\n",
       "   ENASampleNum  ENAExpNumber  ENARunNumber  NCBI_TaxID          NCBI_sciname  \n",
       "0    ERS5501508    ERX4839538    ERR5033191     39288.0  Circaeaster agrestis  \n",
       "1    ERS8701094    ERX8772557    ERR9229774    124848.0    Gynocardia odorata  \n",
       "\n",
       "[2 rows x 48 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "whole_pt = pd.read_csv(sdir + DataSource + '_wholePT.csv')\n",
    "print(whole_pt.shape[0])\n",
    "whole_pt = whole_pt[(whole_pt.ENASampleNum.notnull()) & (whole_pt.ENARunNumber.notnull())]\n",
    "print(whole_pt.shape[0])\n",
    "whole_pt = whole_pt[whole_pt.NCBI_sciname.notnull()]\n",
    "print(whole_pt.shape[0])\n",
    "whole_pt.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Rename and modify genbank files, convert to embl flat file\n",
    "if not os.path.exists(sdir + '/Genbank_files/'):\n",
    "    os.makedirs(sdir + '/Genbank_files/')\n",
    "gb_files = [file for file in os.listdir(sdir + 'geseq/') if file.endswith('.gb')]\n",
    "print(len(gb_files))\n",
    "for idx, row in whole_pt.iterrows():\n",
    "    idx_file = [i for i, x in enumerate([row.Geseq_name in file for file in gb_files]) if x]\n",
    "    for index, record in enumerate(SeqIO.parse(sdir + 'geseq/' + gb_files[idx_file[0]], \"genbank\")):\n",
    "        try:\n",
    "            print(record.id)\n",
    "            record.id = row.ENASampleNum + '_pt'\n",
    "            record.name = row.ENARunNumber + '_pt'\n",
    "            record.description = row.ENARunNumber + ' (' + row.Sample_Name + ') chloroplast, complete genome'\n",
    "            record.version = row.ENARunNumber + '.1'\n",
    "            record.annotations['source'] = row.ENASampleNum + ', ' + row.sci_name + ', isolate'\n",
    "            record.annotations['accessions'] = row.ENASampleNum\n",
    "            record.annotations['organism'] = row.NCBI_sciname\n",
    "            record.annotations[\"data_file_division\"]=\"PLN\";\n",
    "            record.annotations.pop('taxonomy')\n",
    "            record.annotations['references'][0].title = paper_title    \n",
    "            record.annotations['references'][0].authors = paper_authors    \n",
    "            record.annotations['references'][0].journal = paper_journal      \n",
    "            record.annotations['references'][0].pubmed_id = paper_id    \n",
    "            record.features[0].qualifiers['organism'] = row.NCBI_sciname\n",
    "            record.features[0].qualifiers['db_xref'] = ['taxon:' + str(int(row['NCBI_TaxID']))]\n",
    "            SeqIO.write(record,sdir + '/Genbank_files/' + row.Sample_Name + '_pt.gb',format='genbank')\n",
    "        except:\n",
    "            print('issue with ',gb_files[idx_file[0]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def del_features(records, feat_del_ls = ['translation','info','annotator']):\n",
    "    for feature in records.features:\n",
    "        for feat_to_del in feat_del_ls:\n",
    "            if feat_to_del in feature.qualifiers:\n",
    "                del feature.qualifiers[feat_to_del]\n",
    "    return records\n",
    "def del_fragment(records, ls_types = ['gene','CDS','rRNA']):\n",
    "    # list gene names\n",
    "    ls_feat = []\n",
    "    for feature in records.features:\n",
    "        if feature.type in ls_types:\n",
    "            ls_feat.append(feature.qualifiers['gene'][0])\n",
    "    # remove fragments from list\n",
    "#     print(len(ls_feat),end='>')\n",
    "    ls_feat = [ifeat for ifeat in ls_feat if 'fragment' not in ifeat]\n",
    "#     print(len(ls_feat))\n",
    "    \n",
    "    # remove fragment feature if full feature is present\n",
    "    rm_count = 0\n",
    "    for feature in records.features:\n",
    "#         print(feature.type, feature.qualifiers)\n",
    "        if feature.type in ls_types:\n",
    "            if 'fragment' in feature.qualifiers['gene'][0]:\n",
    "#                 print(feature.type, feature.qualifiers,end=':')\n",
    "                if feature.qualifiers['gene'][0].split('-')[0] in ls_feat:\n",
    "                    rm_count += 1\n",
    "#                     print('true')\n",
    "                    records.features.remove(feature)\n",
    "    print('removed',rm_count,'partial features')\n",
    "    return records\n",
    "def del_longfeat(records, ls_types = ['gene','CDS','rRNA'], len_thrs = 5000):\n",
    "    for feature in records.features:\n",
    "        if feature.type in ls_types:\n",
    "            if len(feature)>len_thrs:\n",
    "                print('removing ',feature.qualifiers['gene'],len(feature))\n",
    "                records.features.remove(feature)\n",
    "    return records\n",
    "def gzip_file(in_file):\n",
    "    import gzip\n",
    "    import shutil\n",
    "    f_gz = in_file + '.gz'\n",
    "    with open(in_file, 'rb') as f_in:\n",
    "        with gzip.open(f_gz, 'wb') as f_out:\n",
    "            shutil.copyfileobj(f_in, f_out)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Whole plastomes\n",
    "if not os.path.exists(sdir + '/EMBL_files/'):\n",
    "    os.makedirs(sdir + '/EMBL_files/')\n",
    "commands = []\n",
    "for idx, row in whole_pt.iterrows():\n",
    "    ### Convert .gb to EMBL flat files\n",
    "    file_gb = sdir + '/Genbank_files/' + row.Sample_Name + '_pt.gb'\n",
    "    records = list(SeqIO.parse(file_gb, \"genbank\"))[0]\n",
    "    print(row.Sample_Name,len(records))\n",
    "    records_clean = del_fragment(records); records_clean = del_fragment(records_clean); records_clean = del_fragment(records_clean)\n",
    "    records_clean = del_longfeat(records_clean,len_thrs=10000);\n",
    "    file_embl = sdir + '/EMBL_files/' + row.Sample_Name + '.embl'\n",
    "    SeqIO.write(records_clean, file_gb, \"genbank\")\n",
    "    SeqIO.write(records_clean, file_embl, \"embl\")\n",
    "    gzip_file(file_embl)\n",
    "    ### Chromosome list file\n",
    "    # https://ena-docs.readthedocs.io/en/latest/submit/fileprep/assembly.html?highlight=chloroplast#chromosome-list-file\n",
    "    list_file=row.ENASampleNum + '_pt\tPT\tCircular-Chromosome\tChloroplast'\n",
    "    with open(file_embl.replace('.embl','_CHR.txt'), 'wb') as f:\n",
    "        f.write(list_file.encode(\"ascii\"))\n",
    "    gzip_file(file_embl.replace('.embl','_CHR.txt'))\n",
    "    ### Manifest file \n",
    "    # https://ena-docs.readthedocs.io/en/latest/submit/assembly/genome.html \n",
    "    manifest_file = file_embl.replace('.embl','_manifest.txt')\n",
    "    manifest={\n",
    "    'STUDY':'PRJEB35285', #https://www.ncbi.nlm.nih.gov/bioproject/588607\n",
    "    'SAMPLE':row.ENASampleNum,\n",
    "    'ASSEMBLYNAME':row.ENASampleNum + '_pt',\n",
    "    'ASSEMBLY_TYPE':'clone or isolate',\n",
    "    'COVERAGE':row.Coverage_base,\n",
    "    'PROGRAM':'GetOrganelle v1.7.5 +  GeSeq 2.03',\n",
    "    'PLATFORM':'ILLUMINA',\n",
    "    'MOLECULETYPE':'genomic DNA',\n",
    "    'FLATFILE': row.Sample_Name + '.embl.gz',\n",
    "    'CHROMOSOME_LIST': row.Sample_Name + '_CHR.txt.gz',\n",
    "    'DESCRIPTION': row.NCBI_sciname + ' chloroplast, complete genome'\n",
    "    }\n",
    "    manifest = pd.DataFrame.from_dict(manifest,orient='index')\n",
    "    manifest.to_csv(manifest_file,header=None,sep='\\t')\n",
    "    \n",
    "    # Submission command\n",
    "    commands.append('java -jar C:/Data/PAFTOL/ENA_submissions/webin-cli-4.3.0.jar ' + \n",
    "          '-username Webin-52995 -passwordFile C:/Data/PAFTOL/ENA_submissions/ena_pwd.txt -context genome -manifest ' +\n",
    "          row.Sample_Name + '_manifest.txt' + ' -validate') \n",
    "pd.DataFrame(commands).to_csv(sdir + '/Commands_WholePT.txt',index=False,header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "286\n",
      "213\n",
      "204\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>idPaftol</th>\n",
       "      <th>idSequencing</th>\n",
       "      <th>Sample_Name</th>\n",
       "      <th>Project</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>Taxonomical_Notes</th>\n",
       "      <th>SumContigLength</th>\n",
       "      <th>...</th>\n",
       "      <th>error_nr</th>\n",
       "      <th>Geseq_name</th>\n",
       "      <th>DataSource</th>\n",
       "      <th>ENASampleNum</th>\n",
       "      <th>ENAExpNumber</th>\n",
       "      <th>ENARunNumber</th>\n",
       "      <th>NCBI_TaxID</th>\n",
       "      <th>NCBI_sciname</th>\n",
       "      <th>Np_contigs</th>\n",
       "      <th>Sump_contigs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>6</td>\n",
       "      <td>960</td>\n",
       "      <td>PAFTOL_000960</td>\n",
       "      <td>Pilot</td>\n",
       "      <td>Caryophyllales</td>\n",
       "      <td>Dioncophyllaceae</td>\n",
       "      <td>Triphyophyllum</td>\n",
       "      <td>peltatum</td>\n",
       "      <td>NaN</td>\n",
       "      <td>176331.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pis_000960</td>\n",
       "      <td>PAFTOL</td>\n",
       "      <td>ERS5477005</td>\n",
       "      <td>ERX4815054</td>\n",
       "      <td>ERR5006122</td>\n",
       "      <td>63090.0</td>\n",
       "      <td>Triphyophyllum peltatum</td>\n",
       "      <td>1</td>\n",
       "      <td>141085</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>30</td>\n",
       "      <td>946</td>\n",
       "      <td>PAFTOL_000946</td>\n",
       "      <td>Pilot</td>\n",
       "      <td>Zygophyllales</td>\n",
       "      <td>Zygophyllaceae</td>\n",
       "      <td>Zygophyllum</td>\n",
       "      <td>fabago</td>\n",
       "      <td>NaN</td>\n",
       "      <td>185529.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pis_000946</td>\n",
       "      <td>PAFTOL</td>\n",
       "      <td>ERS4591016</td>\n",
       "      <td>ERX4143467</td>\n",
       "      <td>ERR4180043</td>\n",
       "      <td>66652.0</td>\n",
       "      <td>Zygophyllum fabago</td>\n",
       "      <td>1</td>\n",
       "      <td>103382</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows Ã— 50 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   idPaftol  idSequencing    Sample_Name Project           order  \\\n",
       "0         6           960  PAFTOL_000960   Pilot  Caryophyllales   \n",
       "1        30           946  PAFTOL_000946   Pilot   Zygophyllales   \n",
       "\n",
       "             family           genus   species Taxonomical_Notes  \\\n",
       "0  Dioncophyllaceae  Triphyophyllum  peltatum               NaN   \n",
       "1    Zygophyllaceae     Zygophyllum    fabago               NaN   \n",
       "\n",
       "   SumContigLength  ... error_nr  Geseq_name  DataSource  ENASampleNum  \\\n",
       "0         176331.0  ...      NaN  Pis_000960      PAFTOL    ERS5477005   \n",
       "1         185529.0  ...      NaN  Pis_000946      PAFTOL    ERS4591016   \n",
       "\n",
       "   ENAExpNumber  ENARunNumber  NCBI_TaxID             NCBI_sciname  \\\n",
       "0    ERX4815054    ERR5006122     63090.0  Triphyophyllum peltatum   \n",
       "1    ERX4143467    ERR4180043     66652.0       Zygophyllum fabago   \n",
       "\n",
       "   Np_contigs  Sump_contigs  \n",
       "0           1        141085  \n",
       "1           1        103382  \n",
       "\n",
       "[2 rows x 50 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Long Contigs\n",
    "contigs_pt = pd.read_csv(sdir + DataSource + '_contigsPT.csv')\n",
    "print(contigs_pt.shape[0])\n",
    "contigs_pt = contigs_pt[(contigs_pt.ENASampleNum.notnull()) & (contigs_pt.ENARunNumber.notnull())]\n",
    "print(contigs_pt.shape[0])\n",
    "contigs_pt = contigs_pt[contigs_pt.NCBI_sciname.notnull()]\n",
    "print(contigs_pt.shape[0])\n",
    "contigs_pt.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "629\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_000960-1-scaffold_1-contig_info_1966698+%2C1884030+%2C1967352_1884640+%2C1884030-%2C1976352+%2C-length_141085_GenBank.gb\n",
      "Pis_000960\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_000946-1-scaffold_1-contig_info_890838-%2C890814-%2C890844-%2C890814+%2C890846_890832-%2C-length_103382_GenBank.gb\n",
      "Pis_000946\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_002855-1-scaffold_2-contig_info_5158610+%2C5287124_1588974_5341156_5240704_5281502-%2C5322192_5288256-%2C-length_77746_GenBank.gb\n",
      "Pis_002855\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_003064-1-scaffold_1-contig_info_26170-%2C26200_24214_26206+%2C26196-%2C-length_91365_GenBank.gb\n",
      "Pis_003064\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_001065-1-scaffold_1-contig_info_750856-%2C750908-%2C750906+%2C750908+%2C750894+%2C-length_154282_GenBank.gb\n",
      "Pis_001065\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_003429-1-scaffold_1-contig_info_555578+%2C555596_555532-%2C555584+%2C555596_555532+%2C555604-%2C-length_168561_GenBank.gb\n",
      "Pis_003429\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_001055-1-scaffold_1-contig_info_506604-%2C507666-%2C507560+%2C507666+%2C507810+%2C-length_76142_GenBank.gb\n",
      "Pis_001055\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_002853-1-scaffold_2-contig_info_3022+%2C-length_82315_GenBank.gb\n",
      "Pis_002853\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_002889-1-scaffold_1-contig_info_856084+%2C912515_949189_942779-%2C957189-%2C912515_949189_942779+%2C949165+%2C-length_102540_GenBank.gb\n",
      "Pis_002889\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_002550-1-scaffold_7-contig_info_412+%2C-length_57409_GenBank.gb\n",
      "Pis_002550\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_002551-1-scaffold_5-contig_info_4877+%2C-length_51672_GenBank.gb\n",
      "Pis_002551\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_003110-1-scaffold_1-contig_info_7896493+%2C7921739+%2C7894233+%2C8199566-%2C7785983_8108039+%2C8199566+%2C7894233-%2C7896339+%2C7894233-%2C7896339+%2C7894233-%2C7921739-%2C8150503+%2C-length_152981_GenBank.gb\n",
      "Pis_003110\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_002582-1-scaffold_2-contig_info_3630398+%2C-length_62466_GenBank.gb\n",
      "Pis_002582\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_003621-2-scaffold_2-contig_info_4323097+%2C4460789+%2C3737574+%2C4302271+%2C3737574+%2C4325990_4423618_4429186+%2C4325463_4420393+%2C-length_66060_GenBank.gb\n",
      "Pis_003621\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_000808-1-scaffold_1-contig_info_3191691+%2C3308400-%2C3278194-%2C-length_50380_GenBank.gb\n",
      "Pis_000808\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_000848-1-scaffold_1-contig_info_174200+%2C174318+%2C174230+%2C174318-%2C174314-%2C-length_109749_GenBank.gb\n",
      "Pis_000848\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_007445-1-scaffold_3-contig_info_97-%2C32662+%2C32720+%2C133+%2C177+%2C-length_50498_GenBank.gb\n",
      "Pis_007445\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_001337-1-scaffold_1-contig_info_408174-%2C408232+%2C408212+%2C408232-%2C408192+%2C-length_158758_GenBank.gb\n",
      "Pis_001337\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_002740-1-scaffold_1-contig_info_1787478_1790776+%2C-length_122835_GenBank.gb\n",
      "Pis_002740\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_002509-1-scaffold_7-contig_info_1027615+%2C-length_51911_GenBank.gb\n",
      "Pis_002509\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_001405-1-scaffold_3-contig_info_417208+%2C417296_417340-%2C417262-%2C-length_66557_GenBank.gb\n",
      "Pis_001405\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_002563-1-scaffold_1-contig_info_4176462-%2C4176598-%2C4208637+%2C-length_82241_GenBank.gb\n",
      "Pis_002563\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_002671-1-scaffold_5-contig_info_12377_11794_12133+%2C-length_59542_GenBank.gb\n",
      "Pis_002671\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_002586-1-scaffold_1-contig_info_6954701_6960432_6953303_6957263_6960323_6946330_6957623_6957049+%2C-length_100888_GenBank.gb\n",
      "Pis_002586\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_002584-1-scaffold_4-contig_info_5942778+%2C5990438-%2C5975825+%2C5990438+%2C5976593+%2C-length_108136_GenBank.gb\n",
      "Pis_002584\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_003151-1-scaffold_1-contig_info_252-%2C20746_316+%2C288+%2C20746_316-%2C319+%2C-length_151890_GenBank.gb\n",
      "Pis_003151\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_001113-1-scaffold_2-contig_info_3442137+%2C3525638_3439273_3421797+%2C3427705+%2C3525638_3439273_3421797-%2C3448905-%2C-length_164790_GenBank.gb\n",
      "Pis_001113\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_003730-1-scaffold_1-contig_info_11747210_2849059_12021355+%2C-length_77167_GenBank.gb\n",
      "Pis_003730\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_001187-1-scaffold_2-contig_info_5157016+%2C5260042+%2C5213352-%2C5260042-%2C5176320+%2C-length_82548_GenBank.gb\n",
      "Pis_001187\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_001191-1-scaffold_1-contig_info_516088+%2C618368_627820+%2C624800+%2C626076_602900-%2C615674_618780+%2C626076_602900+%2C624800-%2C618868-%2C-length_94569_GenBank.gb\n",
      "Pis_001191\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_000242-1-scaffold_11-contig_info_2062552+%2C-length_55406_GenBank.gb\n",
      "Pis_000242\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_003856-1-scaffold_1-contig_info_703770_703615-%2C741369_775503+%2C757811-%2C-length_52946_GenBank.gb\n",
      "Pis_003856\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_003867-1-scaffold_1-contig_info_2468301+%2C2895835+%2C2601709_2612389-%2C2895835-%2C2602907-%2C-length_92517_GenBank.gb\n",
      "Pis_003867\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_003938-1-scaffold_2-contig_info_13117047_13241275_13109709_12757144_13278299_13082179_13126519_13123073_13123929+%2C-length_60567_GenBank.gb\n",
      "Pis_003938\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_003961-1-scaffold_1-contig_info_870506-%2C871456_862656-%2C870396-%2C871456_862656+%2C879398+%2C-length_160483_GenBank.gb\n",
      "Pis_003961\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_004050-1-scaffold_1-contig_info_15+%2C445166-%2C443288_370779+%2C-length_51250_GenBank.gb\n",
      "Pis_004050\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_004112-1-scaffold_2-contig_info_431253+%2C437799-%2C420996-%2C437799+%2C431291+%2C-length_151813_GenBank.gb\n",
      "Pis_004112\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_004222-1-scaffold_1-contig_info_117303+%2C111611_129922_123807_127383+%2C130666-%2C111611_129922_123807_127383-%2C132931+%2C-length_132066_GenBank.gb\n",
      "Pis_004222\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_004310-1-scaffold_1-contig_info_3098298-%2C3391096+%2C3390352-%2C-length_52487_GenBank.gb\n",
      "Pis_004310\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_004328-1-scaffold_2-contig_info_41502_41424+%2C-length_88682_GenBank.gb\n",
      "Pis_004328\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_004360-1-scaffold_1-contig_info_134150+%2C134164+%2C134126-%2C134164-%2C134158-%2C-length_157270_GenBank.gb\n",
      "Pis_004360\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_004372-1-scaffold_1-contig_info_9985480_12939076+%2C13507192-%2C12876290+%2C-length_76126_GenBank.gb\n",
      "Pis_004372\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_004397-1-scaffold_4-contig_info_582310+%2C-length_52785_GenBank.gb\n",
      "Pis_004397\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_004443-1-scaffold_1-contig_info_681108+%2C678838_690326+%2C697944-%2C678838_690326-%2C683170+%2C-length_150693_GenBank.gb\n",
      "Pis_004443\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_004448-1-scaffold_1-contig_info_10433416_10434728_10195832_9921906_10684564_10691184_10112138_9170224_9724702_10433402+%2C-length_64344_GenBank.gb\n",
      "Pis_004448\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_004458-1-scaffold_1-contig_info_1570592-%2C1569066_1630550-%2C1630502-%2C1569066_1630550+%2C1696860_1656494-%2C-length_160731_GenBank.gb\n",
      "Pis_004458\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_004466-1-scaffold_1-contig_info_586324-%2C589742-%2C591310-%2C589742+%2C589548-%2C-length_81906_GenBank.gb\n",
      "Pis_004466\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_004470-1-scaffold_1-contig_info_176246+%2C176276+%2C176260+%2C176276-%2C176278-%2C-length_94114_GenBank.gb\n",
      "Pis_004470\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_004692-1-scaffold_1-contig_info_3141504_3066348-%2C2979034_3184542-%2C3179686_2990990+%2C2979034_3184542+%2C3248016+%2C-length_161546_GenBank.gb\n",
      "Pis_004692\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_004704-1-scaffold_4-contig_info_74402_74574+%2C-length_56068_GenBank.gb\n",
      "Pis_004704\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_004741-1-scaffold_3-contig_info_46104+%2C-length_79273_GenBank.gb\n",
      "Pis_004741\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_004743-1-scaffold_1-contig_info_62400+%2C62528+%2C62516+%2C62528-%2C62510+%2C-length_107846_GenBank.gb\n",
      "Pis_004743\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_004756-1-scaffold_3-contig_info_1817204_1766446_1753740_1816994_1819608_1802560_1819638+%2C-length_66908_GenBank.gb\n",
      "Pis_004756\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_004821-1-scaffold_2-contig_info_91328+%2C92188_92190_92134_92156-%2C92202+%2C92188_92190_92134_92156+%2C92192_92036_92118_91698_91692_92182-%2C-length_126339_GenBank.gb\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\ProgramData\\Anaconda3\\lib\\site-packages\\Bio\\GenBank\\__init__.py:1265: BiopythonParserWarning: Couldn't parse feature location: 'join(65657..65662,join(66452..66908,1..17))'\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pis_004821\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_004822-1-scaffold_2-contig_info_95826+%2C23330_98458-%2C97298-%2C14740-%2C11814-%2C14740-%2C11814-%2C14740-%2C98598-%2C98554_96936_98704-%2C-length_53333_GenBank.gb\n",
      "Pis_004822\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_004855-1-scaffold_1-contig_info_109232_109502_109456_109164+%2C-length_63380_GenBank.gb\n",
      "Pis_004855\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_004862-1-scaffold_2-contig_info_13685+%2C20224_24275+%2C19989_124168_22081_13137_19993_111001_20228_13907_1521_1517_19693_19607+%2C-length_109279_GenBank.gb\n",
      "Pis_004862\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_004886-1-scaffold_5-contig_info_353+%2C445+%2C451_161-%2C-length_55410_GenBank.gb\n",
      "Pis_004886\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_004895-1-scaffold_3-contig_info_10575310_10637238_9168948-%2C10654934_10639212_10655560_10698948-%2C10688592_10347066-%2C10654934_10639212_10655560_10698948+%2C10582006+%2C-length_102222_GenBank.gb\n",
      "Pis_004895\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_004897-1-scaffold_1-contig_info_10960378-%2C10865596_10871046_10972244+%2C10971032-%2C10865596_10871046_10972244-%2C10962738_10870410_10962304_10960876_10935464-%2C-length_161269_GenBank.gb\n",
      "Pis_004897\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_005061-1-scaffold_1-contig_info_177-%2C111+%2C36207-%2C36577_231-%2C283+%2C36577_231+%2C36207+%2C36327+%2C36207+%2C36327+%2C36207+%2C111-%2C281+%2C-length_150468_GenBank.gb\n",
      "Pis_005061\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_005062-1-scaffold_1-contig_info_1315172-%2C1322509+%2C1281411-%2C1304330-%2C1281411-%2C1358217-%2C1516220-%2C1358217+%2C1281411+%2C1304330+%2C1281411+%2C1322509-%2C1563522_509+%2C-length_111950_GenBank.gb\n",
      "Pis_005062\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_005065-1-scaffold_1-contig_info_116383-%2C120187-%2C114154-%2C116979+%2C118449_120781-%2C116979-%2C114154+%2C116329+%2C114154+%2C116329+%2C114154+%2C120187+%2C119994-%2C-length_152518_GenBank.gb\n",
      "Pis_005065\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_005067-1-scaffold_1-contig_info_249+%2C-length_121119_GenBank.gb\n",
      "Pis_005067\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_005074-1-scaffold_1-contig_info_663924-%2C653972+%2C592254+%2C820467+%2C796856+%2C820467-%2C592254-%2C686728+%2C592254-%2C686728+%2C592254-%2C653972-%2C690228-%2C-length_103771_GenBank.gb\n",
      "Pis_005074\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_005085-1-scaffold_2-contig_info_68569_68575_71383_67843_63219_67459+%2C-length_53456_GenBank.gb\n",
      "Pis_005085\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_005091-1-scaffold_1-contig_info_357-%2C419-%2C104309+%2C123-%2C413+%2C-length_74323_GenBank.gb\n",
      "Pis_005091\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_005095-1-scaffold_5-contig_info_227+%2C241-%2C55445-%2C55661+%2C55445-%2C55661+%2C55445-%2C219-%2C231-%2C219+%2C55445+%2C241+%2C243+%2C-length_96466_GenBank.gb\n",
      "Pis_005095\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_005096-1-scaffold_4-contig_info_15191748+%2C15062440+%2C14457645+%2C7621919-%2C15212170+%2C14457645+%2C15370622_15517194+%2C15422688+%2C15530590+%2C15347614+%2C15251138+%2C-length_78215_GenBank.gb\n",
      "Pis_005096\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_005199-1-scaffold_1-contig_info_466998-%2C747338_729555-%2C710246+%2C747338_729555+%2C720167+%2C-length_78939_GenBank.gb\n",
      "Pis_005199\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_005239-1-scaffold_5-contig_info_1125121+%2C1164237_1199860-%2C1141945+%2C1164237_1199860+%2C1142507-%2C-length_82459_GenBank.gb\n",
      "Pis_005239\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_005241-1-scaffold_1-contig_info_839743+%2C822876+%2C863835+%2C822876-%2C863797-%2C-length_156108_GenBank.gb\n",
      "Pis_005241\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_005374-1-scaffold_1-contig_info_12414686_12973176_12748484_12973174_9654254+%2C-length_118311_GenBank.gb\n",
      "Pis_005374\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_005376-1-scaffold_9-contig_info_4994174+%2C4981054+%2C5003210_4995406+%2C-length_56458_GenBank.gb\n",
      "Pis_005376\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_005390-1-scaffold_5-contig_info_13337938_12821090_13153396_13131422_13337940_13266672_13266106_13268520_13262360_13138560_13265118_13140412+%2C-length_57911_GenBank.gb\n",
      "Pis_005390\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_005455-1-scaffold_8-contig_info_527426+%2C592490+%2C548714-%2C-length_71928_GenBank.gb\n",
      "Pis_005455\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_005456-1-scaffold_1-contig_info_1329+%2C-length_56024_GenBank.gb\n",
      "Pis_005456\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_005467-1-scaffold_2-contig_info_2674046_2699216-%2C2780904+%2C2684610_2703502-%2C2780904-%2C2674614_2724198_2712718-%2C-length_115905_GenBank.gb\n",
      "Pis_005467\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_005686-1-scaffold_1-contig_info_413681-%2C449936+%2C436284-%2C449936-%2C425291-%2C-length_152538_GenBank.gb\n",
      "Pis_005686\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_005687-1-scaffold_1-contig_info_907631+%2C917126+%2C895304-%2C917126-%2C932591+%2C-length_151075_GenBank.gb\n",
      "Pis_005687\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_005746-1-scaffold_3-contig_info_1147017+%2C-length_107872_GenBank.gb\n",
      "Pis_005746\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_005761-1-scaffold_1-contig_info_1123206+%2C1152611_1169974+%2C1233964-%2C1152611_1169974-%2C1158430_1236016_1189430-%2C-length_150903_GenBank.gb\n",
      "Pis_005761\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_005782-1-scaffold_6-contig_info_4036882+%2C4045122-%2C4061498+%2C-length_57066_GenBank.gb\n",
      "Pis_005782\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_005805-1-scaffold_2-contig_info_2104377+%2C2374615-%2C2123311+%2C-length_85002_GenBank.gb\n",
      "Pis_005805\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_005814-1-scaffold_1-contig_info_571296-%2C608542+%2C640751-%2C608542-%2C585351-%2C-length_146513_GenBank.gb\n",
      "Pis_005814\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_005933-1-scaffold_4-contig_info_638271+%2C676365-%2C659420_639588+%2C-length_53676_GenBank.gb\n",
      "Pis_005933\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_005952-1-scaffold_6-contig_info_209325_208233_198238_249762_210646_258013_209580_38347_148193_221744_108351_212998_230766_202000_206627_169283_203887+%2C-length_51017_GenBank.gb\n",
      "Pis_005952\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_006271-1-scaffold_7-contig_info_1168212_1214252+%2C-length_60492_GenBank.gb\n",
      "Pis_006271"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\ProgramData\\Anaconda3\\lib\\site-packages\\Bio\\GenBank\\__init__.py:1265: BiopythonParserWarning: Couldn't parse feature location: 'join(49674..50064,join(50729..51017,1..40))'\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_006274-1-scaffold_1-contig_info_468528-%2C479306_495366_489444-%2C481272-%2C-length_58696_GenBank.gb\n",
      "Pis_006274\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_006305-1-scaffold_5-contig_info_2175992+%2C-length_54441_GenBank.gb\n",
      "Pis_006305\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_006335-1-scaffold_2-contig_info_63784+%2C-length_71734_GenBank.gb\n",
      "Pis_006335\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_006512-1-scaffold_7-contig_info_1801453_1706240+%2C-length_72264_GenBank.gb\n",
      "Pis_006512\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_006516-1-scaffold_5-contig_info_854945-%2C903783_890908+%2C893941-%2C-length_60216_GenBank.gb\n",
      "Pis_006516\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_006521-1-scaffold_1-contig_info_1933020+%2C2222033_1958909+%2C2073575_2020575-%2C-length_66130_GenBank.gb\n",
      "Pis_006521\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_006527-1-scaffold_1-contig_info_873560_888230_804625+%2C-length_82882_GenBank.gb\n",
      "Pis_006527\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_006529-1-scaffold_10-contig_info_678656_641625_586320+%2C-length_60045_GenBank.gb\n",
      "Pis_006529\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_006558-1-scaffold_3-contig_info_153110_180304_180302_150208_152496_179002_152854_168958_167896_176224_169076+%2C-length_52136_GenBank.gb\n",
      "Pis_006558\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_006574-1-scaffold_3-contig_info_69280_60892_62118_69334+%2C-length_83399_GenBank.gb\n",
      "Pis_006574\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_006760-1-scaffold_5-contig_info_2231172+%2C2218319-%2C2234160-%2C-length_71361_GenBank.gb\n",
      "Pis_006760\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_006838-1-scaffold_2-contig_info_20067+%2C-length_57302_GenBank.gb\n",
      "Pis_006838\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_006845-1-scaffold_3-contig_info_8065+%2C8081-%2C233-%2C8113+%2C8095_8089+%2C-length_76133_GenBank.gb\n",
      "Pis_006845\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_006856-1-scaffold_2-contig_info_39_107_13979_45+%2C-length_57763_GenBank.gb\n",
      "Pis_006856\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_006864-1-scaffold_3-contig_info_3343238_3397074_3344362_3331252_3528591_3353322+%2C-length_85178_GenBank.gb\n",
      "Pis_006864\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_006873-1-scaffold_1-contig_info_518612_247021_678491+%2C696352_602360_582630_555232_576868_615522_527203_691432_691426_691430_480499_688446-%2C653815_674508+%2C-length_62010_GenBank.gb\n",
      "Pis_006873\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_006886-1-scaffold_2-contig_info_59424_61642_59604_62566_59024_62472_54412_62652_62622+%2C-length_51609_GenBank.gb\n",
      "Pis_006886\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_006887-1-scaffold_4-contig_info_1604341_1577421+%2C-length_69091_GenBank.gb\n",
      "Pis_006887\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_006896-1-scaffold_2-contig_info_761874-%2C820875+%2C820271-%2C-length_82545_GenBank.gb\n",
      "Pis_006896\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_006905-1-scaffold_4-contig_info_6660+%2C3445+%2C7108+%2C5295-%2C1929+%2C5295-%2C6609+%2C3445-%2C6666+%2C5793-%2C5479+%2C5793-%2C5479+%2C5793-%2C5479+%2C5793-%2C6672-%2C-length_79469_GenBank.gb\n",
      "Pis_006905\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_006906-1-scaffold_1-contig_info_2591915-%2C2825969_2825237_2589311-%2C2683317_2793146_2815378+%2C2825969_2825237_2589311+%2C2794981_2626360-%2C-length_148093_GenBank.gb\n",
      "Pis_006906\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_006908-1-scaffold_1-contig_info_577_601+%2C36199-%2C40220-%2C36199-%2C40220-%2C36199-%2C641_45+%2C-length_84997_GenBank.gb\n",
      "Pis_006908\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_007027-1-scaffold_2-contig_info_2046362_1649278_1260496_2047240_2030686_1580918_1835630_1438962+%2C-length_53149_GenBank.gb\n",
      "Pis_007027\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_007037-1-scaffold_2-contig_info_155640+%2C156332_155932_152562_154628_155554_155576+%2C155654_154750-%2C156332_155932_152562_154628_155554_155576-%2C155934_154668_155686-%2C-length_98507_GenBank.gb\n",
      "Pis_007037\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_007092-1-scaffold_1-contig_info_1095048_1089782-%2C1092350+%2C1097190+%2C1092350-%2C1096680+%2C-length_158073_GenBank.gb\n",
      "Pis_007092\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_007106-1-scaffold_1-contig_info_5356664+%2C5347242_5343612_5343442-%2C5359136-%2C5347242_5343612_5343442+%2C5360972-%2C-length_158666_GenBank.gb\n",
      "Pis_007106\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_007284-1-scaffold_10-contig_info_2311974-%2C2311919+%2C2343904+%2C-length_51870_GenBank.gb\n",
      "Pis_007284\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_007287-1-scaffold_1-contig_info_5668808+%2C5677410_5755960_5448957_5490975-%2C5757061+%2C-length_57715_GenBank.gb\n",
      "Pis_007287\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_013077-1-scaffold_1-contig_info_401_575+%2C467_509_419+%2C443-%2C467_509_419-%2C581_389_521_42189_317_583_233-%2C-length_168301_GenBank.gb\n",
      "Pis_013077\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_007507-1-scaffold_1-contig_info_1158748+%2C1401105+%2C1414341+%2C-length_53029_GenBank.gb\n",
      "Pis_007507\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_007556-1-scaffold_1-contig_info_332093-%2C341940+%2C341199-%2C-length_50579_GenBank.gb\n",
      "Pis_007556\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_007484-1-scaffold_2-contig_info_89873_89858_89747_89860+%2C-length_114173_GenBank.gb\n",
      "Pis_007484\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_011256-1-scaffold_5-contig_info_2702558+%2C2710320+%2C2709822-%2C-length_54417_GenBank.gb\n",
      "Pis_011256\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_011260-1-scaffold_1-contig_info_598-%2C354034_606_351804_566+%2C308-%2C354034_606_351804_566-%2C629-%2C-length_136712_GenBank.gb\n",
      "Pis_011260\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_011263-1-scaffold_1-contig_info_1289-%2C1377+%2C1035+%2C1377-%2C1317-%2C-length_157717_GenBank.gb\n",
      "Pis_011263\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_011275-1-scaffold_1-contig_info_1364180-%2C1364354_1364366+%2C1364072+%2C1364354_1364366-%2C1364294-%2C-length_129018_GenBank.gb\n",
      "Pis_011275\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_011286-1-scaffold_1-contig_info_465836-%2C466402-%2C466154+%2C-length_54684_GenBank.gb\n",
      "Pis_011286\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_011314-1-scaffold_3-contig_info_7156+%2C-length_56477_GenBank.gb\n",
      "Pis_011314\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_013041-1-scaffold_2-contig_info_13-%2C19917_19921-%2C19897+%2C19917_19921+%2C16953+%2C-length_105490_GenBank.gb\n",
      "Pis_013041\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_008675-1-scaffold_2-contig_info_22954-%2C22964_22962-%2C22994-%2C22964_22962+%2C22976+%2C-length_108535_GenBank.gb\n",
      "Pis_008675\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_008685-1-scaffold_6-contig_info_18416-%2C18490-%2C18472-%2C-length_50369_GenBank.gb\n",
      "Pis_008685\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_008709-1-scaffold_1-contig_info_2578374-%2C2648406+%2C2524658+%2C2566888+%2C2606874+%2C2524658-%2C2606914+%2C2566888-%2C2524658-%2C2648406-%2C2579496+%2C-length_140960_GenBank.gb\n",
      "Pis_008709\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_008715-1-scaffold_2-contig_info_5354434_5343742_5403596_5338024+%2C-length_52943_GenBank.gb\n",
      "Pis_008715\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_008731-1-scaffold_5-contig_info_49854_49820_49660_49874_49850+%2C-length_58564_GenBank.gb\n",
      "Pis_008731\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_008760-1-scaffold_12-contig_info_35047_549-%2C495_507+%2C35113-%2C-length_66538_GenBank.gb\n",
      "Pis_008760\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_008771-1-scaffold_2-contig_info_2374460-%2C2527982_2525924+%2C2447320_2475344_2517672_2470010_2468506_2525608-%2C-length_71449_GenBank.gb\n",
      "Pis_008771\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_008779-1-scaffold_7-contig_info_2232002+%2C2094816_2154224+%2C2243266-%2C-length_65190_GenBank.gb\n",
      "Pis_008779\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_008795-1-scaffold_3-contig_info_109604+%2C112004_111736-%2C112016-%2C-length_66669_GenBank.gb\n",
      "Pis_008795\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_008822-1-scaffold_14-contig_info_267160_263200_266848_266444_267164_266986_267148+%2C-length_50274_GenBank.gb\n",
      "Pis_008822\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_009261-1-scaffold_3-contig_info_6185880_6351238+%2C6064328+%2C6300400+%2C6064328+%2C6300400+%2C6064328+%2C6174186-%2C6112416-%2C6198648-%2C6150806+%2C6303792-%2C-length_77996_GenBank.gb\n",
      "Pis_009261\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_009291-1-scaffold_1-contig_info_1331200-%2C69966+%2C1347572-%2C1424072-%2C69966-%2C1437174+%2C1422404+%2C1437174-%2C69966+%2C1424072+%2C1339358-%2C-length_130278_GenBank.gb\n",
      "Pis_009291\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_009451-1-scaffold_2-contig_info_106406+%2C106648+%2C106656+%2C106648-%2C106418-%2C-length_133608_GenBank.gb\n",
      "Pis_009451\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_013043-1-scaffold_1-contig_info_31-%2C27-%2C23293-%2C27+%2C23221+%2C-length_160331_GenBank.gb\n",
      "Pis_013043\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_009531-1-scaffold_11-contig_info_85916+%2C84444_86132_84418_85050+%2C76536-%2C86136_84006+%2C86108_84450+%2C-length_51050_GenBank.gb\n",
      "Pis_009531\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_009547-1-scaffold_1-contig_info_79136+%2C79318+%2C79304_79298-%2C-length_83949_GenBank.gb\n",
      "Pis_009547\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_013039-1-scaffold_3-contig_info_17602+%2C-length_52974_GenBank.gb\n",
      "Pis_013039\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_009585-1-scaffold_1-contig_info_10964657+%2C11201814-%2C11335420-%2C-length_53795_GenBank.gb\n",
      "Pis_009585\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_009693-1-scaffold_1-contig_info_1301-%2C1774-%2C1776+%2C1774+%2C1726-%2C-length_161367_GenBank.gb\n",
      "Pis_009693\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_009989-1-scaffold_4-contig_info_33492-%2C33640-%2C33596+%2C-length_74016_GenBank.gb\n",
      "Pis_009989\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_009995-1-scaffold_1-contig_info_58632_59162_58894_58930_58988_58528_58108_58822_58780_58924_58534_58906_58570_52998_58954_57344_58884_58760_58858_57116_58936_57670+%2C-length_103187_GenBank.gb\n",
      "Pis_009995\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_010027-1-scaffold_1-contig_info_21334+%2C21316-%2C21246-%2C21316+%2C21340-%2C-length_163168_GenBank.gb\n",
      "Pis_010027\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_013037-1-scaffold_6-contig_info_1291+%2C1303+%2C1013-%2C1303-%2C1301-%2C-length_87900_GenBank.gb\n",
      "Pis_013037\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_018865-1-scaffold_3-contig_info_3309-%2C3629-%2C3353+%2C-length_65101_GenBank.gb\n",
      "Pis_018865\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_010191-1-scaffold_1-contig_info_486-%2C414-%2C492+%2C414+%2C509-%2C-length_142074_GenBank.gb\n",
      "Pis_010191\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_010217-1-scaffold_1-contig_info_676-%2C1332+%2C44837+%2C1332-%2C44239+%2C-length_155401_GenBank.gb\n",
      "Pis_010217\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_010219-1-scaffold_2-contig_info_51+%2C-length_64937_GenBank.gb\n",
      "Pis_010219\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_010223-1-scaffold_1-contig_info_88274-%2C88304-%2C88270-%2C88304+%2C88312-%2C-length_155798_GenBank.gb\n",
      "Pis_010223\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_010245-1-scaffold_1-contig_info_644598_836901-%2C711919_783049-%2C836841+%2C-length_82127_GenBank.gb\n",
      "Pis_010245\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_010257-1-scaffold_7-contig_info_75325+%2C67467+%2C75343-%2C-length_72172_GenBank.gb\n",
      "Pis_010257\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_010295-1-scaffold_1-contig_info_41520+%2C44261+%2C44267-%2C42861-%2C42951+%2C42861+%2C43935+%2C44261+%2C42287+%2C-length_145945_GenBank.gb\n",
      "Pis_010295\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_010597-1-scaffold_1-contig_info_556382_561606_561566_565452_561630_554792_561570_561654_567058+%2C-length_72422_GenBank.gb\n",
      "Pis_010597\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_010621-1-scaffold_2-contig_info_671854+%2C675404_677268_676400-%2C675324-%2C674414-%2C677426_672560_677262+%2C677488-%2C675924-%2C-length_70512_GenBank.gb\n",
      "Pis_010621\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_010639-1-scaffold_9-contig_info_67238_67212_66926_67108_67236+%2C-length_54151_GenBank.gb\n",
      "Pis_010639\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_010651-1-scaffold_9-contig_info_1608186+%2C1608184-%2C1608188-%2C-length_65261_GenBank.gb\n",
      "Pis_010651\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_010655-1-scaffold_1-contig_info_3646464-%2C3770686-%2C3745862-%2C-length_80687_GenBank.gb\n",
      "Pis_010655\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_010661-1-scaffold_3-contig_info_346142_376688_333774_336172+%2C-length_72449_GenBank.gb\n",
      "Pis_010661\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_010711-1-scaffold_1-contig_info_40380+%2C41010_40932-%2C41012-%2C41010_40932+%2C40910+%2C-length_80968_GenBank.gb\n",
      "Pis_010711\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_010713-1-scaffold_15-contig_info_80034_80038+%2C-length_50913_GenBank.gb\n",
      "Pis_010713\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_010723-1-scaffold_3-contig_info_191420+%2C-length_114504_GenBank.gb\n",
      "Pis_010723\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\ProgramData\\Anaconda3\\lib\\site-packages\\Bio\\GenBank\\__init__.py:1265: BiopythonParserWarning: Couldn't parse feature location: 'join(48717..49269,join(50408..50913,1..90))'\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_010813-1-scaffold_5-contig_info_130470+%2C130496-%2C130464+%2C130504+%2C130510+%2C130446+%2C130510-%2C130504-%2C130464-%2C130496+%2C130434-%2C130498-%2C-length_118925_GenBank.gb\n",
      "Pis_010813\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_010889-1-scaffold_3-contig_info_43650+%2C44240+%2C44150+%2C44240-%2C44110-%2C44222+%2C44244-%2C-length_130236_GenBank.gb\n",
      "Pis_010889\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_010915-1-scaffold_1-contig_info_241316-%2C241576_241582-%2C241416-%2C241576_241582+%2C241558_241564-%2C-length_154029_GenBank.gb\n",
      "Pis_010915\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_011991-1-scaffold_2-contig_info_192310+%2C192326+%2C192150-%2C192326-%2C192322+%2C-length_147596_GenBank.gb\n",
      "Pis_011991\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_012011-1-scaffold_7-contig_info_312028+%2C-length_55585_GenBank.gb\n",
      "Pis_012011\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_012015-1-scaffold_1-contig_info_721228-%2C722216+%2C722106-%2C722216-%2C722146+%2C-length_115709_GenBank.gb\n",
      "Pis_012015\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_012047-1-scaffold_1-contig_info_32714-%2C33076+%2C33120-%2C-length_55541_GenBank.gb\n",
      "Pis_012047\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_012077-1-scaffold_1-contig_info_107240+%2C107226_107332+%2C107314+%2C107226_107332-%2C107290+%2C-length_153531_GenBank.gb\n",
      "Pis_012077\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_012079-1-scaffold_1-contig_info_97492+%2C97322_97556-%2C97598+%2C97322_97556+%2C97610+%2C-length_84033_GenBank.gb\n",
      "Pis_012079\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_012125-1-scaffold_9-contig_info_40554_40602_40514+%2C-length_53745_GenBank.gb\n",
      "Pis_012125\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_012157-1-scaffold_5-contig_info_1716770+%2C-length_82262_GenBank.gb\n",
      "Pis_012157\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_012173-1-scaffold_1-contig_info_478204+%2C477758+%2C477820-%2C477758-%2C478618+%2C-length_159582_GenBank.gb\n",
      "Pis_012173\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_016287-1-scaffold_1-contig_info_862716_931869_861598_898999_857380+%2C-length_54505_GenBank.gb\n",
      "Pis_016287\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_016289-1-scaffold_6-contig_info_658765+%2C-length_82689_GenBank.gb\n",
      "Pis_016289\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_021539-1-scaffold_3-contig_info_70331-%2C1069_39_1126-%2C1061_1045-%2C1069_39_1126+%2C70608+%2C-length_151243_GenBank.gb\n",
      "Pis_021539\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_021817-1-scaffold_2-contig_info_1523616+%2C-length_56479_GenBank.gb\n",
      "Pis_021817\n",
      "GeSeqJob-20220304-PAFTOL_contigs_Pis_028437-1-scaffold_4-contig_info_986+%2C-length_62145_GenBank.gb\n",
      "Pis_028437\n"
     ]
    }
   ],
   "source": [
    "# Rename and modify genbank files, convert to embl flat file\n",
    "if not os.path.exists(sdir + '/Genbank_files/'):\n",
    "    os.makedirs(sdir + '/Genbank_files/')\n",
    "gb_files = [file for file in os.listdir(sdir + 'geseq/') if file.endswith('.gb')]\n",
    "print(len(gb_files))\n",
    "for idx, row in contigs_pt.iterrows():\n",
    "    idx_file = [i for i, x in enumerate([row.Geseq_name in file for file in gb_files]) if x]\n",
    "    if len(idx_file)>0:\n",
    "        print(gb_files[idx_file[0]])\n",
    "        for index, record in enumerate(SeqIO.parse(sdir + 'geseq/' + gb_files[idx_file[0]], \"genbank\")):\n",
    "            try:\n",
    "                print(record.id)\n",
    "                record.id = row.ENASampleNum + '_pt'\n",
    "                record.name = row.ENARunNumber + '_pt'\n",
    "                record.description = row.ENARunNumber + ' (' + row.Sample_Name + ') chloroplast, complete genome'\n",
    "                record.version = row.ENARunNumber + '.1'\n",
    "                record.annotations['source'] = row.ENASampleNum + ', ' + row.sci_name + ', isolate'\n",
    "                record.annotations['accessions'] = row.ENASampleNum\n",
    "                record.annotations['organism'] = row.NCBI_sciname\n",
    "                record.annotations[\"data_file_division\"]=\"PLN\";\n",
    "                record.annotations.pop('taxonomy')\n",
    "                record.annotations['references'][0].title = paper_title    \n",
    "                record.annotations['references'][0].authors = paper_authors    \n",
    "                record.annotations['references'][0].journal = paper_journal      \n",
    "                record.annotations['references'][0].pubmed_id = paper_id    \n",
    "                record.features[0].qualifiers['organism'] = row.NCBI_sciname\n",
    "                record.features[0].qualifiers['db_xref'] = ['taxon:' + str(int(row['NCBI_TaxID']))]\n",
    "    #             SeqIO.write(record,sdir + '/Genbank_files/' + row.Sample_Name + '_pt.gb',format='genbank')\n",
    "            except:\n",
    "                print('issue with ',gb_files[idx_file[0]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Long Contigs\n",
    "commands = []\n",
    "for idx, row in contigs_pt.iterrows():\n",
    "    ### Convert .gb to EMBL flat files\n",
    "    file_gb = sdir + '/Genbank_files/' + row.Sample_Name + '_pt.gb'\n",
    "    records = list(SeqIO.parse(file_gb, \"genbank\"))[0]\n",
    "    print(row.Sample_Name,len(records))\n",
    "    records_clean = del_fragment(records); records_clean = del_fragment(records_clean); records_clean = del_fragment(records_clean)\n",
    "    records_clean = del_longfeat(records_clean,len_thrs=10000);\n",
    "    file_embl = sdir + '/EMBL_files/' + row.Sample_Name + '.embl'\n",
    "    SeqIO.write(records_clean, file_gb, \"genbank\")\n",
    "    SeqIO.write(records_clean, file_embl, \"embl\")\n",
    "    gzip_file(file_embl)\n",
    "    ### Chromosome list file\n",
    "    # https://ena-docs.readthedocs.io/en/latest/submit/fileprep/assembly.html?highlight=chloroplast#chromosome-list-file\n",
    "    list_file=row.ENASampleNum + '_pt\tPT\tCircular-Chromosome\tChloroplast'\n",
    "    with open(file_embl.replace('.embl','_CHR.txt'), 'wb') as f:\n",
    "        f.write(list_file.encode(\"ascii\"))\n",
    "    gzip_file(file_embl.replace('.embl','_CHR.txt'))\n",
    "    ### Manifest file \n",
    "    # https://ena-docs.readthedocs.io/en/latest/submit/assembly/genome.html \n",
    "    manifest_file = file_embl.replace('.embl','_manifest.txt')\n",
    "    manifest={\n",
    "    'STUDY':'PRJEB35285', #https://www.ncbi.nlm.nih.gov/bioproject/588607\n",
    "    'SAMPLE':row.ENASampleNum,\n",
    "    'ASSEMBLYNAME':row.ENASampleNum + '_pt',\n",
    "    'ASSEMBLY_TYPE':'clone or isolate',\n",
    "    'COVERAGE':row.Coverage_base,\n",
    "    'PROGRAM':'GetOrganelle v1.7.5 +  GeSeq 2.03',\n",
    "    'PLATFORM':'ILLUMINA',\n",
    "    'MOLECULETYPE':'genomic DNA',\n",
    "    'FLATFILE': row.Sample_Name + '.embl.gz',\n",
    "    'CHROMOSOME_LIST': row.Sample_Name + '_CHR.txt.gz',\n",
    "    'DESCRIPTION': row.NCBI_sciname + ' chloroplast, complete genome'\n",
    "    }\n",
    "    manifest = pd.DataFrame.from_dict(manifest,orient='index')\n",
    "    manifest.to_csv(manifest_file,header=None,sep='\\t')\n",
    "    \n",
    "    # Submission command\n",
    "    commands.append('java -jar C:/Data/PAFTOL/ENA_submissions/webin-cli-4.3.0.jar ' + \n",
    "          '-username Webin-52995 -passwordFile C:/Data/PAFTOL/ENA_submissions/ena_pwd.txt -context genome -manifest ' +\n",
    "          row.Sample_Name + '_manifest.txt' + ' -validate') \n",
    "pd.DataFrame(commands).to_csv(sdir + '/Commands_WholePT.txt',index=False,header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
