{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import SeqIO; from Bio.Seq import Seq; from Bio.SeqRecord import SeqRecord; from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "import os\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# latest paftol export containing ENA sample accessions\n",
    "# https://www.ebi.ac.uk/ena/submit/report/swagger-ui.html#/samples/findAllUsingGET_13\n",
    "pft_file = '../PAFTOL_DB/2022-03-14_paftol_export_stats.csv'\n",
    "cptree_file = 'cpTree_v7/cpTree_v7_tree_clean.csv'\n",
    "org_dir = 'C:/Data/PAFTOL/Organelles/'\n",
    "\n",
    "# DataSource = 'GAP' \n",
    "# ENA_project_number = 'PRJEB49212'\n",
    "DataSource = 'PAFTOL' \n",
    "ENA_project_number = 'PRJEB35285' #https://www.ncbi.nlm.nih.gov/bioproject/588607"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3451 ['PAFTOL_004799', 'PAFTOL_000992']\n"
     ]
    }
   ],
   "source": [
    "## All Validated Samples\n",
    "R = pd.read_csv(cptree_file)\n",
    "R = R[R.DataSource==DataSource]\n",
    "subset = R.Sample_Name.to_list()\n",
    "print(R.shape[0],subset[:2])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load data and subset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "pdir = DataSource + '/'\n",
    "sdir = 'ENA_submissions/' + pdir\n",
    "if not os.path.exists(sdir):\n",
    "    os.makedirs(sdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9715\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>idPaftol</th>\n",
       "      <th>idSequencing</th>\n",
       "      <th>Sample_Name</th>\n",
       "      <th>Project</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>Taxonomical_Notes</th>\n",
       "      <th>SumContigLength</th>\n",
       "      <th>...</th>\n",
       "      <th>Coverage_Kmer</th>\n",
       "      <th>Coverage_base</th>\n",
       "      <th>Run_Time</th>\n",
       "      <th>maxK</th>\n",
       "      <th>NRepeat_Pattern</th>\n",
       "      <th>NPath</th>\n",
       "      <th>Redo_FastPlast</th>\n",
       "      <th>error_pt</th>\n",
       "      <th>error_nr</th>\n",
       "      <th>SeqID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>961</td>\n",
       "      <td>PAFTOL_000961</td>\n",
       "      <td>Pilot</td>\n",
       "      <td>Ranunculales</td>\n",
       "      <td>Circaeasteraceae</td>\n",
       "      <td>Circaeaster</td>\n",
       "      <td>agrestis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>138795.0</td>\n",
       "      <td>...</td>\n",
       "      <td>59.4</td>\n",
       "      <td>90.7</td>\n",
       "      <td>329.43</td>\n",
       "      <td>105.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pis_000961</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4</td>\n",
       "      <td>2571</td>\n",
       "      <td>PAFTOL_002571</td>\n",
       "      <td>Connaraceae</td>\n",
       "      <td>Oxalidales</td>\n",
       "      <td>Connaraceae</td>\n",
       "      <td>Rourea</td>\n",
       "      <td>minor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>213954.0</td>\n",
       "      <td>...</td>\n",
       "      <td>18.5</td>\n",
       "      <td>60.2</td>\n",
       "      <td>747.57</td>\n",
       "      <td>105.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pis_002571</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   idPaftol  idSequencing    Sample_Name      Project         order  \\\n",
       "0         2           961  PAFTOL_000961        Pilot  Ranunculales   \n",
       "1         4          2571  PAFTOL_002571  Connaraceae    Oxalidales   \n",
       "\n",
       "             family        genus   species Taxonomical_Notes  SumContigLength  \\\n",
       "0  Circaeasteraceae  Circaeaster  agrestis               NaN         138795.0   \n",
       "1       Connaraceae       Rourea     minor               NaN         213954.0   \n",
       "\n",
       "   ... Coverage_Kmer  Coverage_base  Run_Time   maxK  NRepeat_Pattern  NPath  \\\n",
       "0  ...          59.4           90.7    329.43  105.0              NaN    NaN   \n",
       "1  ...          18.5           60.2    747.57  105.0              NaN    NaN   \n",
       "\n",
       "   Redo_FastPlast  error_pt  error_nr       SeqID  \n",
       "0           False       NaN       NaN  Pis_000961  \n",
       "1           False       NaN       NaN  Pis_002571  \n",
       "\n",
       "[2 rows x 42 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load GetOrg result table\n",
    "Org_df = pd.read_csv(org_dir + DataSource + '/' + DataSource + '_Organelle_Recovery.csv').astype({'idSequencing':'int'})\n",
    "if DataSource == 'PAFTOL':\n",
    "    Org_df['SeqID'] = Org_df.Sample_Name.str.replace('PAFTOL_','Pis_')\n",
    "elif DataSource in ['GAP']:\n",
    "    Org_df['SeqID'] = Org_df.Sample_Name\n",
    "elif DataSource in ['SRA']:\n",
    "    sra2_pt = pd.read_csv('cpSRA_GetOrg/cpSRA_GetOrg_Organelle_Recovery.csv')\n",
    "    Org_df = pd.concat([Org_df,sra2_pt],ignore_index=True)\n",
    "    Org_df = Org_df.groupby('Sample_Name').head(1)\n",
    "    Org_df['SeqID'] = Org_df.Sample_Name\n",
    "print(Org_df.shape[0])\n",
    "Org_df[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3451 3451\n"
     ]
    }
   ],
   "source": [
    "if subset is not None:\n",
    "    Org_df = Org_df[Org_df.Sample_Name.isin(subset)]\n",
    "    print(Org_df.shape[0],Org_df.Sample_Name.nunique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\ProgramData\\Anaconda3\\lib\\site-packages\\IPython\\core\\interactiveshell.py:3146: DtypeWarning: Columns (29,35,45,57,74,83) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_Name</th>\n",
       "      <th>idPaftol</th>\n",
       "      <th>DataSource</th>\n",
       "      <th>Project</th>\n",
       "      <th>Order</th>\n",
       "      <th>Family</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Species</th>\n",
       "      <th>Taxonomical_Notes</th>\n",
       "      <th>ENASampleNum</th>\n",
       "      <th>idSequencing</th>\n",
       "      <th>ExternalSequenceID</th>\n",
       "      <th>ENAExpNumber</th>\n",
       "      <th>ENARunNumber</th>\n",
       "      <th>NumReads</th>\n",
       "      <th>NumRecoveredGenes</th>\n",
       "      <th>SumContigLength</th>\n",
       "      <th>NCBI_TaxID</th>\n",
       "      <th>NCBI_sciname</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PAFTOL_000564</td>\n",
       "      <td>1</td>\n",
       "      <td>PAFTOL</td>\n",
       "      <td>Asteraceae</td>\n",
       "      <td>Asterales</td>\n",
       "      <td>Asteraceae</td>\n",
       "      <td>Gymnolaena</td>\n",
       "      <td>sp.</td>\n",
       "      <td>Initially recorded as Gymnolaena litoralis but...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>564</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PAFTOL_000961</td>\n",
       "      <td>2</td>\n",
       "      <td>PAFTOL</td>\n",
       "      <td>Pilot</td>\n",
       "      <td>Ranunculales</td>\n",
       "      <td>Circaeasteraceae</td>\n",
       "      <td>Circaeaster</td>\n",
       "      <td>agrestis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERS5501508</td>\n",
       "      <td>961</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERX4839538</td>\n",
       "      <td>ERR5033191</td>\n",
       "      <td>625228.0</td>\n",
       "      <td>332.0</td>\n",
       "      <td>138795.0</td>\n",
       "      <td>39288.0</td>\n",
       "      <td>Circaeaster agrestis</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Sample_Name  idPaftol DataSource     Project         Order  \\\n",
       "0  PAFTOL_000564         1     PAFTOL  Asteraceae     Asterales   \n",
       "1  PAFTOL_000961         2     PAFTOL       Pilot  Ranunculales   \n",
       "\n",
       "             Family        Genus   Species  \\\n",
       "0        Asteraceae   Gymnolaena       sp.   \n",
       "1  Circaeasteraceae  Circaeaster  agrestis   \n",
       "\n",
       "                                   Taxonomical_Notes ENASampleNum  \\\n",
       "0  Initially recorded as Gymnolaena litoralis but...          NaN   \n",
       "1                                                NaN   ERS5501508   \n",
       "\n",
       "   idSequencing ExternalSequenceID ENAExpNumber ENARunNumber  NumReads  \\\n",
       "0           564                NaN          NaN          NaN       NaN   \n",
       "1           961                NaN   ERX4839538   ERR5033191  625228.0   \n",
       "\n",
       "   NumRecoveredGenes  SumContigLength  NCBI_TaxID          NCBI_sciname  \n",
       "0                NaN              NaN         NaN                   NaN  \n",
       "1              332.0         138795.0     39288.0  Circaeaster agrestis  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load paftol export and merge ena sample IDs\n",
    "db = pd.read_csv(pft_file)\n",
    "db = db[db.idPaftol.notnull()].astype({'idPaftol':'int','idSequencing':'int'})\n",
    "db = db[['Sample_Name','idPaftol', 'DataSource', 'Project', 'Order', 'Family', 'Genus', 'Species', 'Taxonomical_Notes', \n",
    "         'ENASampleNum', 'idSequencing', 'ExternalSequenceID', 'ENAExpNumber',\n",
    "       'ENARunNumber', 'NumReads', 'NumRecoveredGenes', 'SumContigLength','NCBI_TaxID', 'NCBI_sciname']]\n",
    "#        'title', 'taxId','id', 'alias', 'firstCreated', 'firstPublic', 'releaseStatus', 'secondaryId', 'scientificName', 'commonName']]\n",
    "db[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'idPaftol': 0, 'idSequencing': 0, 'Sample_Name': 0, 'Project': 0, 'order': 0, 'family': 0, 'genus': 0, 'species': 0, 'Taxonomical_Notes': 2231, 'SumContigLength': 363, 'sci_name': 0, 'log_pt': 0, 'log_nr': 0, 'pt_recovered': 0, 'Nseq_pt': 0, 'Sum_len_pt': 0, 'Nanybase_pt': 0, 'nr_recovered': 0, 'Nseq_nr': 27, 'Sum_len_nr': 27, 'Nanybase_nr': 27, 'Completed_Reads': 0, 'Completed_Extending': 0, 'Completed_Assembly': 0, 'Completed_Slimming': 0, 'Completed_Output': 0, 'Warning_Look': 0, 'Warning_Multi_Structure': 0, 'Warning_Self_Loop': 0, 'Info_Large_Repeats': 0, 'Info_Disentangling_Failed': 0, 'Assembly': 6, 'Coverage_Kmer': 635, 'Coverage_base': 8, 'Run_Time': 5, 'maxK': 6, 'NRepeat_Pattern': 3380, 'NPath': 3422, 'Redo_FastPlast': 0, 'error_pt': 3450, 'error_nr': 3432, 'SeqID': 0, 'DataSource': 0, 'ENASampleNum': 47, 'ENAExpNumber': 0, 'ENARunNumber': 0, 'NCBI_TaxID': 172, 'NCBI_sciname': 172}\n",
      "3451 3451\n"
     ]
    }
   ],
   "source": [
    "Org_df = pd.merge(Org_df,\n",
    "         db[['Sample_Name','DataSource','ENASampleNum','ENAExpNumber','ENARunNumber','NCBI_TaxID','NCBI_sciname']],\n",
    "        how='left',on='Sample_Name').groupby('Sample_Name').head(1)\n",
    "print(Org_df.isna().sum().to_dict())\n",
    "print(Org_df.shape[0],Org_df.Sample_Name.nunique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "geseq_dir = sdir + 'fasta/'\n",
    "if not os.path.exists(geseq_dir):\n",
    "    os.makedirs(geseq_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Whole plastomes\n",
    "### Annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "whole_pt = Org_df[(Org_df.Assembly.isin(['circular genome','circular genome with gaps']) & (Org_df.Sum_len_pt>80000))]\n",
    "print(whole_pt.shape[0],'samples',whole_pt.genus.nunique(),'genera')\n",
    "whole_pt[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Output fasta files \n",
    "for idx, row in whole_pt.iterrows():\n",
    "    seqs = list(SeqIO.parse(org_dir + DataSource + '/' + 'fasta_pt/' + row.Sample_Name + '_pt.fasta',format='fasta'))\n",
    "    # Verif only 1 sequence\n",
    "    if len(seqs) > 1:\n",
    "        print('error, more than 1 sequence')\n",
    "        pass\n",
    "    seq = seqs[0]\n",
    "    print(row.Sample_Name, row.SeqID, row.family, row.sci_name,len(seq),seq.seq.count('N'))\n",
    "    seq.id = row.SeqID\n",
    "    seq.description = ''\n",
    "    SeqIO.write(seq, geseq_dir + row.SeqID + '-circular.fasta',format='fasta')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Submit files to GeSeq https://chlorobox.mpimp-golm.mpg.de/geseq.html and extract zip file in wdir/GeSeq_output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "whole_pt.to_csv(sdir + DataSource + '_wholePT.csv',index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Contigs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(124, 24)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SeqID</th>\n",
       "      <th>Ini_Species</th>\n",
       "      <th>TaxID</th>\n",
       "      <th>mol_type</th>\n",
       "      <th>length</th>\n",
       "      <th>Ini_sci_name</th>\n",
       "      <th>sci_name_query</th>\n",
       "      <th>Similar_match</th>\n",
       "      <th>Duplicates</th>\n",
       "      <th>kew_id</th>\n",
       "      <th>...</th>\n",
       "      <th>sci_name</th>\n",
       "      <th>authors</th>\n",
       "      <th>rank</th>\n",
       "      <th>taxonomic_status</th>\n",
       "      <th>Ini_kew_id</th>\n",
       "      <th>Ini_taxonomic_status</th>\n",
       "      <th>Duplicate_type</th>\n",
       "      <th>Nsp</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_045294.1</td>\n",
       "      <td>Acorus tatarinowii</td>\n",
       "      <td>123564</td>\n",
       "      <td>genomic DNA</td>\n",
       "      <td>153296</td>\n",
       "      <td>Acorus tatarinowii</td>\n",
       "      <td>Acorus tatarinowii</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>77188885-1</td>\n",
       "      <td>...</td>\n",
       "      <td>Acorus calamus var. angustatus</td>\n",
       "      <td>Besser</td>\n",
       "      <td>VARIETY</td>\n",
       "      <td>Accepted</td>\n",
       "      <td>84028-1</td>\n",
       "      <td>Synonym</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>Acorales</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_053912.1</td>\n",
       "      <td>Aponogeton lakhonensis</td>\n",
       "      <td>217757</td>\n",
       "      <td>genomic DNA</td>\n",
       "      <td>154860</td>\n",
       "      <td>Aponogeton lakhonensis</td>\n",
       "      <td>Aponogeton lakhonensis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>82954-1</td>\n",
       "      <td>...</td>\n",
       "      <td>Aponogeton lakhonensis</td>\n",
       "      <td>A.Camus</td>\n",
       "      <td>SPECIES</td>\n",
       "      <td>Accepted</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>62</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>Alismatales</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         SeqID             Ini_Species   TaxID     mol_type  length  \\\n",
       "0  NC_045294.1      Acorus tatarinowii  123564  genomic DNA  153296   \n",
       "1  NC_053912.1  Aponogeton lakhonensis  217757  genomic DNA  154860   \n",
       "\n",
       "             Ini_sci_name          sci_name_query Similar_match  Duplicates  \\\n",
       "0      Acorus tatarinowii      Acorus tatarinowii           NaN       False   \n",
       "1  Aponogeton lakhonensis  Aponogeton lakhonensis           NaN       False   \n",
       "\n",
       "       kew_id  ...                        sci_name  authors     rank  \\\n",
       "0  77188885-1  ...  Acorus calamus var. angustatus   Besser  VARIETY   \n",
       "1     82954-1  ...          Aponogeton lakhonensis  A.Camus  SPECIES   \n",
       "\n",
       "  taxonomic_status Ini_kew_id Ini_taxonomic_status Duplicate_type Nsp  \\\n",
       "0         Accepted    84028-1              Synonym            NaN   2   \n",
       "1         Accepted        NaN                  NaN            NaN  62   \n",
       "\n",
       "           class        order  \n",
       "0  Magnoliopsida     Acorales  \n",
       "1  Magnoliopsida  Alismatales  \n",
       "\n",
       "[2 rows x 24 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "references = pd.read_csv('References/NCBI_plastid_ref/NCBI_plastid_ref_list_SELECTED.csv')\n",
    "print(references.shape)\n",
    "references.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3114 samples 2623 genera\n",
      "3114 samples 2623 genera\n",
      "100 samples 100 genera\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>idPaftol</th>\n",
       "      <th>idSequencing</th>\n",
       "      <th>Project</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>Taxonomical_Notes</th>\n",
       "      <th>SumContigLength</th>\n",
       "      <th>sci_name</th>\n",
       "      <th>...</th>\n",
       "      <th>Redo_FastPlast</th>\n",
       "      <th>error_pt</th>\n",
       "      <th>error_nr</th>\n",
       "      <th>SeqID</th>\n",
       "      <th>DataSource</th>\n",
       "      <th>ENASampleNum</th>\n",
       "      <th>ENAExpNumber</th>\n",
       "      <th>ENARunNumber</th>\n",
       "      <th>NCBI_TaxID</th>\n",
       "      <th>NCBI_sciname</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PAFTOL_007905</th>\n",
       "      <td>4838</td>\n",
       "      <td>7905</td>\n",
       "      <td>Celastraceae</td>\n",
       "      <td>Celastrales</td>\n",
       "      <td>Celastraceae</td>\n",
       "      <td>Euonymus</td>\n",
       "      <td>australianus</td>\n",
       "      <td>Euonymus-Celastrus-Haydenoxylon clade</td>\n",
       "      <td>201111.0</td>\n",
       "      <td>Euonymus australianus</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pis_007905</td>\n",
       "      <td>PAFTOL</td>\n",
       "      <td>ERS8702347</td>\n",
       "      <td>ERX7189840</td>\n",
       "      <td>ERR7619260</td>\n",
       "      <td>1089391.0</td>\n",
       "      <td>Euonymus australianus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PAFTOL_014479</th>\n",
       "      <td>8655</td>\n",
       "      <td>14479</td>\n",
       "      <td>Asteraceae</td>\n",
       "      <td>Asterales</td>\n",
       "      <td>Asteraceae</td>\n",
       "      <td>Olgaea</td>\n",
       "      <td>leucophylla</td>\n",
       "      <td>NaN</td>\n",
       "      <td>182634.0</td>\n",
       "      <td>Olgaea leucophylla</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pis_014479</td>\n",
       "      <td>PAFTOL</td>\n",
       "      <td>ERS8704497</td>\n",
       "      <td>ERX7191722</td>\n",
       "      <td>ERR7621142</td>\n",
       "      <td>591993.0</td>\n",
       "      <td>Olgaea leucophylla</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 47 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               idPaftol  idSequencing       Project        order  \\\n",
       "Sample_Name                                                        \n",
       "PAFTOL_007905      4838          7905  Celastraceae  Celastrales   \n",
       "PAFTOL_014479      8655         14479    Asteraceae    Asterales   \n",
       "\n",
       "                     family     genus       species  \\\n",
       "Sample_Name                                           \n",
       "PAFTOL_007905  Celastraceae  Euonymus  australianus   \n",
       "PAFTOL_014479    Asteraceae    Olgaea   leucophylla   \n",
       "\n",
       "                                   Taxonomical_Notes  SumContigLength  \\\n",
       "Sample_Name                                                             \n",
       "PAFTOL_007905  Euonymus-Celastrus-Haydenoxylon clade         201111.0   \n",
       "PAFTOL_014479                                    NaN         182634.0   \n",
       "\n",
       "                            sci_name  ...  Redo_FastPlast  error_pt  error_nr  \\\n",
       "Sample_Name                           ...                                       \n",
       "PAFTOL_007905  Euonymus australianus  ...           False       NaN       NaN   \n",
       "PAFTOL_014479     Olgaea leucophylla  ...           False       NaN       NaN   \n",
       "\n",
       "                    SeqID  DataSource  ENASampleNum  ENAExpNumber  \\\n",
       "Sample_Name                                                         \n",
       "PAFTOL_007905  Pis_007905      PAFTOL    ERS8702347    ERX7189840   \n",
       "PAFTOL_014479  Pis_014479      PAFTOL    ERS8704497    ERX7191722   \n",
       "\n",
       "               ENARunNumber  NCBI_TaxID           NCBI_sciname  \n",
       "Sample_Name                                                     \n",
       "PAFTOL_007905    ERR7619260   1089391.0  Euonymus australianus  \n",
       "PAFTOL_014479    ERR7621142    591993.0     Olgaea leucophylla  \n",
       "\n",
       "[2 rows x 47 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "incomplete_pt = Org_df[(Org_df.Assembly.isin(['circular genome','circular genome with gaps'])==False) &\n",
    "                   (Org_df.pt_recovered==True)]\n",
    "print(incomplete_pt.shape[0],'samples',incomplete_pt.genus.nunique(),'genera')\n",
    "incomplete_pt = incomplete_pt[incomplete_pt.Sum_len_pt<200000].set_index('Sample_Name')\n",
    "print(incomplete_pt.shape[0],'samples',incomplete_pt.genus.nunique(),'genera')\n",
    "### Random subset for test \n",
    "incomplete_pt = incomplete_pt.sample(n=100)\n",
    "print(incomplete_pt.shape[0],'samples',incomplete_pt.genus.nunique(),'genera')\n",
    "incomplete_pt[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SeqID</th>\n",
       "      <th>Ini_Species</th>\n",
       "      <th>TaxID</th>\n",
       "      <th>mol_type</th>\n",
       "      <th>length</th>\n",
       "      <th>Ini_sci_name</th>\n",
       "      <th>sci_name_query</th>\n",
       "      <th>Similar_match</th>\n",
       "      <th>Duplicates</th>\n",
       "      <th>kew_id</th>\n",
       "      <th>...</th>\n",
       "      <th>sci_name</th>\n",
       "      <th>authors</th>\n",
       "      <th>rank</th>\n",
       "      <th>taxonomic_status</th>\n",
       "      <th>Ini_kew_id</th>\n",
       "      <th>Ini_taxonomic_status</th>\n",
       "      <th>Duplicate_type</th>\n",
       "      <th>Nsp</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>NC_048982.1</td>\n",
       "      <td>Canarium album</td>\n",
       "      <td>300208</td>\n",
       "      <td>genomic DNA</td>\n",
       "      <td>163347</td>\n",
       "      <td>Canarium album</td>\n",
       "      <td>Canarium album</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>127264-1</td>\n",
       "      <td>...</td>\n",
       "      <td>Canarium album</td>\n",
       "      <td>(Lour.) DC.</td>\n",
       "      <td>SPECIES</td>\n",
       "      <td>Accepted</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Same_Taxon</td>\n",
       "      <td>122</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>Sapindales</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>NC_024929.1</td>\n",
       "      <td>Citrus aurantiifolia</td>\n",
       "      <td>159033</td>\n",
       "      <td>genomic DNA</td>\n",
       "      <td>159893</td>\n",
       "      <td>Citrus aurantiifolia</td>\n",
       "      <td>Citrus × aurantiifolia</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>59599-2</td>\n",
       "      <td>...</td>\n",
       "      <td>Citrus × aurantiifolia</td>\n",
       "      <td>(Christm.) Swingle</td>\n",
       "      <td>SPECIES</td>\n",
       "      <td>Accepted</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>94</td>\n",
       "      <td>Magnoliopsida</td>\n",
       "      <td>Sapindales</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           SeqID           Ini_Species   TaxID     mol_type  length  \\\n",
       "109  NC_048982.1        Canarium album  300208  genomic DNA  163347   \n",
       "110  NC_024929.1  Citrus aurantiifolia  159033  genomic DNA  159893   \n",
       "\n",
       "             Ini_sci_name          sci_name_query Similar_match  Duplicates  \\\n",
       "109        Canarium album          Canarium album           NaN        True   \n",
       "110  Citrus aurantiifolia  Citrus × aurantiifolia          True       False   \n",
       "\n",
       "       kew_id  ...                sci_name             authors     rank  \\\n",
       "109  127264-1  ...          Canarium album         (Lour.) DC.  SPECIES   \n",
       "110   59599-2  ...  Citrus × aurantiifolia  (Christm.) Swingle  SPECIES   \n",
       "\n",
       "    taxonomic_status Ini_kew_id Ini_taxonomic_status Duplicate_type  Nsp  \\\n",
       "109         Accepted        NaN                  NaN     Same_Taxon  122   \n",
       "110         Accepted        NaN                  NaN            NaN   94   \n",
       "\n",
       "             class       order  \n",
       "109  Magnoliopsida  Sapindales  \n",
       "110  Magnoliopsida  Sapindales  \n",
       "\n",
       "[2 rows x 24 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "references[references.order=='Sapindales']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "order_dc = {'Gunnerales':'Dilleniales','Picramniales':'Sapindales'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PAFTOL_007905 Celastrales Celastraceae\n",
      "\t Ref (1, 24): NC_045280.1 Celastrales Celastraceae\n",
      "\t 30 / 31 contigs mapped to reference\n",
      "\t sum len concat: 24986 / 25415 98.3 %\n",
      "PAFTOL_014479 Asterales Asteraceae\n",
      "\t Ref (1, 24): NC_044728.1 Asterales Asteraceae\n",
      "\t 30 / 30 contigs mapped to reference\n",
      "\t sum len concat: 36976 / 36976 100.0 %\n",
      "PAFTOL_006622 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 9 / 10 contigs mapped to reference\n",
      "\t sum len concat: 157832 / 158427 99.6 %\n",
      "PAFTOL_010719 Sapindales Meliaceae\n",
      "\t Ref (2, 24): NC_048982.1 Sapindales Burseraceae\n",
      "\t 26 / 26 contigs mapped to reference\n",
      "\t sum len concat: 54070 / 54070 100.0 %\n",
      "PAFTOL_012011 Ericales Sapotaceae\n",
      "\t Ref (2, 24): NC_041127.1 Ericales Styracaceae\n",
      "\t 8 / 8 contigs mapped to reference\n",
      "\t sum len concat: 127068 / 127068 100.0 %\n",
      "PAFTOL_004680 Asparagales Asphodelaceae\n",
      "\t Ref (2, 24): NC_053784.1 Asparagales Asparagaceae\n",
      "\t 47 / 47 contigs mapped to reference\n",
      "\t sum len concat: 58789 / 58789 100.0 %\n",
      "PAFTOL_018841 Rosales Rosaceae\n",
      "\t Ref (2, 24): NC_039731.1 Rosales Cannabaceae\n",
      "\t 44 / 45 contigs mapped to reference\n",
      "\t sum len concat: 54559 / 55177 98.9 %\n",
      "PAFTOL_015265 Rosales Rhamnaceae\n",
      "\t Ref (1, 24): NC_037151.1 Rosales Rhamnaceae\n",
      "\t 43 / 43 contigs mapped to reference\n",
      "\t sum len concat: 72056 / 72056 100.0 %\n",
      "PAFTOL_000975 Vitales Vitaceae\n",
      "\t Ref (1, 24): NC_039803.1 Vitales Vitaceae\n",
      "\t 18 / 19 contigs mapped to reference\n",
      "\t sum len concat: 38085 / 38641 98.6 %\n",
      "PAFTOL_004459 Rosales Moraceae\n",
      "\t Ref (2, 24): NC_039731.1 Rosales Cannabaceae\n",
      "\t 47 / 47 contigs mapped to reference\n",
      "\t sum len concat: 84859 / 84859 100.0 %\n",
      "PAFTOL_006015 Solanales Solanaceae\n",
      "\t Ref (1, 24): NC_039600.1 Solanales Solanaceae\n",
      "\t 41 / 41 contigs mapped to reference\n",
      "\t sum len concat: 47057 / 47057 100.0 %\n",
      "PAFTOL_023499 Brassicales Capparaceae\n",
      "\t Ref (2, 24): NC_049578.1 Brassicales Brassicaceae\n",
      "\t 36 / 36 contigs mapped to reference\n",
      "\t sum len concat: 36448 / 36448 100.0 %\n",
      "PAFTOL_011083 Dioscoreales Nartheciaceae\n",
      "\t Ref (2, 24): NC_036660.1 Dioscoreales Burmanniaceae\n",
      "\t 39 / 47 contigs mapped to reference\n",
      "\t sum len concat: 53318 / 56956 93.6 %\n",
      "PAFTOL_012079 Ericales Sapotaceae\n",
      "\t Ref (2, 24): NC_041127.1 Ericales Styracaceae\n",
      "\t 3 / 3 contigs mapped to reference\n",
      "\t sum len concat: 158668 / 158668 100.0 %\n",
      "PAFTOL_000972 Arecales Dasypogonaceae\n",
      "\t Ref (1, 24): NC_029970.1 Arecales Dasypogonaceae\n",
      "\t 32 / 32 contigs mapped to reference\n",
      "\t sum len concat: 104273 / 104273 100.0 %\n",
      "PAFTOL_007253 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 8 / 8 contigs mapped to reference\n",
      "\t sum len concat: 112384 / 112384 100.0 %\n",
      "PAFTOL_003041 Sapindales Sapindaceae\n",
      "\t Ref (2, 24): NC_048982.1 Sapindales Burseraceae\n",
      "\t 12 / 12 contigs mapped to reference\n",
      "\t sum len concat: 135290 / 135290 100.0 %\n",
      "PAFTOL_004743 Sapindales Sapindaceae\n",
      "\t Ref (2, 24): NC_048982.1 Sapindales Burseraceae\n",
      "\t 2 / 2 contigs mapped to reference\n",
      "\t sum len concat: 159089 / 159089 100.0 %\n",
      "PAFTOL_015149 Asterales Campanulaceae\n",
      "\t Ref (2, 24): NC_044728.1 Asterales Asteraceae\n",
      "\t 43 / 45 contigs mapped to reference\n",
      "\t sum len concat: 86474 / 87511 98.8 %\n",
      "PAFTOL_015615 Ericales Ericaceae\n",
      "\t Ref (2, 24): NC_041127.1 Ericales Styracaceae\n",
      "\t 12 / 16 contigs mapped to reference\n",
      "\t sum len concat: 132079 / 132948 99.3 %\n",
      "PAFTOL_004371 Rosales Moraceae\n",
      "\t Ref (2, 24): NC_039731.1 Rosales Cannabaceae\n",
      "\t 37 / 37 contigs mapped to reference\n",
      "\t sum len concat: 91047 / 91047 100.0 %\n",
      "PAFTOL_007973 Celastrales Celastraceae\n",
      "\t Ref (1, 24): NC_045280.1 Celastrales Celastraceae\n",
      "\t 32 / 32 contigs mapped to reference\n",
      "\t sum len concat: 29297 / 29297 100.0 %\n",
      "PAFTOL_004477 Rosales Moraceae\n",
      "\t Ref (2, 24): NC_039731.1 Rosales Cannabaceae\n",
      "\t 45 / 45 contigs mapped to reference\n",
      "\t sum len concat: 40552 / 40552 100.0 %\n",
      "PAFTOL_029857 Lamiales Orobanchaceae\n",
      "\t Ref (2, 24): NC_053620.1 Lamiales Lamiaceae\n",
      "\t 29 / 29 contigs mapped to reference\n",
      "\t sum len concat: 29062 / 29062 100.0 %\n",
      "PAFTOL_004257 Malpighiales Euphorbiaceae\n",
      "\t Ref (2, 24): NC_030601.1 Malpighiales Erythroxylaceae\n",
      "\t 42 / 44 contigs mapped to reference\n",
      "\t sum len concat: 57598 / 59338 97.1 %\n",
      "PAFTOL_021529 Gentianales Rubiaceae\n",
      "\t Ref (2, 24): NC_054359.1 Gentianales Gentianaceae\n",
      "\t 38 / 45 contigs mapped to reference\n",
      "\t sum len concat: 36659 / 39081 93.8 %\n",
      "PAFTOL_006111 Arecales Arecaceae\n",
      "\t Ref (1, 24): NC_045081.1 Arecales Arecaceae\n",
      "\t 17 / 17 contigs mapped to reference\n",
      "\t sum len concat: 127450 / 127450 100.0 %\n",
      "PAFTOL_010453 Sapindales Anacardiaceae\n",
      "\t Ref (2, 24): NC_048982.1 Sapindales Burseraceae\n",
      "\t 33 / 33 contigs mapped to reference\n",
      "\t sum len concat: 36015 / 36015 100.0 %\n",
      "PAFTOL_019329 Brassicales Cleomaceae\n",
      "\t Ref (2, 24): NC_049578.1 Brassicales Brassicaceae\n",
      "\t 53 / 53 contigs mapped to reference\n",
      "\t sum len concat: 71014 / 71014 100.0 %\n",
      "PAFTOL_019293 Asterales Asteraceae\n",
      "\t Ref (1, 24): NC_044728.1 Asterales Asteraceae\n",
      "\t 40 / 40 contigs mapped to reference\n",
      "\t sum len concat: 107635 / 107635 100.0 %\n",
      "PAFTOL_016305 Saxifragales Hamamelidaceae\n",
      "\t Ref (2, 24): NC_037885.1 Saxifragales Haloragaceae\n",
      "\t 30 / 31 contigs mapped to reference\n",
      "\t sum len concat: 24802 / 25914 95.7 %\n",
      "PAFTOL_002628 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 5 / 5 contigs mapped to reference\n",
      "\t sum len concat: 124883 / 124883 100.0 %\n",
      "PAFTOL_011889 Sapindales Rutaceae\n",
      "\t Ref (1, 24): NC_024929.1 Sapindales Rutaceae\n",
      "\t 35 / 35 contigs mapped to reference\n",
      "\t sum len concat: 49864 / 49864 100.0 %\n",
      "PAFTOL_012985 Malpighiales Peraceae\n",
      "\t Ref (2, 24): NC_030601.1 Malpighiales Erythroxylaceae\n",
      "\t 6 / 6 contigs mapped to reference\n",
      "\t sum len concat: 163713 / 163713 100.0 %\n",
      "PAFTOL_023435 Brassicales Caricaceae\n",
      "\t Ref (1, 24): NC_010323.1 Brassicales Caricaceae\n",
      "\t 56 / 56 contigs mapped to reference\n",
      "\t sum len concat: 68901 / 68901 100.0 %\n",
      "PAFTOL_014813 Asterales Asteraceae\n",
      "\t Ref (1, 24): NC_044728.1 Asterales Asteraceae\n",
      "\t 30 / 30 contigs mapped to reference\n",
      "\t sum len concat: 37113 / 37113 100.0 %\n",
      "PAFTOL_004390 Rosales Urticaceae\n",
      "\t Ref (2, 24): NC_039731.1 Rosales Cannabaceae\n",
      "\t 57 / 57 contigs mapped to reference\n",
      "\t sum len concat: 64543 / 64543 100.0 %\n",
      "PAFTOL_008709 Poales Poaceae\n",
      "\t Ref (1, 24): NC_035050.1 Poales Poaceae\n",
      "\t 1 / 1 contigs mapped to reference\n",
      "\t sum len concat: 140960 / 140960 100.0 %\n",
      "PAFTOL_010063 Commelinales Haemodoraceae\n",
      "\t Ref (1, 24): NC_040160.1 Commelinales Haemodoraceae\n",
      "\t 59 / 59 contigs mapped to reference\n",
      "\t sum len concat: 49309 / 49309 100.0 %\n",
      "PAFTOL_006252 Myrtales Melastomataceae\n",
      "\t Ref (2, 24): NC_042889.1 Myrtales Lythraceae\n",
      "\t 45 / 46 contigs mapped to reference\n",
      "\t sum len concat: 41978 / 42554 98.6 %\n",
      "PAFTOL_011879 Sapindales Rutaceae\n",
      "\t Ref (1, 24): NC_024929.1 Sapindales Rutaceae\n",
      "\t 35 / 35 contigs mapped to reference\n",
      "\t sum len concat: 104578 / 104578 100.0 %\n",
      "PAFTOL_006957 Solanales Solanaceae\n",
      "\t Ref (1, 24): NC_039600.1 Solanales Solanaceae\n",
      "\t 8 / 8 contigs mapped to reference\n",
      "\t sum len concat: 23266 / 23266 100.0 %\n",
      "PAFTOL_004696 Sapindales Sapindaceae\n",
      "\t Ref (2, 24): NC_048982.1 Sapindales Burseraceae\n",
      "\t 17 / 18 contigs mapped to reference\n",
      "\t sum len concat: 130890 / 131888 99.2 %\n",
      "PAFTOL_014801 Asterales Asteraceae\n",
      "\t Ref (1, 24): NC_044728.1 Asterales Asteraceae\n",
      "\t 19 / 20 contigs mapped to reference\n",
      "\t sum len concat: 35029 / 35703 98.1 %\n",
      "PAFTOL_015539 Ericales Ericaceae\n",
      "\t Ref (2, 24): NC_041127.1 Ericales Styracaceae\n",
      "\t 10 / 10 contigs mapped to reference\n",
      "\t sum len concat: 116994 / 116994 100.0 %\n",
      "PAFTOL_008650 Asparagales Asparagaceae\n",
      "\t Ref (1, 24): NC_053784.1 Asparagales Asparagaceae\n",
      "\t 26 / 26 contigs mapped to reference\n",
      "\t sum len concat: 40428 / 40428 100.0 %\n",
      "PAFTOL_007942 Celastrales Celastraceae\n",
      "\t Ref (1, 24): NC_045280.1 Celastrales Celastraceae\n",
      "\t 8 / 8 contigs mapped to reference\n",
      "\t sum len concat: 35390 / 35390 100.0 %\n",
      "PAFTOL_005730 Asterales Asteraceae\n",
      "\t Ref (1, 24): NC_044728.1 Asterales Asteraceae\n",
      "\t 51 / 51 contigs mapped to reference\n",
      "\t sum len concat: 50075 / 50075 100.0 %\n",
      "PAFTOL_008711 Poales Poaceae\n",
      "\t Ref (1, 24): NC_035050.1 Poales Poaceae\n",
      "\t 28 / 28 contigs mapped to reference\n",
      "\t sum len concat: 38258 / 38258 100.0 %\n",
      "PAFTOL_005727 Asterales Asteraceae\n",
      "\t Ref (1, 24): NC_044728.1 Asterales Asteraceae\n",
      "\t 14 / 14 contigs mapped to reference\n",
      "\t sum len concat: 124290 / 124290 100.0 %\n",
      "PAFTOL_006960 Solanales Solanaceae\n",
      "\t Ref (1, 24): NC_039600.1 Solanales Solanaceae\n",
      "\t 23 / 23 contigs mapped to reference\n",
      "\t sum len concat: 50411 / 50411 100.0 %\n",
      "PAFTOL_014513 Asterales Asteraceae\n",
      "\t Ref (1, 24): NC_044728.1 Asterales Asteraceae\n",
      "\t 36 / 37 contigs mapped to reference\n",
      "\t sum len concat: 39159 / 40083 97.7 %\n",
      "PAFTOL_005378 Brassicales Resedaceae\n",
      "\t Ref (2, 24): NC_049578.1 Brassicales Brassicaceae\n",
      "\t 6 / 6 contigs mapped to reference\n",
      "\t sum len concat: 124275 / 124275 100.0 %\n",
      "PAFTOL_006135 Arecales Arecaceae\n",
      "\t Ref (1, 24): NC_045081.1 Arecales Arecaceae\n",
      "\t 47 / 47 contigs mapped to reference\n",
      "\t sum len concat: 69980 / 69980 100.0 %\n",
      "PAFTOL_002655 Asparagales Iridaceae\n",
      "\t Ref (2, 24): NC_053784.1 Asparagales Asparagaceae\n",
      "\t 23 / 24 contigs mapped to reference\n",
      "\t sum len concat: 53351 / 54249 98.3 %\n",
      "PAFTOL_019233 Asterales Asteraceae\n",
      "\t Ref (1, 24): NC_044728.1 Asterales Asteraceae\n",
      "\t 10 / 10 contigs mapped to reference\n",
      "\t sum len concat: 124663 / 124663 100.0 %\n",
      "PAFTOL_015763 Ericales Lecythidaceae\n",
      "\t Ref (2, 24): NC_041127.1 Ericales Styracaceae\n",
      "\t 3 / 3 contigs mapped to reference\n",
      "\t sum len concat: 158435 / 158435 100.0 %\n",
      "PAFTOL_007313 Alismatales Araceae\n",
      "\t Ref (1, 24): NC_035499.1 Alismatales Araceae\n",
      "\t 37 / 37 contigs mapped to reference\n",
      "\t sum len concat: 112479 / 112479 100.0 %\n",
      "PAFTOL_001072 Asparagales Orchidaceae\n",
      "\t Ref (1, 24): NC_045400.1 Asparagales Orchidaceae\n",
      "\t 18 / 18 contigs mapped to reference\n",
      "\t sum len concat: 53718 / 53718 100.0 %\n",
      "PAFTOL_006292 Myrtales Melastomataceae\n",
      "\t Ref (2, 24): NC_042889.1 Myrtales Lythraceae\n",
      "\t 18 / 18 contigs mapped to reference\n",
      "\t sum len concat: 122560 / 122560 100.0 %\n",
      "PAFTOL_015271 Rosales Rosaceae\n",
      "\t Ref (2, 24): NC_039731.1 Rosales Cannabaceae\n",
      "\t 40 / 42 contigs mapped to reference\n",
      "\t sum len concat: 59937 / 60773 98.6 %\n",
      "PAFTOL_006145 Arecales Arecaceae\n",
      "\t Ref (1, 24): NC_045081.1 Arecales Arecaceae\n",
      "\t 35 / 35 contigs mapped to reference\n",
      "\t sum len concat: 113906 / 113906 100.0 %\n",
      "PAFTOL_010261 Apiales Apiaceae\n",
      "\t Ref (1, 24): NC_054296.1 Apiales Apiaceae\n",
      "\t 8 / 8 contigs mapped to reference\n",
      "\t sum len concat: 126997 / 126997 100.0 %\n",
      "PAFTOL_015595 Ericales Ericaceae\n",
      "\t Ref (2, 24): NC_041127.1 Ericales Styracaceae\n",
      "\t 13 / 13 contigs mapped to reference\n",
      "\t sum len concat: 126130 / 126130 100.0 %\n",
      "PAFTOL_005892 Gentianales Rubiaceae\n",
      "\t Ref (2, 24): NC_054359.1 Gentianales Gentianaceae\n",
      "\t 45 / 52 contigs mapped to reference\n",
      "\t sum len concat: 34693 / 37615 92.2 %\n",
      "PAFTOL_006892 Caryophyllales Aizoaceae\n",
      "\t Ref (2, 24): NC_046575.1 Caryophyllales Amaranthaceae\n",
      "\t 24 / 24 contigs mapped to reference\n",
      "\t sum len concat: 130219 / 130219 100.0 %\n",
      "PAFTOL_008832 Cucurbitales Tetramelaceae\n",
      "\t Ref (2, 24): NC_046385.1 Cucurbitales Begoniaceae\n",
      "\t 27 / 27 contigs mapped to reference\n",
      "\t sum len concat: 97135 / 97135 100.0 %\n",
      "PAFTOL_026123 Asparagales Orchidaceae\n",
      "\t Ref (1, 24): NC_045400.1 Asparagales Orchidaceae\n",
      "\t 45 / 45 contigs mapped to reference\n",
      "\t sum len concat: 43327 / 43327 100.0 %\n",
      "PAFTOL_010597 Sapindales Anacardiaceae\n",
      "\t Ref (2, 24): NC_048982.1 Sapindales Burseraceae\n",
      "\t 2 / 2 contigs mapped to reference\n",
      "\t sum len concat: 75794 / 75794 100.0 %\n",
      "PAFTOL_006700 Alismatales Araceae\n",
      "\t Ref (1, 24): NC_035499.1 Alismatales Araceae\n",
      "\t 63 / 63 contigs mapped to reference\n",
      "\t sum len concat: 59971 / 59971 100.0 %\n",
      "PAFTOL_008745 Cucurbitales Cucurbitaceae\n",
      "\t Ref (1, 24): NC_036135.1 Cucurbitales Cucurbitaceae\n",
      "\t 7 / 7 contigs mapped to reference\n",
      "\t sum len concat: 131278 / 131278 100.0 %\n",
      "PAFTOL_007902 Celastrales Celastraceae\n",
      "\t Ref (1, 24): NC_045280.1 Celastrales Celastraceae\n",
      "\t 39 / 41 contigs mapped to reference\n",
      "\t sum len concat: 48697 / 49663 98.1 %\n",
      "PAFTOL_011288 Apiales Apiaceae\n",
      "\t Ref (1, 24): NC_054296.1 Apiales Apiaceae\n",
      "\t 31 / 31 contigs mapped to reference\n",
      "\t sum len concat: 105929 / 105929 100.0 %\n",
      "PAFTOL_016791 Magnoliales Myristicaceae\n",
      "\t Ref (2, 24): NC_043867.1 Magnoliales Annonaceae\n",
      "\t 16 / 16 contigs mapped to reference\n",
      "\t sum len concat: 38078 / 38078 100.0 %\n",
      "PAFTOL_006376 Asparagales Amaryllidaceae\n",
      "\t Ref (2, 24): NC_053784.1 Asparagales Asparagaceae\n",
      "\t 53 / 53 contigs mapped to reference\n",
      "\t sum len concat: 68030 / 68030 100.0 %\n",
      "PAFTOL_001069 Crossosomatales Crossosomataceae\n",
      "\t Ref (1, 24): NC_052922.1 Crossosomatales Staphyleaceae\n",
      "\t 38 / 38 contigs mapped to reference\n",
      "\t sum len concat: 65191 / 65191 100.0 %\n",
      "PAFTOL_016651 Lamiales Lamiaceae\n",
      "\t Ref (1, 24): NC_053620.1 Lamiales Lamiaceae\n",
      "\t 59 / 59 contigs mapped to reference\n",
      "\t sum len concat: 84553 / 84553 100.0 %\n",
      "PAFTOL_008344 Arecales Arecaceae\n",
      "\t Ref (1, 24): NC_045081.1 Arecales Arecaceae\n",
      "\t 35 / 35 contigs mapped to reference\n",
      "\t sum len concat: 35218 / 35218 100.0 %\n",
      "PAFTOL_007444 Fabales Fabaceae\n",
      "\t Ref (1, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 10 / 11 contigs mapped to reference\n",
      "\t sum len concat: 128004 / 128298 99.8 %\n",
      "PAFTOL_008504 Myrtales Melastomataceae\n",
      "\t Ref (2, 24): NC_042889.1 Myrtales Lythraceae\n",
      "\t 41 / 43 contigs mapped to reference\n",
      "\t sum len concat: 49264 / 50282 98.0 %\n",
      "PAFTOL_010553 Sapindales Rutaceae\n",
      "\t Ref (1, 24): NC_024929.1 Sapindales Rutaceae\n",
      "\t 50 / 50 contigs mapped to reference\n",
      "\t sum len concat: 73839 / 73839 100.0 %\n",
      "PAFTOL_004808 Dioscoreales Dioscoreaceae\n",
      "\t Ref (1, 24): NC_034686.1 Dioscoreales Dioscoreaceae\n",
      "\t 26 / 26 contigs mapped to reference\n",
      "\t sum len concat: 35133 / 35133 100.0 %\n",
      "PAFTOL_015261 Rosales Rhamnaceae\n",
      "\t Ref (1, 24): NC_037151.1 Rosales Rhamnaceae\n",
      "\t 33 / 33 contigs mapped to reference\n",
      "\t sum len concat: 42171 / 42171 100.0 %\n",
      "PAFTOL_005317 Oxalidales Connaraceae\n",
      "\t Ref (2, 24): NC_042597.1 Oxalidales Cephalotaceae\n",
      "\t 10 / 10 contigs mapped to reference\n",
      "\t sum len concat: 23592 / 23592 100.0 %\n",
      "PAFTOL_005677 Asterales Asteraceae\n",
      "\t Ref (1, 24): NC_044728.1 Asterales Asteraceae\n",
      "\t 44 / 44 contigs mapped to reference\n",
      "\t sum len concat: 46691 / 46691 100.0 %\n",
      "PAFTOL_026053 Asparagales Orchidaceae\n",
      "\t Ref (1, 24): NC_045400.1 Asparagales Orchidaceae\n",
      "\t 29 / 29 contigs mapped to reference\n",
      "\t sum len concat: 111759 / 111759 100.0 %\n",
      "PAFTOL_029939 Lamiales Plantaginaceae\n",
      "\t Ref (2, 24): NC_053620.1 Lamiales Lamiaceae\n",
      "\t 22 / 22 contigs mapped to reference\n",
      "\t sum len concat: 66914 / 66914 100.0 %\n",
      "PAFTOL_016303 Saxifragales Hamamelidaceae\n",
      "\t Ref (2, 24): NC_037885.1 Saxifragales Haloragaceae\n",
      "\t 39 / 40 contigs mapped to reference\n",
      "\t sum len concat: 43267 / 44505 97.2 %\n",
      "PAFTOL_001347 Ericales Lecythidaceae\n",
      "\t Ref (2, 24): NC_041127.1 Ericales Styracaceae\n",
      "\t 10 / 10 contigs mapped to reference\n",
      "\t sum len concat: 130595 / 130595 100.0 %\n",
      "PAFTOL_011485 Laurales Monimiaceae\n",
      "\t Ref (2, 24): NC_037504.1 Laurales Calycanthaceae\n",
      "\t 41 / 41 contigs mapped to reference\n",
      "\t sum len concat: 81832 / 81832 100.0 %\n",
      "PAFTOL_010109 Lamiales Gesneriaceae\n",
      "\t Ref (2, 24): NC_053620.1 Lamiales Lamiaceae\n",
      "\t 47 / 47 contigs mapped to reference\n",
      "\t sum len concat: 88074 / 88074 100.0 %\n",
      "PAFTOL_004103 Gentianales Apocynaceae\n",
      "\t Ref (2, 24): NC_054359.1 Gentianales Gentianaceae\n",
      "\t 26 / 26 contigs mapped to reference\n",
      "\t sum len concat: 120977 / 120977 100.0 %\n",
      "PAFTOL_023513 Brassicales Tovariaceae\n",
      "\t Ref (2, 24): NC_049578.1 Brassicales Brassicaceae\n",
      "\t 40 / 40 contigs mapped to reference\n",
      "\t sum len concat: 121676 / 121676 100.0 %\n",
      "PAFTOL_015563 Ericales Ericaceae\n",
      "\t Ref (2, 24): NC_041127.1 Ericales Styracaceae\n",
      "\t 15 / 15 contigs mapped to reference\n",
      "\t sum len concat: 132519 / 132519 100.0 %\n",
      "PAFTOL_004832 Picramniales Picramniaceae\n",
      "\t Ref (2, 24): NC_048982.1 Sapindales Burseraceae\n",
      "\t 40 / 43 contigs mapped to reference\n",
      "\t sum len concat: 50682 / 52767 96.0 %\n",
      "PAFTOL_006522 Fabales Surianaceae\n",
      "\t Ref (2, 24): NC_025745.1 Fabales Fabaceae\n",
      "\t 23 / 24 contigs mapped to reference\n",
      "\t sum len concat: 144683 / 145340 99.5 %\n",
      "PAFTOL_011251 Apiales Apiaceae\n",
      "\t Ref (1, 24): NC_054296.1 Apiales Apiaceae\n",
      "\t 46 / 46 contigs mapped to reference\n",
      "\t sum len concat: 96206 / 96206 100.0 %\n",
      "PAFTOL_016943 Asterales Asteraceae\n",
      "\t Ref (1, 24): NC_044728.1 Asterales Asteraceae\n",
      "\t 43 / 43 contigs mapped to reference\n",
      "\t sum len concat: 79161 / 79161 100.0 %\n",
      "PAFTOL_006738 Alismatales Araceae\n",
      "\t Ref (1, 24): NC_035499.1 Alismatales Araceae\n",
      "\t 27 / 27 contigs mapped to reference\n",
      "\t sum len concat: 43154 / 43154 100.0 %\n",
      "PAFTOL_013403 Santalales Loranthaceae\n",
      "\t Ref (1, 24): NC_040862.1 Santalales Loranthaceae\n",
      "\t 15 / 15 contigs mapped to reference\n",
      "\t sum len concat: 113140 / 113140 100.0 %\n"
     ]
    }
   ],
   "source": [
    "commands = []\n",
    "for Sample_Name, row in incomplete_pt.iterrows():\n",
    "    print(Sample_Name,row.order,row.family)\n",
    "    ### Identify best reference\n",
    "    # check if reference in same family\n",
    "    best_ref = references[references.family==row.family]\n",
    "    if best_ref.shape[0]<1:\n",
    "        # otherwise order\n",
    "        best_ref = references[references.order==row.order]\n",
    "        if best_ref.shape[0]<1:\n",
    "            #Otherwise use closest order\n",
    "            best_ref = references[references.order==order_dc[row.order]]\n",
    "    print('\\t Ref',best_ref.shape, end = ': ')\n",
    "    best_ref = best_ref.iloc[0]\n",
    "    print(best_ref.SeqID,best_ref.order,best_ref.family)\n",
    "    incomplete_pt.loc[Sample_Name,'Ref-SeqID'] = best_ref.SeqID\n",
    "    incomplete_pt.loc[Sample_Name,'Ref-order'] = best_ref.order\n",
    "    incomplete_pt.loc[Sample_Name,'Ref-family'] = best_ref.family\n",
    "    \n",
    "    ### launch blastn\n",
    "    # program_path = 'C:/Program Files/NCBI/blast-BLAST_VERSION+/bin/blastn.exe'\n",
    "    program_path = 'blastn'\n",
    "    ref_path = 'References/NCBI_plastid_ref/' + best_ref['SeqID'] + '.fasta'\n",
    "    contigs_path = '../Organelles/' + DataSource + '/fasta_pt/' + Sample_Name + '_pt.fasta'\n",
    "    out_path = 'ENA_submissions/' + DataSource + '/out_blast/'+ Sample_Name + '.blast'\n",
    "    command = program_path + ' -query ' + contigs_path + ' -subject ' + ref_path + \\\n",
    "                ' -outfmt \"6 qseqid sseqid pident length slen qlen mismatch gapopen qstart qend sstart send evalue bitscore\"' + \\\n",
    "                ' -out ' + out_path\n",
    "    commands.append(command)\n",
    "    os.system(command)\n",
    "    \n",
    "    ### read blast output\n",
    "    blast_output = pd.read_table(out_path, sep='\\t',header=None)\n",
    "    blast_output.columns = ['qseqid','sseqid','pident','length','slen','qlen','mismatch','gapopen','qstart','qend','sstart','send','evalue','bitscore']\n",
    "#     print('\\t',blast_output.shape[0],'matched for', blast_output.qseqid.nunique(),'qseqid',end='; ')\n",
    "    incomplete_pt.loc[Sample_Name,'Blast-N_qseqid_match'] = blast_output.qseqid.nunique()\n",
    "    # Keep best match (by bitscore) per queried contig\n",
    "    blast_output = blast_output.sort_values('bitscore',ascending=False).groupby('qseqid').head(1)\n",
    "    blast_output['oriented_sstart'] = blast_output[['sstart','send']].min(axis=1); blast_output['oriented_ssend'] = blast_output[['sstart','send']].max(axis=1)\n",
    "    blast_output['Need_RevComp'] = blast_output.sstart > blast_output.oriented_sstart\n",
    "    blast_output = blast_output.sort_values('oriented_sstart').set_index('qseqid')\n",
    "#     print('Contigs to reverse complement:',blast_output.Need_RevComp.value_counts().to_dict())\n",
    "    \n",
    "    \n",
    "    ### Load contigs and write concatenated fasta\n",
    "    gapN = 100\n",
    "    records = SeqIO.to_dict(SeqIO.parse(contigs_path, \"fasta\"))\n",
    "    print('\\t',blast_output.shape[0],'/',len(records),'contigs mapped to reference')\n",
    "    concat_seq = ''\n",
    "    for qseqid, blast_row in blast_output.iterrows():\n",
    "        if blast_row.Need_RevComp==True:\n",
    "            concat_seq += records[qseqid].seq.reverse_complement()\n",
    "        else:\n",
    "            concat_seq += records[qseqid].seq \n",
    "        concat_seq += Seq(gapN * 'N')\n",
    "    concat_seq = concat_seq[:-gapN] # Remove trailing Ns\n",
    "    print('\\t sum len concat:',blast_output.qlen.sum(),'/',int(row.Sum_len_pt),round(blast_output.qlen.sum()/int(row.Sum_len_pt)*100,1),'%')\n",
    "    incomplete_pt.loc[Sample_Name,'Concat-len'] = blast_output.qlen.sum()\n",
    "    incomplete_pt.loc[Sample_Name,'Concat-%_mapped'] = round(blast_output.qlen.sum()/int(row.Sum_len_pt)*100,1)\n",
    "    incomplete_pt.loc[Sample_Name,'Concat-len_with_Ns'] = len(concat_seq)\n",
    "    Concat_record = SeqRecord(\n",
    "        concat_seq,\n",
    "        id=row.SeqID + '_concat_pt',\n",
    "        name=Sample_Name + '_concat_pt',\n",
    "        description=''\n",
    "    )\n",
    "    SeqIO.write(Concat_record,'ENA_submissions/' + DataSource + '/fasta/' + row.SeqID + '-concat.fasta','fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "incomplete_pt.to_csv(sdir + DataSource + '_concatPT.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ENA submission\n",
    "https://ena-docs.readthedocs.io/en/latest/submit/fileprep/assembly.html?highlight=chloroplast#\n",
    "\n",
    "cd C:\\Data\\PAFTOL\\cpDNA\\ENA_submissions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "paper_title = 'Building the largest, curated plastid genome dataset and phylogeny by leveraging nuclear-oriented sequence data'\n",
    "paper_authors = 'KEVIN LEEMPOEL1, PAUL BAILEY1, ELAINE FRANCOSO1, BERTA GALLEGO1, De-Zhu Li2,3,4, Hong-Tao Li2,3, OLIVIER MAURIN1, SHYAMALI ROY1,' +  \\\n",
    "            'Jun-Bo Yang2,3, Ting-Shuang Yi2,3, ALEXANDRE R. ZUNTINI1, WOLF L. EISERHARDT1,5,' +  \\\n",
    "            'PAUL J. KERSEY1, ILIA J. LEITCH1, FELIX FOREST1 AND WILLIAM J. BAKER1'\n",
    "paper_journal = 'Unpublished'\n",
    "paper_id = 'NA'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "437\n",
      "428\n",
      "410 {'circular genome': 315, 'scaffold': 93, 'circular genome with gaps': 2}\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-48-9115cba50240>:1: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n",
      "  submit_pt = pd.concat([pd.read_csv(sdir + DataSource + '_wholePT.csv'), pd.read_csv(sdir + DataSource + '_concatPT.csv')],ignore_index=True)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Assembly</th>\n",
       "      <th>Blast-N_qseqid_match</th>\n",
       "      <th>Completed_Assembly</th>\n",
       "      <th>Completed_Extending</th>\n",
       "      <th>Completed_Output</th>\n",
       "      <th>Completed_Reads</th>\n",
       "      <th>Completed_Slimming</th>\n",
       "      <th>Concat-%_mapped</th>\n",
       "      <th>Concat-len</th>\n",
       "      <th>Concat-len_with_Ns</th>\n",
       "      <th>...</th>\n",
       "      <th>idPaftol</th>\n",
       "      <th>idSequencing</th>\n",
       "      <th>log_nr</th>\n",
       "      <th>log_pt</th>\n",
       "      <th>maxK</th>\n",
       "      <th>nr_recovered</th>\n",
       "      <th>order</th>\n",
       "      <th>pt_recovered</th>\n",
       "      <th>sci_name</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>circular genome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>961</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>105.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Ranunculales</td>\n",
       "      <td>True</td>\n",
       "      <td>Circaeaster agrestis</td>\n",
       "      <td>agrestis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>circular genome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>11</td>\n",
       "      <td>12989</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>105.0</td>\n",
       "      <td>True</td>\n",
       "      <td>Malpighiales</td>\n",
       "      <td>True</td>\n",
       "      <td>Gynocardia odorata</td>\n",
       "      <td>odorata</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 55 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Assembly  Blast-N_qseqid_match  Completed_Assembly  \\\n",
       "0  circular genome                   NaN                True   \n",
       "1  circular genome                   NaN                True   \n",
       "\n",
       "   Completed_Extending  Completed_Output  Completed_Reads  Completed_Slimming  \\\n",
       "0                 True              True             True                True   \n",
       "1                 True              True             True                True   \n",
       "\n",
       "   Concat-%_mapped  Concat-len  Concat-len_with_Ns  ...  idPaftol  \\\n",
       "0              NaN         NaN                 NaN  ...         2   \n",
       "1              NaN         NaN                 NaN  ...        11   \n",
       "\n",
       "   idSequencing log_nr log_pt   maxK nr_recovered         order  pt_recovered  \\\n",
       "0           961   True   True  105.0         True  Ranunculales          True   \n",
       "1         12989   True   True  105.0         True  Malpighiales          True   \n",
       "\n",
       "               sci_name   species  \n",
       "0  Circaeaster agrestis  agrestis  \n",
       "1    Gynocardia odorata   odorata  \n",
       "\n",
       "[2 rows x 55 columns]"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "submit_pt = pd.concat([pd.read_csv(sdir + DataSource + '_wholePT.csv'), pd.read_csv(sdir + DataSource + '_concatPT.csv')],ignore_index=True)\n",
    "# submit_pt = pd.read_csv(sdir + DataSource + '_concatPT.csv')\n",
    "print(submit_pt.shape[0])\n",
    "submit_pt = submit_pt[(submit_pt.ENASampleNum.notnull()) & (submit_pt.ENARunNumber.notnull())]\n",
    "print(submit_pt.shape[0])\n",
    "submit_pt = submit_pt[submit_pt.NCBI_sciname.notnull()]\n",
    "print(submit_pt.shape[0],submit_pt.Assembly.value_counts().to_dict())\n",
    "submit_pt.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "submit_pt = submit_pt[submit_pt.ENASampleNum=='ERS10966407']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1678470, 4)\n",
      "(1172782, 4)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>len</th>\n",
       "      <th>Sample_Name</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GAP_026547-rrn23</td>\n",
       "      <td>2812</td>\n",
       "      <td>GAP_026547</td>\n",
       "      <td>rrn23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GAP_026547-rrn16</td>\n",
       "      <td>1491</td>\n",
       "      <td>GAP_026547</td>\n",
       "      <td>rrn16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              seqid   len Sample_Name   gene\n",
       "0  GAP_026547-rrn23  2812  GAP_026547  rrn23\n",
       "1  GAP_026547-rrn16  1491  GAP_026547  rrn16"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_seq = pd.read_csv('cpTree_v7/AllSamples_Allgenes_All_SeqTable.csv')\n",
    "print(all_seq.shape)\n",
    "genes_82 = pd.read_table('cp_genes_82.txt',header=None).rename(columns={0:'gene'})\n",
    "all_seq = all_seq[all_seq.gene.isin(genes_82.gene)]\n",
    "print(all_seq.shape)\n",
    "all_seq.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "def list_genes(record, ls_types = ['gene','CDS','rRNA']):\n",
    "    # list gene names\n",
    "    feat_ls = []; idx = 0\n",
    "    for feature in record.features:\n",
    "        if feature.type in ls_types:\n",
    "            feat_ls.append(feature.qualifiers['gene'][0])\n",
    "        idx += 1\n",
    "    return list(set(feat_ls))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "439\n",
      "PAFTOL_029857 GeSeqJob-20220305-120228_Pis_029857_concat_pt_GenBank.gb:1 31862 20 / 15 133 % ['atpA']\n"
     ]
    }
   ],
   "source": [
    "# Rename and modify genbank files, convert to embl flat file\n",
    "if not os.path.exists(sdir + '/Genbank_files/'):\n",
    "    os.makedirs(sdir + '/Genbank_files/')\n",
    "gb_files = [file for file in os.listdir(sdir + 'geseq/') if file.endswith('.gb')]\n",
    "print(len(gb_files))\n",
    "for idx, row in submit_pt.iterrows():\n",
    "    idx_file = [i for i, x in enumerate([row.SeqID in file for file in gb_files]) if x]\n",
    "    print(row.Sample_Name,gb_files[idx_file[0]],end=':')\n",
    "    records = list(SeqIO.parse(sdir + 'geseq/' + gb_files[idx_file[0]], \"genbank\"))\n",
    "    record = records[0]\n",
    "    record_genes = list_genes(record); record_genes = [gene for gene in record_genes if gene in genes_82.gene.to_list()]\n",
    "    blast_genes = all_seq[all_seq.Sample_Name==row.Sample_Name]\n",
    "    print(len(records), len(records[0]), len(record_genes),'/',blast_genes.gene.nunique(),\n",
    "         int(len(record_genes)/blast_genes.gene.nunique()*100),'%',blast_genes[blast_genes.gene.isin(record_genes)==False].gene.unique())\n",
    "    try:\n",
    "        if row.Assembly in ['circular genome','circular genome with gaps']:\n",
    "            record.description = row.ENARunNumber + ' chloroplast, complete genome'\n",
    "        elif row.Assembly in ['scaffold']:\n",
    "            record.description = row.ENARunNumber + ' chloroplast, partial genome (concatenated contigs)'\n",
    "        submit_pt.loc[idx,'ENA_SeqID'] = row.ENARunNumber + '_pt'\n",
    "        record.id = submit_pt.loc[idx,'ENA_SeqID']\n",
    "        record.name = submit_pt.loc[idx,'ENA_SeqID']\n",
    "        record.version = submit_pt.loc[idx,'ENA_SeqID'] + '.1'\n",
    "        record.annotations['source'] = row.ENASampleNum + ', ' + row.sci_name + ', isolate'\n",
    "        record.annotations['accessions'] = row.ENASampleNum\n",
    "        record.annotations['organism'] = row.NCBI_sciname\n",
    "        record.annotations[\"data_file_division\"]=\"PLN\";\n",
    "        record.annotations.pop('taxonomy')\n",
    "        record.annotations['references'][0].title = paper_title    \n",
    "        record.annotations['references'][0].authors = paper_authors    \n",
    "        record.annotations['references'][0].journal = paper_journal      \n",
    "        record.annotations['references'][0].pubmed_id = paper_id    \n",
    "        record.features[0].qualifiers['organism'] = row.NCBI_sciname\n",
    "        record.features[0].qualifiers['db_xref'] = ['taxon:' + str(int(row['NCBI_TaxID']))]\n",
    "        SeqIO.write(record,sdir + 'Genbank_files/' + row.SeqID + '_pt.gb',format='genbank')\n",
    "    except:\n",
    "        print('issue with ',gb_files[idx_file[0]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [],
   "source": [
    "def del_features(record, feat_del_ls = ['translation','info','annotator']):\n",
    "    for feature in record.features:\n",
    "        for feat_to_del in feat_del_ls:\n",
    "            if feat_to_del in feature.qualifiers:\n",
    "                del feature.qualifiers[feat_to_del]\n",
    "    return record\n",
    "\n",
    "def del_fragment(record, ls_types = ['gene','CDS','rRNA']):\n",
    "    # list gene names\n",
    "    ls_feat = []\n",
    "    for feature in record.features:\n",
    "        if feature.type in ls_types:\n",
    "            ls_feat.append(feature.qualifiers['gene'][0])\n",
    "    ls_feat = [ifeat for ifeat in ls_feat if 'fragment' not in ifeat]\n",
    "\n",
    "    # remove fragment feature if full feature is present\n",
    "    rm_count = 0\n",
    "    for feature in record.features:\n",
    "        if feature.type in ls_types:\n",
    "            if 'fragment' in feature.qualifiers['gene'][0]:\n",
    "                if feature.qualifiers['gene'][0].split('-')[0] in ls_feat:\n",
    "                    rm_count += 1\n",
    "                    record.features.remove(feature)\n",
    "    print('removed',rm_count,'partial features')\n",
    "    return record\n",
    "\n",
    "def del_longfeat(record, ls_types = ['gene','CDS','rRNA'], len_thrs = 5000):\n",
    "    for feature in record.features:\n",
    "        if feature.type in ls_types:\n",
    "            if len(feature)>len_thrs:\n",
    "                print('removing ',feature.qualifiers['gene'],len(feature))\n",
    "                record.features.remove(feature)\n",
    "    return record\n",
    "\n",
    "def add_gap_features(record, gapN = 100):\n",
    "    # find gaps in record\n",
    "    gaps_idx = [m.start() for m in re.finditer(gapN * 'N', str(record.seq))]\n",
    "    print(len(gaps_idx),'gaps')\n",
    "    if len(gaps_idx)>0:\n",
    "        ## for each gap, add a sequence feature\n",
    "        for gap_pos in gaps_idx:\n",
    "            # find index of gap in features\n",
    "            features_start = [feat.location.start for feat in record.features]\n",
    "            gap_feature_idx = [i for i in range(len(features_start)) if features_start[i] > gap_pos]\n",
    "            \n",
    "            gap_feat = SeqFeature(FeatureLocation(gap_pos,gap_pos+gapN), strand=1, type='assembly_gap')\n",
    "            gap_feat.qualifiers['gap_type']='within scaffold'; \n",
    "            gap_feat.qualifiers['estimated_length']='unknown'\n",
    "            gap_feat.qualifiers['linkage_evidence']='unspecified'\n",
    "            if len(gap_feature_idx)>0:\n",
    "                gap_feature_idx = gap_feature_idx[0]\n",
    "                record.features.insert(gap_feature_idx, gap_feat)\n",
    "            elif len(gap_feature_idx)==0: #if position is further than any other feature, append at the end of feature list\n",
    "                print(\"END\")\n",
    "                record.features.append(gap_feat)\n",
    "    return record\n",
    "\n",
    "def gzip_file(in_file):\n",
    "    import gzip\n",
    "    import shutil\n",
    "    f_gz = in_file + '.gz'\n",
    "    with open(in_file, 'rb') as f_in:\n",
    "        with gzip.open(f_gz, 'wb') as f_out:\n",
    "            shutil.copyfileobj(f_in, f_out)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pis_029857 ERR9230283_pt > 31862\n",
      "removed 1 partial features\n",
      "removed 1 partial features\n",
      "28 gaps\n"
     ]
    }
   ],
   "source": [
    "# ENA files\n",
    "if not os.path.exists(sdir + '/EMBL_files/'):\n",
    "    os.makedirs(sdir + '/EMBL_files/')\n",
    "commands = []\n",
    "for idx, row in submit_pt.iterrows():\n",
    "    ### Convert .gb to EMBL flat files\n",
    "    file_gb = sdir + 'Genbank_files/' + row.SeqID + '_pt.gb'\n",
    "    print(row.SeqID,row.ENA_SeqID,end=' > ')\n",
    "#     try:\n",
    "    record = list(SeqIO.parse(file_gb, \"genbank\"))[0]\n",
    "    print(len(record))\n",
    "    record_clean = del_features(record = record)\n",
    "    record_clean = del_fragment(record = record_clean); record_clean = del_fragment(record = record_clean);\n",
    "    record_clean = del_longfeat(record = record_clean,len_thrs=8000);\n",
    "    record_clean = add_gap_features(record = record_clean)\n",
    "    file_embl = sdir + '/EMBL_files/' + row.ENA_SeqID + '.embl'\n",
    "    SeqIO.write(record_clean, file_gb, \"genbank\")\n",
    "    SeqIO.write(record_clean, file_embl, \"embl\")\n",
    "    gzip_file(file_embl)\n",
    "\n",
    "    ### Chromosome list file\n",
    "    # https://ena-docs.readthedocs.io/en/latest/submit/fileprep/assembly.html?highlight=chloroplast#chromosome-list-file\n",
    "    if row.Assembly in ['circular genome','circular genome with gaps']:\n",
    "        list_file=row.ENA_SeqID + '\tPT\tCircular-Chromosome\tChloroplast'       \n",
    "    elif row.Assembly in ['scaffold']:\n",
    "        list_file=row.ENA_SeqID + '\tPT\tLinear-Chromosome\tChloroplast'\n",
    "    with open(file_embl.replace('.embl','_CHR.txt'), 'wb') as f:\n",
    "        f.write(list_file.encode(\"ascii\"))\n",
    "    gzip_file(file_embl.replace('.embl','_CHR.txt'))\n",
    "\n",
    "    ### Manifest file \n",
    "    # https://ena-docs.readthedocs.io/en/latest/submit/assembly/genome.html \n",
    "    manifest_file = file_embl.replace('.embl','_manifest.txt')\n",
    "    if row.Assembly in ['circular genome','circular genome with gaps']:\n",
    "        tmp_description=' chloroplast, complete genome' \n",
    "    elif row.Assembly in ['scaffold']:\n",
    "        tmp_description=' chloroplast, partial genome (concatenated contigs)' \n",
    "    manifest={\n",
    "        'STUDY':ENA_project_number,\n",
    "        'SAMPLE':row.ENASampleNum,\n",
    "        'ASSEMBLYNAME':row.ENA_SeqID,\n",
    "        'ASSEMBLY_TYPE':'clone or isolate',\n",
    "        'COVERAGE':row.Coverage_base,\n",
    "        'PROGRAM':'GetOrganelle v1.7.5 +  GeSeq 2.03',\n",
    "        'PLATFORM':'ILLUMINA',\n",
    "        'MOLECULETYPE':'genomic DNA',\n",
    "        'FLATFILE': row.ENA_SeqID + '.embl.gz',\n",
    "        'CHROMOSOME_LIST': row.ENA_SeqID + '_CHR.txt.gz',\n",
    "        'DESCRIPTION': row.NCBI_sciname + tmp_description\n",
    "    }\n",
    "    manifest = pd.DataFrame.from_dict(manifest,orient='index')\n",
    "    manifest.to_csv(manifest_file,header=None,sep='\\t')\n",
    "\n",
    "    # Submission command\n",
    "    commands.append('java -jar C:/Data/PAFTOL/ENA_submissions/webin-cli-4.3.0.jar ' + \n",
    "          '-username Webin-52995 -passwordFile C:/Data/PAFTOL/ENA_submissions/ena_pwd.txt -context genome -manifest ' +\n",
    "          row.ENA_SeqID + '_manifest.txt' + ' -validate') \n",
    "#     except:\n",
    "#         print('issue',row.SeqID)\n",
    "pd.DataFrame(commands).to_csv(sdir + '/ENA_Submission_Commands_' + DataSource + '.txt',index=False,header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ENA_submissions/PAFTOL/Genbank_files/Pis_029857_pt.gb'"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file_gb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[SeqFeature(FeatureLocation(ExactPosition(0), ExactPosition(31862), strand=1), type='source'),\n",
       " SeqFeature(CompoundLocation([FeatureLocation(ExactPosition(31310), ExactPosition(31862), strand=1), FeatureLocation(ExactPosition(0), ExactPosition(6), strand=1)], 'join'), type='gene', location_operator='join'),\n",
       " SeqFeature(CompoundLocation([FeatureLocation(ExactPosition(31310), ExactPosition(31862), strand=1), FeatureLocation(ExactPosition(0), ExactPosition(6), strand=1)], 'join'), type='CDS', location_operator='join'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(615), ExactPosition(1334), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(615), ExactPosition(1334), strand=-1), type='intron'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(1727), ExactPosition(1973), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(1727), ExactPosition(1973), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(2017), ExactPosition(2359), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(2017), ExactPosition(2359), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(2776), ExactPosition(4606), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(2776), ExactPosition(4606), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(5931), ExactPosition(6624), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(5931), ExactPosition(6624), strand=1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(6868), ExactPosition(6942), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(6868), ExactPosition(6942), strand=-1), type='tRNA'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(7099), ExactPosition(7402), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(7099), ExactPosition(7402), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(8038), ExactPosition(9367), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(8038), ExactPosition(9367), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(9774), ExactPosition(11007), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(9774), ExactPosition(11007), strand=1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(11727), ExactPosition(12636), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(11727), ExactPosition(12636), strand=1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(12392), ExactPosition(13305), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(13234), ExactPosition(13305), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(13234), ExactPosition(13305), strand=-1), type='exon'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(12392), ExactPosition(13234), strand=-1), type='intron'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(12653), ExactPosition(12788), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(12653), ExactPosition(12788), strand=1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(14206), ExactPosition(14770), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(14206), ExactPosition(14770), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(15030), ExactPosition(16133), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(15464), ExactPosition(16133), strand=-1), type='intron'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(15464), ExactPosition(16133), strand=-1), type='exon'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(15030), ExactPosition(15464), strand=-1), type='intron'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(16282), ExactPosition(16405), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(16282), ExactPosition(16405), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(17905), ExactPosition(17979), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(17905), ExactPosition(17979), strand=-1), type='tRNA'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(18221), ExactPosition(18342), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(18221), ExactPosition(18342), strand=-1), type='rRNA'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(18593), ExactPosition(18696), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(18593), ExactPosition(18696), strand=-1), type='rRNA'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(18794), ExactPosition(21605), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(18794), ExactPosition(21048), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(18794), ExactPosition(21048), strand=-1), type='rRNA'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(18993), ExactPosition(21605), strand=-1), type='rRNA'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(18993), ExactPosition(21605), strand=-1), type='exon'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(18794), ExactPosition(18993), strand=-1), type='intron'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(21782), ExactPosition(22668), strand=-1), type='gene'),\n",
       " SeqFeature(CompoundLocation([FeatureLocation(ExactPosition(22630), ExactPosition(22668), strand=-1), FeatureLocation(ExactPosition(21782), ExactPosition(21817), strand=-1)], 'join'), type='tRNA', location_operator='join'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(22630), ExactPosition(22668), strand=-1), type='exon'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(21817), ExactPosition(22630), strand=-1), type='intron'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(21782), ExactPosition(21817), strand=-1), type='exon'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(22732), ExactPosition(23752), strand=-1), type='gene'),\n",
       " SeqFeature(CompoundLocation([FeatureLocation(ExactPosition(23715), ExactPosition(23752), strand=-1), FeatureLocation(ExactPosition(22732), ExactPosition(22767), strand=-1)], 'join'), type='tRNA', location_operator='join'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(23715), ExactPosition(23752), strand=-1), type='exon'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(22767), ExactPosition(23715), strand=-1), type='intron'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(22732), ExactPosition(22767), strand=-1), type='exon'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(24050), ExactPosition(25541), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(24050), ExactPosition(25541), strand=-1), type='rRNA'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(25768), ExactPosition(25840), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(25768), ExactPosition(25840), strand=-1), type='tRNA'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(27030), ExactPosition(27824), strand=1), type='gene'),\n",
       " SeqFeature(CompoundLocation([FeatureLocation(ExactPosition(27030), ExactPosition(27262), strand=1), FeatureLocation(ExactPosition(27798), ExactPosition(27824), strand=1)], 'join'), type='CDS', location_operator='join'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(27030), ExactPosition(27262), strand=1), type='exon'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(27262), ExactPosition(27798), strand=1), type='intron'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(27798), ExactPosition(27824), strand=1), type='exon'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(28399), ExactPosition(29090), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(28399), ExactPosition(29090), strand=1), type='intron'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(28902), ExactPosition(28983), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(28902), ExactPosition(28983), strand=1), type='tRNA'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(29671), ExactPosition(30763), strand=-1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(29671), ExactPosition(30763), strand=-1), type='CDS'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(31010), ExactPosition(31292), strand=1), type='gene'),\n",
       " SeqFeature(FeatureLocation(ExactPosition(31010), ExactPosition(31292), strand=1), type='CDS')]"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "record_clean.features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.DataFrame(commands).to_csv(sdir + '/ENA_Submission_Commands_' + DataSource + '.txt',index=False,header=None)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
