{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_qualifier(feature, qualifier):\n",
    "    try:\n",
    "        return feature.qualifiers[qualifier][0]\n",
    "    except:\n",
    "        return np.nan\n",
    "def get_features(gb_file):\n",
    "    seq_dc=[]\n",
    "    for rec in SeqIO.parse(gb_file, \"genbank\"):\n",
    "        for feature in rec.features:\n",
    "            feat_dc = {'type':feature.type}\n",
    "            feat_dc['gene'] = get_qualifier(feature,'gene')\n",
    "            feat_dc['product'] = get_qualifier(feature,'product')\n",
    "#             feat_dc['info'] = get_qualifier(feature,'info')\n",
    "            if feat_dc['gene'] is np.nan:\n",
    "                feat_dc['gene'] = get_qualifier(feature,'note')\n",
    "            feat_dc['len'] = len(feature.location.extract(rec).seq)\n",
    "            feat_dc['location'] = str(feature.location)\n",
    "            feat_dc['seq'] = str(feature.location.extract(rec).seq)\n",
    "            seq_dc.append(feat_dc)\n",
    "    return pd.DataFrame(seq_dc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_fasta(seq_df, sample_stats, fasta_file):\n",
    "    Records = []\n",
    "    for idx, row in seq_df.iterrows():\n",
    "        record = SeqRecord( Seq(row.seq), id=sample_stats['Sample'].replace('-','_') + '-' + row.gene.replace('-','_'), \n",
    "                description= ';product=' + str(row['product']) + ', location=' + str(row['location']) + ';' )\n",
    "        Records.append(record)\n",
    "    SeqIO.write(Records,fasta_file,format='fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "124\n",
      "{'Sample': 'NC_003386.1', 'Nfeatures_raw': 546, 'Ngenes_raw': 129, 'Nfeatures_sel': 208, 'Ngenes_sel': 119, 'Nidentical': 0, 'Nduplicates': 49, 'Nfeatures': 119, 'Ngenes': 119}\n",
      "{'gene': 207, 'CDS': 125, 'exon': 72, 'tRNA': 67, 'intron': 49, 'rRNA': 21, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 81, 'tRNA': 34, 'rRNA': 4}\n",
      "{'Sample': 'NC_005086.1', 'Nfeatures_raw': 467, 'Ngenes_raw': 120, 'Nfeatures_sel': 177, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 38, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 169, 'CDS': 100, 'exon': 76, 'tRNA': 67, 'intron': 40, 'rRNA': 10, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 80, 'tRNA': 31, 'rRNA': 4}\n",
      "{'Sample': 'NC_009962.1', 'Nfeatures_raw': 448, 'Ngenes_raw': 119, 'Nfeatures_sel': 169, 'Ngenes_sel': 114, 'Nidentical': 0, 'Nduplicates': 32, 'Nfeatures': 114, 'Ngenes': 114}\n",
      "{'gene': 160, 'CDS': 95, 'exon': 73, 'tRNA': 64, 'intron': 41, 'rRNA': 10, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 31, 'rRNA': 4}\n",
      "{'Sample': 'NC_010323.1', 'Nfeatures_raw': 455, 'Ngenes_raw': 121, 'Nfeatures_sel': 170, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 32, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 158, 'CDS': 94, 'exon': 80, 'tRNA': 66, 'intron': 42, 'rRNA': 10, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_010654.1', 'Nfeatures_raw': 390, 'Ngenes_raw': 103, 'Nfeatures_sel': 148, 'Ngenes_sel': 96, 'Nidentical': 0, 'Nduplicates': 25, 'Nfeatures': 96, 'Ngenes': 96}\n",
      "{'gene': 146, 'tRNA': 69, 'CDS': 63, 'exon': 56, 'intron': 31, 'rRNA': 20, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 58, 'tRNA': 34, 'rRNA': 4}\n",
      "{'Sample': 'NC_011156.4', 'Nfeatures_raw': 395, 'Ngenes_raw': 117, 'Nfeatures_sel': 160, 'Ngenes_sel': 111, 'Nidentical': 0, 'Nduplicates': 36, 'Nfeatures': 111, 'Ngenes': 111}\n",
      "{'gene': 161, 'CDS': 108, 'tRNA': 46, 'exon': 43, 'intron': 26, 'rRNA': 10, 'source': 1}\n",
      "{'CDS': 78, 'tRNA': 29, 'rRNA': 4}\n",
      "{'Sample': 'NC_012818.1', 'Nfeatures_raw': 552, 'Ngenes_raw': 129, 'Nfeatures_sel': 213, 'Ngenes_sel': 120, 'Nidentical': 0, 'Nduplicates': 56, 'Nfeatures': 120, 'Ngenes': 120}\n",
      "{'gene': 217, 'CDS': 141, 'exon': 67, 'tRNA': 60, 'intron': 42, 'rRNA': 20, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 85, 'tRNA': 31, 'rRNA': 4}\n",
      "{'Sample': 'NC_015605.1', 'Nfeatures_raw': 461, 'Ngenes_raw': 121, 'Nfeatures_sel': 173, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 32, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 161, 'CDS': 96, 'exon': 79, 'tRNA': 65, 'intron': 43, 'rRNA': 12, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\ProgramData\\Anaconda3\\lib\\site-packages\\Bio\\GenBank\\__init__.py:366: BiopythonParserWarning: Attempting to fix invalid location '156533..99712' as it looks like incorrect origin wrapping. Please fix input file, this could have unintended behavior.\n",
      "  BiopythonParserWarning,\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Sample': 'NC_016986.1', 'Nfeatures_raw': 514, 'Ngenes_raw': 131, 'Nfeatures_sel': 197, 'Ngenes_sel': 126, 'Nidentical': 0, 'Nduplicates': 47, 'Nfeatures': 126, 'Ngenes': 126}\n",
      "{'gene': 188, 'CDS': 109, 'exon': 81, 'tRNA': 76, 'intron': 43, 'rRNA': 12, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 85, 'tRNA': 37, 'rRNA': 4}\n",
      "{'Sample': 'NC_017006.1', 'Nfeatures_raw': 507, 'Ngenes_raw': 127, 'Nfeatures_sel': 196, 'Ngenes_sel': 120, 'Nidentical': 0, 'Nduplicates': 44, 'Nfeatures': 120, 'Ngenes': 120}\n",
      "{'gene': 197, 'CDS': 118, 'exon': 64, 'tRNA': 63, 'intron': 38, 'rRNA': 22, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 84, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_020146.1', 'Nfeatures_raw': 548, 'Ngenes_raw': 131, 'Nfeatures_sel': 213, 'Ngenes_sel': 121, 'Nidentical': 0, 'Nduplicates': 54, 'Nfeatures': 121, 'Ngenes': 121}\n",
      "{'gene': 214, 'CDS': 134, 'exon': 67, 'tRNA': 63, 'intron': 43, 'rRNA': 22, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 84, 'tRNA': 33, 'rRNA': 4}\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\ProgramData\\Anaconda3\\lib\\site-packages\\Bio\\GenBank\\__init__.py:366: BiopythonParserWarning: Attempting to fix invalid location '165380..84468' as it looks like incorrect origin wrapping. Please fix input file, this could have unintended behavior.\n",
      "  BiopythonParserWarning,\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Sample': 'NC_020372.1', 'Nfeatures_raw': 452, 'Ngenes_raw': 119, 'Nfeatures_sel': 175, 'Ngenes_sel': 114, 'Nidentical': 0, 'Nduplicates': 41, 'Nfeatures': 114, 'Ngenes': 114}\n",
      "{'gene': 166, 'CDS': 101, 'exon': 69, 'tRNA': 63, 'intron': 36, 'rRNA': 12, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 31, 'rRNA': 4}\n",
      "{'Sample': 'NC_021426.1', 'Nfeatures_raw': 480, 'Ngenes_raw': 122, 'Nfeatures_sel': 178, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 38, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 171, 'CDS': 102, 'exon': 79, 'tRNA': 64, 'intron': 43, 'rRNA': 16, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_022137.1', 'Nfeatures_raw': 540, 'Ngenes_raw': 126, 'Nfeatures_sel': 207, 'Ngenes_sel': 117, 'Nidentical': 0, 'Nduplicates': 55, 'Nfeatures': 117, 'Ngenes': 117}\n",
      "{'gene': 210, 'CDS': 138, 'exon': 69, 'tRNA': 57, 'intron': 41, 'rRNA': 20, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 85, 'tRNA': 28, 'rRNA': 4}\n",
      "{'Sample': 'NC_022812.1', 'Nfeatures_raw': 468, 'Ngenes_raw': 122, 'Nfeatures_sel': 174, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 34, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 164, 'CDS': 96, 'exon': 80, 'tRNA': 66, 'intron': 43, 'rRNA': 14, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_023256.1', 'Nfeatures_raw': 467, 'Ngenes_raw': 123, 'Nfeatures_sel': 175, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 36, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 165, 'CDS': 101, 'exon': 78, 'tRNA': 66, 'intron': 42, 'rRNA': 10, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_024157.1', 'Nfeatures_raw': 532, 'Ngenes_raw': 134, 'Nfeatures_sel': 201, 'Ngenes_sel': 124, 'Nidentical': 0, 'Nduplicates': 49, 'Nfeatures': 124, 'Ngenes': 124}\n",
      "{'gene': 200, 'CDS': 120, 'exon': 75, 'tRNA': 69, 'intron': 45, 'rRNA': 18, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 85, 'tRNA': 35, 'rRNA': 4}\n",
      "{'Sample': 'NC_024158.1', 'Nfeatures_raw': 506, 'Ngenes_raw': 128, 'Nfeatures_sel': 196, 'Ngenes_sel': 118, 'Nidentical': 0, 'Nduplicates': 46, 'Nfeatures': 118, 'Ngenes': 118}\n",
      "{'gene': 196, 'CDS': 126, 'exon': 62, 'tRNA': 58, 'intron': 39, 'rRNA': 20, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 85, 'tRNA': 29, 'rRNA': 4}\n",
      "{'Sample': 'NC_024728.1', 'Nfeatures_raw': 483, 'Ngenes_raw': 121, 'Nfeatures_sel': 185, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 41, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 179, 'CDS': 103, 'exon': 71, 'tRNA': 70, 'intron': 39, 'rRNA': 16, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\ProgramData\\Anaconda3\\lib\\site-packages\\Bio\\GenBank\\__init__.py:366: BiopythonParserWarning: Attempting to fix invalid location '159887..87141' as it looks like incorrect origin wrapping. Please fix input file, this could have unintended behavior.\n",
      "  BiopythonParserWarning,\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Sample': 'NC_024929.1', 'Nfeatures_raw': 483, 'Ngenes_raw': 121, 'Nfeatures_sel': 180, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 37, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 167, 'CDS': 102, 'exon': 85, 'tRNA': 66, 'intron': 44, 'rRNA': 14, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_025745.1', 'Nfeatures_raw': 389, 'Ngenes_raw': 119, 'Nfeatures_sel': 155, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 32, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 149, 'CDS': 100, 'exon': 53, 'tRNA': 49, 'intron': 29, 'rRNA': 8, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_026295.1', 'Nfeatures_raw': 465, 'Ngenes_raw': 125, 'Nfeatures_sel': 181, 'Ngenes_sel': 121, 'Nidentical': 0, 'Nduplicates': 42, 'Nfeatures': 121, 'Ngenes': 121}\n",
      "{'gene': 179, 'CDS': 117, 'exon': 63, 'tRNA': 53, 'intron': 39, 'rRNA': 13, 'source': 1}\n",
      "{'CDS': 83, 'tRNA': 34, 'rRNA': 4}\n",
      "{'Sample': 'NC_026301.1', 'Nfeatures_raw': 405, 'Ngenes_raw': 107, 'Nfeatures_sel': 153, 'Ngenes_sel': 99, 'Nidentical': 0, 'Nduplicates': 31, 'Nfeatures': 99, 'Ngenes': 99}\n",
      "{'gene': 153, 'tRNA': 69, 'CDS': 68, 'exon': 56, 'intron': 30, 'rRNA': 24, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 61, 'tRNA': 34, 'rRNA': 4}\n",
      "{'Sample': 'NC_026786.1', 'Nfeatures_raw': 481, 'Ngenes_raw': 122, 'Nfeatures_sel': 183, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 37, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 173, 'CDS': 104, 'exon': 78, 'tRNA': 70, 'intron': 41, 'rRNA': 10, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_026847.1', 'Nfeatures_raw': 541, 'Ngenes_raw': 112, 'Nfeatures_sel': 207, 'Ngenes_sel': 104, 'Nidentical': 0, 'Nduplicates': 48, 'Nfeatures': 104, 'Ngenes': 104}\n",
      "{'gene': 207, 'CDS': 121, 'exon': 72, 'tRNA': 68, 'intron': 46, 'rRNA': 22, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 73, 'tRNA': 27, 'rRNA': 4}\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\ProgramData\\Anaconda3\\lib\\site-packages\\Bio\\GenBank\\__init__.py:366: BiopythonParserWarning: Attempting to fix invalid location '161640..88578' as it looks like incorrect origin wrapping. Please fix input file, this could have unintended behavior.\n",
      "  BiopythonParserWarning,\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Sample': 'NC_027512.1', 'Nfeatures_raw': 480, 'Ngenes_raw': 126, 'Nfeatures_sel': 187, 'Ngenes_sel': 122, 'Nidentical': 0, 'Nduplicates': 38, 'Nfeatures': 122, 'Ngenes': 122}\n",
      "{'gene': 179, 'CDS': 108, 'exon': 72, 'tRNA': 67, 'intron': 37, 'rRNA': 12, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 86, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_027829.1', 'Nfeatures_raw': 472, 'Ngenes_raw': 105, 'Nfeatures_sel': 177, 'Ngenes_sel': 102, 'Nidentical': 0, 'Nduplicates': 42, 'Nfeatures': 102, 'Ngenes': 102}\n",
      "{'gene': 174, 'CDS': 95, 'exon': 74, 'tRNA': 66, 'intron': 40, 'rRNA': 22, 'source': 1}\n",
      "{'CDS': 68, 'tRNA': 30, 'rRNA': 4}\n",
      "{'Sample': 'NC_029427.1', 'Nfeatures_raw': 454, 'Ngenes_raw': 120, 'Nfeatures_sel': 168, 'Ngenes_sel': 114, 'Nidentical': 0, 'Nduplicates': 32, 'Nfeatures': 114, 'Ngenes': 114}\n",
      "{'gene': 157, 'CDS': 97, 'exon': 80, 'tRNA': 62, 'intron': 43, 'rRNA': 10, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 31, 'rRNA': 4}\n",
      "{'Sample': 'NC_029433.1', 'Nfeatures_raw': 463, 'Ngenes_raw': 121, 'Nfeatures_sel': 174, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 35, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 163, 'CDS': 99, 'exon': 78, 'tRNA': 65, 'intron': 43, 'rRNA': 10, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_029970.1', 'Nfeatures_raw': 473, 'Ngenes_raw': 122, 'Nfeatures_sel': 181, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 38, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 174, 'CDS': 99, 'exon': 72, 'tRNA': 70, 'intron': 39, 'rRNA': 14, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_030601.1', 'Nfeatures_raw': 462, 'Ngenes_raw': 121, 'Nfeatures_sel': 175, 'Ngenes_sel': 114, 'Nidentical': 0, 'Nduplicates': 38, 'Nfeatures': 114, 'Ngenes': 114}\n",
      "{'gene': 164, 'CDS': 100, 'exon': 77, 'tRNA': 65, 'intron': 40, 'rRNA': 11, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 78, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_033874.1', 'Nfeatures_raw': 512, 'Ngenes_raw': 128, 'Nfeatures_sel': 200, 'Ngenes_sel': 122, 'Nidentical': 0, 'Nduplicates': 48, 'Nfeatures': 122, 'Ngenes': 122}\n",
      "{'gene': 193, 'CDS': 116, 'exon': 71, 'tRNA': 68, 'intron': 43, 'rRNA': 16, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 86, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_033878.1', 'Nfeatures_raw': 456, 'Ngenes_raw': 121, 'Nfeatures_sel': 170, 'Ngenes_sel': 114, 'Nidentical': 0, 'Nduplicates': 35, 'Nfeatures': 114, 'Ngenes': 114}\n",
      "{'gene': 161, 'CDS': 96, 'exon': 76, 'tRNA': 65, 'intron': 43, 'rRNA': 10, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 31, 'rRNA': 4}\n",
      "{'Sample': 'NC_034685.1', 'Nfeatures_raw': 435, 'Ngenes_raw': 119, 'Nfeatures_sel': 168, 'Ngenes_sel': 116, 'Nidentical': 0, 'Nduplicates': 33, 'Nfeatures': 116, 'Ngenes': 116}\n",
      "{'gene': 160, 'CDS': 94, 'exon': 65, 'tRNA': 62, 'intron': 39, 'rRNA': 14, 'source': 1}\n",
      "{'CDS': 80, 'tRNA': 32, 'rRNA': 4}\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\ProgramData\\Anaconda3\\lib\\site-packages\\Bio\\GenBank\\__init__.py:366: BiopythonParserWarning: Attempting to fix invalid location '153920..83865' as it looks like incorrect origin wrapping. Please fix input file, this could have unintended behavior.\n",
      "  BiopythonParserWarning,\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Sample': 'NC_034686.1', 'Nfeatures_raw': 470, 'Ngenes_raw': 122, 'Nfeatures_sel': 180, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 35, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 171, 'CDS': 102, 'exon': 73, 'tRNA': 69, 'intron': 40, 'rRNA': 10, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_034942.1', 'Nfeatures_raw': 443, 'Ngenes_raw': 123, 'Nfeatures_sel': 173, 'Ngenes_sel': 120, 'Nidentical': 0, 'Nduplicates': 40, 'Nfeatures': 120, 'Ngenes': 120}\n",
      "{'gene': 169, 'CDS': 110, 'exon': 64, 'tRNA': 56, 'intron': 34, 'rRNA': 9, 'source': 1}\n",
      "{'CDS': 82, 'tRNA': 34, 'rRNA': 4}\n",
      "{'Sample': 'NC_035050.1', 'Nfeatures_raw': 492, 'Ngenes_raw': 120, 'Nfeatures_sel': 195, 'Ngenes_sel': 114, 'Nidentical': 0, 'Nduplicates': 38, 'Nfeatures': 114, 'Ngenes': 114}\n",
      "{'gene': 188, 'CDS': 114, 'tRNA': 70, 'exon': 68, 'intron': 35, 'rRNA': 12, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 78, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_035499.1', 'Nfeatures_raw': 464, 'Ngenes_raw': 121, 'Nfeatures_sel': 178, 'Ngenes_sel': 113, 'Nidentical': 0, 'Nduplicates': 35, 'Nfeatures': 113, 'Ngenes': 113}\n",
      "{'gene': 172, 'CDS': 101, 'exon': 70, 'tRNA': 68, 'intron': 38, 'rRNA': 10, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 30, 'rRNA': 4}\n",
      "{'Sample': 'NC_035545.1', 'Nfeatures_raw': 457, 'Ngenes_raw': 123, 'Nfeatures_sel': 172, 'Ngenes_sel': 116, 'Nidentical': 0, 'Nduplicates': 34, 'Nfeatures': 116, 'Ngenes': 116}\n",
      "{'gene': 162, 'CDS': 97, 'exon': 76, 'tRNA': 66, 'intron': 41, 'rRNA': 10, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 80, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_035807.1', 'Nfeatures_raw': 545, 'Ngenes_raw': 116, 'Nfeatures_sel': 218, 'Ngenes_sel': 107, 'Nidentical': 0, 'Nduplicates': 60, 'Nfeatures': 107, 'Ngenes': 107}\n",
      "{'gene': 221, 'CDS': 148, 'exon': 58, 'tRNA': 53, 'intron': 36, 'rRNA': 24, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 77, 'tRNA': 26, 'rRNA': 4}\n",
      "{'Sample': 'NC_035854.1', 'Nfeatures_raw': 521, 'Ngenes_raw': 124, 'Nfeatures_sel': 215, 'Ngenes_sel': 117, 'Nidentical': 0, 'Nduplicates': 60, 'Nfeatures': 117, 'Ngenes': 117}\n",
      "{'gene': 217, 'CDS': 142, 'tRNA': 61, 'exon': 49, 'intron': 31, 'rRNA': 16, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 85, 'tRNA': 28, 'rRNA': 4}\n",
      "{'Sample': 'NC_036135.1', 'Nfeatures_raw': 447, 'Ngenes_raw': 120, 'Nfeatures_sel': 169, 'Ngenes_sel': 114, 'Nidentical': 0, 'Nduplicates': 34, 'Nfeatures': 114, 'Ngenes': 114}\n",
      "{'gene': 160, 'CDS': 95, 'exon': 73, 'tRNA': 64, 'intron': 40, 'rRNA': 10, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 31, 'rRNA': 4}\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\ProgramData\\Anaconda3\\lib\\site-packages\\Bio\\GenBank\\__init__.py:366: BiopythonParserWarning: Attempting to fix invalid location '154782..89286' as it looks like incorrect origin wrapping. Please fix input file, this could have unintended behavior.\n",
      "  BiopythonParserWarning,\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Sample': 'NC_036154.1', 'Nfeatures_raw': 459, 'Ngenes_raw': 122, 'Nfeatures_sel': 176, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 36, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 169, 'CDS': 97, 'exon': 70, 'tRNA': 70, 'intron': 38, 'rRNA': 10, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_036304.1', 'Nfeatures_raw': 443, 'Ngenes_raw': 123, 'Nfeatures_sel': 168, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 33, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 158, 'CDS': 98, 'exon': 72, 'tRNA': 62, 'intron': 38, 'rRNA': 10, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\ProgramData\\Anaconda3\\lib\\site-packages\\Bio\\GenBank\\__init__.py:366: BiopythonParserWarning: Attempting to fix invalid location '159659..88064' as it looks like incorrect origin wrapping. Please fix input file, this could have unintended behavior.\n",
      "  BiopythonParserWarning,\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Sample': 'NC_036416.1', 'Nfeatures_raw': 476, 'Ngenes_raw': 122, 'Nfeatures_sel': 178, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 36, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 168, 'CDS': 96, 'exon': 80, 'tRNA': 68, 'intron': 43, 'rRNA': 16, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_036660.1', 'Nfeatures_raw': 475, 'Ngenes_raw': 122, 'Nfeatures_sel': 180, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 35, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 172, 'CDS': 100, 'exon': 75, 'tRNA': 68, 'intron': 39, 'rRNA': 16, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 78, 'tRNA': 33, 'rRNA': 4}\n",
      "{'Sample': 'NC_036960.1', 'Nfeatures_raw': 481, 'Ngenes_raw': 123, 'Nfeatures_sel': 180, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 35, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 171, 'CDS': 101, 'exon': 79, 'tRNA': 68, 'intron': 43, 'rRNA': 14, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_037151.1', 'Nfeatures_raw': 447, 'Ngenes_raw': 121, 'Nfeatures_sel': 168, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 31, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 157, 'CDS': 92, 'exon': 76, 'tRNA': 66, 'intron': 41, 'rRNA': 10, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_037504.1', 'Nfeatures_raw': 466, 'Ngenes_raw': 126, 'Nfeatures_sel': 175, 'Ngenes_sel': 117, 'Nidentical': 0, 'Nduplicates': 35, 'Nfeatures': 117, 'Ngenes': 117}\n",
      "{'gene': 167, 'CDS': 99, 'exon': 74, 'tRNA': 65, 'intron': 40, 'rRNA': 16, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 81, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_037845.1', 'Nfeatures_raw': 460, 'Ngenes_raw': 122, 'Nfeatures_sel': 176, 'Ngenes_sel': 116, 'Nidentical': 0, 'Nduplicates': 37, 'Nfeatures': 116, 'Ngenes': 116}\n",
      "{'gene': 167, 'CDS': 98, 'exon': 73, 'tRNA': 68, 'intron': 39, 'rRNA': 10, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 80, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_037885.1', 'Nfeatures_raw': 470, 'Ngenes_raw': 120, 'Nfeatures_sel': 177, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 35, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 167, 'CDS': 99, 'exon': 79, 'tRNA': 66, 'intron': 42, 'rRNA': 12, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\ProgramData\\Anaconda3\\lib\\site-packages\\Bio\\GenBank\\__init__.py:366: BiopythonParserWarning: Attempting to fix invalid location '163091..87267' as it looks like incorrect origin wrapping. Please fix input file, this could have unintended behavior.\n",
      "  BiopythonParserWarning,\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Sample': 'NC_037948.1', 'Nfeatures_raw': 478, 'Ngenes_raw': 122, 'Nfeatures_sel': 178, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 38, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 171, 'CDS': 102, 'exon': 78, 'tRNA': 64, 'intron': 42, 'rRNA': 16, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_038057.1', 'Nfeatures_raw': 463, 'Ngenes_raw': 121, 'Nfeatures_sel': 173, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 34, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 161, 'CDS': 96, 'exon': 81, 'tRNA': 67, 'intron': 43, 'rRNA': 10, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_038074.1', 'Nfeatures_raw': 527, 'Ngenes_raw': 123, 'Nfeatures_sel': 204, 'Ngenes_sel': 122, 'Nidentical': 0, 'Nduplicates': 51, 'Nfeatures': 122, 'Ngenes': 122}\n",
      "{'gene': 206, 'CDS': 126, 'exon': 71, 'tRNA': 64, 'intron': 41, 'rRNA': 18, 'source': 1}\n",
      "{'CDS': 85, 'tRNA': 33, 'rRNA': 4}\n",
      "{'Sample': 'NC_038093.1', 'Nfeatures_raw': 468, 'Ngenes_raw': 122, 'Nfeatures_sel': 174, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 37, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 163, 'CDS': 97, 'exon': 81, 'tRNA': 65, 'intron': 43, 'rRNA': 14, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_039155.1', 'Nfeatures_raw': 466, 'Ngenes_raw': 125, 'Nfeatures_sel': 183, 'Ngenes_sel': 123, 'Nidentical': 0, 'Nduplicates': 41, 'Nfeatures': 123, 'Ngenes': 123}\n",
      "{'gene': 185, 'CDS': 107, 'tRNA': 64, 'exon': 61, 'intron': 32, 'rRNA': 16, 'source': 1}\n",
      "{'CDS': 85, 'tRNA': 34, 'rRNA': 4}\n",
      "{'Sample': 'NC_039395.1', 'Nfeatures_raw': 462, 'Ngenes_raw': 122, 'Nfeatures_sel': 174, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 37, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 166, 'CDS': 97, 'exon': 75, 'tRNA': 65, 'intron': 40, 'rRNA': 14, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\ProgramData\\Anaconda3\\lib\\site-packages\\Bio\\GenBank\\__init__.py:366: BiopythonParserWarning: Attempting to fix invalid location '159531..87293' as it looks like incorrect origin wrapping. Please fix input file, this could have unintended behavior.\n",
      "  BiopythonParserWarning,\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Sample': 'NC_039424.1', 'Nfeatures_raw': 461, 'Ngenes_raw': 121, 'Nfeatures_sel': 177, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 36, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 168, 'CDS': 96, 'exon': 72, 'tRNA': 71, 'intron': 39, 'rRNA': 10, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_039600.1', 'Nfeatures_raw': 475, 'Ngenes_raw': 119, 'Nfeatures_sel': 179, 'Ngenes_sel': 114, 'Nidentical': 0, 'Nduplicates': 40, 'Nfeatures': 114, 'Ngenes': 114}\n",
      "{'gene': 168, 'CDS': 101, 'exon': 79, 'tRNA': 67, 'intron': 44, 'rRNA': 11, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 31, 'rRNA': 4}\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\ProgramData\\Anaconda3\\lib\\site-packages\\Bio\\GenBank\\__init__.py:366: BiopythonParserWarning: Attempting to fix invalid location '157771..87110' as it looks like incorrect origin wrapping. Please fix input file, this could have unintended behavior.\n",
      "  BiopythonParserWarning,\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Sample': 'NC_039627.1', 'Nfeatures_raw': 492, 'Ngenes_raw': 124, 'Nfeatures_sel': 182, 'Ngenes_sel': 116, 'Nidentical': 0, 'Nduplicates': 37, 'Nfeatures': 116, 'Ngenes': 116}\n",
      "{'gene': 176, 'CDS': 101, 'exon': 81, 'tRNA': 67, 'intron': 42, 'rRNA': 20, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 80, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_039676.1', 'Nfeatures_raw': 487, 'Ngenes_raw': 123, 'Nfeatures_sel': 185, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 40, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 176, 'CDS': 104, 'exon': 77, 'tRNA': 70, 'intron': 41, 'rRNA': 14, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_039731.1', 'Nfeatures_raw': 461, 'Ngenes_raw': 121, 'Nfeatures_sel': 173, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 35, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 161, 'CDS': 98, 'exon': 79, 'tRNA': 65, 'intron': 43, 'rRNA': 10, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\ProgramData\\Anaconda3\\lib\\site-packages\\Bio\\GenBank\\__init__.py:366: BiopythonParserWarning: Attempting to fix invalid location '160932..89144' as it looks like incorrect origin wrapping. Please fix input file, this could have unintended behavior.\n",
      "  BiopythonParserWarning,\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Sample': 'NC_039803.1', 'Nfeatures_raw': 455, 'Ngenes_raw': 122, 'Nfeatures_sel': 169, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 33, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 158, 'CDS': 95, 'exon': 79, 'tRNA': 65, 'intron': 43, 'rRNA': 10, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_039815.1', 'Nfeatures_raw': 489, 'Ngenes_raw': 123, 'Nfeatures_sel': 187, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 41, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 180, 'CDS': 105, 'exon': 74, 'tRNA': 71, 'intron': 40, 'rRNA': 14, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_040160.1', 'Nfeatures_raw': 369, 'Ngenes_raw': 117, 'Nfeatures_sel': 146, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 25, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 139, 'CDS': 89, 'exon': 54, 'tRNA': 52, 'intron': 29, 'rRNA': 5, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_040219.1', 'Nfeatures_raw': 502, 'Ngenes_raw': 122, 'Nfeatures_sel': 206, 'Ngenes_sel': 114, 'Nidentical': 0, 'Nduplicates': 52, 'Nfeatures': 114, 'Ngenes': 114}\n",
      "{'gene': 206, 'CDS': 137, 'tRNA': 58, 'exon': 47, 'intron': 30, 'rRNA': 19, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 84, 'tRNA': 26, 'rRNA': 4}\n",
      "{'Sample': 'NC_040862.1', 'Nfeatures_raw': 401, 'Ngenes_raw': 99, 'Nfeatures_sel': 158, 'Ngenes_sel': 96, 'Nidentical': 0, 'Nduplicates': 37, 'Nfeatures': 96, 'Ngenes': 96}\n",
      "{'gene': 154, 'CDS': 80, 'tRNA': 64, 'exon': 52, 'intron': 34, 'rRNA': 16, 'source': 1}\n",
      "{'CDS': 64, 'tRNA': 28, 'rRNA': 4}\n",
      "{'Sample': 'NC_040983.1', 'Nfeatures_raw': 478, 'Ngenes_raw': 119, 'Nfeatures_sel': 180, 'Ngenes_sel': 113, 'Nidentical': 0, 'Nduplicates': 38, 'Nfeatures': 113, 'Ngenes': 113}\n",
      "{'gene': 171, 'CDS': 103, 'exon': 79, 'tRNA': 68, 'intron': 42, 'rRNA': 10, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 78, 'tRNA': 31, 'rRNA': 4}\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\ProgramData\\Anaconda3\\lib\\site-packages\\Bio\\GenBank\\__init__.py:366: BiopythonParserWarning: Attempting to fix invalid location '136363..5491' as it looks like incorrect origin wrapping. Please fix input file, this could have unintended behavior.\n",
      "  BiopythonParserWarning,\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Sample': 'NC_041000.1', 'Nfeatures_raw': 539, 'Ngenes_raw': 127, 'Nfeatures_sel': 205, 'Ngenes_sel': 119, 'Nidentical': 0, 'Nduplicates': 48, 'Nfeatures': 119, 'Ngenes': 119}\n",
      "{'gene': 207, 'CDS': 135, 'exon': 70, 'tRNA': 57, 'intron': 45, 'rRNA': 20, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 85, 'tRNA': 30, 'rRNA': 4}\n",
      "{'Sample': 'NC_041127.1', 'Nfeatures_raw': 444, 'Ngenes_raw': 123, 'Nfeatures_sel': 168, 'Ngenes_sel': 116, 'Nidentical': 0, 'Nduplicates': 32, 'Nfeatures': 116, 'Ngenes': 116}\n",
      "{'gene': 158, 'CDS': 95, 'exon': 73, 'tRNA': 64, 'intron': 39, 'rRNA': 10, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 80, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_041261.1', 'Nfeatures_raw': 520, 'Ngenes_raw': 124, 'Nfeatures_sel': 191, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 35, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 188, 'CDS': 109, 'exon': 84, 'tRNA': 67, 'intron': 45, 'rRNA': 22, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_041275.1', 'Nfeatures_raw': 455, 'Ngenes_raw': 122, 'Nfeatures_sel': 172, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 35, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 161, 'CDS': 99, 'exon': 75, 'tRNA': 64, 'intron': 41, 'rRNA': 10, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_041281.1', 'Nfeatures_raw': 451, 'Ngenes_raw': 121, 'Nfeatures_sel': 169, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 33, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 157, 'CDS': 95, 'exon': 78, 'tRNA': 64, 'intron': 42, 'rRNA': 10, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_041452.1', 'Nfeatures_raw': 484, 'Ngenes_raw': 121, 'Nfeatures_sel': 180, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 36, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 170, 'CDS': 97, 'exon': 83, 'tRNA': 71, 'intron': 44, 'rRNA': 14, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_041466.1', 'Nfeatures_raw': 394, 'Ngenes_raw': 123, 'Nfeatures_sel': 159, 'Ngenes_sel': 119, 'Nidentical': 0, 'Nduplicates': 31, 'Nfeatures': 119, 'Ngenes': 119}\n",
      "{'gene': 161, 'CDS': 102, 'tRNA': 48, 'exon': 43, 'intron': 26, 'rRNA': 13, 'source': 1}\n",
      "{'CDS': 83, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_041475.1', 'Nfeatures_raw': 451, 'Ngenes_raw': 123, 'Nfeatures_sel': 168, 'Ngenes_sel': 116, 'Nidentical': 0, 'Nduplicates': 31, 'Nfeatures': 116, 'Ngenes': 116}\n",
      "{'gene': 158, 'CDS': 95, 'exon': 78, 'tRNA': 64, 'intron': 41, 'rRNA': 10, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 33, 'rRNA': 4}\n",
      "{'Sample': 'NC_041481.1', 'Nfeatures_raw': 450, 'Ngenes_raw': 120, 'Nfeatures_sel': 171, 'Ngenes_sel': 114, 'Nidentical': 0, 'Nduplicates': 34, 'Nfeatures': 114, 'Ngenes': 114}\n",
      "{'gene': 162, 'CDS': 96, 'exon': 71, 'tRNA': 63, 'intron': 39, 'rRNA': 14, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 31, 'rRNA': 4}\n",
      "{'Sample': 'NC_041492.1', 'Nfeatures_raw': 464, 'Ngenes_raw': 121, 'Nfeatures_sel': 173, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 36, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 162, 'CDS': 96, 'exon': 80, 'tRNA': 67, 'intron': 44, 'rRNA': 10, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_041640.1', 'Nfeatures_raw': 403, 'Ngenes_raw': 98, 'Nfeatures_sel': 180, 'Ngenes_sel': 95, 'Nidentical': 0, 'Nduplicates': 55, 'Nfeatures': 95, 'Ngenes': 95}\n",
      "{'gene': 184, 'CDS': 139, 'tRNA': 23, 'exon': 22, 'rRNA': 22, 'intron': 12, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 12, 'rRNA': 4}\n",
      "{'Sample': 'NC_042211.1', 'Nfeatures_raw': 442, 'Ngenes_raw': 121, 'Nfeatures_sel': 169, 'Ngenes_sel': 112, 'Nidentical': 0, 'Nduplicates': 32, 'Nfeatures': 112, 'Ngenes': 112}\n",
      "{'gene': 164, 'CDS': 89, 'tRNA': 69, 'exon': 66, 'intron': 35, 'rRNA': 14, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 77, 'tRNA': 31, 'rRNA': 4}\n",
      "{'Sample': 'NC_042450.1', 'Nfeatures_raw': 458, 'Ngenes_raw': 122, 'Nfeatures_sel': 171, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 33, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 162, 'CDS': 95, 'exon': 77, 'tRNA': 64, 'intron': 41, 'rRNA': 14, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_042597.1', 'Nfeatures_raw': 448, 'Ngenes_raw': 116, 'Nfeatures_sel': 167, 'Ngenes_sel': 108, 'Nidentical': 0, 'Nduplicates': 32, 'Nfeatures': 108, 'Ngenes': 108}\n",
      "{'gene': 159, 'CDS': 93, 'exon': 76, 'tRNA': 66, 'intron': 39, 'rRNA': 10, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 72, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_042740.1', 'Nfeatures_raw': 462, 'Ngenes_raw': 123, 'Nfeatures_sel': 171, 'Ngenes_sel': 114, 'Nidentical': 0, 'Nduplicates': 35, 'Nfeatures': 114, 'Ngenes': 114}\n",
      "{'gene': 165, 'CDS': 99, 'exon': 76, 'tRNA': 63, 'intron': 41, 'rRNA': 13, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 31, 'rRNA': 4}\n",
      "{'Sample': 'NC_042889.1', 'Nfeatures_raw': 466, 'Ngenes_raw': 121, 'Nfeatures_sel': 182, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 42, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 173, 'CDS': 103, 'exon': 68, 'tRNA': 65, 'intron': 36, 'rRNA': 16, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\ProgramData\\Anaconda3\\lib\\site-packages\\Bio\\GenBank\\__init__.py:366: BiopythonParserWarning: Attempting to fix invalid location '160919..88145' as it looks like incorrect origin wrapping. Please fix input file, this could have unintended behavior.\n",
      "  BiopythonParserWarning,\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Sample': 'NC_042940.1', 'Nfeatures_raw': 510, 'Ngenes_raw': 122, 'Nfeatures_sel': 192, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 42, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 181, 'CDS': 111, 'exon': 83, 'tRNA': 68, 'intron': 48, 'rRNA': 14, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_043796.1', 'Nfeatures_raw': 487, 'Ngenes_raw': 124, 'Nfeatures_sel': 183, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 35, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 175, 'CDS': 107, 'exon': 77, 'tRNA': 64, 'intron': 43, 'rRNA': 16, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 78, 'tRNA': 33, 'rRNA': 4}\n",
      "{'Sample': 'NC_043800.1', 'Nfeatures_raw': 464, 'Ngenes_raw': 122, 'Nfeatures_sel': 174, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 32, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 165, 'CDS': 96, 'exon': 77, 'tRNA': 66, 'intron': 41, 'rRNA': 14, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_043867.1', 'Nfeatures_raw': 479, 'Ngenes_raw': 124, 'Nfeatures_sel': 181, 'Ngenes_sel': 117, 'Nidentical': 0, 'Nduplicates': 35, 'Nfeatures': 117, 'Ngenes': 117}\n",
      "{'gene': 172, 'CDS': 104, 'exon': 76, 'tRNA': 65, 'intron': 41, 'rRNA': 16, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 81, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_044728.1', 'Nfeatures_raw': 447, 'Ngenes_raw': 122, 'Nfeatures_sel': 169, 'Ngenes_sel': 116, 'Nidentical': 0, 'Nduplicates': 35, 'Nfeatures': 116, 'Ngenes': 116}\n",
      "{'gene': 159, 'CDS': 95, 'exon': 73, 'tRNA': 64, 'intron': 41, 'rRNA': 10, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 33, 'rRNA': 4}\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\ProgramData\\Anaconda3\\lib\\site-packages\\Bio\\GenBank\\__init__.py:366: BiopythonParserWarning: Attempting to fix invalid location '156852..86364' as it looks like incorrect origin wrapping. Please fix input file, this could have unintended behavior.\n",
      "  BiopythonParserWarning,\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Sample': 'NC_044826.1', 'Nfeatures_raw': 461, 'Ngenes_raw': 121, 'Nfeatures_sel': 173, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 35, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 161, 'CDS': 98, 'exon': 79, 'tRNA': 65, 'intron': 43, 'rRNA': 10, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_044828.1', 'Nfeatures_raw': 459, 'Ngenes_raw': 123, 'Nfeatures_sel': 172, 'Ngenes_sel': 116, 'Nidentical': 0, 'Nduplicates': 33, 'Nfeatures': 116, 'Ngenes': 116}\n",
      "{'gene': 161, 'CDS': 97, 'exon': 78, 'tRNA': 66, 'intron': 42, 'rRNA': 10, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 80, 'tRNA': 32, 'rRNA': 4}\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\ProgramData\\Anaconda3\\lib\\site-packages\\Bio\\GenBank\\__init__.py:366: BiopythonParserWarning: Attempting to fix invalid location '158724..86383' as it looks like incorrect origin wrapping. Please fix input file, this could have unintended behavior.\n",
      "  BiopythonParserWarning,\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Sample': 'NC_045081.1', 'Nfeatures_raw': 469, 'Ngenes_raw': 121, 'Nfeatures_sel': 182, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 41, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 174, 'CDS': 101, 'tRNA': 71, 'exon': 70, 'intron': 38, 'rRNA': 10, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_045235.1', 'Nfeatures_raw': 466, 'Ngenes_raw': 119, 'Nfeatures_sel': 176, 'Ngenes_sel': 112, 'Nidentical': 0, 'Nduplicates': 42, 'Nfeatures': 112, 'Ngenes': 112}\n",
      "{'gene': 164, 'CDS': 103, 'exon': 77, 'tRNA': 65, 'intron': 42, 'rRNA': 10, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 77, 'tRNA': 31, 'rRNA': 4}\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\ProgramData\\Anaconda3\\lib\\site-packages\\Bio\\GenBank\\__init__.py:366: BiopythonParserWarning: Attempting to fix invalid location '157850..87258' as it looks like incorrect origin wrapping. Please fix input file, this could have unintended behavior.\n",
      "  BiopythonParserWarning,\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Sample': 'NC_045274.1', 'Nfeatures_raw': 462, 'Ngenes_raw': 121, 'Nfeatures_sel': 173, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 36, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 163, 'CDS': 95, 'exon': 78, 'tRNA': 68, 'intron': 43, 'rRNA': 10, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_045280.1', 'Nfeatures_raw': 484, 'Ngenes_raw': 120, 'Nfeatures_sel': 184, 'Ngenes_sel': 113, 'Nidentical': 0, 'Nduplicates': 39, 'Nfeatures': 113, 'Ngenes': 113}\n",
      "{'gene': 177, 'CDS': 103, 'exon': 74, 'tRNA': 69, 'intron': 42, 'rRNA': 14, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 77, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_045294.1', 'Nfeatures_raw': 497, 'Ngenes_raw': 124, 'Nfeatures_sel': 186, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 38, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 179, 'CDS': 104, 'exon': 80, 'tRNA': 69, 'intron': 42, 'rRNA': 18, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_045400.1', 'Nfeatures_raw': 459, 'Ngenes_raw': 119, 'Nfeatures_sel': 177, 'Ngenes_sel': 114, 'Nidentical': 0, 'Nduplicates': 38, 'Nfeatures': 114, 'Ngenes': 114}\n",
      "{'gene': 168, 'CDS': 100, 'exon': 69, 'tRNA': 67, 'intron': 40, 'rRNA': 10, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 31, 'rRNA': 4}\n",
      "{'Sample': 'NC_046385.1', 'Nfeatures_raw': 465, 'Ngenes_raw': 119, 'Nfeatures_sel': 176, 'Ngenes_sel': 113, 'Nidentical': 0, 'Nduplicates': 41, 'Nfeatures': 113, 'Ngenes': 113}\n",
      "{'gene': 166, 'CDS': 99, 'exon': 77, 'tRNA': 67, 'intron': 41, 'rRNA': 10, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 30, 'rRNA': 4}\n",
      "{'Sample': 'NC_046458.1', 'Nfeatures_raw': 478, 'Ngenes_raw': 121, 'Nfeatures_sel': 184, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 39, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 175, 'CDS': 106, 'exon': 74, 'tRNA': 68, 'intron': 40, 'rRNA': 10, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_046575.1', 'Nfeatures_raw': 457, 'Ngenes_raw': 123, 'Nfeatures_sel': 174, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 36, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 167, 'CDS': 98, 'exon': 70, 'tRNA': 65, 'intron': 38, 'rRNA': 14, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_046602.1', 'Nfeatures_raw': 467, 'Ngenes_raw': 120, 'Nfeatures_sel': 181, 'Ngenes_sel': 114, 'Nidentical': 0, 'Nduplicates': 40, 'Nfeatures': 114, 'Ngenes': 114}\n",
      "{'gene': 174, 'CDS': 103, 'exon': 69, 'tRNA': 66, 'intron': 36, 'rRNA': 14, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 31, 'rRNA': 4}\n",
      "{'Sample': 'NC_048982.1', 'Nfeatures_raw': 460, 'Ngenes_raw': 119, 'Nfeatures_sel': 173, 'Ngenes_sel': 114, 'Nidentical': 0, 'Nduplicates': 37, 'Nfeatures': 114, 'Ngenes': 114}\n",
      "{'gene': 162, 'CDS': 98, 'exon': 78, 'tRNA': 65, 'intron': 42, 'rRNA': 10, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 31, 'rRNA': 4}\n",
      "{'Sample': 'NC_049157.1', 'Nfeatures_raw': 453, 'Ngenes_raw': 121, 'Nfeatures_sel': 169, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 31, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 157, 'CDS': 94, 'exon': 79, 'tRNA': 65, 'intron': 43, 'rRNA': 10, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_049578.1', 'Nfeatures_raw': 469, 'Ngenes_raw': 120, 'Nfeatures_sel': 174, 'Ngenes_sel': 114, 'Nidentical': 0, 'Nduplicates': 38, 'Nfeatures': 114, 'Ngenes': 114}\n",
      "{'gene': 165, 'CDS': 95, 'exon': 81, 'tRNA': 70, 'intron': 43, 'rRNA': 10, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 31, 'rRNA': 4}\n",
      "{'Sample': 'NC_050211.1', 'Nfeatures_raw': 475, 'Ngenes_raw': 123, 'Nfeatures_sel': 177, 'Ngenes_sel': 116, 'Nidentical': 0, 'Nduplicates': 34, 'Nfeatures': 116, 'Ngenes': 116}\n",
      "{'gene': 169, 'CDS': 99, 'exon': 78, 'tRNA': 66, 'intron': 42, 'rRNA': 16, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 80, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_050270.1', 'Nfeatures_raw': 443, 'Ngenes_raw': 121, 'Nfeatures_sel': 168, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 33, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 157, 'CDS': 93, 'exon': 74, 'tRNA': 65, 'intron': 39, 'rRNA': 10, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_050853.1', 'Nfeatures_raw': 463, 'Ngenes_raw': 123, 'Nfeatures_sel': 172, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 33, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 162, 'CDS': 98, 'exon': 79, 'tRNA': 66, 'intron': 43, 'rRNA': 10, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_050922.1', 'Nfeatures_raw': 434, 'Ngenes_raw': 122, 'Nfeatures_sel': 166, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 35, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 159, 'CDS': 94, 'exon': 66, 'tRNA': 63, 'intron': 37, 'rRNA': 10, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_050986.1', 'Nfeatures_raw': 476, 'Ngenes_raw': 121, 'Nfeatures_sel': 181, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 34, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 173, 'CDS': 102, 'exon': 75, 'tRNA': 67, 'intron': 40, 'rRNA': 14, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 80, 'tRNA': 31, 'rRNA': 4}\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\ProgramData\\Anaconda3\\lib\\site-packages\\Bio\\GenBank\\__init__.py:366: BiopythonParserWarning: Attempting to fix invalid location '158598..0' as it looks like incorrect origin wrapping. Please fix input file, this could have unintended behavior.\n",
      "  BiopythonParserWarning,\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Sample': 'NC_051913.1', 'Nfeatures_raw': 471, 'Ngenes_raw': 123, 'Nfeatures_sel': 171, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 33, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 165, 'CDS': 97, 'exon': 80, 'tRNA': 63, 'intron': 45, 'rRNA': 16, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_051971.1', 'Nfeatures_raw': 470, 'Ngenes_raw': 120, 'Nfeatures_sel': 177, 'Ngenes_sel': 114, 'Nidentical': 0, 'Nduplicates': 37, 'Nfeatures': 114, 'Ngenes': 114}\n",
      "{'gene': 166, 'CDS': 97, 'exon': 79, 'tRNA': 68, 'intron': 43, 'rRNA': 12, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 78, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_051978.1', 'Nfeatures_raw': 518, 'Ngenes_raw': 129, 'Nfeatures_sel': 194, 'Ngenes_sel': 122, 'Nidentical': 0, 'Nduplicates': 46, 'Nfeatures': 122, 'Ngenes': 122}\n",
      "{'gene': 193, 'CDS': 114, 'exon': 74, 'tRNA': 66, 'intron': 46, 'rRNA': 20, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 85, 'tRNA': 33, 'rRNA': 4}\n",
      "{'Sample': 'NC_052855.1', 'Nfeatures_raw': 466, 'Ngenes_raw': 117, 'Nfeatures_sel': 181, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 43, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 173, 'CDS': 106, 'exon': 72, 'tRNA': 65, 'intron': 39, 'rRNA': 10, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_052912.1', 'Nfeatures_raw': 461, 'Ngenes_raw': 120, 'Nfeatures_sel': 177, 'Ngenes_sel': 113, 'Nidentical': 0, 'Nduplicates': 37, 'Nfeatures': 113, 'Ngenes': 113}\n",
      "{'gene': 170, 'CDS': 100, 'exon': 68, 'tRNA': 66, 'intron': 38, 'rRNA': 14, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 77, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_052922.1', 'Nfeatures_raw': 460, 'Ngenes_raw': 120, 'Nfeatures_sel': 173, 'Ngenes_sel': 114, 'Nidentical': 0, 'Nduplicates': 33, 'Nfeatures': 114, 'Ngenes': 114}\n",
      "{'gene': 162, 'CDS': 101, 'exon': 78, 'tRNA': 63, 'intron': 41, 'rRNA': 10, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 31, 'rRNA': 4}\n",
      "{'Sample': 'NC_053567.1', 'Nfeatures_raw': 489, 'Ngenes_raw': 128, 'Nfeatures_sel': 188, 'Ngenes_sel': 123, 'Nidentical': 0, 'Nduplicates': 39, 'Nfeatures': 123, 'Ngenes': 123}\n",
      "{'gene': 179, 'CDS': 108, 'exon': 76, 'tRNA': 68, 'intron': 41, 'rRNA': 12, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 86, 'tRNA': 33, 'rRNA': 4}\n",
      "{'Sample': 'NC_053620.1', 'Nfeatures_raw': 458, 'Ngenes_raw': 122, 'Nfeatures_sel': 172, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 34, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 162, 'CDS': 95, 'exon': 77, 'tRNA': 65, 'intron': 40, 'rRNA': 14, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_053780.1', 'Nfeatures_raw': 447, 'Ngenes_raw': 119, 'Nfeatures_sel': 167, 'Ngenes_sel': 113, 'Nidentical': 0, 'Nduplicates': 35, 'Nfeatures': 113, 'Ngenes': 113}\n",
      "{'gene': 162, 'CDS': 90, 'exon': 70, 'tRNA': 65, 'intron': 39, 'rRNA': 16, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 78, 'tRNA': 31, 'rRNA': 4}\n",
      "{'Sample': 'NC_053784.1', 'Nfeatures_raw': 467, 'Ngenes_raw': 121, 'Nfeatures_sel': 178, 'Ngenes_sel': 115, 'Nidentical': 0, 'Nduplicates': 38, 'Nfeatures': 115, 'Ngenes': 115}\n",
      "{'gene': 168, 'CDS': 97, 'exon': 76, 'tRNA': 71, 'intron': 40, 'rRNA': 10, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 32, 'rRNA': 4}\n",
      "{'Sample': 'NC_053912.1', 'Nfeatures_raw': 453, 'Ngenes_raw': 120, 'Nfeatures_sel': 170, 'Ngenes_sel': 114, 'Nidentical': 0, 'Nduplicates': 33, 'Nfeatures': 114, 'Ngenes': 114}\n",
      "{'gene': 161, 'CDS': 94, 'exon': 75, 'tRNA': 67, 'intron': 41, 'rRNA': 10, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 31, 'rRNA': 4}\n",
      "{'Sample': 'NC_054296.1', 'Nfeatures_raw': 457, 'Ngenes_raw': 120, 'Nfeatures_sel': 171, 'Ngenes_sel': 114, 'Nidentical': 0, 'Nduplicates': 33, 'Nfeatures': 114, 'Ngenes': 114}\n",
      "{'gene': 162, 'CDS': 96, 'exon': 77, 'tRNA': 65, 'intron': 41, 'rRNA': 11, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 31, 'rRNA': 4}\n",
      "{'Sample': 'NC_054306.1', 'Nfeatures_raw': 490, 'Ngenes_raw': 121, 'Nfeatures_sel': 181, 'Ngenes_sel': 114, 'Nidentical': 0, 'Nduplicates': 38, 'Nfeatures': 114, 'Ngenes': 114}\n",
      "{'gene': 171, 'CDS': 102, 'exon': 84, 'tRNA': 68, 'intron': 46, 'rRNA': 14, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 31, 'rRNA': 4}\n",
      "{'Sample': 'NC_054357.1', 'Nfeatures_raw': 436, 'Ngenes_raw': 108, 'Nfeatures_sel': 168, 'Ngenes_sel': 101, 'Nidentical': 0, 'Nduplicates': 39, 'Nfeatures': 101, 'Ngenes': 101}\n",
      "{'gene': 173, 'CDS': 97, 'tRNA': 59, 'exon': 51, 'intron': 30, 'rRNA': 21, 'misc_feature': 2, 'repeat_region': 2, 'source': 1}\n",
      "{'CDS': 68, 'tRNA': 29, 'rRNA': 4}\n",
      "{'Sample': 'NC_054359.1', 'Nfeatures_raw': 445, 'Ngenes_raw': 116, 'Nfeatures_sel': 168, 'Ngenes_sel': 114, 'Nidentical': 0, 'Nduplicates': 34, 'Nfeatures': 114, 'Ngenes': 114}\n",
      "{'gene': 158, 'CDS': 92, 'exon': 76, 'tRNA': 66, 'intron': 42, 'rRNA': 10, 'source': 1}\n",
      "{'CDS': 79, 'tRNA': 31, 'rRNA': 4}\n"
     ]
    }
   ],
   "source": [
    "ref_dir = 'References/'\n",
    "all_gb_files = [ref_dir + file for file in os.listdir(ref_dir) if file.endswith(\".gb\")]\n",
    "# all_gb_files = ['References/NC_003386.1.gb']\n",
    "print(len(all_gb_files))\n",
    "\n",
    "Samples_ls = []\n",
    "for gb_file in all_gb_files:\n",
    "    # Get features\n",
    "    seq_df = get_features(gb_file)\n",
    "\n",
    "    # Get raw feature stats\n",
    "    sample_stats={'Sample':gb_file.split('/')[-1].replace('.gb','')}\n",
    "    sample_stats['Nfeatures_raw']=seq_df.shape[0];sample_stats['Ngenes_raw']=seq_df.gene.nunique()\n",
    "    feat_raw_stats = seq_df.groupby('type').size().sort_values(ascending=False).to_dict()\n",
    "\n",
    "    # Removing fragments\n",
    "    seq_df = seq_df[seq_df.gene.str.contains('-fragment')==False]\n",
    "    # Keeping CDS, rRNA, repeat_region\n",
    "    accepted_types = ['CDS','rRNA','tRNA'] # Could add repeat_region\n",
    "    seq_gene = seq_df[seq_df.type=='gene']\n",
    "    seq_df = seq_df[seq_df.type.isin(accepted_types)]\n",
    "    seq_gene = seq_gene[seq_gene.gene.isin(seq_df.gene)==False]\n",
    "    sample_stats['Nfeatures_sel']=seq_df.shape[0];sample_stats['Ngenes_sel']=seq_df.gene.nunique()\n",
    "\n",
    "    # Removing duplicates\n",
    "    sample_stats['Nidentical'] = seq_df[seq_df.duplicated(keep=False)].gene.nunique()\n",
    "    sample_stats['Nduplicates'] = seq_df[seq_df.duplicated(subset='gene',keep=False)].gene.nunique()\n",
    "    seq_df = seq_df.sort_values('len',ascending=False).groupby('gene').head(1) # Keep longest sequence per gene\n",
    "    sample_stats['Nfeatures']=seq_df.shape[0];sample_stats['Ngenes']=seq_df.gene.nunique()\n",
    "    Samples_ls.append(sample_stats)\n",
    "    print(sample_stats)\n",
    "    print(feat_raw_stats)\n",
    "    print(seq_df.groupby('type').size().sort_values(ascending=False).to_dict())\n",
    "\n",
    "    seq_df = seq_df.sort_values(['type','gene']).reset_index(drop=True)\n",
    "    write_fasta(seq_df, sample_stats, fasta_file = gb_file.replace('.gb','_CDS.fasta'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ngenes: {0.1: 113, 0.5: 115, 0.9: 119}\n",
      "Nduplicates: {0.1: 32, 0.5: 36, 0.9: 47}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Nfeatures_raw</th>\n",
       "      <th>Ngenes_raw</th>\n",
       "      <th>Nfeatures_sel</th>\n",
       "      <th>Ngenes_sel</th>\n",
       "      <th>Nidentical</th>\n",
       "      <th>Nduplicates</th>\n",
       "      <th>Nfeatures</th>\n",
       "      <th>Ngenes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_003386.1</td>\n",
       "      <td>546</td>\n",
       "      <td>129</td>\n",
       "      <td>208</td>\n",
       "      <td>119</td>\n",
       "      <td>0</td>\n",
       "      <td>49</td>\n",
       "      <td>119</td>\n",
       "      <td>119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_005086.1</td>\n",
       "      <td>467</td>\n",
       "      <td>120</td>\n",
       "      <td>177</td>\n",
       "      <td>115</td>\n",
       "      <td>0</td>\n",
       "      <td>38</td>\n",
       "      <td>115</td>\n",
       "      <td>115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_009962.1</td>\n",
       "      <td>448</td>\n",
       "      <td>119</td>\n",
       "      <td>169</td>\n",
       "      <td>114</td>\n",
       "      <td>0</td>\n",
       "      <td>32</td>\n",
       "      <td>114</td>\n",
       "      <td>114</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_010323.1</td>\n",
       "      <td>455</td>\n",
       "      <td>121</td>\n",
       "      <td>170</td>\n",
       "      <td>115</td>\n",
       "      <td>0</td>\n",
       "      <td>32</td>\n",
       "      <td>115</td>\n",
       "      <td>115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_010654.1</td>\n",
       "      <td>390</td>\n",
       "      <td>103</td>\n",
       "      <td>148</td>\n",
       "      <td>96</td>\n",
       "      <td>0</td>\n",
       "      <td>25</td>\n",
       "      <td>96</td>\n",
       "      <td>96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>NC_053912.1</td>\n",
       "      <td>453</td>\n",
       "      <td>120</td>\n",
       "      <td>170</td>\n",
       "      <td>114</td>\n",
       "      <td>0</td>\n",
       "      <td>33</td>\n",
       "      <td>114</td>\n",
       "      <td>114</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>NC_054296.1</td>\n",
       "      <td>457</td>\n",
       "      <td>120</td>\n",
       "      <td>171</td>\n",
       "      <td>114</td>\n",
       "      <td>0</td>\n",
       "      <td>33</td>\n",
       "      <td>114</td>\n",
       "      <td>114</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>NC_054306.1</td>\n",
       "      <td>490</td>\n",
       "      <td>121</td>\n",
       "      <td>181</td>\n",
       "      <td>114</td>\n",
       "      <td>0</td>\n",
       "      <td>38</td>\n",
       "      <td>114</td>\n",
       "      <td>114</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>NC_054357.1</td>\n",
       "      <td>436</td>\n",
       "      <td>108</td>\n",
       "      <td>168</td>\n",
       "      <td>101</td>\n",
       "      <td>0</td>\n",
       "      <td>39</td>\n",
       "      <td>101</td>\n",
       "      <td>101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>NC_054359.1</td>\n",
       "      <td>445</td>\n",
       "      <td>116</td>\n",
       "      <td>168</td>\n",
       "      <td>114</td>\n",
       "      <td>0</td>\n",
       "      <td>34</td>\n",
       "      <td>114</td>\n",
       "      <td>114</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>124 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Sample  Nfeatures_raw  Ngenes_raw  Nfeatures_sel  Ngenes_sel  \\\n",
       "0    NC_003386.1            546         129            208         119   \n",
       "1    NC_005086.1            467         120            177         115   \n",
       "2    NC_009962.1            448         119            169         114   \n",
       "3    NC_010323.1            455         121            170         115   \n",
       "4    NC_010654.1            390         103            148          96   \n",
       "..           ...            ...         ...            ...         ...   \n",
       "119  NC_053912.1            453         120            170         114   \n",
       "120  NC_054296.1            457         120            171         114   \n",
       "121  NC_054306.1            490         121            181         114   \n",
       "122  NC_054357.1            436         108            168         101   \n",
       "123  NC_054359.1            445         116            168         114   \n",
       "\n",
       "     Nidentical  Nduplicates  Nfeatures  Ngenes  \n",
       "0             0           49        119     119  \n",
       "1             0           38        115     115  \n",
       "2             0           32        114     114  \n",
       "3             0           32        115     115  \n",
       "4             0           25         96      96  \n",
       "..          ...          ...        ...     ...  \n",
       "119           0           33        114     114  \n",
       "120           0           33        114     114  \n",
       "121           0           38        114     114  \n",
       "122           0           39        101     101  \n",
       "123           0           34        114     114  \n",
       "\n",
       "[124 rows x 9 columns]"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Samples_stats = pd.DataFrame(Samples_ls)\n",
    "Samples_stats.to_csv(ref_dir + 'Ref_stats.csv',index=False)\n",
    "print('Ngenes:',Samples_stats.Ngenes.quantile([.1,.5,.9]).astype(int).to_dict())\n",
    "print('Nduplicates:',Samples_stats.Nduplicates.quantile([.1,.5,.9]).astype(int).to_dict())\n",
    "Samples_stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "124\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "14213"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Concat .fasta files in a single fasta\n",
    "all_fa_files = [file for file in os.listdir(ref_dir) if file.endswith(\"_CDS.fasta\")]\n",
    "print(len(all_fa_files))\n",
    "all_seqs=[]; ls_id=[]\n",
    "for ifa in all_fa_files:\n",
    "    for record in SeqIO.parse(ref_dir + ifa,format='fasta'):\n",
    "        all_seqs.append(record)\n",
    "SeqIO.write(all_seqs,ref_dir + 'cpDNA_RefCDS_raw.fasta',format='fasta')\n",
    "len(all_seqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[126649.0, 127202.0, 125052.0, 125582.0]\n",
      "446.0 29052.0\n"
     ]
    }
   ],
   "source": [
    "# https://blog.finxter.com/how-to-extract-numbers-from-a-string-in-python/\n",
    "import re\n",
    "def get_loc(sentence):\n",
    "    return [float(s) for s in re.findall(r'-?\\d+\\.?\\d*', sentence)]\n",
    "print(get_loc('join{[126649:127202](-), [125052:125582](-)}\t'))\n",
    "\n",
    "ls_loc=[76120.0, 76234.0, 105286.0, 105518.0, 104729.0, 104829.0]\n",
    "def len_exons(ls_loc):\n",
    "    len_exon=0;\n",
    "    for i in range(0,len(ls_loc),2):\n",
    "        exon = ls_loc[i:i+2]\n",
    "        len_exon += max(exon) - min(exon)\n",
    "    return len_exon\n",
    "def len_introns(ls_loc):\n",
    "    len_intron=0;\n",
    "    if len(ls_loc)/4>0:\n",
    "        for i in range(1,len(ls_loc),4):\n",
    "            intron = ls_loc[i:i+2]\n",
    "            len_intron += max(intron) - min(intron)\n",
    "    return len_intron\n",
    "print(len_exons(ls_loc),len_introns(ls_loc))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>len</th>\n",
       "      <th>join</th>\n",
       "      <th>forward</th>\n",
       "      <th>reverse</th>\n",
       "      <th>location</th>\n",
       "      <th>Accession</th>\n",
       "      <th>gene</th>\n",
       "      <th>ls_loc</th>\n",
       "      <th>N_exons</th>\n",
       "      <th>len_exons</th>\n",
       "      <th>len_introns</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_003386.1-accD</td>\n",
       "      <td>929</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[57912:58841](+)</td>\n",
       "      <td>NC_003386.1</td>\n",
       "      <td>accD</td>\n",
       "      <td>[57912.0, 58841.0]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>929.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_003386.1-atpA</td>\n",
       "      <td>1527</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[15691:17218](-)</td>\n",
       "      <td>NC_003386.1</td>\n",
       "      <td>atpA</td>\n",
       "      <td>[15691.0, 17218.0]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1527.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              seqid   len   join  forward  reverse          location  \\\n",
       "0  NC_003386.1-accD   929  False     True    False  [57912:58841](+)   \n",
       "1  NC_003386.1-atpA  1527  False    False     True  [15691:17218](-)   \n",
       "\n",
       "     Accession  gene              ls_loc  N_exons  len_exons  len_introns  \n",
       "0  NC_003386.1  accD  [57912.0, 58841.0]      1.0      929.0          0.0  \n",
       "1  NC_003386.1  atpA  [15691.0, 17218.0]      1.0     1527.0          0.0  "
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_stats={}\n",
    "for record in all_seqs:\n",
    "    location = record.description.split('location=')[1].split(';')[0]\n",
    "    ref_stats[record.id]={'len':len(record),'join':'join' in location,'forward':'+' in location,'reverse':'-' in location,\n",
    "                         'location':location}\n",
    "ref_stats = pd.DataFrame.from_dict(ref_stats,orient='index').reset_index().rename(columns={'index':'seqid'})\n",
    "ref_stats[['Accession','gene']]=ref_stats.seqid.str.split('-',expand=True)\n",
    "ref_stats['ls_loc'] = ref_stats.apply(lambda row: get_loc(row['location']),axis=1)\n",
    "ref_stats['N_exons'] = ref_stats.apply(lambda row: len(row['ls_loc'])/2,axis=1)\n",
    "ref_stats['len_exons'] = ref_stats.apply(lambda row: len_exons(row['ls_loc']),axis=1)\n",
    "ref_stats['len_introns'] = ref_stats.apply(lambda row: len_introns(row['ls_loc']),axis=1)\n",
    "ref_stats[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>len_min</th>\n",
       "      <th>len_median</th>\n",
       "      <th>len_max</th>\n",
       "      <th>Accession_count</th>\n",
       "      <th>forward_sum</th>\n",
       "      <th>reverse_sum</th>\n",
       "      <th>join_sum</th>\n",
       "      <th>N_exons_mean</th>\n",
       "      <th>len_exons_median</th>\n",
       "      <th>len_introns_median</th>\n",
       "      <th>len_introns_max</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>atpF</th>\n",
       "      <td>424</td>\n",
       "      <td>555.0</td>\n",
       "      <td>576</td>\n",
       "      <td>124</td>\n",
       "      <td>13</td>\n",
       "      <td>111</td>\n",
       "      <td>123</td>\n",
       "      <td>1.991935</td>\n",
       "      <td>555.0</td>\n",
       "      <td>1279.5</td>\n",
       "      <td>2338.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>clpP1</th>\n",
       "      <td>273</td>\n",
       "      <td>591.0</td>\n",
       "      <td>948</td>\n",
       "      <td>121</td>\n",
       "      <td>10</td>\n",
       "      <td>111</td>\n",
       "      <td>106</td>\n",
       "      <td>2.727273</td>\n",
       "      <td>591.0</td>\n",
       "      <td>1160.0</td>\n",
       "      <td>1954.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ndhA</th>\n",
       "      <td>301</td>\n",
       "      <td>1092.0</td>\n",
       "      <td>1119</td>\n",
       "      <td>116</td>\n",
       "      <td>23</td>\n",
       "      <td>93</td>\n",
       "      <td>113</td>\n",
       "      <td>1.974138</td>\n",
       "      <td>1092.0</td>\n",
       "      <td>2154.0</td>\n",
       "      <td>2553.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ndhB</th>\n",
       "      <td>311</td>\n",
       "      <td>1533.0</td>\n",
       "      <td>1569</td>\n",
       "      <td>119</td>\n",
       "      <td>46</td>\n",
       "      <td>73</td>\n",
       "      <td>114</td>\n",
       "      <td>1.957983</td>\n",
       "      <td>1533.0</td>\n",
       "      <td>2203.0</td>\n",
       "      <td>2295.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pafI</th>\n",
       "      <td>228</td>\n",
       "      <td>507.0</td>\n",
       "      <td>543</td>\n",
       "      <td>124</td>\n",
       "      <td>11</td>\n",
       "      <td>113</td>\n",
       "      <td>121</td>\n",
       "      <td>2.943548</td>\n",
       "      <td>507.0</td>\n",
       "      <td>1081.0</td>\n",
       "      <td>1907.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>petB</th>\n",
       "      <td>489</td>\n",
       "      <td>648.0</td>\n",
       "      <td>708</td>\n",
       "      <td>124</td>\n",
       "      <td>116</td>\n",
       "      <td>8</td>\n",
       "      <td>113</td>\n",
       "      <td>1.911290</td>\n",
       "      <td>648.0</td>\n",
       "      <td>781.5</td>\n",
       "      <td>1469.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>petD</th>\n",
       "      <td>479</td>\n",
       "      <td>483.0</td>\n",
       "      <td>539</td>\n",
       "      <td>124</td>\n",
       "      <td>116</td>\n",
       "      <td>8</td>\n",
       "      <td>90</td>\n",
       "      <td>1.725806</td>\n",
       "      <td>483.0</td>\n",
       "      <td>716.0</td>\n",
       "      <td>1226.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rpl16</th>\n",
       "      <td>9</td>\n",
       "      <td>408.0</td>\n",
       "      <td>492</td>\n",
       "      <td>124</td>\n",
       "      <td>8</td>\n",
       "      <td>116</td>\n",
       "      <td>102</td>\n",
       "      <td>1.822581</td>\n",
       "      <td>408.0</td>\n",
       "      <td>1406.0</td>\n",
       "      <td>2195.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rpl2</th>\n",
       "      <td>392</td>\n",
       "      <td>825.0</td>\n",
       "      <td>849</td>\n",
       "      <td>123</td>\n",
       "      <td>57</td>\n",
       "      <td>66</td>\n",
       "      <td>117</td>\n",
       "      <td>1.959350</td>\n",
       "      <td>825.0</td>\n",
       "      <td>1285.0</td>\n",
       "      <td>158597.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rpoC1</th>\n",
       "      <td>1359</td>\n",
       "      <td>2049.0</td>\n",
       "      <td>2118</td>\n",
       "      <td>124</td>\n",
       "      <td>14</td>\n",
       "      <td>110</td>\n",
       "      <td>118</td>\n",
       "      <td>1.951613</td>\n",
       "      <td>2049.0</td>\n",
       "      <td>2795.0</td>\n",
       "      <td>3127.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rps12</th>\n",
       "      <td>232</td>\n",
       "      <td>372.0</td>\n",
       "      <td>806</td>\n",
       "      <td>124</td>\n",
       "      <td>81</td>\n",
       "      <td>115</td>\n",
       "      <td>119</td>\n",
       "      <td>2.887097</td>\n",
       "      <td>372.0</td>\n",
       "      <td>69194.5</td>\n",
       "      <td>135516.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rps16</th>\n",
       "      <td>40</td>\n",
       "      <td>261.0</td>\n",
       "      <td>534</td>\n",
       "      <td>111</td>\n",
       "      <td>5</td>\n",
       "      <td>106</td>\n",
       "      <td>90</td>\n",
       "      <td>1.810811</td>\n",
       "      <td>261.0</td>\n",
       "      <td>1104.0</td>\n",
       "      <td>1822.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnA_UGC</th>\n",
       "      <td>35</td>\n",
       "      <td>105.0</td>\n",
       "      <td>435</td>\n",
       "      <td>123</td>\n",
       "      <td>58</td>\n",
       "      <td>65</td>\n",
       "      <td>14</td>\n",
       "      <td>1.113821</td>\n",
       "      <td>105.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>885.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnC_ACA</th>\n",
       "      <td>72</td>\n",
       "      <td>95.0</td>\n",
       "      <td>97</td>\n",
       "      <td>59</td>\n",
       "      <td>3</td>\n",
       "      <td>56</td>\n",
       "      <td>57</td>\n",
       "      <td>1.966102</td>\n",
       "      <td>95.0</td>\n",
       "      <td>663.0</td>\n",
       "      <td>776.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnE_UUC</th>\n",
       "      <td>73</td>\n",
       "      <td>75.0</td>\n",
       "      <td>80</td>\n",
       "      <td>124</td>\n",
       "      <td>17</td>\n",
       "      <td>107</td>\n",
       "      <td>10</td>\n",
       "      <td>1.080645</td>\n",
       "      <td>75.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1044.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnF_AAA</th>\n",
       "      <td>80</td>\n",
       "      <td>93.0</td>\n",
       "      <td>93</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>1.857143</td>\n",
       "      <td>93.0</td>\n",
       "      <td>657.0</td>\n",
       "      <td>688.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnG_UCC</th>\n",
       "      <td>23</td>\n",
       "      <td>71.0</td>\n",
       "      <td>92</td>\n",
       "      <td>119</td>\n",
       "      <td>108</td>\n",
       "      <td>11</td>\n",
       "      <td>116</td>\n",
       "      <td>1.974790</td>\n",
       "      <td>71.0</td>\n",
       "      <td>723.0</td>\n",
       "      <td>954.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnI_AAU</th>\n",
       "      <td>73</td>\n",
       "      <td>91.5</td>\n",
       "      <td>92</td>\n",
       "      <td>10</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>1.900000</td>\n",
       "      <td>91.5</td>\n",
       "      <td>631.0</td>\n",
       "      <td>725.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnI_GAU</th>\n",
       "      <td>72</td>\n",
       "      <td>88.0</td>\n",
       "      <td>295</td>\n",
       "      <td>121</td>\n",
       "      <td>58</td>\n",
       "      <td>63</td>\n",
       "      <td>35</td>\n",
       "      <td>1.289256</td>\n",
       "      <td>88.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1141.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnK_UUU</th>\n",
       "      <td>37</td>\n",
       "      <td>72.0</td>\n",
       "      <td>90</td>\n",
       "      <td>115</td>\n",
       "      <td>6</td>\n",
       "      <td>109</td>\n",
       "      <td>112</td>\n",
       "      <td>1.973913</td>\n",
       "      <td>72.0</td>\n",
       "      <td>2582.0</td>\n",
       "      <td>4056.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnL_UAA</th>\n",
       "      <td>34</td>\n",
       "      <td>85.0</td>\n",
       "      <td>472</td>\n",
       "      <td>123</td>\n",
       "      <td>112</td>\n",
       "      <td>11</td>\n",
       "      <td>112</td>\n",
       "      <td>1.910569</td>\n",
       "      <td>85.0</td>\n",
       "      <td>510.0</td>\n",
       "      <td>1300.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnS_CGA</th>\n",
       "      <td>70</td>\n",
       "      <td>92.0</td>\n",
       "      <td>94</td>\n",
       "      <td>84</td>\n",
       "      <td>75</td>\n",
       "      <td>9</td>\n",
       "      <td>77</td>\n",
       "      <td>1.916667</td>\n",
       "      <td>92.0</td>\n",
       "      <td>698.0</td>\n",
       "      <td>933.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnT_CGU</th>\n",
       "      <td>77</td>\n",
       "      <td>78.0</td>\n",
       "      <td>91</td>\n",
       "      <td>23</td>\n",
       "      <td>21</td>\n",
       "      <td>2</td>\n",
       "      <td>22</td>\n",
       "      <td>1.956522</td>\n",
       "      <td>78.0</td>\n",
       "      <td>718.0</td>\n",
       "      <td>932.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnV_UAC</th>\n",
       "      <td>37</td>\n",
       "      <td>74.0</td>\n",
       "      <td>285</td>\n",
       "      <td>118</td>\n",
       "      <td>10</td>\n",
       "      <td>108</td>\n",
       "      <td>104</td>\n",
       "      <td>1.881356</td>\n",
       "      <td>74.0</td>\n",
       "      <td>659.0</td>\n",
       "      <td>775.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          len_min  len_median  len_max  Accession_count  forward_sum  \\\n",
       "gene                                                                   \n",
       "atpF          424       555.0      576              124           13   \n",
       "clpP1         273       591.0      948              121           10   \n",
       "ndhA          301      1092.0     1119              116           23   \n",
       "ndhB          311      1533.0     1569              119           46   \n",
       "pafI          228       507.0      543              124           11   \n",
       "petB          489       648.0      708              124          116   \n",
       "petD          479       483.0      539              124          116   \n",
       "rpl16           9       408.0      492              124            8   \n",
       "rpl2          392       825.0      849              123           57   \n",
       "rpoC1        1359      2049.0     2118              124           14   \n",
       "rps12         232       372.0      806              124           81   \n",
       "rps16          40       261.0      534              111            5   \n",
       "trnA_UGC       35       105.0      435              123           58   \n",
       "trnC_ACA       72        95.0       97               59            3   \n",
       "trnE_UUC       73        75.0       80              124           17   \n",
       "trnF_AAA       80        93.0       93                7            1   \n",
       "trnG_UCC       23        71.0       92              119          108   \n",
       "trnI_AAU       73        91.5       92               10            4   \n",
       "trnI_GAU       72        88.0      295              121           58   \n",
       "trnK_UUU       37        72.0       90              115            6   \n",
       "trnL_UAA       34        85.0      472              123          112   \n",
       "trnS_CGA       70        92.0       94               84           75   \n",
       "trnT_CGU       77        78.0       91               23           21   \n",
       "trnV_UAC       37        74.0      285              118           10   \n",
       "\n",
       "          reverse_sum  join_sum  N_exons_mean  len_exons_median  \\\n",
       "gene                                                              \n",
       "atpF              111       123      1.991935             555.0   \n",
       "clpP1             111       106      2.727273             591.0   \n",
       "ndhA               93       113      1.974138            1092.0   \n",
       "ndhB               73       114      1.957983            1533.0   \n",
       "pafI              113       121      2.943548             507.0   \n",
       "petB                8       113      1.911290             648.0   \n",
       "petD                8        90      1.725806             483.0   \n",
       "rpl16             116       102      1.822581             408.0   \n",
       "rpl2               66       117      1.959350             825.0   \n",
       "rpoC1             110       118      1.951613            2049.0   \n",
       "rps12             115       119      2.887097             372.0   \n",
       "rps16             106        90      1.810811             261.0   \n",
       "trnA_UGC           65        14      1.113821             105.0   \n",
       "trnC_ACA           56        57      1.966102              95.0   \n",
       "trnE_UUC          107        10      1.080645              75.0   \n",
       "trnF_AAA            6         6      1.857143              93.0   \n",
       "trnG_UCC           11       116      1.974790              71.0   \n",
       "trnI_AAU            6         9      1.900000              91.5   \n",
       "trnI_GAU           63        35      1.289256              88.0   \n",
       "trnK_UUU          109       112      1.973913              72.0   \n",
       "trnL_UAA           11       112      1.910569              85.0   \n",
       "trnS_CGA            9        77      1.916667              92.0   \n",
       "trnT_CGU            2        22      1.956522              78.0   \n",
       "trnV_UAC          108       104      1.881356              74.0   \n",
       "\n",
       "          len_introns_median  len_introns_max  \n",
       "gene                                           \n",
       "atpF                  1279.5           2338.0  \n",
       "clpP1                 1160.0           1954.0  \n",
       "ndhA                  2154.0           2553.0  \n",
       "ndhB                  2203.0           2295.0  \n",
       "pafI                  1081.0           1907.0  \n",
       "petB                   781.5           1469.0  \n",
       "petD                   716.0           1226.0  \n",
       "rpl16                 1406.0           2195.0  \n",
       "rpl2                  1285.0         158597.0  \n",
       "rpoC1                 2795.0           3127.0  \n",
       "rps12                69194.5         135516.0  \n",
       "rps16                 1104.0           1822.0  \n",
       "trnA_UGC                 0.0            885.0  \n",
       "trnC_ACA               663.0            776.0  \n",
       "trnE_UUC                 0.0           1044.0  \n",
       "trnF_AAA               657.0            688.0  \n",
       "trnG_UCC               723.0            954.0  \n",
       "trnI_AAU               631.0            725.0  \n",
       "trnI_GAU                 0.0           1141.0  \n",
       "trnK_UUU              2582.0           4056.0  \n",
       "trnL_UAA               510.0           1300.0  \n",
       "trnS_CGA               698.0            933.0  \n",
       "trnT_CGU               718.0            932.0  \n",
       "trnV_UAC               659.0            775.0  "
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_stats = ref_stats.groupby('gene').agg({'len':['min','median','max'],'Accession':['count'],'forward':['sum'],\n",
    "                            'reverse':['sum'],'join':['sum'],'N_exons':['mean'],'len_exons':['median'],'len_introns':['median','max']})\n",
    "gene_stats.columns = ['_'.join(col).strip() for col in gene_stats.columns.values]\n",
    "gene_stats.reset_index().to_csv(ref_dir + 'gene_stats.csv',index=False)\n",
    "gene_stats[gene_stats.join_sum>5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "144 {'rpoC2': 124, 'rps12': 124, 'atpA': 124, 'rpoC1': 124, 'rpoB': 124, 'rpoA': 124, 'rpl36': 124, 'rpl33': 124, 'rpl20': 124, 'rpl16': 124, 'rpl14': 124, 'rbcL': 124, 'psbZ': 124, 'psbT': 124, 'psbM': 124, 'psbL': 124, 'psbK': 124, 'psbI': 124, 'psbH': 124, 'psbF': 124, 'psbE': 124, 'rps11': 124, 'rps14': 124, 'psbC': 124, 'rps18': 124, 'trnY_GUA': 124, 'trnW_CCA': 124, 'trnR_ACG': 124, 'trnQ_UUG': 124, 'trnM_CAU': 124, 'trnH_GUG': 124, 'trnE_UUC': 124, 'trnD_GUC': 124, 'trnC_GCA': 124, 'rrn5': 124, 'rrn4.5': 124, 'rrn23': 124, 'rrn16': 124, 'rps8': 124, 'rps7': 124, 'rps4': 124, 'rps3': 124, 'rps2': 124, 'rps19': 124, 'psbD': 124, 'ycf2': 124, 'psbB': 124, 'petG': 124, 'ccsA': 124, 'atpI': 124, 'pafI': 124, 'pafII': 124, 'pbf1': 124, 'atpH': 124, 'atpF': 124, 'petA': 124, 'petB': 124, 'petD': 124, 'psbA': 124, 'psaA': 124, 'psaB': 124, 'psaC': 124, 'atpB': 124, 'petN': 124, 'psaI': 124, 'psaJ': 124, 'atpE': 124, 'trnS_GCU': 123, 'trnL_UAG': 123, 'trnF_GAA': 123, 'trnL_UAA': 123, 'trnS_UGA': 123, 'trnN_GUU': 123, 'trnP_UGG': 123, 'trnA_UGC': 123, 'trnR_UCU': 123, 'cemA': 123, 'psbJ': 123, 'rpl22': 123, 'rpl2': 123, 'petL': 123, 'trnS_GGA': 122, 'matK': 122, 'ycf1': 122, 'trnI_GAU': 121, 'rpl23': 121, 'trnG_GCC': 121, 'trnL_CAA': 121, 'clpP1': 121, 'rps15': 120, 'trnT_GGU': 119, 'ndhD': 119, 'ndhB': 119, 'trnG_UCC': 119, 'accD': 118, 'trnV_UAC': 118, 'infA': 118, 'ndhC': 117, 'ndhE': 117, 'ndhG': 117, 'ndhK': 116, 'trnV_GAC': 116, 'ndhH': 116, 'ndhF': 116, 'ndhA': 116, 'trnT_UGU': 115, 'rpl32': 115, 'trnK_UUU': 115, 'ndhJ': 115, 'ndhI': 114, 'rps16': 111, 'trnS_CGA': 84, 'trnC_ACA': 59, 'psaM': 29, 'psb30': 27, 'trnP_GGG': 24, 'chlL': 24, 'chlN': 23, 'trnT_CGU': 23, 'chlB': 23, 'rpl21': 19, 'trnR_CCG': 16, 'trnI_AAU': 10, 'trnF_AAA': 7, 'trnfM_CAU': 5, 'trnY_AUA': 4, 'cysT': 4, 'trnS_AGA': 3, 'trnTERM_UUA': 3, 'trnnull_NNN': 3, 'trnR_UCG': 3, 'trnO_CUA': 3, 'trnL_AAG': 2, 'trnI_UAU': 2, 'trnG_ACC': 2, 'trnD_AUC': 2, 'trnK_CUU': 1, 'trnQ_CUG': 1, 'cysA': 1, 'trnA_AGC': 1, 'trnR_CCU': 1, 'trnT_AGU': 1, 'trnS_ACU': 1, 'trnL_CAG': 1}\n"
     ]
    }
   ],
   "source": [
    "print(gene_stats.shape[0],gene_stats.Accession_count.sort_values(ascending=False).to_dict())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NC_003386.1-cemA removed 1515 690.0\n",
      "NC_003386.1-trnL_UAA removed 472 85.0\n",
      "NC_005086.1-trnV_UAC removed 168 74.0\n",
      "NC_010323.1-infA removed 105 234.0\n",
      "NC_010654.1-trnS_UGA removed 402 93.0\n",
      "NC_011156.4-cysT removed 170 474.0\n",
      "NC_011156.4-ndhB removed 311 1533.0\n",
      "NC_011156.4-ndhD removed 114 1503.0\n",
      "NC_011156.4-rpl32 removed 456 171.0\n",
      "NC_011156.4-rps16 removed 66 261.0\n",
      "NC_011156.4-trnL_UAA removed 35 85.0\n",
      "NC_012818.1-cemA removed 1521 690.0\n",
      "NC_012818.1-clpP1 removed 273 591.0\n",
      "NC_012818.1-rpoB removed 696 3213.0\n",
      "NC_016986.1-rpl21 removed 39 360.0\n",
      "NC_016986.1-rpl23 removed 81 282.0\n",
      "NC_016986.1-trnS_AGA removed 336 108.0\n",
      "NC_017006.1-cemA removed 1518 690.0\n",
      "NC_017006.1-trnG_UCC removed 23 71.0\n",
      "NC_020146.1-cemA removed 1428 690.0\n",
      "NC_020146.1-rpl16 removed 81 408.0\n",
      "NC_020146.1-ycf2 removed 3409 6862.5\n",
      "NC_020146.1-trnL_UAA removed 345 85.0\n",
      "NC_022137.1-cemA removed 1533 690.0\n",
      "NC_022137.1-rps11 removed 168 417.0\n",
      "NC_024157.1-cemA removed 1521 690.0\n",
      "NC_024157.1-trnI_GAU removed 295 88.0\n",
      "NC_024158.1-cemA removed 1527 690.0\n",
      "NC_024158.1-rpoB removed 453 3213.0\n",
      "NC_024158.1-rps11 removed 156 417.0\n",
      "NC_024158.1-trnL_UAA removed 34 85.0\n",
      "NC_024929.1-infA removed 102 234.0\n",
      "NC_025745.1-infA removed 88 234.0\n",
      "NC_025745.1-rps16 removed 69 261.0\n",
      "NC_026295.1-accD removed 651 1476.0\n",
      "NC_026295.1-rps16 removed 69 261.0\n",
      "NC_026295.1-ycf1 removed 2155 5515.5\n",
      "NC_026295.1-trnL_UAA removed 372 85.0\n",
      "NC_026295.1-trnV_UAC removed 264 74.0\n",
      "NC_026301.1-chlL removed 267 874.5\n",
      "NC_026301.1-trnS_UGA removed 430 93.0\n",
      "NC_026301.1-trnV_UAC removed 265 74.0\n",
      "NC_026847.1-accD removed 718 1476.0\n",
      "NC_026847.1-ndhA removed 301 1092.0\n",
      "NC_026847.1-ndhD removed 234 1503.0\n",
      "NC_026847.1-rpl16 removed 9 408.0\n",
      "NC_026847.1-rpl23 removed 723 282.0\n",
      "NC_026847.1-rps12 removed 806 372.0\n",
      "NC_026847.1-ycf1 removed 466 5515.5\n",
      "NC_027829.1-ndhB removed 360 1533.0\n",
      "NC_027829.1-rpl33 removed 69 201.0\n",
      "NC_030601.1-rps7 removed 123 468.0\n",
      "NC_033874.1-cemA removed 1566 690.0\n",
      "NC_033874.1-ycf2 removed 3404 6862.5\n",
      "NC_033874.1-trnV_UAC removed 285 74.0\n",
      "NC_034686.1-rps16 removed 534 261.0\n",
      "NC_034942.1-ycf1 removed 2472 5515.5\n",
      "NC_034942.1-trnS_UGA removed 521 93.0\n",
      "NC_035050.1-ycf1 removed 807 5515.5\n",
      "NC_035499.1-cemA removed 174 690.0\n",
      "NC_035807.1-cemA removed 1515 690.0\n",
      "NC_035807.1-pafI removed 228 507.0\n",
      "NC_035807.1-trnL_UAA removed 34 85.0\n",
      "NC_035854.1-cemA removed 1500 690.0\n",
      "NC_035854.1-pafI removed 228 507.0\n",
      "NC_035854.1-rpoB removed 813 3213.0\n",
      "NC_035854.1-rps11 removed 168 417.0\n",
      "NC_035854.1-trnL_UAA removed 34 85.0\n",
      "NC_037151.1-infA removed 60 234.0\n",
      "NC_037504.1-trnS_UGA removed 447 93.0\n",
      "NC_037845.1-trnH_GUG removed 160572 75.0\n",
      "NC_038074.1-cemA removed 1500 690.0\n",
      "NC_038074.1-infA removed 78 234.0\n",
      "NC_038074.1-ndhJ removed 129 477.0\n",
      "NC_038074.1-rps16 removed 68 261.0\n",
      "NC_038074.1-rps2 removed 30 711.0\n",
      "NC_039155.1-accD removed 731 1476.0\n",
      "NC_039155.1-trnV_UAC removed 265 74.0\n",
      "NC_039395.1-infA removed 75 234.0\n",
      "NC_040219.1-cemA removed 1536 690.0\n",
      "NC_040219.1-ndhB removed 701 1533.0\n",
      "NC_040219.1-pafI removed 228 507.0\n",
      "NC_040219.1-trnL_UAA removed 34 85.0\n",
      "NC_040862.1-trnA_UGC removed 435 105.0\n",
      "NC_040862.1-trnH_GUG removed 121677 75.0\n",
      "NC_040862.1-trnL_UAA removed 366 85.0\n",
      "NC_041000.1-cemA removed 1473 690.0\n",
      "NC_041000.1-petA removed 12 963.0\n",
      "NC_041000.1-rpl16 removed 9 408.0\n",
      "NC_041000.1-ycf1 removed 2641 5515.5\n",
      "NC_041000.1-trnG_UCC removed 23 71.0\n",
      "NC_041261.1-accD removed 633 1476.0\n",
      "NC_041261.1-rpl23 removed 120 282.0\n",
      "NC_041466.1-accD removed 650 1476.0\n",
      "NC_041466.1-rps16 removed 524 261.0\n",
      "NC_041466.1-ycf1 removed 2246 5515.5\n",
      "NC_041466.1-trnA_UGC removed 35 105.0\n",
      "NC_041481.1-rpl2 removed 392 825.0\n",
      "NC_041640.1-rpl36 removed 600 114.0\n",
      "NC_042211.1-ycf1 removed 1257 5515.5\n",
      "NC_042597.1-infA removed 69 234.0\n",
      "NC_042597.1-ndhD removed 250 1503.0\n",
      "NC_042597.1-ndhG removed 162 531.0\n",
      "NC_042940.1-infA removed 102 234.0\n",
      "NC_042940.1-trnL_UAA removed 428 85.0\n",
      "NC_043796.1-ycf1 removed 744 5515.5\n",
      "NC_043796.1-trnA_UGC removed 302 105.0\n",
      "NC_043800.1-infA removed 24 234.0\n",
      "NC_045235.1-infA removed 108 234.0\n",
      "NC_045280.1-ndhA removed 469 1092.0\n",
      "NC_045280.1-ndhF removed 1055 2226.0\n",
      "NC_045280.1-ndhH removed 207 1182.0\n",
      "NC_045280.1-rpl16 removed 9 408.0\n",
      "NC_045280.1-rps16 removed 40 261.0\n",
      "NC_045280.1-trnH_GUG removed 148627 75.0\n",
      "NC_045294.1-accD removed 266 1476.0\n",
      "NC_046575.1-rpl23 removed 135 282.0\n",
      "NC_050270.1-infA removed 54 234.0\n",
      "NC_050270.1-rpl22 removed 141 426.0\n",
      "NC_051971.1-infA removed 78 234.0\n",
      "NC_051971.1-rps16 removed 40 261.0\n",
      "NC_051978.1-cemA removed 1524 690.0\n",
      "NC_052912.1-infA removed 108 234.0\n",
      "NC_054296.1-trnH_GUG removed 155479 75.0\n",
      "NC_054357.1-ycf1 removed 1496 5515.5\n",
      "NC_054357.1-trnS_UGA removed 375 93.0\n",
      "NC_054359.1-trnH_GUG removed 153356 75.0\n",
      "NC_054359.1-trnL_UAA removed 468 85.0\n",
      "14085\n"
     ]
    }
   ],
   "source": [
    "# Remove outliers\n",
    "for idx, record in enumerate(all_seqs):\n",
    "    if (len(record) > gene_stats.loc[record.id.split('-')[1],:]['len_median']*2) | \\\n",
    "        (len(record) < gene_stats.loc[record.id.split('-')[1],:]['len_median']/2):\n",
    "        print(record.id,'removed',len(record),gene_stats.loc[record.id.split('-')[1],:]['len_median'])\n",
    "        del all_seqs[idx]\n",
    "print(len(all_seqs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "144\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>len_min</th>\n",
       "      <th>len_median</th>\n",
       "      <th>len_max</th>\n",
       "      <th>Accession_count</th>\n",
       "      <th>forward_sum</th>\n",
       "      <th>reverse_sum</th>\n",
       "      <th>join_sum</th>\n",
       "      <th>N_exons_mean</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>accD</th>\n",
       "      <td>746</td>\n",
       "      <td>1476.0</td>\n",
       "      <td>1893</td>\n",
       "      <td>112</td>\n",
       "      <td>101</td>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>1.008929</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>atpA</th>\n",
       "      <td>882</td>\n",
       "      <td>1524.0</td>\n",
       "      <td>1545</td>\n",
       "      <td>124</td>\n",
       "      <td>14</td>\n",
       "      <td>110</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      len_min  len_median  len_max  Accession_count  forward_sum  reverse_sum  \\\n",
       "gene                                                                            \n",
       "accD      746      1476.0     1893              112          101           11   \n",
       "atpA      882      1524.0     1545              124           14          110   \n",
       "\n",
       "      join_sum  N_exons_mean  \n",
       "gene                          \n",
       "accD         1      1.008929  \n",
       "atpA         0      1.000000  "
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_stats={}\n",
    "for record in all_seqs:\n",
    "    location = record.description.split('location=')[1].split(';')[0]\n",
    "    ref_stats[record.id]={'len':len(record),'join':'join' in location,'forward':'+' in location,'reverse':'-' in location,\n",
    "                         'location':location}\n",
    "ref_stats = pd.DataFrame.from_dict(ref_stats,orient='index').reset_index().rename(columns={'index':'seqid'})\n",
    "ref_stats[['Accession','gene']]=ref_stats.seqid.str.split('-',expand=True)\n",
    "ref_stats['ls_loc'] = ref_stats.apply(lambda row: get_loc(row['location']),axis=1)\n",
    "ref_stats['N_exons'] = ref_stats.apply(lambda row: len(row['ls_loc'])/2,axis=1)\n",
    "gene_stats = ref_stats.groupby('gene').agg({'len':['min','median','max'],'Accession':['count'],\n",
    "                                           'forward':['sum'],'reverse':['sum'],'join':['sum'],'N_exons':['mean']})\n",
    "gene_stats.columns = ['_'.join(col).strip() for col in gene_stats.columns.values]\n",
    "gene_stats.reset_index().to_csv(ref_dir + 'gene_stats.csv',index=False)\n",
    "print(gene_stats.shape[0])\n",
    "gene_stats[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "113 {'psaI': 124, 'psaB': 124, 'rps19': 124, 'psaJ': 124, 'psbA': 124, 'psbB': 124, 'psbC': 124, 'psbD': 124, 'psbE': 124, 'psbF': 124, 'psbH': 124, 'psbI': 124, 'psbK': 124, 'psbL': 124, 'psbM': 124, 'psbT': 124, 'psbZ': 124, 'rbcL': 124, 'rpl14': 124, 'rrn4.5': 124, 'rrn23': 124, 'rpl20': 124, 'atpA': 124, 'rrn16': 124, 'rps8': 124, 'rpoA': 124, 'rpoC1': 124, 'rpoC2': 124, 'rps4': 124, 'rps14': 124, 'rps3': 124, 'psaC': 124, 'rps18': 124, 'psaA': 124, 'trnE_UUC': 124, 'atpB': 124, 'atpE': 124, 'atpF': 124, 'atpH': 124, 'atpI': 124, 'ccsA': 124, 'trnY_GUA': 124, 'trnW_CCA': 124, 'trnR_ACG': 124, 'petN': 124, 'trnM_CAU': 124, 'trnQ_UUG': 124, 'pbf1': 124, 'petG': 124, 'trnD_GUC': 124, 'trnC_GCA': 124, 'petD': 124, 'petB': 124, 'rrn5': 124, 'pafII': 124, 'trnF_GAA': 123, 'rps2': 123, 'petL': 123, 'rps12': 123, 'petA': 123, 'rps7': 123, 'rpl33': 123, 'rpl36': 123, 'trnL_UAG': 123, 'trnS_GCU': 123, 'trnR_UCU': 123, 'trnP_UGG': 123, 'trnN_GUU': 123, 'psbJ': 123, 'trnS_GGA': 122, 'rpl22': 122, 'ycf2': 122, 'matK': 122, 'rpl2': 122, 'trnL_CAA': 121, 'trnG_GCC': 121, 'rps11': 121, 'rpoB': 121, 'pafI': 121, 'clpP1': 120, 'rpl16': 120, 'trnA_UGC': 120, 'rps15': 120, 'trnI_GAU': 120, 'trnH_GUG': 119, 'trnT_GGU': 119, 'trnS_UGA': 118, 'ndhE': 117, 'trnG_UCC': 117, 'ndhC': 117, 'rpl23': 117, 'trnV_GAC': 116, 'ndhK': 116, 'ndhB': 116, 'ndhG': 116, 'ndhD': 116, 'ndhF': 115, 'ndhH': 115, 'trnK_UUU': 115, 'trnT_UGU': 115, 'ndhA': 114, 'rpl32': 114, 'ndhJ': 114, 'ndhI': 114, 'ycf1': 113, 'trnV_UAC': 113, 'accD': 112, 'trnL_UAA': 112, 'cemA': 108, 'infA': 105, 'rps16': 103, 'trnS_CGA': 84, 'trnC_ACA': 59}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>len_min</th>\n",
       "      <th>len_median</th>\n",
       "      <th>len_max</th>\n",
       "      <th>Accession_count</th>\n",
       "      <th>forward_sum</th>\n",
       "      <th>reverse_sum</th>\n",
       "      <th>join_sum</th>\n",
       "      <th>N_exons_mean</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>accD</th>\n",
       "      <td>746</td>\n",
       "      <td>1476</td>\n",
       "      <td>1893</td>\n",
       "      <td>112</td>\n",
       "      <td>101</td>\n",
       "      <td>11</td>\n",
       "      <td>1</td>\n",
       "      <td>1.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>atpA</th>\n",
       "      <td>882</td>\n",
       "      <td>1524</td>\n",
       "      <td>1545</td>\n",
       "      <td>124</td>\n",
       "      <td>14</td>\n",
       "      <td>110</td>\n",
       "      <td>0</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>atpB</th>\n",
       "      <td>849</td>\n",
       "      <td>1497</td>\n",
       "      <td>1520</td>\n",
       "      <td>124</td>\n",
       "      <td>12</td>\n",
       "      <td>112</td>\n",
       "      <td>0</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>atpE</th>\n",
       "      <td>396</td>\n",
       "      <td>402</td>\n",
       "      <td>450</td>\n",
       "      <td>124</td>\n",
       "      <td>13</td>\n",
       "      <td>111</td>\n",
       "      <td>0</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>atpF</th>\n",
       "      <td>424</td>\n",
       "      <td>555</td>\n",
       "      <td>576</td>\n",
       "      <td>124</td>\n",
       "      <td>13</td>\n",
       "      <td>111</td>\n",
       "      <td>123</td>\n",
       "      <td>1.99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnV_UAC</th>\n",
       "      <td>37</td>\n",
       "      <td>74</td>\n",
       "      <td>119</td>\n",
       "      <td>113</td>\n",
       "      <td>9</td>\n",
       "      <td>104</td>\n",
       "      <td>104</td>\n",
       "      <td>1.92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnW_CCA</th>\n",
       "      <td>74</td>\n",
       "      <td>74</td>\n",
       "      <td>76</td>\n",
       "      <td>124</td>\n",
       "      <td>11</td>\n",
       "      <td>113</td>\n",
       "      <td>1</td>\n",
       "      <td>1.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnY_GUA</th>\n",
       "      <td>82</td>\n",
       "      <td>86</td>\n",
       "      <td>89</td>\n",
       "      <td>124</td>\n",
       "      <td>16</td>\n",
       "      <td>108</td>\n",
       "      <td>0</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ycf1</th>\n",
       "      <td>3164</td>\n",
       "      <td>5538</td>\n",
       "      <td>6951</td>\n",
       "      <td>113</td>\n",
       "      <td>29</td>\n",
       "      <td>84</td>\n",
       "      <td>1</td>\n",
       "      <td>1.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ycf2</th>\n",
       "      <td>255</td>\n",
       "      <td>6865</td>\n",
       "      <td>7374</td>\n",
       "      <td>122</td>\n",
       "      <td>67</td>\n",
       "      <td>55</td>\n",
       "      <td>0</td>\n",
       "      <td>1.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>113 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          len_min  len_median  len_max  Accession_count  forward_sum  \\\n",
       "gene                                                                   \n",
       "accD          746        1476     1893              112          101   \n",
       "atpA          882        1524     1545              124           14   \n",
       "atpB          849        1497     1520              124           12   \n",
       "atpE          396         402      450              124           13   \n",
       "atpF          424         555      576              124           13   \n",
       "...           ...         ...      ...              ...          ...   \n",
       "trnV_UAC       37          74      119              113            9   \n",
       "trnW_CCA       74          74       76              124           11   \n",
       "trnY_GUA       82          86       89              124           16   \n",
       "ycf1         3164        5538     6951              113           29   \n",
       "ycf2          255        6865     7374              122           67   \n",
       "\n",
       "          reverse_sum  join_sum  N_exons_mean  \n",
       "gene                                           \n",
       "accD               11         1          1.01  \n",
       "atpA              110         0          1.00  \n",
       "atpB              112         0          1.00  \n",
       "atpE              111         0          1.00  \n",
       "atpF              111       123          1.99  \n",
       "...               ...       ...           ...  \n",
       "trnV_UAC          104       104          1.92  \n",
       "trnW_CCA          113         1          1.01  \n",
       "trnY_GUA          108         0          1.00  \n",
       "ycf1               84         1          1.01  \n",
       "ycf2               55         0          1.00  \n",
       "\n",
       "[113 rows x 8 columns]"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_stats_clean = gene_stats[gene_stats.Accession_count>(gene_stats.Accession_count.max()*0.33)].astype({'len_median':'int'})\n",
    "gene_stats_clean['N_exons_mean'] = gene_stats_clean['N_exons_mean'].round(2)\n",
    "gene_stats_clean.reset_index().to_csv(ref_dir + 'cleaned_gene_stats.csv',index=False)\n",
    "print(gene_stats_clean.shape[0],gene_stats_clean.Accession_count.sort_values(ascending=False).to_dict())\n",
    "gene_stats_clean"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "psaM removed, rpl21 removed, trnF_AAA removed, trnP_GGG removed, trnR_CCG removed, trnT_CGU removed, psb30 removed, trnS_AGA removed, psb30 removed, trnG_ACC removed, trnP_GGG removed, trnR_CCG removed, chlB removed, chlN removed, psaM removed, trnP_GGG removed, trnR_CCG removed, trnY_AUA removed, trnfM_CAU removed, chlB removed, chlN removed, psaM removed, rpl21 removed, trnP_GGG removed, trnR_UCG removed, trnT_AGU removed, chlB removed, chlN removed, psaM removed, trnP_GGG removed, trnR_CCG removed, trnY_AUA removed, trnfM_CAU removed, chlB removed, chlN removed, psaM removed, rpl21 removed, trnP_GGG removed, trnR_CCG removed, chlB removed, chlN removed, psaM removed, rpl21 removed, trnD_AUC removed, trnP_GGG removed, trnR_CCG removed, chlB removed, chlN removed, psaM removed, rpl21 removed, trnA_AGC removed, trnP_GGG removed, chlB removed, chlN removed, psaM removed, rpl21 removed, trnK_CUU removed, trnP_GGG removed, trnR_CCG removed, trnT_CGU removed, chlB removed, chlN removed, psaM removed, rpl21 removed, trnP_GGG removed, trnR_CCG removed, trnI_AAU removed, trnO_CUA removed, trnT_CGU removed, trnY_AUA removed, chlB removed, chlN removed, psaM removed, trnI_AAU removed, trnP_GGG removed, trnTERM_UUA removed, psb30 removed, trnG_ACC removed, trnP_GGG removed, trnR_CCG removed, trnT_CGU removed, trnS_ACU removed, chlB removed, chlN removed, cysT removed, psaM removed, rpl21 removed, trnP_GGG removed, trnR_CCG removed, trnfM_CAU removed, trnT_CGU removed, trnT_CGU removed, chlB removed, chlN removed, cysT removed, psaM removed, rpl21 removed, trnP_GGG removed, trnR_CCG removed, psaM removed, trnT_CGU removed, chlB removed, chlN removed, psaM removed, trnO_CUA removed, trnR_CCG removed, trnT_CGU removed, trnI_AAU removed, psaM removed, chlB removed, chlN removed, psb30 removed, rpl21 removed, trnP_GGG removed, chlB removed, chlN removed, psaM removed, rpl21 removed, trnP_GGG removed, trnnull_NNN removed, trnT_CGU removed, trnT_CGU removed, trnTERM_UUA removed, psaM removed, psb30 removed, trnI_AAU removed, trnI_AAU removed, chlB removed, chlN removed, psaM removed, rpl21 removed, trnP_GGG removed, trnQ_CUG removed, trnT_CGU removed, trnnull_NNN removed, chlB removed, chlN removed, psaM removed, rpl21 removed, trnF_AAA removed, trnP_GGG removed, trnR_CCG removed, trnT_CGU removed, psaM removed, trnF_AAA removed, trnF_AAA removed, trnT_CGU removed, chlB removed, chlN removed, psb30 removed, rpl21 removed, trnP_GGG removed, trnY_AUA removed, chlB removed, chlN removed, psaM removed, rpl21 removed, trnL_CAG removed, trnP_GGG removed, trnR_CCG removed, trnT_CGU removed, psaM removed, trnF_AAA removed, trnT_CGU removed, trnI_AAU removed, trnS_AGA removed, trnT_CGU removed, chlB removed, chlN removed, psaM removed, trnI_AAU removed, trnI_UAU removed, trnP_GGG removed, trnR_UCG removed, trnI_UAU removed, chlB removed, chlN removed, cysA removed, psb30 removed, rpl21 removed, trnR_CCU removed, trnfM_CAU removed, trnI_AAU removed, trnI_AAU removed, trnL_AAG removed, trnI_AAU removed, psaM removed, trnT_CGU removed, psaM removed, trnF_AAA removed, trnT_CGU removed, trnR_UCG removed, psaM removed, trnTERM_UUA removed, psb30 removed, trnO_CUA removed, trnF_AAA removed, chlB removed, chlN removed, psaM removed, rpl21 removed, trnD_AUC removed, trnP_GGG removed, trnR_CCG removed, trnL_AAG removed, chlB removed, chlN removed, cysT removed, psaM removed, rpl21 removed, trnP_GGG removed, trnR_CCG removed, trnT_CGU removed, trnfM_CAU removed, trnT_CGU removed, trnT_CGU removed, trnT_CGU removed, chlB removed, chlN removed, psaM removed, 13865\n"
     ]
    }
   ],
   "source": [
    "# Remove genes not kept\n",
    "for idx, record in enumerate(all_seqs):\n",
    "    if (record.id.split('-')[1] not in gene_stats_clean.index):\n",
    "        print(record.id,'removed',end=', ')\n",
    "        del all_seqs[idx]\n",
    "print(len(all_seqs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13632\n"
     ]
    }
   ],
   "source": [
    "# Remove duplicates\n",
    "ls_id = []\n",
    "for idx, record in enumerate(all_seqs):\n",
    "    if record.id in ls_id:\n",
    "        print(record.id,'removed',end=', ')\n",
    "        del all_seqs[idx]\n",
    "    else:\n",
    "        ls_id.append(record.id)\n",
    "print(len(all_seqs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "13632"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SeqIO.write(all_seqs,ref_dir + 'cpDNA_RefCDS.fasta',format='fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['atpF', 'clpP1', 'ndhA', 'ndhB', 'pafI', 'petB', 'petD', 'rpl16',\n",
      "       'rpl2', 'rpoC1', 'rps12', 'rps16', 'trnC_ACA', 'trnG_UCC', 'trnK_UUU',\n",
      "       'trnL_UAA', 'trnS_CGA', 'trnV_UAC'],\n",
      "      dtype='object', name='gene')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>len_min</th>\n",
       "      <th>len_median</th>\n",
       "      <th>len_max</th>\n",
       "      <th>Accession_count</th>\n",
       "      <th>forward_sum</th>\n",
       "      <th>reverse_sum</th>\n",
       "      <th>join_sum</th>\n",
       "      <th>N_exons_mean</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>atpF</th>\n",
       "      <td>424</td>\n",
       "      <td>555</td>\n",
       "      <td>576</td>\n",
       "      <td>124</td>\n",
       "      <td>13</td>\n",
       "      <td>111</td>\n",
       "      <td>123</td>\n",
       "      <td>1.99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>clpP1</th>\n",
       "      <td>323</td>\n",
       "      <td>591</td>\n",
       "      <td>948</td>\n",
       "      <td>120</td>\n",
       "      <td>10</td>\n",
       "      <td>110</td>\n",
       "      <td>106</td>\n",
       "      <td>2.74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ndhA</th>\n",
       "      <td>574</td>\n",
       "      <td>1092</td>\n",
       "      <td>1119</td>\n",
       "      <td>114</td>\n",
       "      <td>22</td>\n",
       "      <td>92</td>\n",
       "      <td>113</td>\n",
       "      <td>1.99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ndhB</th>\n",
       "      <td>773</td>\n",
       "      <td>1533</td>\n",
       "      <td>1569</td>\n",
       "      <td>116</td>\n",
       "      <td>44</td>\n",
       "      <td>72</td>\n",
       "      <td>113</td>\n",
       "      <td>1.97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pafI</th>\n",
       "      <td>283</td>\n",
       "      <td>507</td>\n",
       "      <td>543</td>\n",
       "      <td>121</td>\n",
       "      <td>11</td>\n",
       "      <td>110</td>\n",
       "      <td>121</td>\n",
       "      <td>2.99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>petB</th>\n",
       "      <td>489</td>\n",
       "      <td>648</td>\n",
       "      <td>708</td>\n",
       "      <td>124</td>\n",
       "      <td>116</td>\n",
       "      <td>8</td>\n",
       "      <td>113</td>\n",
       "      <td>1.91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>petD</th>\n",
       "      <td>479</td>\n",
       "      <td>483</td>\n",
       "      <td>539</td>\n",
       "      <td>124</td>\n",
       "      <td>116</td>\n",
       "      <td>8</td>\n",
       "      <td>90</td>\n",
       "      <td>1.73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rpl16</th>\n",
       "      <td>392</td>\n",
       "      <td>408</td>\n",
       "      <td>492</td>\n",
       "      <td>120</td>\n",
       "      <td>7</td>\n",
       "      <td>113</td>\n",
       "      <td>102</td>\n",
       "      <td>1.85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rpl2</th>\n",
       "      <td>462</td>\n",
       "      <td>825</td>\n",
       "      <td>849</td>\n",
       "      <td>122</td>\n",
       "      <td>57</td>\n",
       "      <td>65</td>\n",
       "      <td>117</td>\n",
       "      <td>1.97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rpoC1</th>\n",
       "      <td>1359</td>\n",
       "      <td>2049</td>\n",
       "      <td>2118</td>\n",
       "      <td>124</td>\n",
       "      <td>14</td>\n",
       "      <td>110</td>\n",
       "      <td>118</td>\n",
       "      <td>1.95</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rps12</th>\n",
       "      <td>232</td>\n",
       "      <td>372</td>\n",
       "      <td>723</td>\n",
       "      <td>123</td>\n",
       "      <td>80</td>\n",
       "      <td>114</td>\n",
       "      <td>118</td>\n",
       "      <td>2.89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rps16</th>\n",
       "      <td>143</td>\n",
       "      <td>261</td>\n",
       "      <td>276</td>\n",
       "      <td>103</td>\n",
       "      <td>3</td>\n",
       "      <td>100</td>\n",
       "      <td>90</td>\n",
       "      <td>1.87</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnC_ACA</th>\n",
       "      <td>72</td>\n",
       "      <td>95</td>\n",
       "      <td>97</td>\n",
       "      <td>59</td>\n",
       "      <td>3</td>\n",
       "      <td>56</td>\n",
       "      <td>57</td>\n",
       "      <td>1.97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnG_UCC</th>\n",
       "      <td>48</td>\n",
       "      <td>71</td>\n",
       "      <td>92</td>\n",
       "      <td>117</td>\n",
       "      <td>106</td>\n",
       "      <td>11</td>\n",
       "      <td>116</td>\n",
       "      <td>1.99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnK_UUU</th>\n",
       "      <td>37</td>\n",
       "      <td>72</td>\n",
       "      <td>90</td>\n",
       "      <td>115</td>\n",
       "      <td>6</td>\n",
       "      <td>109</td>\n",
       "      <td>112</td>\n",
       "      <td>1.97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnL_UAA</th>\n",
       "      <td>84</td>\n",
       "      <td>85</td>\n",
       "      <td>89</td>\n",
       "      <td>112</td>\n",
       "      <td>103</td>\n",
       "      <td>9</td>\n",
       "      <td>112</td>\n",
       "      <td>2.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnS_CGA</th>\n",
       "      <td>70</td>\n",
       "      <td>92</td>\n",
       "      <td>94</td>\n",
       "      <td>84</td>\n",
       "      <td>75</td>\n",
       "      <td>9</td>\n",
       "      <td>77</td>\n",
       "      <td>1.92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>trnV_UAC</th>\n",
       "      <td>37</td>\n",
       "      <td>74</td>\n",
       "      <td>119</td>\n",
       "      <td>113</td>\n",
       "      <td>9</td>\n",
       "      <td>104</td>\n",
       "      <td>104</td>\n",
       "      <td>1.92</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          len_min  len_median  len_max  Accession_count  forward_sum  \\\n",
       "gene                                                                   \n",
       "atpF          424         555      576              124           13   \n",
       "clpP1         323         591      948              120           10   \n",
       "ndhA          574        1092     1119              114           22   \n",
       "ndhB          773        1533     1569              116           44   \n",
       "pafI          283         507      543              121           11   \n",
       "petB          489         648      708              124          116   \n",
       "petD          479         483      539              124          116   \n",
       "rpl16         392         408      492              120            7   \n",
       "rpl2          462         825      849              122           57   \n",
       "rpoC1        1359        2049     2118              124           14   \n",
       "rps12         232         372      723              123           80   \n",
       "rps16         143         261      276              103            3   \n",
       "trnC_ACA       72          95       97               59            3   \n",
       "trnG_UCC       48          71       92              117          106   \n",
       "trnK_UUU       37          72       90              115            6   \n",
       "trnL_UAA       84          85       89              112          103   \n",
       "trnS_CGA       70          92       94               84           75   \n",
       "trnV_UAC       37          74      119              113            9   \n",
       "\n",
       "          reverse_sum  join_sum  N_exons_mean  \n",
       "gene                                           \n",
       "atpF              111       123          1.99  \n",
       "clpP1             110       106          2.74  \n",
       "ndhA               92       113          1.99  \n",
       "ndhB               72       113          1.97  \n",
       "pafI              110       121          2.99  \n",
       "petB                8       113          1.91  \n",
       "petD                8        90          1.73  \n",
       "rpl16             113       102          1.85  \n",
       "rpl2               65       117          1.97  \n",
       "rpoC1             110       118          1.95  \n",
       "rps12             114       118          2.89  \n",
       "rps16             100        90          1.87  \n",
       "trnC_ACA           56        57          1.97  \n",
       "trnG_UCC           11       116          1.99  \n",
       "trnK_UUU          109       112          1.97  \n",
       "trnL_UAA            9       112          2.00  \n",
       "trnS_CGA            9        77          1.92  \n",
       "trnV_UAC          104       104          1.92  "
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Multi_Exons = gene_stats_clean[(gene_stats_clean.N_exons_mean>1.33)]\n",
    "print(Multi_Exons.index)\n",
    "Multi_Exons"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run 'makeblastdb -in References/cpDNA_RefCDS.fasta -dbtype nucl -parse_seqids'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "blastn  -query ../GetOrganelles/PAFTOL/fasta_pt/PAFTOL_016977_pt.fasta  -db References/cpDNA_RefGenes.fasta -outfmt \"6 qseqid sseqid pident length mismatch gapopen qlen qstart qend slen sstart send evalue bitscore qseq\" -out PAFTOL_016977.blast -max_target_seqs 5 -num_threads 1 -gapopen 0 -gapextend 0\n",
    "\n",
    "python targets_from_blast.py PAFTOL_016977.blast PAFTOL_016977_ptGenes.fasta"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
