{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "reduced-assistant",
   "metadata": {},
   "outputs": [],
   "source": [
    "#python scripts/targets_from_blast.py Blastn/PAFTOL_012529_cpRefGenes.blastn cpGenes_by_sample/PAFTOL_012529_cpGenes.fasta 50"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "recognized-watershed",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq; from Bio.SeqRecord import SeqRecord"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "answering-reset",
   "metadata": {},
   "outputs": [],
   "source": [
    "def blast_results(blast_file):\n",
    "    #load blast result\n",
    "    sample_blast=pd.read_table(blast_file, sep='\\t', header=None) \n",
    "    \n",
    "    # Because references are blasted against the contigs, I'm swapping query and subject fields to make the analysis clearer.\n",
    "    qseqid='sseqid'; sseqid='qseqid'; qlen='slen'; qstart='sstart'; \n",
    "    qend='send'; slen='qlen'; sstart='qstart'; send='qend'\n",
    "    sample_blast.columns=[qseqid, sseqid, 'pident', 'length', 'mismatch', 'gapopen', qlen, qstart, \n",
    "                              qend, slen, sstart, send, 'evalue', 'bitscore','qseq']\n",
    "    # Now query is the contigs and subject is the references\n",
    "\n",
    "    #split reference id to get ref species and ref gene\n",
    "    sample_blast[['ref_sp', 'starget']] = sample_blast['sseqid'].str.split('-', n=1, expand=True)\n",
    "    \n",
    "    # Get direction of match and reverse complement if necessary\n",
    "    sample_blast['qdir'] = (sample_blast.qstart<sample_blast.qend).replace({True:'For',False:'Rev'})\n",
    "    sample_blast['sdir'] = (sample_blast.sstart<sample_blast.send).replace({True:'For',False:'Rev'})\n",
    "    for idx, row in sample_blast.iterrows():\n",
    "        if (row.qdir=='Rev'):\n",
    "            sample_blast.loc[idx,'qseq'] = str(Seq(row['qseq']).reverse_complement())\n",
    "            tmp = row['qstart']; sample_blast.loc[idx,'qstart'] = row['qend']; sample_blast.loc[idx,'qend']=tmp\n",
    "        elif (row.sdir=='Rev'):\n",
    "            sample_blast.loc[idx,'qseq'] = str(Seq(row['qseq']).reverse_complement())\n",
    "            tmp = row['sstart']; sample_blast.loc[idx,'sstart'] = row['send']; sample_blast.loc[idx,'send']=tmp\n",
    "    \n",
    "    #calculate overlap\n",
    "    sample_blast['scover']=round((sample_blast.send-sample_blast.sstart+1)/sample_blast.slen*100,1)\n",
    "    sample_blast['qcover']=round((sample_blast.qend-sample_blast.qstart+1)/sample_blast.qlen*100,1)\n",
    "    sample_blast['overlap']=False\n",
    "    \n",
    "    return sample_blast[['qseqid', 'sseqid', 'starget', 'ref_sp', 'pident', 'evalue', \n",
    "                         'bitscore','length', 'qlen', 'slen', 'qstart', 'qend', 'sstart','send','qdir','sdir',\n",
    "                        'scover', 'qcover', 'overlap', 'qseq']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "exclusive-possible",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make dictionary of nucleotides\n",
    "def seq_str2dc(seq_dc):\n",
    "    return dict(zip(list(np.array([i for i in range(seq_dc['sstart'],seq_dc['send']+1)])), seq_dc['qseq']))\n",
    "\n",
    "def overlap_resolve(df):\n",
    "    df = df.sort_values('bitscore',ascending=False)\n",
    "    best = df.to_dict('records')[0] # Take best bitscore as backbone\n",
    "    best['seq_dc'] = seq_str2dc(best)\n",
    "    current_length = best['length'];\n",
    "    # Iterate over other matches to complete the reference (s)\n",
    "    for idx, row in df.iterrows():\n",
    "        row_dc = row.to_dict()\n",
    "        row_dc['seq_dc'] = seq_str2dc(row_dc)\n",
    "        best['seq_dc'] = {**row_dc['seq_dc'], **best['seq_dc']} # Add only new positions\n",
    "        # update bitscore if length as increased\n",
    "        best['length'] = len(best['seq_dc'])\n",
    "        if best['length']>current_length:\n",
    "            update_len =  best['length'] - current_length\n",
    "            update_ratio = update_len/row_dc['length']\n",
    "            best['bitscore'] = round(best['bitscore'] + (update_ratio*row_dc['bitscore']),0)\n",
    "            best['pident'] = round( ((best['pident']*(current_length)) + (update_len*row_dc['pident'])) / best['length'],0)\n",
    "            best['scover'] = round((best['length'])/best['slen']*100,1)\n",
    "            best['overlap']=True\n",
    "            current_length = best['length'];\n",
    "    best['qseq'] = \"\".join([N[1] for N in sorted(best['seq_dc'].items())])\n",
    "    del best[\"seq_dc\"]\n",
    "    return best\n",
    "\n",
    "# test_df = pd.DataFrame([{'Sample':'seq3','pident':90,'bitscore':10,'sstart':1,'send':6,'slen':25,'qseq':'AGAGAG'},\n",
    "#                         {'Sample':'seq1','pident':80,'bitscore':50,'sstart':10,'send':20,'slen':25,'qseq':'ACACACACACA'},\n",
    "#                        {'Sample':'seq4','pident':80,'bitscore':9,'sstart':19,'send':25,'slen':25,'qseq':'XXXXXXX'},\n",
    "#                         {'Sample':'seq2','pident':90,'bitscore':25,'sstart':20,'send':25,'slen':25,'qseq':'CTCTCT'}])\n",
    "# test_df['length'] = test_df.qseq.str.len()\n",
    "# test_df['scover'] = test_df.length/test_df.slen*100\n",
    "# best = overlap_resolve(test_df)\n",
    "# print(test_df)\n",
    "# print(best)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "forty-surrey",
   "metadata": {},
   "outputs": [],
   "source": [
    "def writefasta(df, fasta_out, sample_name):\n",
    "    out_seqs = []\n",
    "    for index, iseq in df.iterrows():\n",
    "        id = sample_name + '-' + iseq.starget\n",
    "        description = '{organism-gene:' + iseq.ref_sp + '-' + iseq.starget + ', originalID:' \\\n",
    "        + iseq.qseqid.replace(',','') + ', length:' + str(iseq.length) + ', pident:' + str(round(iseq.pident,2)) + ', bitscore:' \\\n",
    "        + str(iseq.bitscore) + ', reference_coverage:' + str(iseq.scover) + ', overlap:' + str(iseq.overlap) + '}'\n",
    "#         + ', sstart:' + str(iseq.sstart) + ', send:' + str(iseq.send) + ', qstart:' + str(iseq.qstart) + ', qend:' + str(iseq.qend) + '}'\n",
    "        # + ';cut2codon:' + str(iseq.cut2codon) \n",
    "        \n",
    "        # Remove insertions\n",
    "        seq = Seq(iseq.qseq.replace('-', ''))\n",
    "        record = SeqRecord(seq=seq,id=id,description=description)\n",
    "        \n",
    "        out_seqs.append(record)\n",
    "        \n",
    "    # Write sequences\n",
    "    SeqIO.write(out_seqs,fasta_out,format='fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "relevant-snapshot",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " PAFTOL_006071 ../tmp_cpDNA/PAFTOL_006071_CPgenes.blastn ../tmp_cpDNA/PAFTOL_006071_CPgenes.fasta\n",
      "All entries: 4467 72\n",
      "qdir  sdir\n",
      "For   For     2368\n",
      "Rev   For     2099\n",
      "dtype: int64\n",
      "13 sseqid with >1 match 4 genes, 0 known split genes\n",
      "restricted: 13 sseqid with >1 match in 4 genes\n",
      "After filtering: 21 21 5514 overlap: {False: 21}\n"
     ]
    }
   ],
   "source": [
    "# Main\n",
    "if __name__ == \"__main__\":\n",
    "    input_blast_file=sys.argv[1]\n",
    "    fasta_out=sys.argv[2]\n",
    "#     input_blast_file= '../tmp_cpDNA/' + 'PAFTOL_006071_CPgenes.blastn'\n",
    "#     fasta_out= '../tmp_cpDNA/' + 'PAFTOL_006071_CPgenes.fasta'\n",
    "    \n",
    "    sample_name=input_blast_file.split('/')[-1].replace('_CPgenes.blastn','')\n",
    "    min_cov=50\n",
    "    min_pid=70\n",
    "    resolve_overlap=True\n",
    "    restrict_genes=True\n",
    "    \n",
    "    \n",
    "    print('\\n',sample_name, input_blast_file, fasta_out)\n",
    "    \n",
    "    if os.path.exists(input_blast_file):\n",
    "        if os.stat(input_blast_file).st_size != 0:\n",
    "            # Load blast output\n",
    "            df_blast=blast_results(blast_file=input_blast_file)\n",
    "            print('All entries:',df_blast.shape[0], df_blast.starget.nunique())\n",
    "            print(df_blast.groupby(['qdir','sdir']).size())\n",
    "            \n",
    "            ### Resolve overlaps: genes with multiple exons AND (if restrict_genes==False) genes with multiple matching contigs\n",
    "            if resolve_overlap:\n",
    "                genes_ex = ['atpF', 'clpP1', 'ndhA', 'ndhB', 'pafI', 'petB', 'petD', 'rpl16',\n",
    "                   'rpl2', 'rpoC1', 'rps12', 'rps16', 'trnC_ACA', 'trnG_UCC', 'trnK_UUU',\n",
    "                   'trnL_UAA', 'trnS_CGA', 'trnV_UAC']\n",
    "\n",
    "                # sseqid with multiple entries\n",
    "                sseqid_ex = df_blast.groupby('sseqid').size().to_frame().rename(columns={0:'N'}).reset_index();\n",
    "                sseqid_ex = sseqid_ex[sseqid_ex.N>1]\n",
    "                if sseqid_ex.shape[0]>0:\n",
    "                    sseqid_ex[['ref_sp', 'starget']] = sseqid_ex['sseqid'].str.split('-', n=1, expand=True)\n",
    "                    print(sseqid_ex.shape[0],'sseqid with >1 match',sseqid_ex.starget.nunique(),'genes,',\n",
    "                          sseqid_ex[sseqid_ex.starget.isin(genes_ex)].starget.nunique(),'known split genes')\n",
    "                    if restrict_genes:\n",
    "                        sseqid_ex = sseqid_ex[sseqid_ex.starget.isin(genes_ex)]\n",
    "                        \n",
    "                # Resolve overlap\n",
    "                if sseqid_ex.shape[0]>0:\n",
    "                    print('restricted:',sseqid_ex.shape[0],'sseqid with >1 match in',sseqid_ex.starget.nunique(),'genes')\n",
    "                    ls_overlap = []\n",
    "                    for sseqid in sseqid_ex.sseqid:\n",
    "                        ls_overlap.append(overlap_resolve(df_blast[df_blast.sseqid==sseqid]))\n",
    "                    df_blast = pd.concat([df_blast, pd.DataFrame(ls_overlap)])\n",
    "            \n",
    "            ### Keep best match for each gene (starget)\n",
    "            df_best = df_blast.sort_values('bitscore',ascending=False).groupby('starget').head(1).reset_index(drop=True)\n",
    "            df_best = df_best[(df_best.scover>=min_cov) & (df_best.pident>=min_pid)]\n",
    "            print('After filtering:',df_best.shape[0], df_best.starget.nunique(), df_best.length.sum(),\n",
    "                 'overlap:',df_best.groupby('overlap').size().to_dict())\n",
    "                                             \n",
    "            if df_best.shape[0]>0:\n",
    "                # Write fasta\n",
    "                writefasta(df = df_best, fasta_out = fasta_out, sample_name = sample_name)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
